---
title: "Homology group visualization"
date: "`r Sys.time()`"
format: 
  html: 
    embed-resources: true
params:
  tree:
    label: "Phylogenetic tree"
    value: kmer_upgma
    input: select
    choices: [kmer_upgma, kmer_nj, core_snp_ml, ani_nj]
  name:
    label: "Analysis name"
    input: text
    value: "HG_pav"
  
---

***

## Initial setup

```{r}
#| label: setup
#| warning: false

suppressPackageStartupMessages(library(tidyverse))
suppressPackageStartupMessages(library(configr))
suppressPackageStartupMessages(library(ggpubr))
suppressPackageStartupMessages(library(here))
suppressPackageStartupMessages(library(ape))
suppressPackageStartupMessages(library(treeio))
suppressPackageStartupMessages(library(ggtree))
suppressPackageStartupMessages(library(ComplexHeatmap))
suppressPackageStartupMessages(library(circlize))
suppressPackageStartupMessages(library(viridisLite))
suppressPackageStartupMessages(library(openxlsx))
suppressPackageStartupMessages(library(GenomicRanges))
suppressPackageStartupMessages(library(org.Pectobacterium.spp.pan.eg.db))

## visualize homology groups PAV for given homology group sets

rm(list = ls())

source("https://raw.githubusercontent.com/lakhanp1/omics_utils/main/RScripts/utils.R")
source("scripts/utils/config_functions.R")
source("scripts/utils/phylogeny_functions.R")
source("scripts/utils/heatmap_utils.R")
source("scripts/utils/homology_groups.R")
################################################################################
set.seed(124)

analysisName <- "assay_HG_pav"

setIds <- c("LZI", "TIR", "assay_FN")

treeMethod <- "kmer_upgma" # ani_upgma, kmer_upgma

confs <- prefix_config_paths(
  conf = suppressWarnings(configr::read.config(file = "project_config.yaml")),
  dir = "."
)

outDir <- confs$analysis$insilico_assay$dir
outPrefix <- file.path(outDir, analysisName)

pangenome <- confs$data$pangenomes$pectobacterium.v2$name
panConf <- confs$data$pangenomes[[pangenome]]

panOrgDb <- org.Pectobacterium.spp.pan.eg.db

# add any homology group set file created to this vector
# `setIds` are searched in these files to identify respective homology groups
# for plotting
hgSetFiles <- c(
  confs$analysis$homology_groups$files$hg_sets,
  confs$analysis$association$files$pheno_specific_groups,
  confs$analysis$prophages$preprocessing$files$raw_prophage_hg
)

```

## Import data

### Sample metadata

```{r}
sampleInfo <- get_metadata(file = panConf$files$metadata, genus = confs$genus)

rawTree <- import_tree(
  file = confs$analysis$phylogeny[[treeMethod]]$files$tree, phylo = TRUE
)

speciesOrder <- suppressMessages(
  readr::read_tsv(confs$analysis$phylogeny[[treeMethod]]$files$species_order)
)

## add species order factor levels to SpeciesName column
sampleInfo %<>% dplyr::mutate(
  SpeciesName = forcats::fct_relevel(SpeciesName, !!!rev(speciesOrder$SpeciesName))
)

```


### Import homology group information

```{r}
# prepare homology groups data for plotting
hgSets <- NULL

for (st in hgSetFiles) {
  df <- suppressMessages(readr::read_tsv(st)) %>%
    dplyr::mutate(
      hgs = stringr::str_split(hgs, ";"),
      setSource = st
    ) %>%
    dplyr::select(id, hgs, setSource)
  
  
  hgSets <- dplyr::bind_rows(hgSets, df)
}



if (any(duplicated(hgSets$id))) {
  stop(
    "duplicate homology group set identifiers found in the data:",
    names(which(table(hgSets$id) > 1))
  )
}

# read prophage HGs stored locally
hgL <- dplyr::left_join(
  x = tibble::tibble(id = setIds), y = hgSets, by = "id"
) %>%
  purrr::transpose() %>%
  purrr::set_names(nm = purrr::map(., "id"))

```

## Data visualization

```{r}
#| column: page
#| fig-height: 8
#| fig-width: 16
#| out-width: '200%'
#| layout-valign: top

htList <- species_key_heatmap(
  genomes = rawTree$tip.label, speciesInfo = sampleInfo
)

htList@heatmap_param$width <- unit(12, "cm")

# prepare homology group PAV matrix from pan.db
hgPavMat <- homology_groups_mat(
  pandb = panOrgDb, type = "pav",
  groups = purrr::map(hgL, "hgs") %>% unlist(use.names = F) %>% unique()
)

hgPavMat <- hgPavMat[rawTree$tip.label, ]

for (id in names(hgL)) {
  
  hgMat <- hgPavMat[, hgL[[id]]$hgs, drop = FALSE]
  
  ht <- homology_group_heatmap(
    mat = hgMat, phy = rawTree,
    # width = unit(10, "cm"),
    name = id, column_title = id,
    use_raster = TRUE, raster_quality = 3
  )
  
  ht@column_dend_param$cluster <- FALSE
  ht@column_names_param$show <- FALSE
  
  htList <- htList + ht
}


```

```{r}
#| fig-height: 8
#| fig-width: 12
#| out-width: '100%'
#| layout-valign: top
#| column: page
#| echo: false

pdf(file = paste(outPrefix, ".hgs.pdf", sep = ""), width = 12, height = 8)
ComplexHeatmap::draw(
  object = htList,
  main_heatmap = hgL[[1]]$id,
  row_dend_side = "left",
  merge_legends = TRUE,
  heatmap_legend_side = "bottom"
  
)
dev.off()

ComplexHeatmap::draw(
  object = htList,
  main_heatmap = hgL[[1]]$id,
  row_dend_side = "left",
  merge_legends = TRUE,
  heatmap_legend_side = "bottom"
)
```




