<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>

<meta charset="utf-8">
<meta name="generator" content="quarto-1.5.57">

<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">

<meta name="author" content="Lakhansing Pardeshi">

<title>pangenome_analysis – Pectobacterium pangenome data analysis</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
  vertical-align: middle;
}
/* CSS for syntax highlighting */
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { display: inline-block; text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
  }
pre.numberSource { margin-left: 3em;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
</style>


<script src="../../site_libs/quarto-nav/quarto-nav.js"></script>
<script src="../../site_libs/quarto-nav/headroom.min.js"></script>
<script src="../../site_libs/clipboard/clipboard.min.js"></script>
<script src="../../site_libs/quarto-search/autocomplete.umd.js"></script>
<script src="../../site_libs/quarto-search/fuse.min.js"></script>
<script src="../../site_libs/quarto-search/quarto-search.js"></script>
<meta name="quarto:offset" content="../../">
<link href="../../scripts/analysis/HG_summary.html" rel="next">
<link href="../../scripts/notebooks/pangenome_construction.html" rel="prev">
<script src="../../site_libs/quarto-html/quarto.js"></script>
<script src="../../site_libs/quarto-html/popper.min.js"></script>
<script src="../../site_libs/quarto-html/tippy.umd.min.js"></script>
<script src="../../site_libs/quarto-html/anchor.min.js"></script>
<link href="../../site_libs/quarto-html/tippy.css" rel="stylesheet">
<link href="../../site_libs/quarto-html/quarto-syntax-highlighting.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="../../site_libs/bootstrap/bootstrap.min.js"></script>
<link href="../../site_libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="../../site_libs/bootstrap/bootstrap.min.css" rel="stylesheet" id="quarto-bootstrap" data-mode="light">
<script id="quarto-search-options" type="application/json">{
  "location": "sidebar",
  "copy-button": false,
  "collapse-after": 3,
  "panel-placement": "start",
  "type": "textbox",
  "limit": 50,
  "keyboard-shortcut": [
    "f",
    "/",
    "s"
  ],
  "show-item-context": false,
  "language": {
    "search-no-results-text": "No results",
    "search-matching-documents-text": "matching documents",
    "search-copy-link-title": "Copy link to search",
    "search-hide-matches-text": "Hide additional matches",
    "search-more-match-text": "more match in this document",
    "search-more-matches-text": "more matches in this document",
    "search-clear-button-title": "Clear",
    "search-text-placeholder": "",
    "search-detached-cancel-button-title": "Cancel",
    "search-submit-button-title": "Submit",
    "search-label": "Search"
  }
}</script>


<link rel="stylesheet" href="../../scripts/utils/quarto_style.css">
<meta property="og:title" content="Pectobacterium pangenome data analysis">
<meta property="og:site_name" content="Pectobacterium pangenome data analysis">
</head>

<body class="nav-sidebar docked">

<div id="quarto-search-results"></div>
  <header id="quarto-header" class="headroom fixed-top">
  <nav class="quarto-secondary-nav">
    <div class="container-fluid d-flex">
      <button type="button" class="quarto-btn-toggle btn" data-bs-toggle="collapse" role="button" data-bs-target=".quarto-sidebar-collapse-item" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
        <i class="bi bi-layout-text-sidebar-reverse"></i>
      </button>
        <nav class="quarto-page-breadcrumbs" aria-label="breadcrumb"><ol class="breadcrumb"><li class="breadcrumb-item"><a href="../../scripts/analysis/pangenome_metadata_viz.html">Pangenome</a></li><li class="breadcrumb-item"><a href="../../scripts/notebooks/pangenome_analysis.html">Analysis</a></li></ol></nav>
        <a class="flex-grow-1" role="navigation" data-bs-toggle="collapse" data-bs-target=".quarto-sidebar-collapse-item" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">      
        </a>
      <button type="button" class="btn quarto-search-button" aria-label="Search" onclick="window.quartoOpenSearch();">
        <i class="bi bi-search"></i>
      </button>
    </div>
  </nav>
</header>
<!-- content -->
<div id="quarto-content" class="quarto-container page-columns page-rows-contents page-layout-article">
<!-- sidebar -->
  <nav id="quarto-sidebar" class="sidebar collapse collapse-horizontal quarto-sidebar-collapse-item sidebar-navigation docked overflow-auto">
    <div class="pt-lg-2 mt-2 text-left sidebar-header">
    <div class="sidebar-title mb-0 py-0">
      <a href="../../">Pectobacterium pangenome data analysis</a> 
        <div class="sidebar-tools-main">
    <div class="dropdown">
      <a href="" title="" id="quarto-navigation-tool-dropdown-0" class="quarto-navigation-tool dropdown-toggle px-1" data-bs-toggle="dropdown" aria-expanded="false" role="link" aria-label=""><i class="bi bi-github"></i></a>
      <ul class="dropdown-menu" aria-labelledby="quarto-navigation-tool-dropdown-0">
          <li>
            <a class="dropdown-item sidebar-tools-main-item" href="https://github.com/lakhanp1/Pectobacterium_pangenome">
            Source Code
            </a>
          </li>
          <li>
            <a class="dropdown-item sidebar-tools-main-item" href="https://github.com/lakhanp1/Pectobacterium_pangenome/issues/new">
            Report a bug
            </a>
          </li>
      </ul>
    </div>
    <a href="https://x.com/lakhanp01" title="" class="quarto-navigation-tool px-1" aria-label=""><i class="bi bi-twitter"></i></a>
</div>
    </div>
      </div>
        <div class="mt-2 flex-shrink-0 align-items-center">
        <div class="sidebar-search">
        <div id="quarto-search" class="" title="Search"></div>
        </div>
        </div>
    <div class="sidebar-menu-container"> 
    <ul class="list-unstyled mt-1">
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../scripts/notebooks/index.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Pectobacterium genus pangenome</span></a>
  </div>
</li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-1" role="navigation" aria-expanded="true">
 <span class="menu-text">Preprocessing</span></a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-1" role="navigation" aria-expanded="true" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-1" class="collapse list-unstyled sidebar-section depth1 show">  
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../scripts/notebooks/preprocessing.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Preprocessing</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../scripts/analysis/ani_clustering.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">ANI</span></a>
  </div>
</li>
      </ul>
  </li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-2" role="navigation" aria-expanded="true">
 <span class="menu-text">Pangenome</span></a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-2" role="navigation" aria-expanded="true" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-2" class="collapse list-unstyled sidebar-section depth1 show">  
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../scripts/analysis/pangenome_metadata_viz.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Pangenome metadata</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../scripts/notebooks/pangenome_construction.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Build</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../scripts/notebooks/pangenome_analysis.html" class="sidebar-item-text sidebar-link active">
 <span class="menu-text">Analysis</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../scripts/analysis/HG_summary.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Pangenome summary</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../scripts/analysis/phylo_viz_tree.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Phylogeny</span></a>
  </div>
</li>
      </ul>
  </li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-3" role="navigation" aria-expanded="true">
 <span class="menu-text">Prophages</span></a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-3" role="navigation" aria-expanded="true" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-3" class="collapse list-unstyled sidebar-section depth1 show">  
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../scripts/notebooks/prophage_analysis.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Analysis</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../scripts/analysis/prophage_summary.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Prophage summary</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../scripts/analysis/synteny_demo.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Syntenic Jaccard index</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../scripts/analysis/prophage_clustering.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Clustering</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../scripts/analysis/prophage_cluster_summary.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Clustering summary</span></a>
  </div>
</li>
          <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start collapsed" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-4" role="navigation" aria-expanded="false">
 <span class="menu-text">Prophage clusters</span></a>
          <a class="sidebar-item-toggle text-start collapsed" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-4" role="navigation" aria-expanded="false" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-4" class="collapse list-unstyled sidebar-section depth2 ">  
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../scripts/ms_figures/phage_grp_45.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Phage group 45</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../scripts/ms_figures/phage_grp_46.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Phage group 46</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../scripts/ms_figures/phage_grp_107.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Phage group 107</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../scripts/ms_figures/phage_grp_71.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Phage group 71</span></a>
  </div>
</li>
      </ul>
  </li>
      </ul>
  </li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../scripts/ms_figures/supplementary_tables.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Supplementary tables</span></a>
  </div>
</li>
    </ul>
    </div>
</nav>
<div id="quarto-sidebar-glass" class="quarto-sidebar-collapse-item" data-bs-toggle="collapse" data-bs-target=".quarto-sidebar-collapse-item"></div>
<!-- margin-sidebar -->
    <div id="quarto-margin-sidebar" class="sidebar margin-sidebar">
        <nav id="TOC" role="doc-toc" class="toc-active">
    <h2 id="toc-title">On this page</h2>
   
  <ul>
  <li><a href="#downstream-analysis-of-pangenome-data" id="toc-downstream-analysis-of-pangenome-data" class="nav-link active" data-scroll-target="#downstream-analysis-of-pangenome-data">Downstream analysis of pangenome data</a>
  <ul class="collapse">
  <li><a href="#setup" id="toc-setup" class="nav-link" data-scroll-target="#setup">Setup</a></li>
  <li><a href="#use-subset-of-genomes-to-determine-pangenome-structure" id="toc-use-subset-of-genomes-to-determine-pangenome-structure" class="nav-link" data-scroll-target="#use-subset-of-genomes-to-determine-pangenome-structure">Use subset of genomes to determine pangenome structure</a></li>
  <li><a href="#visualize-pangenome-metadata" id="toc-visualize-pangenome-metadata" class="nav-link" data-scroll-target="#visualize-pangenome-metadata">Visualize pangenome metadata</a></li>
  <li><a href="#phylogeny-analysis" id="toc-phylogeny-analysis" class="nav-link" data-scroll-target="#phylogeny-analysis">Phylogeny analysis</a></li>
  <li><a href="#homology-groups-analysis" id="toc-homology-groups-analysis" class="nav-link" data-scroll-target="#homology-groups-analysis">Homology groups analysis</a>
  <ul class="collapse">
  <li><a href="#go-enrichment-of-homology-groups-categories" id="toc-go-enrichment-of-homology-groups-categories" class="nav-link" data-scroll-target="#go-enrichment-of-homology-groups-categories">GO enrichment of homology groups categories</a></li>
  <li><a href="#summarize-homology-groups" id="toc-summarize-homology-groups" class="nav-link" data-scroll-target="#summarize-homology-groups">Summarize homology groups</a></li>
  <li><a href="#visualize-tandemly-mapped-homology-groups-sets-on-the-pangenome" id="toc-visualize-tandemly-mapped-homology-groups-sets-on-the-pangenome" class="nav-link" data-scroll-target="#visualize-tandemly-mapped-homology-groups-sets-on-the-pangenome">Visualize tandemly mapped homology groups sets on the pangenome</a></li>
  <li><a href="#visualize-homology-groups-across-pagenome-for-go-terms-of-interest" id="toc-visualize-homology-groups-across-pagenome-for-go-terms-of-interest" class="nav-link" data-scroll-target="#visualize-homology-groups-across-pagenome-for-go-terms-of-interest">Visualize homology groups across pagenome for GO terms of interest</a></li>
  </ul></li>
  </ul></li>
  </ul>
<div class="toc-actions"><ul><li><a href="https://github.com/lakhanp1/Pectobacterium_pangenome/issues/new" class="toc-action"><i class="bi bi-github"></i>Report an issue</a></li></ul></div></nav>
    </div>
<!-- main -->
<main class="content" id="quarto-document-content"><div class="quarto-title-block"><nav class="quarto-page-breadcrumbs quarto-title-breadcrumbs d-none d-lg-block" aria-label="breadcrumb"><ol class="breadcrumb"><li class="breadcrumb-item"><a href="../../scripts/analysis/pangenome_metadata_viz.html">Pangenome</a></li><li class="breadcrumb-item"><a href="../../scripts/notebooks/pangenome_analysis.html">Analysis</a></li></ol></nav><div class="quarto-title-tools-only"><h1></h1><button type="button" class="btn code-tools-button" id="quarto-code-tools-source"><i class="bi"></i> Code</button></div></div>

<header id="title-block-header" class="quarto-title-block default">
<div class="quarto-title">
</div>



<div class="quarto-title-meta">

    <div>
    <div class="quarto-title-meta-heading">Author</div>
    <div class="quarto-title-meta-contents">
             <p>Lakhansing Pardeshi </p>
          </div>
  </div>
    
  
    
  </div>
  


</header>


<section id="downstream-analysis-of-pangenome-data" class="level1">
<h1>Downstream analysis of pangenome data</h1>
<section id="setup" class="level2">
<h2 class="anchored" data-anchor-id="setup">Setup</h2>
<div class="sourceCode" id="cb1"><pre class="sourceCode numberSource bash number-lines code-with-copy"><code class="sourceCode bash"><span id="cb1-1"><a href="#cb1-1"></a><span class="co">#!/usr/bin/env bash</span></span>
<span id="cb1-2"><a href="#cb1-2"></a></span>
<span id="cb1-3"><a href="#cb1-3"></a><span class="bu">shopt</span> <span class="at">-s</span> expand_aliases</span>
<span id="cb1-4"><a href="#cb1-4"></a><span class="bu">source</span> ~/.bash_aliases</span>
<span id="cb1-5"><a href="#cb1-5"></a></span>
<span id="cb1-6"><a href="#cb1-6"></a><span class="co"># set -e</span></span>
<span id="cb1-7"><a href="#cb1-7"></a><span class="co"># set -u</span></span>
<span id="cb1-8"><a href="#cb1-8"></a><span class="bu">set</span> <span class="at">-o</span> pipefail</span>
<span id="cb1-9"><a href="#cb1-9"></a></span>
<span id="cb1-10"><a href="#cb1-10"></a><span class="co"># source scripts/utils/setup_analysis.sh 'pectobacterium.v2'</span></span>
<span id="cb1-11"><a href="#cb1-11"></a><span class="bu">source</span> scripts/utils/setup_analysis.sh <span class="va">$@</span></span>
<span id="cb1-12"><a href="#cb1-12"></a></span>
<span id="cb1-13"><a href="#cb1-13"></a><span class="cf">if</span> <span class="bu">[</span> <span class="ot">-z</span> <span class="va">${pan_db</span><span class="op">+</span>x<span class="va">}</span> <span class="bu">]</span><span class="kw">;</span></span>
<span id="cb1-14"><a href="#cb1-14"></a><span class="cf">then</span></span>
<span id="cb1-15"><a href="#cb1-15"></a>    <span class="bu">echo</span> <span class="st">"</span><span class="dt">\$</span><span class="st">pan_db is unset"</span></span>
<span id="cb1-16"><a href="#cb1-16"></a>    <span class="ex">error_exit</span> 1</span>
<span id="cb1-17"><a href="#cb1-17"></a><span class="cf">fi</span></span>
<span id="cb1-18"><a href="#cb1-18"></a></span>
<span id="cb1-19"><a href="#cb1-19"></a><span class="bu">source</span> <span class="va">$TOOLS_PATH</span>/miniconda3/etc/profile.d/conda.sh</span>
<span id="cb1-20"><a href="#cb1-20"></a><span class="ex">conda</span> activate pantools_master</span>
<span id="cb1-21"><a href="#cb1-21"></a><span class="bu">export</span> <span class="va">PANTOOLS</span><span class="op">=</span><span class="st">"</span><span class="va">$PANTOOLS_4_1</span><span class="st">"</span></span>
<span id="cb1-22"><a href="#cb1-22"></a></span>
<span id="cb1-23"><a href="#cb1-23"></a><span class="co">######################################################################</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</section>
<section id="use-subset-of-genomes-to-determine-pangenome-structure" class="level2">
<h2 class="anchored" data-anchor-id="use-subset-of-genomes-to-determine-pangenome-structure">Use subset of genomes to determine pangenome structure</h2>
<div class="sourceCode" id="cb2"><pre class="sourceCode numberSource bash number-lines code-with-copy"><code class="sourceCode bash"><span id="cb2-1"><a href="#cb2-1"></a><span class="co">## Pangenome structure for species with more than 20 genomes</span></span>
<span id="cb2-2"><a href="#cb2-2"></a><span class="cf">for</span> sp <span class="kw">in</span> <span class="kw">`</span><span class="fu">awk</span> <span class="at">-F</span> <span class="st">"\t"</span> <span class="st">'{ if (NR != 1 &amp;&amp; $2 &gt;= 20) {print $1} }'</span> <span class="va">$PANGENOME_DIR</span>/analysis_configs/species_wise_genomes.tab<span class="kw">`</span></span>
<span id="cb2-3"><a href="#cb2-3"></a><span class="cf">do</span></span>
<span id="cb2-4"><a href="#cb2-4"></a>    <span class="va">genomes</span><span class="op">=</span><span class="kw">`</span><span class="fu">grep</span> <span class="st">"^</span><span class="va">${sp}</span><span class="st">\b"</span> <span class="va">$PANGENOME_DIR</span>/analysis_configs/species_wise_genomes.tab <span class="kw">|</span> <span class="fu">cut</span> <span class="at">-f3</span><span class="kw">`</span></span>
<span id="cb2-5"><a href="#cb2-5"></a>    </span>
<span id="cb2-6"><a href="#cb2-6"></a>    <span class="ex">process_start</span> <span class="st">"pangenome_structure for genome of species </span><span class="va">${sp}</span><span class="st">: </span><span class="va">$genomes</span><span class="st">"</span></span>
<span id="cb2-7"><a href="#cb2-7"></a>    <span class="va">$PANTOOLS</span> pangenome_structure <span class="at">-t</span> 20 <span class="at">--include</span> <span class="va">${genomes}</span> <span class="va">${pan_db}</span></span>
<span id="cb2-8"><a href="#cb2-8"></a>    <span class="ex">error_exit</span> <span class="va">$?</span></span>
<span id="cb2-9"><a href="#cb2-9"></a></span>
<span id="cb2-10"><a href="#cb2-10"></a>    <span class="ex">Rscript</span> <span class="va">${pan_db}</span>/pangenome_size/gene/pangenome_growth.R</span>
<span id="cb2-11"><a href="#cb2-11"></a>    <span class="ex">Rscript</span> <span class="va">${pan_db}</span>/pangenome_size/gene/gains_losses_median_or_average.R</span>
<span id="cb2-12"><a href="#cb2-12"></a>    <span class="ex">Rscript</span> <span class="va">${pan_db}</span>/pangenome_size/gene/gains_losses_median_and_average.R</span>
<span id="cb2-13"><a href="#cb2-13"></a>    <span class="co"># Rscript ${pan_db}/pangenome_size/gene/heaps_law.R</span></span>
<span id="cb2-14"><a href="#cb2-14"></a></span>
<span id="cb2-15"><a href="#cb2-15"></a>    <span class="co">## move results to a folder</span></span>
<span id="cb2-16"><a href="#cb2-16"></a>    <span class="va">psDir</span><span class="op">=</span><span class="va">${pan_db}</span>/pangenome_size/gene.<span class="va">${sp}</span></span>
<span id="cb2-17"><a href="#cb2-17"></a>    <span class="bu">[</span> <span class="ot">-d</span> <span class="va">${psDir}</span> <span class="bu">]</span> <span class="kw">&amp;&amp;</span> <span class="fu">rm</span> <span class="at">-r</span> <span class="va">${psDir}</span></span>
<span id="cb2-18"><a href="#cb2-18"></a>    <span class="fu">mv</span> <span class="va">${pan_db}</span>/pangenome_size/gene <span class="va">${psDir}</span></span>
<span id="cb2-19"><a href="#cb2-19"></a><span class="cf">done</span></span>
<span id="cb2-20"><a href="#cb2-20"></a></span>
<span id="cb2-21"><a href="#cb2-21"></a><span class="co">## update the path in heaps_law.R script</span></span>
<span id="cb2-22"><a href="#cb2-22"></a><span class="bu">cd</span> <span class="va">${pan_db}</span>/pangenome_size</span>
<span id="cb2-23"><a href="#cb2-23"></a><span class="cf">for</span> i <span class="kw">in</span> gene.<span class="pp">*</span></span>
<span id="cb2-24"><a href="#cb2-24"></a><span class="cf">do</span></span>
<span id="cb2-25"><a href="#cb2-25"></a>    <span class="fu">sed</span> <span class="at">-i.bak</span> <span class="st">"s/pangenome_size\/gene/pangenome_size\/gene.</span><span class="va">$i</span><span class="st">/g"</span> <span class="va">$i</span>/heaps_law.R</span>
<span id="cb2-26"><a href="#cb2-26"></a><span class="cf">done</span></span>
<span id="cb2-27"><a href="#cb2-27"></a><span class="bu">cd</span> <span class="va">$PROJECT_DIR</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</section>
<section id="visualize-pangenome-metadata" class="level2">
<h2 class="anchored" data-anchor-id="visualize-pangenome-metadata">Visualize pangenome metadata</h2>
<div class="sourceCode" id="cb3"><pre class="sourceCode numberSource bash number-lines code-with-copy"><code class="sourceCode bash"><span id="cb3-1"><a href="#cb3-1"></a><span class="ex">quarto</span> render scripts/analysis/pangenome_metadata.viz.qmd</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</section>
<section id="phylogeny-analysis" class="level2">
<h2 class="anchored" data-anchor-id="phylogeny-analysis">Phylogeny analysis</h2>
<div class="sourceCode" id="cb4"><pre class="sourceCode numberSource bash number-lines code-with-copy"><code class="sourceCode bash"><span id="cb4-1"><a href="#cb4-1"></a><span class="co">## UPGMA tree</span></span>
<span id="cb4-2"><a href="#cb4-2"></a><span class="ex">./scripts/analysis/phylo_process_newicks.R</span> <span class="dt">\</span></span>
<span id="cb4-3"><a href="#cb4-3"></a>--tree ./analysis/pangenome_v2/phylogeny/ani/ANI_UPGMA.pangenome.newick <span class="dt">\</span></span>
<span id="cb4-4"><a href="#cb4-4"></a>--name <span class="st">"ani_upgma"</span> <span class="at">--config</span> project_config.yaml <span class="at">--save_leaf_order</span></span>
<span id="cb4-5"><a href="#cb4-5"></a></span>
<span id="cb4-6"><a href="#cb4-6"></a><span class="co">## Neighbor joining tree</span></span>
<span id="cb4-7"><a href="#cb4-7"></a><span class="ex">./scripts/analysis/phylo_process_newicks.R</span> <span class="dt">\</span></span>
<span id="cb4-8"><a href="#cb4-8"></a>--tree ./analysis/pangenome_v2/phylogeny/ani/ANI_NJ.pangenome.newick <span class="dt">\</span></span>
<span id="cb4-9"><a href="#cb4-9"></a>--name <span class="st">"ani_nj"</span> <span class="at">--config</span> project_config.yaml <span class="at">--save_leaf_order</span></span>
<span id="cb4-10"><a href="#cb4-10"></a></span>
<span id="cb4-11"><a href="#cb4-11"></a><span class="co">## gene distance NJ tree</span></span>
<span id="cb4-12"><a href="#cb4-12"></a><span class="ex">./scripts/analysis/phylo_process_newicks.R</span> <span class="dt">\</span></span>
<span id="cb4-13"><a href="#cb4-13"></a>--tree ./data/pangenomes/pectobacterium.v2/pectobacterium.v2.DB/gene_classification.100.0/gene_distance.tree <span class="dt">\</span></span>
<span id="cb4-14"><a href="#cb4-14"></a>--name <span class="st">"gene_nj"</span> <span class="at">--config</span> project_config.yaml </span>
<span id="cb4-15"><a href="#cb4-15"></a></span>
<span id="cb4-16"><a href="#cb4-16"></a><span class="co">## kmer distance UPGMA tree</span></span>
<span id="cb4-17"><a href="#cb4-17"></a><span class="ex">./scripts/analysis/phylo_process_newicks.R</span> <span class="dt">\</span></span>
<span id="cb4-18"><a href="#cb4-18"></a>--tree ./data/pangenomes/pectobacterium.v2/pectobacterium.v2.DB/kmer_classification.100.0/kmer.upgma.newick <span class="dt">\</span></span>
<span id="cb4-19"><a href="#cb4-19"></a>--name <span class="st">"kmer_upgma"</span> <span class="at">--config</span> project_config.yaml <span class="at">--save_leaf_order</span></span>
<span id="cb4-20"><a href="#cb4-20"></a></span>
<span id="cb4-21"><a href="#cb4-21"></a><span class="co">## kmer distance NJ tree</span></span>
<span id="cb4-22"><a href="#cb4-22"></a><span class="ex">./scripts/analysis/phylo_process_newicks.R</span> <span class="dt">\</span></span>
<span id="cb4-23"><a href="#cb4-23"></a>--tree ./data/pangenomes/pectobacterium.v2/pectobacterium.v2.DB/kmer_classification.100.0/kmer.nj.newick <span class="dt">\</span></span>
<span id="cb4-24"><a href="#cb4-24"></a>--name <span class="st">"kmer_nj"</span> <span class="at">--config</span> project_config.yaml <span class="at">--save_leaf_order</span></span>
<span id="cb4-25"><a href="#cb4-25"></a></span>
<span id="cb4-26"><a href="#cb4-26"></a><span class="co">## core-SNP ML tree</span></span>
<span id="cb4-27"><a href="#cb4-27"></a><span class="ex">./scripts/analysis/phylo_process_newicks.R</span> <span class="dt">\</span></span>
<span id="cb4-28"><a href="#cb4-28"></a>--tree ./data/pangenomes/pectobacterium.v2/pectobacterium.v2.DB/core_snp_tree/informative.fasta.treefile <span class="dt">\</span></span>
<span id="cb4-29"><a href="#cb4-29"></a>--name <span class="st">"core_snp_ml"</span> <span class="at">--config</span> project_config.yaml <span class="at">--save_leaf_order</span> <span class="at">--bootstrap</span></span>
<span id="cb4-30"><a href="#cb4-30"></a></span>
<span id="cb4-31"><a href="#cb4-31"></a><span class="co">## process homology groups</span></span>
<span id="cb4-32"><a href="#cb4-32"></a><span class="ex">Rscript</span> scripts/analysis/homology_group_process.R </span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<p>Visualize core-SNP phylogenetic tree with metadata.</p>
<div class="sourceCode" id="cb5"><pre class="sourceCode numberSource bash number-lines code-with-copy"><code class="sourceCode bash"><span id="cb5-1"><a href="#cb5-1"></a><span class="ex">quarto</span> render scripts/analysis/phylo_viz_tree.qmd</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</section>
<section id="homology-groups-analysis" class="level2">
<h2 class="anchored" data-anchor-id="homology-groups-analysis">Homology groups analysis</h2>
<section id="go-enrichment-of-homology-groups-categories" class="level3">
<h3 class="anchored" data-anchor-id="go-enrichment-of-homology-groups-categories">GO enrichment of homology groups categories</h3>
<p>This script will use the pangenome <code>org.db</code> object (built <a href="../../scripts/notebooks/pangenome_construction.html#build-pangenome-org-db-object">here</a>) and perform GO enrichment using <code>BioConductor::topGO</code> package.</p>
<div class="sourceCode" id="cb6"><pre class="sourceCode numberSource bash number-lines code-with-copy"><code class="sourceCode bash"><span id="cb6-1"><a href="#cb6-1"></a><span class="ex">Rscript</span> scripts/analysis/HG_GO_enrichment.R</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</section>
<section id="summarize-homology-groups" class="level3">
<h3 class="anchored" data-anchor-id="summarize-homology-groups">Summarize homology groups</h3>
<div class="sourceCode" id="cb7"><pre class="sourceCode numberSource bash number-lines code-with-copy"><code class="sourceCode bash"><span id="cb7-1"><a href="#cb7-1"></a><span class="ex">quarto</span> render scripts/analysis/HG_summary.qmd </span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</section>
<section id="visualize-tandemly-mapped-homology-groups-sets-on-the-pangenome" class="level3">
<h3 class="anchored" data-anchor-id="visualize-tandemly-mapped-homology-groups-sets-on-the-pangenome">Visualize tandemly mapped homology groups sets on the pangenome</h3>
<div class="sourceCode" id="cb8"><pre class="sourceCode numberSource bash number-lines code-with-copy"><code class="sourceCode bash"><span id="cb8-1"><a href="#cb8-1"></a><span class="ex">quarto</span> render scripts/analysis/HG_tandem_match_viz.qmd <span class="at">--execute-dir</span> ./</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</section>
<section id="visualize-homology-groups-across-pagenome-for-go-terms-of-interest" class="level3">
<h3 class="anchored" data-anchor-id="visualize-homology-groups-across-pagenome-for-go-terms-of-interest">Visualize homology groups across pagenome for GO terms of interest</h3>
<p>to-do: convert this to script with arguments</p>
<div class="sourceCode" id="cb9"><pre class="sourceCode numberSource bash number-lines code-with-copy"><code class="sourceCode bash"><span id="cb9-1"><a href="#cb9-1"></a><span class="ex">quarto</span> render scripts/analysis/homology_groups_go_viz.qmd <span class="at">--execute-dir</span> ./</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<section id="viral-dna-integration-related-biological-processes" class="level4">
<h4 class="anchored" data-anchor-id="viral-dna-integration-related-biological-processes">viral DNA integration related biological processes</h4>
<div class="sourceCode" id="cb10"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb10-1"><a href="#cb10-1"></a>analysisName <span class="ot">&lt;-</span> <span class="st">"virulence_GO"</span></span>
<span id="cb10-2"><a href="#cb10-2"></a></span>
<span id="cb10-3"><a href="#cb10-3"></a>goIds <span class="ot">&lt;-</span> <span class="fu">c</span>(</span>
<span id="cb10-4"><a href="#cb10-4"></a>  <span class="st">"GO:0046718"</span>, <span class="st">"GO:0044826"</span>, <span class="st">"GO:0019069"</span>, <span class="st">"GO:0044659"</span>, <span class="st">"GO:0075713"</span>,</span>
<span id="cb10-5"><a href="#cb10-5"></a>  <span class="st">"GO:0044660"</span>, <span class="st">"GO:0019076"</span>, <span class="st">"GO:0019058"</span></span>
<span id="cb10-6"><a href="#cb10-6"></a>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</section>
<section id="secretion-systems-in-pectobacterium-spp." class="level4">
<h4 class="anchored" data-anchor-id="secretion-systems-in-pectobacterium-spp.">secretion systems in <em>Pectobacterium</em> spp.</h4>
<div class="sourceCode" id="cb11"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb11-1"><a href="#cb11-1"></a>analysisName <span class="ot">&lt;-</span> <span class="st">"secretion_sys"</span></span>
<span id="cb11-2"><a href="#cb11-2"></a></span>
<span id="cb11-3"><a href="#cb11-3"></a>goIds <span class="ot">&lt;-</span> <span class="fu">c</span>(</span>
<span id="cb11-4"><a href="#cb11-4"></a>  <span class="st">"GO:0030253"</span>, <span class="st">"GO:0015628"</span>, <span class="st">"GO:0030254"</span>, <span class="st">"GO:0030255"</span>, <span class="st">"GO:0046819"</span>,</span>
<span id="cb11-5"><a href="#cb11-5"></a>  <span class="st">"GO:0033103"</span>, <span class="st">"GO:0044315"</span></span>
<span id="cb11-6"><a href="#cb11-6"></a>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</section>
<section id="homology-groups-for-lipopolysaccharide-biosynthetic-process" class="level4">
<h4 class="anchored" data-anchor-id="homology-groups-for-lipopolysaccharide-biosynthetic-process">Homology groups for <code>lipopolysaccharide biosynthetic process</code></h4>
<div class="sourceCode" id="cb12"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb12-1"><a href="#cb12-1"></a>analysisName <span class="ot">&lt;-</span> <span class="st">"lps_syn"</span></span>
<span id="cb12-2"><a href="#cb12-2"></a></span>
<span id="cb12-3"><a href="#cb12-3"></a>goIds <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">"GO:0009103"</span>, <span class="st">"GO:0009244"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>


<!-- -->

</section>
</section>
</section>
</section>

<a onclick="window.scrollTo(0, 0); return false;" role="button" id="quarto-back-to-top"><i class="bi bi-arrow-up"></i> Back to top</a></main> <!-- /main -->
<script id="quarto-html-after-body" type="application/javascript">
window.document.addEventListener("DOMContentLoaded", function (event) {
  const toggleBodyColorMode = (bsSheetEl) => {
    const mode = bsSheetEl.getAttribute("data-mode");
    const bodyEl = window.document.querySelector("body");
    if (mode === "dark") {
      bodyEl.classList.add("quarto-dark");
      bodyEl.classList.remove("quarto-light");
    } else {
      bodyEl.classList.add("quarto-light");
      bodyEl.classList.remove("quarto-dark");
    }
  }
  const toggleBodyColorPrimary = () => {
    const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
    if (bsSheetEl) {
      toggleBodyColorMode(bsSheetEl);
    }
  }
  toggleBodyColorPrimary();  
  const icon = "";
  const anchorJS = new window.AnchorJS();
  anchorJS.options = {
    placement: 'right',
    icon: icon
  };
  anchorJS.add('.anchored');
  const isCodeAnnotation = (el) => {
    for (const clz of el.classList) {
      if (clz.startsWith('code-annotation-')) {                     
        return true;
      }
    }
    return false;
  }
  const onCopySuccess = function(e) {
    // button target
    const button = e.trigger;
    // don't keep focus
    button.blur();
    // flash "checked"
    button.classList.add('code-copy-button-checked');
    var currentTitle = button.getAttribute("title");
    button.setAttribute("title", "Copied!");
    let tooltip;
    if (window.bootstrap) {
      button.setAttribute("data-bs-toggle", "tooltip");
      button.setAttribute("data-bs-placement", "left");
      button.setAttribute("data-bs-title", "Copied!");
      tooltip = new bootstrap.Tooltip(button, 
        { trigger: "manual", 
          customClass: "code-copy-button-tooltip",
          offset: [0, -8]});
      tooltip.show();    
    }
    setTimeout(function() {
      if (tooltip) {
        tooltip.hide();
        button.removeAttribute("data-bs-title");
        button.removeAttribute("data-bs-toggle");
        button.removeAttribute("data-bs-placement");
      }
      button.setAttribute("title", currentTitle);
      button.classList.remove('code-copy-button-checked');
    }, 1000);
    // clear code selection
    e.clearSelection();
  }
  const getTextToCopy = function(trigger) {
      const codeEl = trigger.previousElementSibling.cloneNode(true);
      for (const childEl of codeEl.children) {
        if (isCodeAnnotation(childEl)) {
          childEl.remove();
        }
      }
      return codeEl.innerText;
  }
  const clipboard = new window.ClipboardJS('.code-copy-button:not([data-in-quarto-modal])', {
    text: getTextToCopy
  });
  clipboard.on('success', onCopySuccess);
  if (window.document.getElementById('quarto-embedded-source-code-modal')) {
    // For code content inside modals, clipBoardJS needs to be initialized with a container option
    // TODO: Check when it could be a function (https://github.com/zenorocha/clipboard.js/issues/860)
    const clipboardModal = new window.ClipboardJS('.code-copy-button[data-in-quarto-modal]', {
      text: getTextToCopy,
      container: window.document.getElementById('quarto-embedded-source-code-modal')
    });
    clipboardModal.on('success', onCopySuccess);
  }
  const viewSource = window.document.getElementById('quarto-view-source') ||
                     window.document.getElementById('quarto-code-tools-source');
  if (viewSource) {
    const sourceUrl = viewSource.getAttribute("data-quarto-source-url");
    viewSource.addEventListener("click", function(e) {
      if (sourceUrl) {
        // rstudio viewer pane
        if (/\bcapabilities=\b/.test(window.location)) {
          window.open(sourceUrl);
        } else {
          window.location.href = sourceUrl;
        }
      } else {
        const modal = new bootstrap.Modal(document.getElementById('quarto-embedded-source-code-modal'));
        modal.show();
      }
      return false;
    });
  }
  function toggleCodeHandler(show) {
    return function(e) {
      const detailsSrc = window.document.querySelectorAll(".cell > details > .sourceCode");
      for (let i=0; i<detailsSrc.length; i++) {
        const details = detailsSrc[i].parentElement;
        if (show) {
          details.open = true;
        } else {
          details.removeAttribute("open");
        }
      }
      const cellCodeDivs = window.document.querySelectorAll(".cell > .sourceCode");
      const fromCls = show ? "hidden" : "unhidden";
      const toCls = show ? "unhidden" : "hidden";
      for (let i=0; i<cellCodeDivs.length; i++) {
        const codeDiv = cellCodeDivs[i];
        if (codeDiv.classList.contains(fromCls)) {
          codeDiv.classList.remove(fromCls);
          codeDiv.classList.add(toCls);
        } 
      }
      return false;
    }
  }
  const hideAllCode = window.document.getElementById("quarto-hide-all-code");
  if (hideAllCode) {
    hideAllCode.addEventListener("click", toggleCodeHandler(false));
  }
  const showAllCode = window.document.getElementById("quarto-show-all-code");
  if (showAllCode) {
    showAllCode.addEventListener("click", toggleCodeHandler(true));
  }
    var localhostRegex = new RegExp(/^(?:http|https):\/\/localhost\:?[0-9]*\//);
    var mailtoRegex = new RegExp(/^mailto:/);
      var filterRegex = new RegExp('/' + window.location.host + '/');
    var isInternal = (href) => {
        return filterRegex.test(href) || localhostRegex.test(href) || mailtoRegex.test(href);
    }
    // Inspect non-navigation links and adorn them if external
 	var links = window.document.querySelectorAll('a[href]:not(.nav-link):not(.navbar-brand):not(.toc-action):not(.sidebar-link):not(.sidebar-item-toggle):not(.pagination-link):not(.no-external):not([aria-hidden]):not(.dropdown-item):not(.quarto-navigation-tool):not(.about-link)');
    for (var i=0; i<links.length; i++) {
      const link = links[i];
      if (!isInternal(link.href)) {
        // undo the damage that might have been done by quarto-nav.js in the case of
        // links that we want to consider external
        if (link.dataset.originalHref !== undefined) {
          link.href = link.dataset.originalHref;
        }
      }
    }
  function tippyHover(el, contentFn, onTriggerFn, onUntriggerFn) {
    const config = {
      allowHTML: true,
      maxWidth: 500,
      delay: 100,
      arrow: false,
      appendTo: function(el) {
          return el.parentElement;
      },
      interactive: true,
      interactiveBorder: 10,
      theme: 'quarto',
      placement: 'bottom-start',
    };
    if (contentFn) {
      config.content = contentFn;
    }
    if (onTriggerFn) {
      config.onTrigger = onTriggerFn;
    }
    if (onUntriggerFn) {
      config.onUntrigger = onUntriggerFn;
    }
    window.tippy(el, config); 
  }
  const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
  for (var i=0; i<noterefs.length; i++) {
    const ref = noterefs[i];
    tippyHover(ref, function() {
      // use id or data attribute instead here
      let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
      try { href = new URL(href).hash; } catch {}
      const id = href.replace(/^#\/?/, "");
      const note = window.document.getElementById(id);
      if (note) {
        return note.innerHTML;
      } else {
        return "";
      }
    });
  }
  const xrefs = window.document.querySelectorAll('a.quarto-xref');
  const processXRef = (id, note) => {
    // Strip column container classes
    const stripColumnClz = (el) => {
      el.classList.remove("page-full", "page-columns");
      if (el.children) {
        for (const child of el.children) {
          stripColumnClz(child);
        }
      }
    }
    stripColumnClz(note)
    if (id === null || id.startsWith('sec-')) {
      // Special case sections, only their first couple elements
      const container = document.createElement("div");
      if (note.children && note.children.length > 2) {
        container.appendChild(note.children[0].cloneNode(true));
        for (let i = 1; i < note.children.length; i++) {
          const child = note.children[i];
          if (child.tagName === "P" && child.innerText === "") {
            continue;
          } else {
            container.appendChild(child.cloneNode(true));
            break;
          }
        }
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(container);
        }
        return container.innerHTML
      } else {
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(note);
        }
        return note.innerHTML;
      }
    } else {
      // Remove any anchor links if they are present
      const anchorLink = note.querySelector('a.anchorjs-link');
      if (anchorLink) {
        anchorLink.remove();
      }
      if (window.Quarto?.typesetMath) {
        window.Quarto.typesetMath(note);
      }
      // TODO in 1.5, we should make sure this works without a callout special case
      if (note.classList.contains("callout")) {
        return note.outerHTML;
      } else {
        return note.innerHTML;
      }
    }
  }
  for (var i=0; i<xrefs.length; i++) {
    const xref = xrefs[i];
    tippyHover(xref, undefined, function(instance) {
      instance.disable();
      let url = xref.getAttribute('href');
      let hash = undefined; 
      if (url.startsWith('#')) {
        hash = url;
      } else {
        try { hash = new URL(url).hash; } catch {}
      }
      if (hash) {
        const id = hash.replace(/^#\/?/, "");
        const note = window.document.getElementById(id);
        if (note !== null) {
          try {
            const html = processXRef(id, note.cloneNode(true));
            instance.setContent(html);
          } finally {
            instance.enable();
            instance.show();
          }
        } else {
          // See if we can fetch this
          fetch(url.split('#')[0])
          .then(res => res.text())
          .then(html => {
            const parser = new DOMParser();
            const htmlDoc = parser.parseFromString(html, "text/html");
            const note = htmlDoc.getElementById(id);
            if (note !== null) {
              const html = processXRef(id, note);
              instance.setContent(html);
            } 
          }).finally(() => {
            instance.enable();
            instance.show();
          });
        }
      } else {
        // See if we can fetch a full url (with no hash to target)
        // This is a special case and we should probably do some content thinning / targeting
        fetch(url)
        .then(res => res.text())
        .then(html => {
          const parser = new DOMParser();
          const htmlDoc = parser.parseFromString(html, "text/html");
          const note = htmlDoc.querySelector('main.content');
          if (note !== null) {
            // This should only happen for chapter cross references
            // (since there is no id in the URL)
            // remove the first header
            if (note.children.length > 0 && note.children[0].tagName === "HEADER") {
              note.children[0].remove();
            }
            const html = processXRef(null, note);
            instance.setContent(html);
          } 
        }).finally(() => {
          instance.enable();
          instance.show();
        });
      }
    }, function(instance) {
    });
  }
      let selectedAnnoteEl;
      const selectorForAnnotation = ( cell, annotation) => {
        let cellAttr = 'data-code-cell="' + cell + '"';
        let lineAttr = 'data-code-annotation="' +  annotation + '"';
        const selector = 'span[' + cellAttr + '][' + lineAttr + ']';
        return selector;
      }
      const selectCodeLines = (annoteEl) => {
        const doc = window.document;
        const targetCell = annoteEl.getAttribute("data-target-cell");
        const targetAnnotation = annoteEl.getAttribute("data-target-annotation");
        const annoteSpan = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
        const lines = annoteSpan.getAttribute("data-code-lines").split(",");
        const lineIds = lines.map((line) => {
          return targetCell + "-" + line;
        })
        let top = null;
        let height = null;
        let parent = null;
        if (lineIds.length > 0) {
            //compute the position of the single el (top and bottom and make a div)
            const el = window.document.getElementById(lineIds[0]);
            top = el.offsetTop;
            height = el.offsetHeight;
            parent = el.parentElement.parentElement;
          if (lineIds.length > 1) {
            const lastEl = window.document.getElementById(lineIds[lineIds.length - 1]);
            const bottom = lastEl.offsetTop + lastEl.offsetHeight;
            height = bottom - top;
          }
          if (top !== null && height !== null && parent !== null) {
            // cook up a div (if necessary) and position it 
            let div = window.document.getElementById("code-annotation-line-highlight");
            if (div === null) {
              div = window.document.createElement("div");
              div.setAttribute("id", "code-annotation-line-highlight");
              div.style.position = 'absolute';
              parent.appendChild(div);
            }
            div.style.top = top - 2 + "px";
            div.style.height = height + 4 + "px";
            div.style.left = 0;
            let gutterDiv = window.document.getElementById("code-annotation-line-highlight-gutter");
            if (gutterDiv === null) {
              gutterDiv = window.document.createElement("div");
              gutterDiv.setAttribute("id", "code-annotation-line-highlight-gutter");
              gutterDiv.style.position = 'absolute';
              const codeCell = window.document.getElementById(targetCell);
              const gutter = codeCell.querySelector('.code-annotation-gutter');
              gutter.appendChild(gutterDiv);
            }
            gutterDiv.style.top = top - 2 + "px";
            gutterDiv.style.height = height + 4 + "px";
          }
          selectedAnnoteEl = annoteEl;
        }
      };
      const unselectCodeLines = () => {
        const elementsIds = ["code-annotation-line-highlight", "code-annotation-line-highlight-gutter"];
        elementsIds.forEach((elId) => {
          const div = window.document.getElementById(elId);
          if (div) {
            div.remove();
          }
        });
        selectedAnnoteEl = undefined;
      };
        // Handle positioning of the toggle
    window.addEventListener(
      "resize",
      throttle(() => {
        elRect = undefined;
        if (selectedAnnoteEl) {
          selectCodeLines(selectedAnnoteEl);
        }
      }, 10)
    );
    function throttle(fn, ms) {
    let throttle = false;
    let timer;
      return (...args) => {
        if(!throttle) { // first call gets through
            fn.apply(this, args);
            throttle = true;
        } else { // all the others get throttled
            if(timer) clearTimeout(timer); // cancel #2
            timer = setTimeout(() => {
              fn.apply(this, args);
              timer = throttle = false;
            }, ms);
        }
      };
    }
      // Attach click handler to the DT
      const annoteDls = window.document.querySelectorAll('dt[data-target-cell]');
      for (const annoteDlNode of annoteDls) {
        annoteDlNode.addEventListener('click', (event) => {
          const clickedEl = event.target;
          if (clickedEl !== selectedAnnoteEl) {
            unselectCodeLines();
            const activeEl = window.document.querySelector('dt[data-target-cell].code-annotation-active');
            if (activeEl) {
              activeEl.classList.remove('code-annotation-active');
            }
            selectCodeLines(clickedEl);
            clickedEl.classList.add('code-annotation-active');
          } else {
            // Unselect the line
            unselectCodeLines();
            clickedEl.classList.remove('code-annotation-active');
          }
        });
      }
  const findCites = (el) => {
    const parentEl = el.parentElement;
    if (parentEl) {
      const cites = parentEl.dataset.cites;
      if (cites) {
        return {
          el,
          cites: cites.split(' ')
        };
      } else {
        return findCites(el.parentElement)
      }
    } else {
      return undefined;
    }
  };
  var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
  for (var i=0; i<bibliorefs.length; i++) {
    const ref = bibliorefs[i];
    const citeInfo = findCites(ref);
    if (citeInfo) {
      tippyHover(citeInfo.el, function() {
        var popup = window.document.createElement('div');
        citeInfo.cites.forEach(function(cite) {
          var citeDiv = window.document.createElement('div');
          citeDiv.classList.add('hanging-indent');
          citeDiv.classList.add('csl-entry');
          var biblioDiv = window.document.getElementById('ref-' + cite);
          if (biblioDiv) {
            citeDiv.innerHTML = biblioDiv.innerHTML;
          }
          popup.appendChild(citeDiv);
        });
        return popup.innerHTML;
      });
    }
  }
});
</script>
<nav class="page-navigation">
  <div class="nav-page nav-page-previous">
      <a href="../../scripts/notebooks/pangenome_construction.html" class="pagination-link" aria-label="Build">
        <i class="bi bi-arrow-left-short"></i> <span class="nav-page-text">Build</span>
      </a>          
  </div>
  <div class="nav-page nav-page-next">
      <a href="../../scripts/analysis/HG_summary.html" class="pagination-link" aria-label="Pangenome summary">
        <span class="nav-page-text">Pangenome summary</span> <i class="bi bi-arrow-right-short"></i>
      </a>
  </div>
</nav><div class="modal fade" id="quarto-embedded-source-code-modal" tabindex="-1" aria-labelledby="quarto-embedded-source-code-modal-label" aria-hidden="true"><div class="modal-dialog modal-dialog-scrollable"><div class="modal-content"><div class="modal-header"><h5 class="modal-title" id="quarto-embedded-source-code-modal-label">Source Code</h5><button class="btn-close" data-bs-dismiss="modal"></button></div><div class="modal-body"><div class="">
<div class="sourceCode" id="cb13" data-shortcodes="false"><pre class="sourceCode numberSource markdown number-lines code-with-copy"><code class="sourceCode markdown"><span id="cb13-1"><a href="#cb13-1"></a><span class="fu"># Downstream analysis of pangenome data</span></span>
<span id="cb13-2"><a href="#cb13-2"></a></span>
<span id="cb13-3"><a href="#cb13-3"></a><span class="fu">## Setup</span></span>
<span id="cb13-4"><a href="#cb13-4"></a></span>
<span id="cb13-5"><a href="#cb13-5"></a><span class="in">``` bash</span></span>
<span id="cb13-6"><a href="#cb13-6"></a><span class="co">#!/usr/bin/env bash</span></span>
<span id="cb13-7"><a href="#cb13-7"></a></span>
<span id="cb13-8"><a href="#cb13-8"></a><span class="bu">shopt</span> <span class="at">-s</span> expand_aliases</span>
<span id="cb13-9"><a href="#cb13-9"></a><span class="bu">source</span> ~/.bash_aliases</span>
<span id="cb13-10"><a href="#cb13-10"></a></span>
<span id="cb13-11"><a href="#cb13-11"></a><span class="co"># set -e</span></span>
<span id="cb13-12"><a href="#cb13-12"></a><span class="co"># set -u</span></span>
<span id="cb13-13"><a href="#cb13-13"></a><span class="bu">set</span> <span class="at">-o</span> pipefail</span>
<span id="cb13-14"><a href="#cb13-14"></a></span>
<span id="cb13-15"><a href="#cb13-15"></a><span class="co"># source scripts/utils/setup_analysis.sh 'pectobacterium.v2'</span></span>
<span id="cb13-16"><a href="#cb13-16"></a><span class="bu">source</span> scripts/utils/setup_analysis.sh <span class="va">$@</span></span>
<span id="cb13-17"><a href="#cb13-17"></a></span>
<span id="cb13-18"><a href="#cb13-18"></a><span class="cf">if</span> <span class="bu">[</span> <span class="ot">-z</span> <span class="va">${pan_db</span><span class="op">+</span>x<span class="va">}</span> <span class="bu">]</span><span class="kw">;</span></span>
<span id="cb13-19"><a href="#cb13-19"></a><span class="cf">then</span></span>
<span id="cb13-20"><a href="#cb13-20"></a>    <span class="bu">echo</span> <span class="st">"</span><span class="dt">\$</span><span class="st">pan_db is unset"</span></span>
<span id="cb13-21"><a href="#cb13-21"></a>    <span class="ex">error_exit</span> 1</span>
<span id="cb13-22"><a href="#cb13-22"></a><span class="cf">fi</span></span>
<span id="cb13-23"><a href="#cb13-23"></a></span>
<span id="cb13-24"><a href="#cb13-24"></a><span class="bu">source</span> <span class="va">$TOOLS_PATH</span>/miniconda3/etc/profile.d/conda.sh</span>
<span id="cb13-25"><a href="#cb13-25"></a><span class="ex">conda</span> activate pantools_master</span>
<span id="cb13-26"><a href="#cb13-26"></a><span class="bu">export</span> <span class="va">PANTOOLS</span><span class="op">=</span><span class="st">"</span><span class="va">$PANTOOLS_4_1</span><span class="st">"</span></span>
<span id="cb13-27"><a href="#cb13-27"></a></span>
<span id="cb13-28"><a href="#cb13-28"></a><span class="co">######################################################################</span></span>
<span id="cb13-29"><a href="#cb13-29"></a><span class="kw">```</span></span>
<span id="cb13-30"><a href="#cb13-30"></a></span>
<span id="cb13-31"><a href="#cb13-31"></a><span class="co">## Use subset of genomes to determine pangenome structure</span></span>
<span id="cb13-32"><a href="#cb13-32"></a></span>
<span id="cb13-33"><a href="#cb13-33"></a><span class="kw">```</span> bash</span>
<span id="cb13-34"><a href="#cb13-34"></a><span class="co">## Pangenome structure for species with more than 20 genomes</span></span>
<span id="cb13-35"><a href="#cb13-35"></a><span class="cf">for</span> sp <span class="kw">in</span> <span class="kw">`</span><span class="fu">awk</span> <span class="at">-F</span> <span class="st">"\t"</span> <span class="st">'{ if (NR != 1 &amp;&amp; $2 &gt;= 20) {print $1} }'</span> <span class="va">$PANGENOME_DIR</span>/analysis_configs/species_wise_genomes.tab<span class="kw">`</span></span>
<span id="cb13-36"><a href="#cb13-36"></a><span class="cf">do</span></span>
<span id="cb13-37"><a href="#cb13-37"></a>    <span class="va">genomes</span><span class="op">=</span><span class="kw">`</span><span class="fu">grep</span> <span class="st">"^</span><span class="va">${sp}</span><span class="st">\b"</span> <span class="va">$PANGENOME_DIR</span>/analysis_configs/species_wise_genomes.tab <span class="kw">|</span> <span class="fu">cut</span> <span class="at">-f3</span><span class="kw">`</span></span>
<span id="cb13-38"><a href="#cb13-38"></a>    </span>
<span id="cb13-39"><a href="#cb13-39"></a>    <span class="ex">process_start</span> <span class="st">"pangenome_structure for genome of species </span><span class="va">${sp}</span><span class="st">: </span><span class="va">$genomes</span><span class="st">"</span></span>
<span id="cb13-40"><a href="#cb13-40"></a>    <span class="va">$PANTOOLS</span> pangenome_structure <span class="at">-t</span> 20 <span class="at">--include</span> <span class="va">${genomes}</span> <span class="va">${pan_db}</span></span>
<span id="cb13-41"><a href="#cb13-41"></a>    <span class="ex">error_exit</span> <span class="va">$?</span></span>
<span id="cb13-42"><a href="#cb13-42"></a></span>
<span id="cb13-43"><a href="#cb13-43"></a>    <span class="ex">Rscript</span> <span class="va">${pan_db}</span>/pangenome_size/gene/pangenome_growth.R</span>
<span id="cb13-44"><a href="#cb13-44"></a>    <span class="ex">Rscript</span> <span class="va">${pan_db}</span>/pangenome_size/gene/gains_losses_median_or_average.R</span>
<span id="cb13-45"><a href="#cb13-45"></a>    <span class="ex">Rscript</span> <span class="va">${pan_db}</span>/pangenome_size/gene/gains_losses_median_and_average.R</span>
<span id="cb13-46"><a href="#cb13-46"></a>    <span class="co"># Rscript ${pan_db}/pangenome_size/gene/heaps_law.R</span></span>
<span id="cb13-47"><a href="#cb13-47"></a></span>
<span id="cb13-48"><a href="#cb13-48"></a>    <span class="co">## move results to a folder</span></span>
<span id="cb13-49"><a href="#cb13-49"></a>    <span class="va">psDir</span><span class="op">=</span><span class="va">${pan_db}</span>/pangenome_size/gene.<span class="va">${sp}</span></span>
<span id="cb13-50"><a href="#cb13-50"></a>    <span class="bu">[</span> <span class="ot">-d</span> <span class="va">${psDir}</span> <span class="bu">]</span> <span class="kw">&amp;&amp;</span> <span class="fu">rm</span> <span class="at">-r</span> <span class="va">${psDir}</span></span>
<span id="cb13-51"><a href="#cb13-51"></a>    <span class="fu">mv</span> <span class="va">${pan_db}</span>/pangenome_size/gene <span class="va">${psDir}</span></span>
<span id="cb13-52"><a href="#cb13-52"></a><span class="cf">done</span></span>
<span id="cb13-53"><a href="#cb13-53"></a></span>
<span id="cb13-54"><a href="#cb13-54"></a><span class="co">## update the path in heaps_law.R script</span></span>
<span id="cb13-55"><a href="#cb13-55"></a><span class="bu">cd</span> <span class="va">${pan_db}</span>/pangenome_size</span>
<span id="cb13-56"><a href="#cb13-56"></a><span class="cf">for</span> i <span class="kw">in</span> gene.<span class="pp">*</span></span>
<span id="cb13-57"><a href="#cb13-57"></a><span class="cf">do</span></span>
<span id="cb13-58"><a href="#cb13-58"></a>    <span class="fu">sed</span> <span class="at">-i.bak</span> <span class="st">"s/pangenome_size\/gene/pangenome_size\/gene.</span><span class="va">$i</span><span class="st">/g"</span> <span class="va">$i</span>/heaps_law.R</span>
<span id="cb13-59"><a href="#cb13-59"></a><span class="cf">done</span></span>
<span id="cb13-60"><a href="#cb13-60"></a><span class="bu">cd</span> <span class="va">$PROJECT_DIR</span></span>
<span id="cb13-61"><a href="#cb13-61"></a></span>
<span id="cb13-62"><a href="#cb13-62"></a><span class="kw">```</span></span>
<span id="cb13-63"><a href="#cb13-63"></a></span>
<span id="cb13-64"><a href="#cb13-64"></a><span class="co">## Visualize pangenome metadata</span></span>
<span id="cb13-65"><a href="#cb13-65"></a></span>
<span id="cb13-66"><a href="#cb13-66"></a><span class="kw">```</span>bash</span>
<span id="cb13-67"><a href="#cb13-67"></a><span class="ex">quarto</span> render scripts/analysis/pangenome_metadata.viz.qmd</span>
<span id="cb13-68"><a href="#cb13-68"></a><span class="kw">```</span></span>
<span id="cb13-69"><a href="#cb13-69"></a></span>
<span id="cb13-70"><a href="#cb13-70"></a><span class="co">## Phylogeny analysis </span></span>
<span id="cb13-71"><a href="#cb13-71"></a></span>
<span id="cb13-72"><a href="#cb13-72"></a><span class="kw">```</span> bash</span>
<span id="cb13-73"><a href="#cb13-73"></a><span class="co">## UPGMA tree</span></span>
<span id="cb13-74"><a href="#cb13-74"></a><span class="ex">./scripts/analysis/phylo_process_newicks.R</span> <span class="dt">\</span></span>
<span id="cb13-75"><a href="#cb13-75"></a>--tree ./analysis/pangenome_v2/phylogeny/ani/ANI_UPGMA.pangenome.newick <span class="dt">\</span></span>
<span id="cb13-76"><a href="#cb13-76"></a>--name <span class="st">"ani_upgma"</span> <span class="at">--config</span> project_config.yaml <span class="at">--save_leaf_order</span></span>
<span id="cb13-77"><a href="#cb13-77"></a></span>
<span id="cb13-78"><a href="#cb13-78"></a><span class="co">## Neighbor joining tree</span></span>
<span id="cb13-79"><a href="#cb13-79"></a><span class="ex">./scripts/analysis/phylo_process_newicks.R</span> <span class="dt">\</span></span>
<span id="cb13-80"><a href="#cb13-80"></a>--tree ./analysis/pangenome_v2/phylogeny/ani/ANI_NJ.pangenome.newick <span class="dt">\</span></span>
<span id="cb13-81"><a href="#cb13-81"></a>--name <span class="st">"ani_nj"</span> <span class="at">--config</span> project_config.yaml <span class="at">--save_leaf_order</span></span>
<span id="cb13-82"><a href="#cb13-82"></a></span>
<span id="cb13-83"><a href="#cb13-83"></a><span class="co">## gene distance NJ tree</span></span>
<span id="cb13-84"><a href="#cb13-84"></a><span class="ex">./scripts/analysis/phylo_process_newicks.R</span> <span class="dt">\</span></span>
<span id="cb13-85"><a href="#cb13-85"></a>--tree ./data/pangenomes/pectobacterium.v2/pectobacterium.v2.DB/gene_classification.100.0/gene_distance.tree <span class="dt">\</span></span>
<span id="cb13-86"><a href="#cb13-86"></a>--name <span class="st">"gene_nj"</span> <span class="at">--config</span> project_config.yaml </span>
<span id="cb13-87"><a href="#cb13-87"></a></span>
<span id="cb13-88"><a href="#cb13-88"></a><span class="co">## kmer distance UPGMA tree</span></span>
<span id="cb13-89"><a href="#cb13-89"></a><span class="ex">./scripts/analysis/phylo_process_newicks.R</span> <span class="dt">\</span></span>
<span id="cb13-90"><a href="#cb13-90"></a>--tree ./data/pangenomes/pectobacterium.v2/pectobacterium.v2.DB/kmer_classification.100.0/kmer.upgma.newick <span class="dt">\</span></span>
<span id="cb13-91"><a href="#cb13-91"></a>--name <span class="st">"kmer_upgma"</span> <span class="at">--config</span> project_config.yaml <span class="at">--save_leaf_order</span></span>
<span id="cb13-92"><a href="#cb13-92"></a></span>
<span id="cb13-93"><a href="#cb13-93"></a><span class="co">## kmer distance NJ tree</span></span>
<span id="cb13-94"><a href="#cb13-94"></a><span class="ex">./scripts/analysis/phylo_process_newicks.R</span> <span class="dt">\</span></span>
<span id="cb13-95"><a href="#cb13-95"></a>--tree ./data/pangenomes/pectobacterium.v2/pectobacterium.v2.DB/kmer_classification.100.0/kmer.nj.newick <span class="dt">\</span></span>
<span id="cb13-96"><a href="#cb13-96"></a>--name <span class="st">"kmer_nj"</span> <span class="at">--config</span> project_config.yaml <span class="at">--save_leaf_order</span></span>
<span id="cb13-97"><a href="#cb13-97"></a></span>
<span id="cb13-98"><a href="#cb13-98"></a><span class="co">## core-SNP ML tree</span></span>
<span id="cb13-99"><a href="#cb13-99"></a><span class="ex">./scripts/analysis/phylo_process_newicks.R</span> <span class="dt">\</span></span>
<span id="cb13-100"><a href="#cb13-100"></a>--tree ./data/pangenomes/pectobacterium.v2/pectobacterium.v2.DB/core_snp_tree/informative.fasta.treefile <span class="dt">\</span></span>
<span id="cb13-101"><a href="#cb13-101"></a>--name <span class="st">"core_snp_ml"</span> <span class="at">--config</span> project_config.yaml <span class="at">--save_leaf_order</span> <span class="at">--bootstrap</span></span>
<span id="cb13-102"><a href="#cb13-102"></a></span>
<span id="cb13-103"><a href="#cb13-103"></a><span class="co">## process homology groups</span></span>
<span id="cb13-104"><a href="#cb13-104"></a><span class="ex">Rscript</span> scripts/analysis/homology_group_process.R </span>
<span id="cb13-105"><a href="#cb13-105"></a></span>
<span id="cb13-106"><a href="#cb13-106"></a><span class="kw">```</span></span>
<span id="cb13-107"><a href="#cb13-107"></a></span>
<span id="cb13-108"><a href="#cb13-108"></a><span class="ex">Visualize</span> core-SNP phylogenetic tree with metadata.</span>
<span id="cb13-109"><a href="#cb13-109"></a></span>
<span id="cb13-110"><a href="#cb13-110"></a><span class="kw">```</span>bash</span>
<span id="cb13-111"><a href="#cb13-111"></a><span class="ex">quarto</span> render scripts/analysis/phylo_viz_tree.qmd</span>
<span id="cb13-112"><a href="#cb13-112"></a><span class="kw">```</span></span>
<span id="cb13-113"><a href="#cb13-113"></a></span>
<span id="cb13-114"><a href="#cb13-114"></a><span class="co">## Homology groups analysis</span></span>
<span id="cb13-115"><a href="#cb13-115"></a></span>
<span id="cb13-116"><a href="#cb13-116"></a><span class="co">### GO enrichment of homology groups categories</span></span>
<span id="cb13-117"><a href="#cb13-117"></a></span>
<span id="cb13-118"><a href="#cb13-118"></a><span class="ex">This</span> script will use the pangenome <span class="kw">`</span>org.db<span class="kw">`</span> <span class="ex">object</span> <span class="er">(</span><span class="ex">built</span> <span class="pp">[</span><span class="ss">here</span><span class="pp">]</span><span class="er">(</span><span class="ex">pangenome_construction.md#build-pangenome-org-db-object</span><span class="kw">))</span></span>
<span id="cb13-119"><a href="#cb13-119"></a><span class="ex">and</span> perform GO enrichment using <span class="kw">`</span>BioConductor::topGO<span class="kw">`</span> <span class="ex">package.</span></span>
<span id="cb13-120"><a href="#cb13-120"></a></span>
<span id="cb13-121"><a href="#cb13-121"></a><span class="kw">```</span>bash</span>
<span id="cb13-122"><a href="#cb13-122"></a><span class="ex">Rscript</span> scripts/analysis/HG_GO_enrichment.R</span>
<span id="cb13-123"><a href="#cb13-123"></a><span class="kw">```</span></span>
<span id="cb13-124"><a href="#cb13-124"></a></span>
<span id="cb13-125"><a href="#cb13-125"></a><span class="co">### Summarize homology groups</span></span>
<span id="cb13-126"><a href="#cb13-126"></a></span>
<span id="cb13-127"><a href="#cb13-127"></a><span class="kw">```</span>bash</span>
<span id="cb13-128"><a href="#cb13-128"></a><span class="ex">quarto</span> render scripts/analysis/HG_summary.qmd </span>
<span id="cb13-129"><a href="#cb13-129"></a><span class="kw">```</span></span>
<span id="cb13-130"><a href="#cb13-130"></a></span>
<span id="cb13-131"><a href="#cb13-131"></a><span class="co">### Visualize tandemly mapped homology groups sets on the pangenome</span></span>
<span id="cb13-132"><a href="#cb13-132"></a></span>
<span id="cb13-133"><a href="#cb13-133"></a><span class="kw">```</span>bash</span>
<span id="cb13-134"><a href="#cb13-134"></a><span class="ex">quarto</span> render scripts/analysis/HG_tandem_match_viz.qmd <span class="at">--execute-dir</span> ./</span>
<span id="cb13-135"><a href="#cb13-135"></a><span class="kw">```</span></span>
<span id="cb13-136"><a href="#cb13-136"></a></span>
<span id="cb13-137"><a href="#cb13-137"></a><span class="co">### Visualize homology groups across pagenome for GO terms of interest</span></span>
<span id="cb13-138"><a href="#cb13-138"></a></span>
<span id="cb13-139"><a href="#cb13-139"></a><span class="ex">to-do:</span> convert this to script with arguments</span>
<span id="cb13-140"><a href="#cb13-140"></a></span>
<span id="cb13-141"><a href="#cb13-141"></a><span class="kw">```</span>bash</span>
<span id="cb13-142"><a href="#cb13-142"></a><span class="ex">quarto</span> render scripts/analysis/homology_groups_go_viz.qmd <span class="at">--execute-dir</span> ./</span>
<span id="cb13-143"><a href="#cb13-143"></a><span class="kw">```</span></span>
<span id="cb13-144"><a href="#cb13-144"></a></span>
<span id="cb13-145"><a href="#cb13-145"></a><span class="co">#### viral DNA integration related biological processes</span></span>
<span id="cb13-146"><a href="#cb13-146"></a></span>
<span id="cb13-147"><a href="#cb13-147"></a><span class="kw">```</span>r</span>
<span id="cb13-148"><a href="#cb13-148"></a><span class="ex">analysisName</span> <span class="op">&lt;</span>- <span class="st">"virulence_GO"</span></span>
<span id="cb13-149"><a href="#cb13-149"></a></span>
<span id="cb13-150"><a href="#cb13-150"></a><span class="ex">goIds</span> <span class="op">&lt;</span>- c<span class="er">(</span></span>
<span id="cb13-151"><a href="#cb13-151"></a>  <span class="st">"GO:0046718"</span><span class="ex">,</span> <span class="st">"GO:0044826"</span>, <span class="st">"GO:0019069"</span>, <span class="st">"GO:0044659"</span>, <span class="st">"GO:0075713"</span>,</span>
<span id="cb13-152"><a href="#cb13-152"></a>  <span class="st">"GO:0044660"</span><span class="ex">,</span> <span class="st">"GO:0019076"</span>, <span class="st">"GO:0019058"</span></span>
<span id="cb13-153"><a href="#cb13-153"></a><span class="kw">)</span></span>
<span id="cb13-154"><a href="#cb13-154"></a><span class="kw">```</span></span>
<span id="cb13-155"><a href="#cb13-155"></a></span>
<span id="cb13-156"><a href="#cb13-156"></a><span class="co">#### secretion systems in *Pectobacterium* spp.</span></span>
<span id="cb13-157"><a href="#cb13-157"></a></span>
<span id="cb13-158"><a href="#cb13-158"></a><span class="kw">```</span>r</span>
<span id="cb13-159"><a href="#cb13-159"></a><span class="ex">analysisName</span> <span class="op">&lt;</span>- <span class="st">"secretion_sys"</span></span>
<span id="cb13-160"><a href="#cb13-160"></a></span>
<span id="cb13-161"><a href="#cb13-161"></a><span class="ex">goIds</span> <span class="op">&lt;</span>- c<span class="er">(</span></span>
<span id="cb13-162"><a href="#cb13-162"></a>  <span class="st">"GO:0030253"</span><span class="ex">,</span> <span class="st">"GO:0015628"</span>, <span class="st">"GO:0030254"</span>, <span class="st">"GO:0030255"</span>, <span class="st">"GO:0046819"</span>,</span>
<span id="cb13-163"><a href="#cb13-163"></a>  <span class="st">"GO:0033103"</span><span class="ex">,</span> <span class="st">"GO:0044315"</span></span>
<span id="cb13-164"><a href="#cb13-164"></a><span class="kw">)</span></span>
<span id="cb13-165"><a href="#cb13-165"></a><span class="kw">```</span></span>
<span id="cb13-166"><a href="#cb13-166"></a></span>
<span id="cb13-167"><a href="#cb13-167"></a><span class="co">#### Homology groups for </span><span class="kw">`</span>lipopolysaccharide biosynthetic process<span class="kw">`</span></span>
<span id="cb13-168"><a href="#cb13-168"></a></span>
<span id="cb13-169"><a href="#cb13-169"></a><span class="kw">```</span>r</span>
<span id="cb13-170"><a href="#cb13-170"></a><span class="ex">analysisName</span> <span class="op">&lt;</span>- <span class="st">"lps_syn"</span></span>
<span id="cb13-171"><a href="#cb13-171"></a></span>
<span id="cb13-172"><a href="#cb13-172"></a><span class="ex">goIds</span> <span class="op">&lt;</span>- c<span class="er">(</span><span class="st">"GO:0009103"</span><span class="ex">,</span> <span class="st">"GO:0009244"</span><span class="kw">)</span></span>
<span id="cb13-173"><a href="#cb13-173"></a><span class="kw">```</span></span></code><button title="Copy to Clipboard" class="code-copy-button" data-in-quarto-modal=""><i class="bi"></i></button></pre></div>
</div></div></div></div></div>
</div> <!-- /content -->




<footer class="footer"><div class="nav-footer"><div class="nav-footer-center"><div class="toc-actions d-sm-block d-md-none"><ul><li><a href="https://github.com/lakhanp1/Pectobacterium_pangenome/issues/new" class="toc-action"><i class="bi bi-github"></i>Report an issue</a></li></ul></div></div></div></footer></body></html>