<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>

<meta charset="utf-8">
<meta name="generator" content="quarto-1.5.57">

<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">

<meta name="author" content="Lakhansing Pardeshi">

<title>preprocessing â€“ Pectobacterium pangenome data analysis</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
  vertical-align: middle;
}
/* CSS for syntax highlighting */
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { display: inline-block; text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
  }
pre.numberSource { margin-left: 3em;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
</style>


<script src="../../site_libs/quarto-nav/quarto-nav.js"></script>
<script src="../../site_libs/quarto-nav/headroom.min.js"></script>
<script src="../../site_libs/clipboard/clipboard.min.js"></script>
<script src="../../site_libs/quarto-search/autocomplete.umd.js"></script>
<script src="../../site_libs/quarto-search/fuse.min.js"></script>
<script src="../../site_libs/quarto-search/quarto-search.js"></script>
<meta name="quarto:offset" content="../../">
<link href="../../scripts/analysis/ani_clustering.html" rel="next">
<link href="../../scripts/notebooks/index.html" rel="prev">
<script src="../../site_libs/quarto-html/quarto.js"></script>
<script src="../../site_libs/quarto-html/popper.min.js"></script>
<script src="../../site_libs/quarto-html/tippy.umd.min.js"></script>
<script src="../../site_libs/quarto-html/anchor.min.js"></script>
<link href="../../site_libs/quarto-html/tippy.css" rel="stylesheet">
<link href="../../site_libs/quarto-html/quarto-syntax-highlighting.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="../../site_libs/bootstrap/bootstrap.min.js"></script>
<link href="../../site_libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="../../site_libs/bootstrap/bootstrap.min.css" rel="stylesheet" id="quarto-bootstrap" data-mode="light">
<script id="quarto-search-options" type="application/json">{
  "location": "sidebar",
  "copy-button": false,
  "collapse-after": 3,
  "panel-placement": "start",
  "type": "textbox",
  "limit": 50,
  "keyboard-shortcut": [
    "f",
    "/",
    "s"
  ],
  "show-item-context": false,
  "language": {
    "search-no-results-text": "No results",
    "search-matching-documents-text": "matching documents",
    "search-copy-link-title": "Copy link to search",
    "search-hide-matches-text": "Hide additional matches",
    "search-more-match-text": "more match in this document",
    "search-more-matches-text": "more matches in this document",
    "search-clear-button-title": "Clear",
    "search-text-placeholder": "",
    "search-detached-cancel-button-title": "Cancel",
    "search-submit-button-title": "Submit",
    "search-label": "Search"
  }
}</script>


<link rel="stylesheet" href="../../scripts/utils/quarto_style.css">
<meta property="og:title" content="Pectobacterium pangenome data analysis">
<meta property="og:site_name" content="Pectobacterium pangenome data analysis">
</head>

<body class="nav-sidebar docked">

<div id="quarto-search-results"></div>
  <header id="quarto-header" class="headroom fixed-top">
  <nav class="quarto-secondary-nav">
    <div class="container-fluid d-flex">
      <button type="button" class="quarto-btn-toggle btn" data-bs-toggle="collapse" role="button" data-bs-target=".quarto-sidebar-collapse-item" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
        <i class="bi bi-layout-text-sidebar-reverse"></i>
      </button>
        <nav class="quarto-page-breadcrumbs" aria-label="breadcrumb"><ol class="breadcrumb"><li class="breadcrumb-item"><a href="../../scripts/notebooks/preprocessing.html">Preprocessing</a></li><li class="breadcrumb-item"><a href="../../scripts/notebooks/preprocessing.html">Preprocessing</a></li></ol></nav>
        <a class="flex-grow-1" role="navigation" data-bs-toggle="collapse" data-bs-target=".quarto-sidebar-collapse-item" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">      
        </a>
      <button type="button" class="btn quarto-search-button" aria-label="Search" onclick="window.quartoOpenSearch();">
        <i class="bi bi-search"></i>
      </button>
    </div>
  </nav>
</header>
<!-- content -->
<div id="quarto-content" class="quarto-container page-columns page-rows-contents page-layout-article">
<!-- sidebar -->
  <nav id="quarto-sidebar" class="sidebar collapse collapse-horizontal quarto-sidebar-collapse-item sidebar-navigation docked overflow-auto">
    <div class="pt-lg-2 mt-2 text-left sidebar-header">
    <div class="sidebar-title mb-0 py-0">
      <a href="../../">Pectobacterium pangenome data analysis</a> 
        <div class="sidebar-tools-main">
    <div class="dropdown">
      <a href="" title="" id="quarto-navigation-tool-dropdown-0" class="quarto-navigation-tool dropdown-toggle px-1" data-bs-toggle="dropdown" aria-expanded="false" role="link" aria-label=""><i class="bi bi-github"></i></a>
      <ul class="dropdown-menu" aria-labelledby="quarto-navigation-tool-dropdown-0">
          <li>
            <a class="dropdown-item sidebar-tools-main-item" href="https://github.com/lakhanp1/Pectobacterium_pangenome">
            Source Code
            </a>
          </li>
          <li>
            <a class="dropdown-item sidebar-tools-main-item" href="https://github.com/lakhanp1/Pectobacterium_pangenome/issues/new">
            Report a bug
            </a>
          </li>
      </ul>
    </div>
    <a href="https://x.com/lakhanp01" title="" class="quarto-navigation-tool px-1" aria-label=""><i class="bi bi-twitter"></i></a>
</div>
    </div>
      </div>
        <div class="mt-2 flex-shrink-0 align-items-center">
        <div class="sidebar-search">
        <div id="quarto-search" class="" title="Search"></div>
        </div>
        </div>
    <div class="sidebar-menu-container"> 
    <ul class="list-unstyled mt-1">
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../scripts/notebooks/index.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Pectobacterium genus pangenome</span></a>
  </div>
</li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-1" role="navigation" aria-expanded="true">
 <span class="menu-text">Preprocessing</span></a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-1" role="navigation" aria-expanded="true" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-1" class="collapse list-unstyled sidebar-section depth1 show">  
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../scripts/notebooks/preprocessing.html" class="sidebar-item-text sidebar-link active">
 <span class="menu-text">Preprocessing</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../scripts/analysis/ani_clustering.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">ANI</span></a>
  </div>
</li>
      </ul>
  </li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-2" role="navigation" aria-expanded="true">
 <span class="menu-text">Pangenome</span></a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-2" role="navigation" aria-expanded="true" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-2" class="collapse list-unstyled sidebar-section depth1 show">  
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../scripts/analysis/pangenome_metadata_viz.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Pangenome metadata</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../scripts/notebooks/pangenome_construction.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Build</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../scripts/notebooks/pangenome_analysis.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Analysis</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../scripts/analysis/HG_summary.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Pangenome summary</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../scripts/analysis/phylo_viz_tree.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Phylogeny</span></a>
  </div>
</li>
      </ul>
  </li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-3" role="navigation" aria-expanded="true">
 <span class="menu-text">Prophages</span></a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-3" role="navigation" aria-expanded="true" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-3" class="collapse list-unstyled sidebar-section depth1 show">  
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../scripts/notebooks/prophage_analysis.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Analysis</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../scripts/analysis/prophage_summary.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Prophage summary</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../scripts/analysis/synteny_demo.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Syntenic Jaccard index</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../scripts/analysis/prophage_clustering.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Clustering</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../scripts/analysis/prophage_cluster_summary.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Clustering summary</span></a>
  </div>
</li>
          <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start collapsed" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-4" role="navigation" aria-expanded="false">
 <span class="menu-text">Prophage clusters</span></a>
          <a class="sidebar-item-toggle text-start collapsed" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-4" role="navigation" aria-expanded="false" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-4" class="collapse list-unstyled sidebar-section depth2 ">  
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../scripts/ms_figures/phage_grp_46.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Phage group 46</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../scripts/ms_figures/phage_grp_45.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Phage group 45</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../scripts/ms_figures/phage_grp_71.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Phage group 71</span></a>
  </div>
</li>
      </ul>
  </li>
      </ul>
  </li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../scripts/ms_figures/supplementary_tables.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Supplementary tables</span></a>
  </div>
</li>
    </ul>
    </div>
</nav>
<div id="quarto-sidebar-glass" class="quarto-sidebar-collapse-item" data-bs-toggle="collapse" data-bs-target=".quarto-sidebar-collapse-item"></div>
<!-- margin-sidebar -->
    <div id="quarto-margin-sidebar" class="sidebar margin-sidebar">
        <nav id="TOC" role="doc-toc" class="toc-active">
    <h2 id="toc-title">On this page</h2>
   
  <ul>
  <li><a href="#preprocessing" id="toc-preprocessing" class="nav-link active" data-scroll-target="#preprocessing">Preprocessing</a>
  <ul class="collapse">
  <li><a href="#data-download-from-ncbi-using-eutils" id="toc-data-download-from-ncbi-using-eutils" class="nav-link" data-scroll-target="#data-download-from-ncbi-using-eutils">Data download from NCBI using EUtils</a></li>
  <li><a href="#check-if-there-are-any-new-assemblies-released-on-ncbi" id="toc-check-if-there-are-any-new-assemblies-released-on-ncbi" class="nav-link" data-scroll-target="#check-if-there-are-any-new-assemblies-released-on-ncbi">Check if there are any new assemblies released on NCBI</a></li>
  <li><a href="#run-standard-annotations-for-all-the-genomes" id="toc-run-standard-annotations-for-all-the-genomes" class="nav-link" data-scroll-target="#run-standard-annotations-for-all-the-genomes">Run standard annotations for all the genomes</a>
  <ul class="collapse">
  <li><a href="#prokka-annotation-for-bacterial-genomes" id="toc-prokka-annotation-for-bacterial-genomes" class="nav-link" data-scroll-target="#prokka-annotation-for-bacterial-genomes">Prokka annotation for bacterial genomes</a></li>
  <li><a href="#busco-protein-and-genome-assembly-evaluation" id="toc-busco-protein-and-genome-assembly-evaluation" class="nav-link" data-scroll-target="#busco-protein-and-genome-assembly-evaluation">BUSCO protein and genome assembly evaluation</a></li>
  <li><a href="#quast-assembly-evaluation" id="toc-quast-assembly-evaluation" class="nav-link" data-scroll-target="#quast-assembly-evaluation">QUAST assembly evaluation</a></li>
  <li><a href="#interproscan-functional-annotation" id="toc-interproscan-functional-annotation" class="nav-link" data-scroll-target="#interproscan-functional-annotation">InterProScan functional annotation</a></li>
  <li><a href="#eggnog-annotation" id="toc-eggnog-annotation" class="nav-link" data-scroll-target="#eggnog-annotation">eggNOG annotation</a></li>
  <li><a href="#multiqc" id="toc-multiqc" class="nav-link" data-scroll-target="#multiqc">MultiQC</a></li>
  <li><a href="#fastani" id="toc-fastani" class="nav-link" data-scroll-target="#fastani">fastANI</a></li>
  <li><a href="#defense-finder" id="toc-defense-finder" class="nav-link" data-scroll-target="#defense-finder">Defense-finder</a></li>
  </ul></li>
  <li><a href="#mmseq2-database" id="toc-mmseq2-database" class="nav-link" data-scroll-target="#mmseq2-database">MMSeq2 database</a></li>
  <li><a href="#prepare-data-for-pangenome-construction" id="toc-prepare-data-for-pangenome-construction" class="nav-link" data-scroll-target="#prepare-data-for-pangenome-construction">Prepare data for pangenome construction</a>
  <ul class="collapse">
  <li><a href="#extract-metadata-from-ncbi-xml-files" id="toc-extract-metadata-from-ncbi-xml-files" class="nav-link" data-scroll-target="#extract-metadata-from-ncbi-xml-files">Extract metadata from NCBI XML files</a></li>
  <li><a href="#correct-taxonomy" id="toc-correct-taxonomy" class="nav-link" data-scroll-target="#correct-taxonomy">Correct taxonomy</a></li>
  <li><a href="#prepare-final-metadata-table-with-updated-taxonomy" id="toc-prepare-final-metadata-table-with-updated-taxonomy" class="nav-link" data-scroll-target="#prepare-final-metadata-table-with-updated-taxonomy">Prepare final metadata table with updated taxonomy</a></li>
  <li><a href="#prepare-pantools-input-files" id="toc-prepare-pantools-input-files" class="nav-link" data-scroll-target="#prepare-pantools-input-files">Prepare PanTools input files</a></li>
  </ul></li>
  </ul></li>
  </ul>
<div class="toc-actions"><ul><li><a href="https://github.com/lakhanp1/Pectobacterium_pangenome/issues/new" class="toc-action"><i class="bi bi-github"></i>Report an issue</a></li></ul></div></nav>
    </div>
<!-- main -->
<main class="content" id="quarto-document-content"><div class="quarto-title-block"><nav class="quarto-page-breadcrumbs quarto-title-breadcrumbs d-none d-lg-block" aria-label="breadcrumb"><ol class="breadcrumb"><li class="breadcrumb-item"><a href="../../scripts/notebooks/preprocessing.html">Preprocessing</a></li><li class="breadcrumb-item"><a href="../../scripts/notebooks/preprocessing.html">Preprocessing</a></li></ol></nav><div class="quarto-title-tools-only"><h1></h1><button type="button" class="btn code-tools-button" id="quarto-code-tools-source"><i class="bi"></i> Code</button></div></div>

<header id="title-block-header" class="quarto-title-block default">
<div class="quarto-title">
</div>



<div class="quarto-title-meta">

    <div>
    <div class="quarto-title-meta-heading">Author</div>
    <div class="quarto-title-meta-contents">
             <p>Lakhansing Pardeshi </p>
          </div>
  </div>
    
  
    
  </div>
  


</header>


<section id="preprocessing" class="level1">
<h1>Preprocessing</h1>
<section id="data-download-from-ncbi-using-eutils" class="level2">
<h2 class="anchored" data-anchor-id="data-download-from-ncbi-using-eutils">Data download from NCBI using EUtils</h2>
<div class="sourceCode" id="cb1"><pre class="sourceCode numberSource bash number-lines code-with-copy"><code class="sourceCode bash"><span id="cb1-1"><a href="#cb1-1"></a><span class="co">## number of records available in NCBI</span></span>
<span id="cb1-2"><a href="#cb1-2"></a><span class="ex">esearch</span> <span class="at">-db</span> assembly <span class="at">-query</span> <span class="st">"pectobacterium [ORGN]"</span> <span class="at">-email</span> <span class="st">"lakhansing.pardeshi@wur.nl"</span></span>
<span id="cb1-3"><a href="#cb1-3"></a></span>
<span id="cb1-4"><a href="#cb1-4"></a><span class="co">## download the Assembly DocumentSummary file</span></span>
<span id="cb1-5"><a href="#cb1-5"></a><span class="ex">esearch</span> <span class="at">-db</span> assembly <span class="at">-query</span> <span class="st">"pectobacterium [ORGN]"</span> <span class="at">-email</span> <span class="st">"lakhansing.pardeshi@wur.nl"</span> <span class="kw">|</span> <span class="dt">\</span></span>
<span id="cb1-6"><a href="#cb1-6"></a><span class="ex">efetch</span> <span class="at">-format</span> docsum <span class="op">&gt;</span> data/reference_data/assembly_docsum.xml</span>
<span id="cb1-7"><a href="#cb1-7"></a></span>
<span id="cb1-8"><a href="#cb1-8"></a><span class="co">## download associated BioSample DocumentSummary file</span></span>
<span id="cb1-9"><a href="#cb1-9"></a><span class="ex">esearch</span> <span class="at">-db</span> assembly <span class="at">-query</span> <span class="st">"pectobacterium [ORGN]"</span> <span class="at">-email</span> <span class="st">"lakhansing.pardeshi@wur.nl"</span> <span class="kw">|</span> <span class="dt">\</span></span>
<span id="cb1-10"><a href="#cb1-10"></a><span class="ex">elink</span> <span class="at">-target</span> biosample  <span class="kw">|</span> <span class="dt">\</span></span>
<span id="cb1-11"><a href="#cb1-11"></a><span class="ex">efetch</span> <span class="at">-format</span> docsum <span class="op">&gt;</span> data/reference_data/biosample_docsum.xml</span>
<span id="cb1-12"><a href="#cb1-12"></a></span>
<span id="cb1-13"><a href="#cb1-13"></a></span>
<span id="cb1-14"><a href="#cb1-14"></a><span class="co">## get all FPT paths for NCBI assemblies</span></span>
<span id="cb1-15"><a href="#cb1-15"></a><span class="ex">esearch</span> <span class="at">-db</span> assembly <span class="at">-query</span> <span class="st">"pectobacterium [ORGN]"</span> <span class="kw">|</span> <span class="dt">\</span></span>
<span id="cb1-16"><a href="#cb1-16"></a><span class="ex">efetch</span> <span class="at">-format</span> docsum <span class="at">-email</span> <span class="st">"lakhansing.pardeshi@wur.nl"</span> <span class="kw">|</span> <span class="dt">\</span></span>
<span id="cb1-17"><a href="#cb1-17"></a><span class="ex">xtract</span> <span class="at">-pattern</span> DocumentSummary <span class="dt">\</span></span>
<span id="cb1-18"><a href="#cb1-18"></a>-tab <span class="st">","</span> <span class="at">-element</span> AssemblyAccession AssemblyName FtpPath_RefSeq FtpPath_GenBank <span class="kw">|</span> <span class="dt">\</span></span>
<span id="cb1-19"><a href="#cb1-19"></a><span class="fu">awk</span> <span class="st">'</span></span>
<span id="cb1-20"><a href="#cb1-20"></a><span class="st">    BEGIN { FS=","};</span></span>
<span id="cb1-21"><a href="#cb1-21"></a><span class="st">    {</span></span>
<span id="cb1-22"><a href="#cb1-22"></a><span class="st">    if($4=="") file=$3; else file=$3;</span></span>
<span id="cb1-23"><a href="#cb1-23"></a><span class="st">    gsub(/ +/, "_", $0);</span></span>
<span id="cb1-24"><a href="#cb1-24"></a><span class="st">    printf "%s/%s_%s_genomic.fna.gz\n", file, $1, $2 &gt; "data/reference_data/ncbi_assembly_ftp.txt";</span></span>
<span id="cb1-25"><a href="#cb1-25"></a><span class="st">    printf "%s_%s\n", $1, $2 &gt; "data/reference_data/ncbi_assembly_ids.txt";</span></span>
<span id="cb1-26"><a href="#cb1-26"></a><span class="st">    }</span></span>
<span id="cb1-27"><a href="#cb1-27"></a><span class="st">'</span></span>
<span id="cb1-28"><a href="#cb1-28"></a></span>
<span id="cb1-29"><a href="#cb1-29"></a><span class="co">## NCBI prokaryotes ANI file</span></span>
<span id="cb1-30"><a href="#cb1-30"></a><span class="fu">wget</span> <span class="at">--timestamping</span> https://ftp.ncbi.nlm.nih.gov/genomes/ASSEMBLY_REPORTS/ANI_report_prokaryotes.txt <span class="at">-P</span> data/other/</span>
<span id="cb1-31"><a href="#cb1-31"></a></span>
<span id="cb1-32"><a href="#cb1-32"></a></span>
<span id="cb1-33"><a href="#cb1-33"></a><span class="fu">cat</span> data/reference_data/local_assembly_ids.txt data/reference_data/ncbi_assembly_ids.txt <span class="dt">\</span></span>
<span id="cb1-34"><a href="#cb1-34"></a><span class="op">&gt;</span> data/reference_data/pre_qc_assembly_ids.txt</span>
<span id="cb1-35"><a href="#cb1-35"></a></span>
<span id="cb1-36"><a href="#cb1-36"></a><span class="fu">mkdir</span> data/reference_data/batches</span>
<span id="cb1-37"><a href="#cb1-37"></a><span class="fu">split</span> <span class="at">-a</span> 2 <span class="at">-d</span> <span class="at">-n</span> r/8 data/reference_data/pre_qc_assembly_ids.txt data/reference_data/batches/temp_ids_batch.</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</section>
<section id="check-if-there-are-any-new-assemblies-released-on-ncbi" class="level2">
<h2 class="anchored" data-anchor-id="check-if-there-are-any-new-assemblies-released-on-ncbi">Check if there are any new assemblies released on NCBI</h2>
<div class="sourceCode" id="cb2"><pre class="sourceCode numberSource bash number-lines code-with-copy"><code class="sourceCode bash"><span id="cb2-1"><a href="#cb2-1"></a><span class="co">## check and extract new assembly IDs</span></span>
<span id="cb2-2"><a href="#cb2-2"></a><span class="cf">for</span> sampleId <span class="kw">in</span> <span class="kw">`</span><span class="fu">cat</span> data/reference_data/pre_qc_assembly_ids.txt<span class="kw">`</span></span>
<span id="cb2-3"><a href="#cb2-3"></a><span class="cf">do</span></span>
<span id="cb2-4"><a href="#cb2-4"></a>    <span class="cf">if</span> <span class="ot">! </span><span class="fu">find</span> data/genomes/ <span class="at">-name</span> <span class="va">${sampleId}</span><span class="pp">*</span> <span class="kw">|</span> <span class="fu">grep</span> <span class="at">-q</span> <span class="st">"."</span></span>
<span id="cb2-5"><a href="#cb2-5"></a>    <span class="cf">then</span></span>
<span id="cb2-6"><a href="#cb2-6"></a>        <span class="bu">echo</span> <span class="va">$sampleId</span></span>
<span id="cb2-7"><a href="#cb2-7"></a>    <span class="cf">fi</span></span>
<span id="cb2-8"><a href="#cb2-8"></a><span class="cf">done</span> <span class="op">&gt;</span> data/reference_data/temp_assembly_ids.txt</span>
<span id="cb2-9"><a href="#cb2-9"></a></span>
<span id="cb2-10"><a href="#cb2-10"></a></span>
<span id="cb2-11"><a href="#cb2-11"></a><span class="co">## download and unzip the FASTA files</span></span>
<span id="cb2-12"><a href="#cb2-12"></a><span class="cf">for</span> sampleId <span class="kw">in</span> <span class="kw">`</span><span class="fu">cat</span> data/reference_data/temp_assembly_ids.txt<span class="kw">`</span></span>
<span id="cb2-13"><a href="#cb2-13"></a><span class="cf">do</span></span>
<span id="cb2-14"><a href="#cb2-14"></a>    <span class="fu">grep</span> <span class="st">"/</span><span class="va">$sampleId</span><span class="st">/"</span> data/reference_data/ncbi_assembly_ftp.txt</span>
<span id="cb2-15"><a href="#cb2-15"></a><span class="cf">done</span> <span class="kw">|</span> <span class="dt">\</span></span>
<span id="cb2-16"><a href="#cb2-16"></a><span class="fu">xargs</span> <span class="at">-I</span> {} wget <span class="at">--timestamping</span> {} <span class="at">-P</span> data/genomes/ncbi/</span>
<span id="cb2-17"><a href="#cb2-17"></a></span>
<span id="cb2-18"><a href="#cb2-18"></a><span class="fu">gzip</span> <span class="at">-d</span> data/genomes/ncbi/<span class="pp">*</span>.gz</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</section>
<section id="run-standard-annotations-for-all-the-genomes" class="level2">
<h2 class="anchored" data-anchor-id="run-standard-annotations-for-all-the-genomes">Run standard annotations for all the genomes</h2>
<section id="prokka-annotation-for-bacterial-genomes" class="level3">
<h3 class="anchored" data-anchor-id="prokka-annotation-for-bacterial-genomes">Prokka annotation for bacterial genomes</h3>
<div class="sourceCode" id="cb3"><pre class="sourceCode numberSource bash number-lines code-with-copy"><code class="sourceCode bash"><span id="cb3-1"><a href="#cb3-1"></a><span class="fu">nohup</span> <span class="dt">\</span></span>
<span id="cb3-2"><a href="#cb3-2"></a>ls data/genomes/<span class="pp">?(</span>NCBI<span class="pp">|</span>WUR<span class="pp">|</span>NIVIP<span class="pp">|</span>NAK<span class="pp">)</span>/<span class="pp">*</span>.<span class="pp">?(</span>fa<span class="pp">|</span>fna<span class="pp">|</span>fasta<span class="pp">)</span> <span class="kw">|</span> <span class="dt">\</span></span>
<span id="cb3-3"><a href="#cb3-3"></a><span class="ex">parallel</span> <span class="at">--gnu</span> <span class="at">--keep-order</span> <span class="at">--jobs</span> 1 <span class="at">--halt</span> now,fail=1 <span class="dt">\</span></span>
<span id="cb3-4"><a href="#cb3-4"></a>--results logs/prokka/{/.} <span class="at">--joblog</span> logs/prokka/parallel.log <span class="dt">\</span></span>
<span id="cb3-5"><a href="#cb3-5"></a>scripts/preprocessing/prokka_annotation.sh {} <span class="dt">\</span></span>
<span id="cb3-6"><a href="#cb3-6"></a><span class="op">&gt;</span>logs/prokka/nohup_batch00.out <span class="dv">2</span><span class="op">&gt;&amp;</span><span class="dv">1</span> <span class="kw">&amp;</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<p>Processing in parallel is failing without any specific error. Hence the data will be processed in serial mode. Further debugging is required to understand the parallel processing failure. For now, running GNU <code>parallel</code> with <code>--jobs 1</code> setting in serial mode.</p>
<section id="remove-the-fasta-sequence-at-the-end-of-prokka-gff-files" class="level4">
<h4 class="anchored" data-anchor-id="remove-the-fasta-sequence-at-the-end-of-prokka-gff-files">Remove the FASTA sequence at the end of prokka gff files</h4>
<div class="sourceCode" id="cb4"><pre class="sourceCode numberSource bash number-lines code-with-copy"><code class="sourceCode bash"><span id="cb4-1"><a href="#cb4-1"></a><span class="cf">for</span> i <span class="kw">in</span> <span class="kw">`</span><span class="fu">cat</span> data/reference_data/pre_qc_assembly_ids.txt<span class="kw">`</span></span>
<span id="cb4-2"><a href="#cb4-2"></a><span class="cf">do</span></span>
<span id="cb4-3"><a href="#cb4-3"></a><span class="fu">sed</span> <span class="at">-n</span> <span class="st">'1,/##FASTA/ {/##FASTA/!p}'</span> data/prokka_annotation/<span class="va">${i}</span>/<span class="va">${i}</span>.gff <span class="op">&gt;</span> <span class="dt">\</span></span>
<span id="cb4-4"><a href="#cb4-4"></a>data/prokka_annotation/<span class="va">${i}</span>/<span class="va">${i}</span>.gff3</span>
<span id="cb4-5"><a href="#cb4-5"></a><span class="cf">done</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</section>
<section id="fasta-file-indexing" class="level4">
<h4 class="anchored" data-anchor-id="fasta-file-indexing">FASTA file indexing</h4>
<div class="sourceCode" id="cb5"><pre class="sourceCode numberSource bash number-lines code-with-copy"><code class="sourceCode bash"><span id="cb5-1"><a href="#cb5-1"></a><span class="bu">printf</span> <span class="st">''</span> <span class="op">&gt;</span> analysis/QC/assembly_chr_size.txt</span>
<span id="cb5-2"><a href="#cb5-2"></a><span class="bu">printf</span> <span class="st">''</span> <span class="op">&gt;</span> data/reference_data/pre_qc_genomes_fa.list</span>
<span id="cb5-3"><a href="#cb5-3"></a><span class="ex">conda</span> activate omics_py37</span>
<span id="cb5-4"><a href="#cb5-4"></a></span>
<span id="cb5-5"><a href="#cb5-5"></a><span class="cf">for</span> i <span class="kw">in</span> <span class="kw">`</span><span class="fu">cat</span> data/reference_data/pre_qc_assembly_ids.txt<span class="kw">`</span></span>
<span id="cb5-6"><a href="#cb5-6"></a><span class="cf">do</span></span>
<span id="cb5-7"><a href="#cb5-7"></a>    <span class="ex">samtools</span> faidx data/prokka_annotation/<span class="va">${i}</span>/<span class="va">${i}</span>.fna</span>
<span id="cb5-8"><a href="#cb5-8"></a>    <span class="fu">sort</span> <span class="at">-r</span> <span class="at">-n</span> <span class="at">-k</span> 2,2 data/prokka_annotation/<span class="va">${i}</span>/<span class="va">${i}</span>.fna.fai <span class="kw">|</span> <span class="dt">\</span></span>
<span id="cb5-9"><a href="#cb5-9"></a>    <span class="fu">awk</span>  <span class="at">-v</span> i=<span class="va">${i}</span> <span class="st">'{print i, "\t", $1, "\t", $2}'</span> <span class="op">&gt;&gt;</span> analysis/QC/assembly_chr_size.txt</span>
<span id="cb5-10"><a href="#cb5-10"></a>    <span class="fu">ls</span> data/prokka_annotation/<span class="va">${i}</span>/<span class="va">${i}</span>.fna <span class="op">&gt;&gt;</span> data/reference_data/pre_qc_genomes_fa.list</span>
<span id="cb5-11"><a href="#cb5-11"></a></span>
<span id="cb5-12"><a href="#cb5-12"></a>    <span class="ex">faToTwoBit</span> data/prokka_annotation/<span class="va">${i}</span>/<span class="va">${i}</span>.fna data/prokka_annotation/<span class="va">${i}</span>/<span class="va">${i}</span>.2bit</span>
<span id="cb5-13"><a href="#cb5-13"></a></span>
<span id="cb5-14"><a href="#cb5-14"></a><span class="cf">done</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</section>
</section>
<section id="busco-protein-and-genome-assembly-evaluation" class="level3">
<h3 class="anchored" data-anchor-id="busco-protein-and-genome-assembly-evaluation">BUSCO protein and genome assembly evaluation</h3>
<div class="sourceCode" id="cb6"><pre class="sourceCode numberSource bash number-lines code-with-copy"><code class="sourceCode bash"><span id="cb6-1"><a href="#cb6-1"></a><span class="co">## !GNU parallel on single server</span></span>
<span id="cb6-2"><a href="#cb6-2"></a><span class="fu">nohup</span> <span class="dt">\</span></span>
<span id="cb6-3"><a href="#cb6-3"></a>cat data/reference_data/temp_assembly_ids.txt <span class="kw">|</span> <span class="dt">\</span></span>
<span id="cb6-4"><a href="#cb6-4"></a><span class="ex">parallel</span> <span class="at">--jobs</span> 1 <span class="at">--workdir</span> <span class="va">$PWD</span> <span class="at">--halt</span> now,fail=1 <span class="at">--keep-order</span> <span class="dt">\</span></span>
<span id="cb6-5"><a href="#cb6-5"></a>--results logs/busco/{} <span class="at">--joblog</span> logs/busco/parallel.log <span class="dt">\</span></span>
<span id="cb6-6"><a href="#cb6-6"></a><span class="va">$PWD</span>/scripts/preprocessing/qc_busco.sh {} <span class="dt">\</span></span>
<span id="cb6-7"><a href="#cb6-7"></a><span class="op">&gt;</span>logs/busco/nohup03.out <span class="dv">2</span><span class="op">&gt;&amp;</span><span class="dv">1</span> <span class="kw">&amp;</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<p>This is failing with <code>parallel</code>, most likely because of the number of open files exceeding the <code>ulimit</code>. Need to debug further. For now, running GNU <code>parallel</code> with <code>--jobs 1</code> setting is serial mode.</p>
</section>
<section id="quast-assembly-evaluation" class="level3">
<h3 class="anchored" data-anchor-id="quast-assembly-evaluation">QUAST assembly evaluation</h3>
<div class="sourceCode" id="cb7"><pre class="sourceCode numberSource bash number-lines code-with-copy"><code class="sourceCode bash"><span id="cb7-1"><a href="#cb7-1"></a><span class="fu">nohup</span> <span class="dt">\</span></span>
<span id="cb7-2"><a href="#cb7-2"></a>cat data/reference_data/temp_assembly_ids.txt <span class="kw">|</span> <span class="dt">\</span></span>
<span id="cb7-3"><a href="#cb7-3"></a><span class="ex">parallel</span> <span class="at">--jobs</span> 6 <span class="at">--workdir</span> <span class="va">$PWD</span> <span class="at">--halt</span> now,fail=1 <span class="dt">\</span></span>
<span id="cb7-4"><a href="#cb7-4"></a>--keep-order <span class="at">--results</span> logs/quast/{} <span class="dt">\</span></span>
<span id="cb7-5"><a href="#cb7-5"></a>--joblog logs/quast/parallel.log <span class="dt">\</span></span>
<span id="cb7-6"><a href="#cb7-6"></a>./scripts/preprocessing/qc_quast.sh {} <span class="dt">\</span></span>
<span id="cb7-7"><a href="#cb7-7"></a><span class="op">&gt;&gt;</span>logs/quast/nohup.out <span class="dv">2</span><span class="op">&gt;&amp;</span><span class="dv">1</span> <span class="kw">&amp;</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</section>
<section id="interproscan-functional-annotation" class="level3">
<h3 class="anchored" data-anchor-id="interproscan-functional-annotation">InterProScan functional annotation</h3>
<div class="sourceCode" id="cb8"><pre class="sourceCode numberSource bash number-lines code-with-copy"><code class="sourceCode bash"><span id="cb8-1"><a href="#cb8-1"></a><span class="co">## GNU parallel on single server</span></span>
<span id="cb8-2"><a href="#cb8-2"></a><span class="fu">nohup</span> <span class="dt">\</span></span>
<span id="cb8-3"><a href="#cb8-3"></a>cat data/reference_data/temp_assembly_ids.txt <span class="kw">|</span> <span class="dt">\</span></span>
<span id="cb8-4"><a href="#cb8-4"></a><span class="ex">parallel</span> <span class="at">--jobs</span> 6 <span class="at">--workdir</span> <span class="va">$PWD</span> <span class="at">--halt</span> now,fail=1 <span class="dt">\</span></span>
<span id="cb8-5"><a href="#cb8-5"></a>--keep-order <span class="at">--results</span> logs/interproscan/{} <span class="dt">\</span></span>
<span id="cb8-6"><a href="#cb8-6"></a>--joblog logs/interproscan/parallel_batch06.log <span class="dt">\</span></span>
<span id="cb8-7"><a href="#cb8-7"></a>./scripts/preprocessing/interproscan_annotation.sh {} <span class="dt">\</span></span>
<span id="cb8-8"><a href="#cb8-8"></a><span class="op">&gt;&gt;</span>logs/interproscan/nohup_batch07.out <span class="dv">2</span><span class="op">&gt;&amp;</span><span class="dv">1</span> <span class="kw">&amp;</span></span>
<span id="cb8-9"><a href="#cb8-9"></a></span>
<span id="cb8-10"><a href="#cb8-10"></a></span>
<span id="cb8-11"><a href="#cb8-11"></a><span class="co">## GNU parallel: submit jobs to another server</span></span>
<span id="cb8-12"><a href="#cb8-12"></a><span class="fu">head</span> <span class="at">-n</span> 10 data/reference_data/sample_subsets/genebank_ass.txt <span class="kw">|</span> <span class="dt">\</span></span>
<span id="cb8-13"><a href="#cb8-13"></a><span class="ex">env_parallel</span> <span class="at">--jobs</span> 4 <span class="at">--workdir</span> <span class="va">$PWD</span> <span class="at">--halt</span> now,fail=1 <span class="dt">\</span></span>
<span id="cb8-14"><a href="#cb8-14"></a>--keep-order <span class="at">--results</span> logs/interproscan/{} <span class="dt">\</span></span>
<span id="cb8-15"><a href="#cb8-15"></a>--joblog logs/interproscan/parallel.log <span class="dt">\</span></span>
<span id="cb8-16"><a href="#cb8-16"></a>--sshlogin 4/waterman.bioinformatics.nl <span class="at">--cleanup</span> <span class="dt">\</span></span>
<span id="cb8-17"><a href="#cb8-17"></a>--env error_exit <span class="at">--env</span> process_start <span class="at">--env</span> TOOLS_PATH <span class="at">--env</span> LUSTRE_HOME <span class="dt">\</span></span>
<span id="cb8-18"><a href="#cb8-18"></a>./scripts/preprocessing/interproscan_annotation.sh {}</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</section>
<section id="eggnog-annotation" class="level3">
<h3 class="anchored" data-anchor-id="eggnog-annotation">eggNOG annotation</h3>
<div class="sourceCode" id="cb9"><pre class="sourceCode numberSource bash number-lines code-with-copy"><code class="sourceCode bash"><span id="cb9-1"><a href="#cb9-1"></a><span class="ex">conda</span> activate eggnog</span>
<span id="cb9-2"><a href="#cb9-2"></a></span>
<span id="cb9-3"><a href="#cb9-3"></a><span class="co">## download eggNOG data</span></span>
<span id="cb9-4"><a href="#cb9-4"></a><span class="ex">download_eggnog_data.py</span> <span class="at">-M</span> <span class="at">-F</span> <span class="at">-H</span> <span class="at">-d</span> 1 <span class="at">--dbname</span> root <span class="at">-f</span> <span class="at">-y</span></span>
<span id="cb9-5"><a href="#cb9-5"></a><span class="ex">download_eggnog_data.py</span> <span class="at">-D</span> <span class="at">-H</span> <span class="at">-d</span> 2759 <span class="at">--dbname</span> eukaryota <span class="at">-q</span> <span class="at">-y</span></span>
<span id="cb9-6"><a href="#cb9-6"></a><span class="ex">download_eggnog_data.py</span> <span class="at">-D</span> <span class="at">-H</span> <span class="at">-d</span> 2 <span class="at">--dbname</span> bacteria <span class="at">-q</span> <span class="at">-y</span></span>
<span id="cb9-7"><a href="#cb9-7"></a><span class="ex">download_eggnog_data.py</span> <span class="at">-D</span> <span class="at">-H</span> <span class="at">-d</span> 10239 <span class="at">--dbname</span> viruses <span class="at">-q</span> <span class="at">-y</span></span>
<span id="cb9-8"><a href="#cb9-8"></a><span class="ex">download_eggnog_data.py</span> <span class="at">-D</span> <span class="at">-H</span> <span class="at">-d</span> 4751 <span class="at">--dbname</span> fungi <span class="at">-q</span> <span class="at">-y</span></span>
<span id="cb9-9"><a href="#cb9-9"></a><span class="ex">download_eggnog_data.py</span> <span class="at">-D</span> <span class="at">-H</span> <span class="at">-d</span> 2157 <span class="at">--dbname</span> archea <span class="at">-q</span> <span class="at">-y</span></span>
<span id="cb9-10"><a href="#cb9-10"></a></span>
<span id="cb9-11"><a href="#cb9-11"></a><span class="ex">mmseqs</span> createindex <span class="st">"</span><span class="va">$EGGNOG_DATA_DIR</span><span class="st">"</span>/mmseqs/mmseqs.db /local_scratch/parde001/tmp</span>
<span id="cb9-12"><a href="#cb9-12"></a></span>
<span id="cb9-13"><a href="#cb9-13"></a><span class="co">## create mmseq database for bacteria, archea and viruses</span></span>
<span id="cb9-14"><a href="#cb9-14"></a><span class="ex">create_dbs.py</span> <span class="at">-m</span> mmseqs <span class="at">--dbname</span> bact_arch_vir <span class="at">--taxids</span> 2,2157,10239 <span class="at">-y</span></span>
<span id="cb9-15"><a href="#cb9-15"></a></span>
<span id="cb9-16"><a href="#cb9-16"></a><span class="co">## annotate for eggNOG clusters</span></span>
<span id="cb9-17"><a href="#cb9-17"></a><span class="cf">for</span> i <span class="kw">in</span> <span class="kw">`</span><span class="fu">cat</span> data/reference_data/assembly_ids.txt<span class="kw">`</span></span>
<span id="cb9-18"><a href="#cb9-18"></a><span class="cf">do</span></span>
<span id="cb9-19"><a href="#cb9-19"></a>    <span class="va">file_aa</span><span class="op">=</span><span class="st">"data/prokka_annotation/</span><span class="va">${i}</span><span class="st">/</span><span class="va">${i}</span><span class="st">.faa"</span></span>
<span id="cb9-20"><a href="#cb9-20"></a>    </span>
<span id="cb9-21"><a href="#cb9-21"></a>    <span class="ex">emapper.py</span> <span class="at">--cpu</span> 20 <span class="at">-m</span> mmseqs <span class="at">--dbmem</span> <span class="at">--go_evidence</span> all  <span class="dt">\</span></span>
<span id="cb9-22"><a href="#cb9-22"></a>    <span class="at">-i</span> <span class="va">${file_aa}</span> <span class="at">-o</span> <span class="va">${i}</span> <span class="at">--output_dir</span> data/eggnog <span class="dt">\</span></span>
<span id="cb9-23"><a href="#cb9-23"></a>    <span class="at">--scratch_dir</span> /local_scratch/parde001/ <span class="at">--temp_dir</span> /local_scratch/parde001/tmp</span>
<span id="cb9-24"><a href="#cb9-24"></a></span>
<span id="cb9-25"><a href="#cb9-25"></a>    <span class="ex">error_exit</span> <span class="va">$?</span></span>
<span id="cb9-26"><a href="#cb9-26"></a><span class="cf">done</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</section>
<section id="multiqc" class="level3">
<h3 class="anchored" data-anchor-id="multiqc">MultiQC</h3>
<div class="sourceCode" id="cb10"><pre class="sourceCode numberSource bash number-lines code-with-copy"><code class="sourceCode bash"><span id="cb10-1"><a href="#cb10-1"></a><span class="co">## QUAST MultiQC</span></span>
<span id="cb10-2"><a href="#cb10-2"></a><span class="fu">nohup</span> <span class="dt">\</span></span>
<span id="cb10-3"><a href="#cb10-3"></a>multiqc <span class="at">-f</span> <span class="at">--filename</span> quast_multiqc <span class="at">--interactive</span> <span class="at">--title</span> <span class="st">"QUAST report"</span> <span class="dt">\</span></span>
<span id="cb10-4"><a href="#cb10-4"></a>--outdir analysis/multiqc/ <span class="at">--module</span> quast data/quast/ <span class="dt">\</span></span>
<span id="cb10-5"><a href="#cb10-5"></a><span class="op">&gt;&gt;</span>nohup.out <span class="dv">2</span><span class="op">&gt;&amp;</span><span class="dv">1</span> <span class="kw">&amp;</span></span>
<span id="cb10-6"><a href="#cb10-6"></a></span>
<span id="cb10-7"><a href="#cb10-7"></a><span class="co">## BUSCO protein MultiQC</span></span>
<span id="cb10-8"><a href="#cb10-8"></a><span class="fu">nohup</span> <span class="dt">\</span></span>
<span id="cb10-9"><a href="#cb10-9"></a>multiqc <span class="at">-f</span> <span class="at">--filename</span> busco_prot_multiqc <span class="at">--interactive</span> <span class="at">--title</span> <span class="st">"BUSCO report"</span> <span class="dt">\</span></span>
<span id="cb10-10"><a href="#cb10-10"></a>--outdir analysis/multiqc/ <span class="at">--module</span> busco data/busco.prot/ <span class="dt">\</span></span>
<span id="cb10-11"><a href="#cb10-11"></a><span class="op">&gt;&gt;</span>nohup.out <span class="dv">2</span><span class="op">&gt;&amp;</span><span class="dv">1</span> <span class="kw">&amp;</span></span>
<span id="cb10-12"><a href="#cb10-12"></a></span>
<span id="cb10-13"><a href="#cb10-13"></a><span class="co">## BUSCO genome MultiQC</span></span>
<span id="cb10-14"><a href="#cb10-14"></a><span class="fu">nohup</span> <span class="dt">\</span></span>
<span id="cb10-15"><a href="#cb10-15"></a>multiqc <span class="at">-f</span> <span class="at">--filename</span> busco_geno_multiqc <span class="at">--interactive</span> <span class="at">--title</span> <span class="st">"BUSCO report"</span> <span class="dt">\</span></span>
<span id="cb10-16"><a href="#cb10-16"></a>--outdir analysis/multiqc/ <span class="at">--module</span> busco data/busco.geno/ <span class="dt">\</span></span>
<span id="cb10-17"><a href="#cb10-17"></a><span class="op">&gt;&gt;</span>nohup.out <span class="dv">2</span><span class="op">&gt;&amp;</span><span class="dv">1</span> <span class="kw">&amp;</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</section>
<section id="fastani" class="level3">
<h3 class="anchored" data-anchor-id="fastani">fastANI</h3>
<div class="sourceCode" id="cb11"><pre class="sourceCode numberSource bash number-lines code-with-copy"><code class="sourceCode bash"><span id="cb11-1"><a href="#cb11-1"></a><span class="ex">process_start</span> <span class="st">"ANI on all genomes"</span></span>
<span id="cb11-2"><a href="#cb11-2"></a><span class="fu">nohup</span> bash scripts/preprocessing/ANI_processing.sh <span class="dt">\</span></span>
<span id="cb11-3"><a href="#cb11-3"></a>data/reference_data/pre_qc_genomes_fa.list <span class="dt">\</span></span>
<span id="cb11-4"><a href="#cb11-4"></a>analysis/02_fastANI/ANI_results <span class="op">&gt;</span>logs/fastANI.log <span class="dv">2</span><span class="op">&gt;&amp;</span><span class="dv">1</span> <span class="kw">&amp;</span></span>
<span id="cb11-5"><a href="#cb11-5"></a><span class="ex">error_exit</span> <span class="va">$?</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</section>
<section id="defense-finder" class="level3">
<h3 class="anchored" data-anchor-id="defense-finder">Defense-finder</h3>
<div class="sourceCode" id="cb12"><pre class="sourceCode numberSource bash number-lines code-with-copy"><code class="sourceCode bash"><span id="cb12-1"><a href="#cb12-1"></a><span class="ex">conda</span> activate defensefinder</span>
<span id="cb12-2"><a href="#cb12-2"></a></span>
<span id="cb12-3"><a href="#cb12-3"></a><span class="cf">for</span> i <span class="kw">in</span> <span class="kw">`</span><span class="fu">cat</span> data/reference_data/assembly_ids.txt<span class="kw">`</span></span>
<span id="cb12-4"><a href="#cb12-4"></a><span class="cf">do</span></span>
<span id="cb12-5"><a href="#cb12-5"></a>    <span class="va">file_aa</span><span class="op">=</span><span class="st">"data/prokka_annotation/</span><span class="va">${i}</span><span class="st">/</span><span class="va">${i}</span><span class="st">.faa"</span></span>
<span id="cb12-6"><a href="#cb12-6"></a>    <span class="va">out_dir</span><span class="op">=</span><span class="st">"data/defense_finder/</span><span class="va">${i}</span><span class="st">"</span></span>
<span id="cb12-7"><a href="#cb12-7"></a>    <span class="fu">mkdir</span> <span class="at">-p</span> <span class="va">${out_dir}</span></span>
<span id="cb12-8"><a href="#cb12-8"></a></span>
<span id="cb12-9"><a href="#cb12-9"></a>    <span class="ex">process_start</span> <span class="st">"defense-finder: </span><span class="va">${i}</span><span class="st">"</span></span>
<span id="cb12-10"><a href="#cb12-10"></a></span>
<span id="cb12-11"><a href="#cb12-11"></a>    <span class="ex">defense-finder</span> run <span class="at">--workers</span> 20 <span class="at">-o</span> <span class="va">${out_dir}</span> <span class="dt">\</span></span>
<span id="cb12-12"><a href="#cb12-12"></a>    <span class="at">--models-dir</span> <span class="va">$TOOLS_PATH</span>/defense-finder-models <span class="va">${file_aa}</span></span>
<span id="cb12-13"><a href="#cb12-13"></a></span>
<span id="cb12-14"><a href="#cb12-14"></a>    <span class="ex">error_exit</span> <span class="va">$?</span></span>
<span id="cb12-15"><a href="#cb12-15"></a><span class="cf">done</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<p>Combine Defense-finder output into single file</p>
<div class="sourceCode" id="cb13"><pre class="sourceCode numberSource bash number-lines code-with-copy"><code class="sourceCode bash"><span id="cb13-1"><a href="#cb13-1"></a><span class="co"># combine systems</span></span>
<span id="cb13-2"><a href="#cb13-2"></a><span class="cf">for</span> i <span class="kw">in</span> <span class="kw">`</span><span class="fu">cat</span> data/reference_data/assembly_ids.txt<span class="kw">`</span></span>
<span id="cb13-3"><a href="#cb13-3"></a><span class="cf">do</span></span>
<span id="cb13-4"><a href="#cb13-4"></a>    <span class="fu">awk</span> <span class="at">-v</span> var=<span class="st">"</span><span class="va">${i}</span><span class="st">"</span> <span class="st">'!/^sys_id/ {print var "\t" $0}'</span> data/defense_finder/<span class="va">${i}</span>/<span class="va">${i}</span>_defense_finder_systems.tsv</span>
<span id="cb13-5"><a href="#cb13-5"></a><span class="cf">done</span> <span class="op">&gt;</span> data/defense_finder/defense_systems.tsv</span>
<span id="cb13-6"><a href="#cb13-6"></a></span>
<span id="cb13-7"><a href="#cb13-7"></a><span class="co"># combine genes</span></span>
<span id="cb13-8"><a href="#cb13-8"></a><span class="cf">for</span> i <span class="kw">in</span> <span class="kw">`</span><span class="fu">cat</span> data/reference_data/assembly_ids.txt<span class="kw">`</span></span>
<span id="cb13-9"><a href="#cb13-9"></a><span class="cf">do</span></span>
<span id="cb13-10"><a href="#cb13-10"></a>    <span class="fu">awk</span> <span class="at">-v</span> var=<span class="st">"</span><span class="va">${i}</span><span class="st">"</span> <span class="st">'NR&gt;1 {print var "\t" $0}'</span> data/defense_finder/<span class="va">${i}</span>/<span class="va">${i}</span>_defense_finder_genes.tsv</span>
<span id="cb13-11"><a href="#cb13-11"></a><span class="cf">done</span> <span class="op">&gt;</span> data/defense_finder/defense_systems_genes.tsv</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</section>
</section>
<section id="mmseq2-database" class="level2">
<h2 class="anchored" data-anchor-id="mmseq2-database">MMSeq2 database</h2>
<pre><code>conda activate mmseq2

mmseqs createdb data/pangenomes/pectobacterium.v2/mmseq_panproteome_db</code></pre>
</section>
<section id="prepare-data-for-pangenome-construction" class="level2">
<h2 class="anchored" data-anchor-id="prepare-data-for-pangenome-construction">Prepare data for pangenome construction</h2>
<section id="extract-metadata-from-ncbi-xml-files" class="level3">
<h3 class="anchored" data-anchor-id="extract-metadata-from-ncbi-xml-files">Extract metadata from NCBI XML files</h3>
<div class="sourceCode" id="cb15"><pre class="sourceCode numberSource bash number-lines code-with-copy"><code class="sourceCode bash"><span id="cb15-1"><a href="#cb15-1"></a><span class="ex">Rscript</span> scripts/preprocessing/raw_prebuild_metadata.R</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</section>
<section id="correct-taxonomy" class="level3">
<h3 class="anchored" data-anchor-id="correct-taxonomy">Correct taxonomy</h3>
<div class="sourceCode" id="cb16"><pre class="sourceCode numberSource bash number-lines code-with-copy"><code class="sourceCode bash"><span id="cb16-1"><a href="#cb16-1"></a><span class="ex">Rscript</span> scripts/preprocessing/taxonomy_check_inhouse.R</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</section>
<section id="prepare-final-metadata-table-with-updated-taxonomy" class="level3">
<h3 class="anchored" data-anchor-id="prepare-final-metadata-table-with-updated-taxonomy">Prepare final metadata table with updated taxonomy</h3>
<div class="sourceCode" id="cb17"><pre class="sourceCode numberSource bash number-lines code-with-copy"><code class="sourceCode bash"><span id="cb17-1"><a href="#cb17-1"></a><span class="ex">Rscript</span> scripts/preprocessing/build_metadata.R</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</section>
<section id="prepare-pantools-input-files" class="level3">
<h3 class="anchored" data-anchor-id="prepare-pantools-input-files">Prepare PanTools input files</h3>
<div class="sourceCode" id="cb18"><pre class="sourceCode numberSource bash number-lines code-with-copy"><code class="sourceCode bash"><span id="cb18-1"><a href="#cb18-1"></a><span class="ex">Rscript</span> scripts/preprocessing/prepare_PanTools_input.R</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>


<!-- -->

</section>
</section>
</section>

<a onclick="window.scrollTo(0, 0); return false;" role="button" id="quarto-back-to-top"><i class="bi bi-arrow-up"></i> Back to top</a></main> <!-- /main -->
<script id="quarto-html-after-body" type="application/javascript">
window.document.addEventListener("DOMContentLoaded", function (event) {
  const toggleBodyColorMode = (bsSheetEl) => {
    const mode = bsSheetEl.getAttribute("data-mode");
    const bodyEl = window.document.querySelector("body");
    if (mode === "dark") {
      bodyEl.classList.add("quarto-dark");
      bodyEl.classList.remove("quarto-light");
    } else {
      bodyEl.classList.add("quarto-light");
      bodyEl.classList.remove("quarto-dark");
    }
  }
  const toggleBodyColorPrimary = () => {
    const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
    if (bsSheetEl) {
      toggleBodyColorMode(bsSheetEl);
    }
  }
  toggleBodyColorPrimary();  
  const icon = "î§‹";
  const anchorJS = new window.AnchorJS();
  anchorJS.options = {
    placement: 'right',
    icon: icon
  };
  anchorJS.add('.anchored');
  const isCodeAnnotation = (el) => {
    for (const clz of el.classList) {
      if (clz.startsWith('code-annotation-')) {                     
        return true;
      }
    }
    return false;
  }
  const onCopySuccess = function(e) {
    // button target
    const button = e.trigger;
    // don't keep focus
    button.blur();
    // flash "checked"
    button.classList.add('code-copy-button-checked');
    var currentTitle = button.getAttribute("title");
    button.setAttribute("title", "Copied!");
    let tooltip;
    if (window.bootstrap) {
      button.setAttribute("data-bs-toggle", "tooltip");
      button.setAttribute("data-bs-placement", "left");
      button.setAttribute("data-bs-title", "Copied!");
      tooltip = new bootstrap.Tooltip(button, 
        { trigger: "manual", 
          customClass: "code-copy-button-tooltip",
          offset: [0, -8]});
      tooltip.show();    
    }
    setTimeout(function() {
      if (tooltip) {
        tooltip.hide();
        button.removeAttribute("data-bs-title");
        button.removeAttribute("data-bs-toggle");
        button.removeAttribute("data-bs-placement");
      }
      button.setAttribute("title", currentTitle);
      button.classList.remove('code-copy-button-checked');
    }, 1000);
    // clear code selection
    e.clearSelection();
  }
  const getTextToCopy = function(trigger) {
      const codeEl = trigger.previousElementSibling.cloneNode(true);
      for (const childEl of codeEl.children) {
        if (isCodeAnnotation(childEl)) {
          childEl.remove();
        }
      }
      return codeEl.innerText;
  }
  const clipboard = new window.ClipboardJS('.code-copy-button:not([data-in-quarto-modal])', {
    text: getTextToCopy
  });
  clipboard.on('success', onCopySuccess);
  if (window.document.getElementById('quarto-embedded-source-code-modal')) {
    // For code content inside modals, clipBoardJS needs to be initialized with a container option
    // TODO: Check when it could be a function (https://github.com/zenorocha/clipboard.js/issues/860)
    const clipboardModal = new window.ClipboardJS('.code-copy-button[data-in-quarto-modal]', {
      text: getTextToCopy,
      container: window.document.getElementById('quarto-embedded-source-code-modal')
    });
    clipboardModal.on('success', onCopySuccess);
  }
  const viewSource = window.document.getElementById('quarto-view-source') ||
                     window.document.getElementById('quarto-code-tools-source');
  if (viewSource) {
    const sourceUrl = viewSource.getAttribute("data-quarto-source-url");
    viewSource.addEventListener("click", function(e) {
      if (sourceUrl) {
        // rstudio viewer pane
        if (/\bcapabilities=\b/.test(window.location)) {
          window.open(sourceUrl);
        } else {
          window.location.href = sourceUrl;
        }
      } else {
        const modal = new bootstrap.Modal(document.getElementById('quarto-embedded-source-code-modal'));
        modal.show();
      }
      return false;
    });
  }
  function toggleCodeHandler(show) {
    return function(e) {
      const detailsSrc = window.document.querySelectorAll(".cell > details > .sourceCode");
      for (let i=0; i<detailsSrc.length; i++) {
        const details = detailsSrc[i].parentElement;
        if (show) {
          details.open = true;
        } else {
          details.removeAttribute("open");
        }
      }
      const cellCodeDivs = window.document.querySelectorAll(".cell > .sourceCode");
      const fromCls = show ? "hidden" : "unhidden";
      const toCls = show ? "unhidden" : "hidden";
      for (let i=0; i<cellCodeDivs.length; i++) {
        const codeDiv = cellCodeDivs[i];
        if (codeDiv.classList.contains(fromCls)) {
          codeDiv.classList.remove(fromCls);
          codeDiv.classList.add(toCls);
        } 
      }
      return false;
    }
  }
  const hideAllCode = window.document.getElementById("quarto-hide-all-code");
  if (hideAllCode) {
    hideAllCode.addEventListener("click", toggleCodeHandler(false));
  }
  const showAllCode = window.document.getElementById("quarto-show-all-code");
  if (showAllCode) {
    showAllCode.addEventListener("click", toggleCodeHandler(true));
  }
    var localhostRegex = new RegExp(/^(?:http|https):\/\/localhost\:?[0-9]*\//);
    var mailtoRegex = new RegExp(/^mailto:/);
      var filterRegex = new RegExp('/' + window.location.host + '/');
    var isInternal = (href) => {
        return filterRegex.test(href) || localhostRegex.test(href) || mailtoRegex.test(href);
    }
    // Inspect non-navigation links and adorn them if external
 	var links = window.document.querySelectorAll('a[href]:not(.nav-link):not(.navbar-brand):not(.toc-action):not(.sidebar-link):not(.sidebar-item-toggle):not(.pagination-link):not(.no-external):not([aria-hidden]):not(.dropdown-item):not(.quarto-navigation-tool):not(.about-link)');
    for (var i=0; i<links.length; i++) {
      const link = links[i];
      if (!isInternal(link.href)) {
        // undo the damage that might have been done by quarto-nav.js in the case of
        // links that we want to consider external
        if (link.dataset.originalHref !== undefined) {
          link.href = link.dataset.originalHref;
        }
      }
    }
  function tippyHover(el, contentFn, onTriggerFn, onUntriggerFn) {
    const config = {
      allowHTML: true,
      maxWidth: 500,
      delay: 100,
      arrow: false,
      appendTo: function(el) {
          return el.parentElement;
      },
      interactive: true,
      interactiveBorder: 10,
      theme: 'quarto',
      placement: 'bottom-start',
    };
    if (contentFn) {
      config.content = contentFn;
    }
    if (onTriggerFn) {
      config.onTrigger = onTriggerFn;
    }
    if (onUntriggerFn) {
      config.onUntrigger = onUntriggerFn;
    }
    window.tippy(el, config); 
  }
  const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
  for (var i=0; i<noterefs.length; i++) {
    const ref = noterefs[i];
    tippyHover(ref, function() {
      // use id or data attribute instead here
      let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
      try { href = new URL(href).hash; } catch {}
      const id = href.replace(/^#\/?/, "");
      const note = window.document.getElementById(id);
      if (note) {
        return note.innerHTML;
      } else {
        return "";
      }
    });
  }
  const xrefs = window.document.querySelectorAll('a.quarto-xref');
  const processXRef = (id, note) => {
    // Strip column container classes
    const stripColumnClz = (el) => {
      el.classList.remove("page-full", "page-columns");
      if (el.children) {
        for (const child of el.children) {
          stripColumnClz(child);
        }
      }
    }
    stripColumnClz(note)
    if (id === null || id.startsWith('sec-')) {
      // Special case sections, only their first couple elements
      const container = document.createElement("div");
      if (note.children && note.children.length > 2) {
        container.appendChild(note.children[0].cloneNode(true));
        for (let i = 1; i < note.children.length; i++) {
          const child = note.children[i];
          if (child.tagName === "P" && child.innerText === "") {
            continue;
          } else {
            container.appendChild(child.cloneNode(true));
            break;
          }
        }
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(container);
        }
        return container.innerHTML
      } else {
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(note);
        }
        return note.innerHTML;
      }
    } else {
      // Remove any anchor links if they are present
      const anchorLink = note.querySelector('a.anchorjs-link');
      if (anchorLink) {
        anchorLink.remove();
      }
      if (window.Quarto?.typesetMath) {
        window.Quarto.typesetMath(note);
      }
      // TODO in 1.5, we should make sure this works without a callout special case
      if (note.classList.contains("callout")) {
        return note.outerHTML;
      } else {
        return note.innerHTML;
      }
    }
  }
  for (var i=0; i<xrefs.length; i++) {
    const xref = xrefs[i];
    tippyHover(xref, undefined, function(instance) {
      instance.disable();
      let url = xref.getAttribute('href');
      let hash = undefined; 
      if (url.startsWith('#')) {
        hash = url;
      } else {
        try { hash = new URL(url).hash; } catch {}
      }
      if (hash) {
        const id = hash.replace(/^#\/?/, "");
        const note = window.document.getElementById(id);
        if (note !== null) {
          try {
            const html = processXRef(id, note.cloneNode(true));
            instance.setContent(html);
          } finally {
            instance.enable();
            instance.show();
          }
        } else {
          // See if we can fetch this
          fetch(url.split('#')[0])
          .then(res => res.text())
          .then(html => {
            const parser = new DOMParser();
            const htmlDoc = parser.parseFromString(html, "text/html");
            const note = htmlDoc.getElementById(id);
            if (note !== null) {
              const html = processXRef(id, note);
              instance.setContent(html);
            } 
          }).finally(() => {
            instance.enable();
            instance.show();
          });
        }
      } else {
        // See if we can fetch a full url (with no hash to target)
        // This is a special case and we should probably do some content thinning / targeting
        fetch(url)
        .then(res => res.text())
        .then(html => {
          const parser = new DOMParser();
          const htmlDoc = parser.parseFromString(html, "text/html");
          const note = htmlDoc.querySelector('main.content');
          if (note !== null) {
            // This should only happen for chapter cross references
            // (since there is no id in the URL)
            // remove the first header
            if (note.children.length > 0 && note.children[0].tagName === "HEADER") {
              note.children[0].remove();
            }
            const html = processXRef(null, note);
            instance.setContent(html);
          } 
        }).finally(() => {
          instance.enable();
          instance.show();
        });
      }
    }, function(instance) {
    });
  }
      let selectedAnnoteEl;
      const selectorForAnnotation = ( cell, annotation) => {
        let cellAttr = 'data-code-cell="' + cell + '"';
        let lineAttr = 'data-code-annotation="' +  annotation + '"';
        const selector = 'span[' + cellAttr + '][' + lineAttr + ']';
        return selector;
      }
      const selectCodeLines = (annoteEl) => {
        const doc = window.document;
        const targetCell = annoteEl.getAttribute("data-target-cell");
        const targetAnnotation = annoteEl.getAttribute("data-target-annotation");
        const annoteSpan = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
        const lines = annoteSpan.getAttribute("data-code-lines").split(",");
        const lineIds = lines.map((line) => {
          return targetCell + "-" + line;
        })
        let top = null;
        let height = null;
        let parent = null;
        if (lineIds.length > 0) {
            //compute the position of the single el (top and bottom and make a div)
            const el = window.document.getElementById(lineIds[0]);
            top = el.offsetTop;
            height = el.offsetHeight;
            parent = el.parentElement.parentElement;
          if (lineIds.length > 1) {
            const lastEl = window.document.getElementById(lineIds[lineIds.length - 1]);
            const bottom = lastEl.offsetTop + lastEl.offsetHeight;
            height = bottom - top;
          }
          if (top !== null && height !== null && parent !== null) {
            // cook up a div (if necessary) and position it 
            let div = window.document.getElementById("code-annotation-line-highlight");
            if (div === null) {
              div = window.document.createElement("div");
              div.setAttribute("id", "code-annotation-line-highlight");
              div.style.position = 'absolute';
              parent.appendChild(div);
            }
            div.style.top = top - 2 + "px";
            div.style.height = height + 4 + "px";
            div.style.left = 0;
            let gutterDiv = window.document.getElementById("code-annotation-line-highlight-gutter");
            if (gutterDiv === null) {
              gutterDiv = window.document.createElement("div");
              gutterDiv.setAttribute("id", "code-annotation-line-highlight-gutter");
              gutterDiv.style.position = 'absolute';
              const codeCell = window.document.getElementById(targetCell);
              const gutter = codeCell.querySelector('.code-annotation-gutter');
              gutter.appendChild(gutterDiv);
            }
            gutterDiv.style.top = top - 2 + "px";
            gutterDiv.style.height = height + 4 + "px";
          }
          selectedAnnoteEl = annoteEl;
        }
      };
      const unselectCodeLines = () => {
        const elementsIds = ["code-annotation-line-highlight", "code-annotation-line-highlight-gutter"];
        elementsIds.forEach((elId) => {
          const div = window.document.getElementById(elId);
          if (div) {
            div.remove();
          }
        });
        selectedAnnoteEl = undefined;
      };
        // Handle positioning of the toggle
    window.addEventListener(
      "resize",
      throttle(() => {
        elRect = undefined;
        if (selectedAnnoteEl) {
          selectCodeLines(selectedAnnoteEl);
        }
      }, 10)
    );
    function throttle(fn, ms) {
    let throttle = false;
    let timer;
      return (...args) => {
        if(!throttle) { // first call gets through
            fn.apply(this, args);
            throttle = true;
        } else { // all the others get throttled
            if(timer) clearTimeout(timer); // cancel #2
            timer = setTimeout(() => {
              fn.apply(this, args);
              timer = throttle = false;
            }, ms);
        }
      };
    }
      // Attach click handler to the DT
      const annoteDls = window.document.querySelectorAll('dt[data-target-cell]');
      for (const annoteDlNode of annoteDls) {
        annoteDlNode.addEventListener('click', (event) => {
          const clickedEl = event.target;
          if (clickedEl !== selectedAnnoteEl) {
            unselectCodeLines();
            const activeEl = window.document.querySelector('dt[data-target-cell].code-annotation-active');
            if (activeEl) {
              activeEl.classList.remove('code-annotation-active');
            }
            selectCodeLines(clickedEl);
            clickedEl.classList.add('code-annotation-active');
          } else {
            // Unselect the line
            unselectCodeLines();
            clickedEl.classList.remove('code-annotation-active');
          }
        });
      }
  const findCites = (el) => {
    const parentEl = el.parentElement;
    if (parentEl) {
      const cites = parentEl.dataset.cites;
      if (cites) {
        return {
          el,
          cites: cites.split(' ')
        };
      } else {
        return findCites(el.parentElement)
      }
    } else {
      return undefined;
    }
  };
  var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
  for (var i=0; i<bibliorefs.length; i++) {
    const ref = bibliorefs[i];
    const citeInfo = findCites(ref);
    if (citeInfo) {
      tippyHover(citeInfo.el, function() {
        var popup = window.document.createElement('div');
        citeInfo.cites.forEach(function(cite) {
          var citeDiv = window.document.createElement('div');
          citeDiv.classList.add('hanging-indent');
          citeDiv.classList.add('csl-entry');
          var biblioDiv = window.document.getElementById('ref-' + cite);
          if (biblioDiv) {
            citeDiv.innerHTML = biblioDiv.innerHTML;
          }
          popup.appendChild(citeDiv);
        });
        return popup.innerHTML;
      });
    }
  }
});
</script>
<nav class="page-navigation">
  <div class="nav-page nav-page-previous">
      <a href="../../scripts/notebooks/index.html" class="pagination-link" aria-label="Pectobacterium genus pangenome">
        <i class="bi bi-arrow-left-short"></i> <span class="nav-page-text">Pectobacterium genus pangenome</span>
      </a>          
  </div>
  <div class="nav-page nav-page-next">
      <a href="../../scripts/analysis/ani_clustering.html" class="pagination-link" aria-label="ANI">
        <span class="nav-page-text">ANI</span> <i class="bi bi-arrow-right-short"></i>
      </a>
  </div>
</nav><div class="modal fade" id="quarto-embedded-source-code-modal" tabindex="-1" aria-labelledby="quarto-embedded-source-code-modal-label" aria-hidden="true"><div class="modal-dialog modal-dialog-scrollable"><div class="modal-content"><div class="modal-header"><h5 class="modal-title" id="quarto-embedded-source-code-modal-label">Source Code</h5><button class="btn-close" data-bs-dismiss="modal"></button></div><div class="modal-body"><div class="">
<div class="sourceCode" id="cb19" data-shortcodes="false"><pre class="sourceCode numberSource markdown number-lines code-with-copy"><code class="sourceCode markdown"><span id="cb19-1"><a href="#cb19-1"></a><span class="fu"># Preprocessing</span></span>
<span id="cb19-2"><a href="#cb19-2"></a></span>
<span id="cb19-3"><a href="#cb19-3"></a><span class="fu">## Data download from NCBI using EUtils</span></span>
<span id="cb19-4"><a href="#cb19-4"></a></span>
<span id="cb19-5"><a href="#cb19-5"></a><span class="in">``` bash</span></span>
<span id="cb19-6"><a href="#cb19-6"></a><span class="co">## number of records available in NCBI</span></span>
<span id="cb19-7"><a href="#cb19-7"></a><span class="ex">esearch</span> <span class="at">-db</span> assembly <span class="at">-query</span> <span class="st">"pectobacterium [ORGN]"</span> <span class="at">-email</span> <span class="st">"lakhansing.pardeshi@wur.nl"</span></span>
<span id="cb19-8"><a href="#cb19-8"></a></span>
<span id="cb19-9"><a href="#cb19-9"></a><span class="co">## download the Assembly DocumentSummary file</span></span>
<span id="cb19-10"><a href="#cb19-10"></a><span class="ex">esearch</span> <span class="at">-db</span> assembly <span class="at">-query</span> <span class="st">"pectobacterium [ORGN]"</span> <span class="at">-email</span> <span class="st">"lakhansing.pardeshi@wur.nl"</span> <span class="kw">|</span> <span class="dt">\</span></span>
<span id="cb19-11"><a href="#cb19-11"></a><span class="ex">efetch</span> <span class="at">-format</span> docsum <span class="op">&gt;</span> data/reference_data/assembly_docsum.xml</span>
<span id="cb19-12"><a href="#cb19-12"></a></span>
<span id="cb19-13"><a href="#cb19-13"></a><span class="co">## download associated BioSample DocumentSummary file</span></span>
<span id="cb19-14"><a href="#cb19-14"></a><span class="ex">esearch</span> <span class="at">-db</span> assembly <span class="at">-query</span> <span class="st">"pectobacterium [ORGN]"</span> <span class="at">-email</span> <span class="st">"lakhansing.pardeshi@wur.nl"</span> <span class="kw">|</span> <span class="dt">\</span></span>
<span id="cb19-15"><a href="#cb19-15"></a><span class="ex">elink</span> <span class="at">-target</span> biosample  <span class="kw">|</span> <span class="dt">\</span></span>
<span id="cb19-16"><a href="#cb19-16"></a><span class="ex">efetch</span> <span class="at">-format</span> docsum <span class="op">&gt;</span> data/reference_data/biosample_docsum.xml</span>
<span id="cb19-17"><a href="#cb19-17"></a></span>
<span id="cb19-18"><a href="#cb19-18"></a></span>
<span id="cb19-19"><a href="#cb19-19"></a><span class="co">## get all FPT paths for NCBI assemblies</span></span>
<span id="cb19-20"><a href="#cb19-20"></a><span class="ex">esearch</span> <span class="at">-db</span> assembly <span class="at">-query</span> <span class="st">"pectobacterium [ORGN]"</span> <span class="kw">|</span> <span class="dt">\</span></span>
<span id="cb19-21"><a href="#cb19-21"></a><span class="ex">efetch</span> <span class="at">-format</span> docsum <span class="at">-email</span> <span class="st">"lakhansing.pardeshi@wur.nl"</span> <span class="kw">|</span> <span class="dt">\</span></span>
<span id="cb19-22"><a href="#cb19-22"></a><span class="ex">xtract</span> <span class="at">-pattern</span> DocumentSummary <span class="dt">\</span></span>
<span id="cb19-23"><a href="#cb19-23"></a>-tab <span class="st">","</span> <span class="at">-element</span> AssemblyAccession AssemblyName FtpPath_RefSeq FtpPath_GenBank <span class="kw">|</span> <span class="dt">\</span></span>
<span id="cb19-24"><a href="#cb19-24"></a><span class="fu">awk</span> <span class="st">'</span></span>
<span id="cb19-25"><a href="#cb19-25"></a><span class="st">    BEGIN { FS=","};</span></span>
<span id="cb19-26"><a href="#cb19-26"></a><span class="st">    {</span></span>
<span id="cb19-27"><a href="#cb19-27"></a><span class="st">    if($4=="") file=$3; else file=$3;</span></span>
<span id="cb19-28"><a href="#cb19-28"></a><span class="st">    gsub(/ +/, "_", $0);</span></span>
<span id="cb19-29"><a href="#cb19-29"></a><span class="st">    printf "%s/%s_%s_genomic.fna.gz\n", file, $1, $2 &gt; "data/reference_data/ncbi_assembly_ftp.txt";</span></span>
<span id="cb19-30"><a href="#cb19-30"></a><span class="st">    printf "%s_%s\n", $1, $2 &gt; "data/reference_data/ncbi_assembly_ids.txt";</span></span>
<span id="cb19-31"><a href="#cb19-31"></a><span class="st">    }</span></span>
<span id="cb19-32"><a href="#cb19-32"></a><span class="st">'</span></span>
<span id="cb19-33"><a href="#cb19-33"></a></span>
<span id="cb19-34"><a href="#cb19-34"></a><span class="co">## NCBI prokaryotes ANI file</span></span>
<span id="cb19-35"><a href="#cb19-35"></a><span class="fu">wget</span> <span class="at">--timestamping</span> https://ftp.ncbi.nlm.nih.gov/genomes/ASSEMBLY_REPORTS/ANI_report_prokaryotes.txt <span class="at">-P</span> data/other/</span>
<span id="cb19-36"><a href="#cb19-36"></a></span>
<span id="cb19-37"><a href="#cb19-37"></a></span>
<span id="cb19-38"><a href="#cb19-38"></a><span class="fu">cat</span> data/reference_data/local_assembly_ids.txt data/reference_data/ncbi_assembly_ids.txt <span class="dt">\</span></span>
<span id="cb19-39"><a href="#cb19-39"></a><span class="op">&gt;</span> data/reference_data/pre_qc_assembly_ids.txt</span>
<span id="cb19-40"><a href="#cb19-40"></a></span>
<span id="cb19-41"><a href="#cb19-41"></a><span class="fu">mkdir</span> data/reference_data/batches</span>
<span id="cb19-42"><a href="#cb19-42"></a><span class="fu">split</span> <span class="at">-a</span> 2 <span class="at">-d</span> <span class="at">-n</span> r/8 data/reference_data/pre_qc_assembly_ids.txt data/reference_data/batches/temp_ids_batch.</span>
<span id="cb19-43"><a href="#cb19-43"></a></span>
<span id="cb19-44"><a href="#cb19-44"></a><span class="in">```</span></span>
<span id="cb19-45"><a href="#cb19-45"></a></span>
<span id="cb19-46"><a href="#cb19-46"></a><span class="fu">## Check if there are any new assemblies released on NCBI</span></span>
<span id="cb19-47"><a href="#cb19-47"></a></span>
<span id="cb19-48"><a href="#cb19-48"></a><span class="in">``` bash</span></span>
<span id="cb19-49"><a href="#cb19-49"></a><span class="co">## check and extract new assembly IDs</span></span>
<span id="cb19-50"><a href="#cb19-50"></a><span class="cf">for</span> sampleId <span class="kw">in</span> <span class="kw">`</span><span class="fu">cat</span> data/reference_data/pre_qc_assembly_ids.txt<span class="kw">`</span></span>
<span id="cb19-51"><a href="#cb19-51"></a><span class="cf">do</span></span>
<span id="cb19-52"><a href="#cb19-52"></a>    <span class="cf">if</span> <span class="ot">! </span><span class="fu">find</span> data/genomes/ <span class="at">-name</span> <span class="va">${sampleId}</span><span class="pp">*</span> <span class="kw">|</span> <span class="fu">grep</span> <span class="at">-q</span> <span class="st">"."</span></span>
<span id="cb19-53"><a href="#cb19-53"></a>    <span class="cf">then</span></span>
<span id="cb19-54"><a href="#cb19-54"></a>        <span class="bu">echo</span> <span class="va">$sampleId</span></span>
<span id="cb19-55"><a href="#cb19-55"></a>    <span class="cf">fi</span></span>
<span id="cb19-56"><a href="#cb19-56"></a><span class="cf">done</span> <span class="op">&gt;</span> data/reference_data/temp_assembly_ids.txt</span>
<span id="cb19-57"><a href="#cb19-57"></a></span>
<span id="cb19-58"><a href="#cb19-58"></a></span>
<span id="cb19-59"><a href="#cb19-59"></a><span class="co">## download and unzip the FASTA files</span></span>
<span id="cb19-60"><a href="#cb19-60"></a><span class="cf">for</span> sampleId <span class="kw">in</span> <span class="kw">`</span><span class="fu">cat</span> data/reference_data/temp_assembly_ids.txt<span class="kw">`</span></span>
<span id="cb19-61"><a href="#cb19-61"></a><span class="cf">do</span></span>
<span id="cb19-62"><a href="#cb19-62"></a>    <span class="fu">grep</span> <span class="st">"/</span><span class="va">$sampleId</span><span class="st">/"</span> data/reference_data/ncbi_assembly_ftp.txt</span>
<span id="cb19-63"><a href="#cb19-63"></a><span class="cf">done</span> <span class="kw">|</span> <span class="dt">\</span></span>
<span id="cb19-64"><a href="#cb19-64"></a><span class="fu">xargs</span> <span class="at">-I</span> {} wget <span class="at">--timestamping</span> {} <span class="at">-P</span> data/genomes/ncbi/</span>
<span id="cb19-65"><a href="#cb19-65"></a></span>
<span id="cb19-66"><a href="#cb19-66"></a><span class="fu">gzip</span> <span class="at">-d</span> data/genomes/ncbi/<span class="pp">*</span>.gz</span>
<span id="cb19-67"><a href="#cb19-67"></a></span>
<span id="cb19-68"><a href="#cb19-68"></a><span class="in">```</span></span>
<span id="cb19-69"><a href="#cb19-69"></a></span>
<span id="cb19-70"><a href="#cb19-70"></a><span class="fu">## Run standard annotations for all the genomes</span></span>
<span id="cb19-71"><a href="#cb19-71"></a></span>
<span id="cb19-72"><a href="#cb19-72"></a><span class="fu">### Prokka annotation for bacterial genomes</span></span>
<span id="cb19-73"><a href="#cb19-73"></a></span>
<span id="cb19-74"><a href="#cb19-74"></a><span class="in">``` bash</span></span>
<span id="cb19-75"><a href="#cb19-75"></a><span class="fu">nohup</span> <span class="dt">\</span></span>
<span id="cb19-76"><a href="#cb19-76"></a>ls data/genomes/<span class="pp">?(</span>NCBI<span class="pp">|</span>WUR<span class="pp">|</span>NIVIP<span class="pp">|</span>NAK<span class="pp">)</span>/<span class="pp">*</span>.<span class="pp">?(</span>fa<span class="pp">|</span>fna<span class="pp">|</span>fasta<span class="pp">)</span> <span class="kw">|</span> <span class="dt">\</span></span>
<span id="cb19-77"><a href="#cb19-77"></a><span class="ex">parallel</span> <span class="at">--gnu</span> <span class="at">--keep-order</span> <span class="at">--jobs</span> 1 <span class="at">--halt</span> now,fail=1 <span class="dt">\</span></span>
<span id="cb19-78"><a href="#cb19-78"></a>--results logs/prokka/{/.} <span class="at">--joblog</span> logs/prokka/parallel.log <span class="dt">\</span></span>
<span id="cb19-79"><a href="#cb19-79"></a>scripts/preprocessing/prokka_annotation.sh {} <span class="dt">\</span></span>
<span id="cb19-80"><a href="#cb19-80"></a><span class="op">&gt;</span>logs/prokka/nohup_batch00.out <span class="dv">2</span><span class="op">&gt;&amp;</span><span class="dv">1</span> <span class="kw">&amp;</span></span>
<span id="cb19-81"><a href="#cb19-81"></a></span>
<span id="cb19-82"><a href="#cb19-82"></a><span class="in">```</span></span>
<span id="cb19-83"><a href="#cb19-83"></a></span>
<span id="cb19-84"><a href="#cb19-84"></a>Processing in parallel is failing without any specific error. Hence the</span>
<span id="cb19-85"><a href="#cb19-85"></a>data will be processed in serial mode. Further debugging is required to</span>
<span id="cb19-86"><a href="#cb19-86"></a>understand the parallel processing failure. For now, running GNU <span class="in">`parallel`</span></span>
<span id="cb19-87"><a href="#cb19-87"></a>with <span class="in">`--jobs 1`</span> setting in serial mode.</span>
<span id="cb19-88"><a href="#cb19-88"></a></span>
<span id="cb19-89"><a href="#cb19-89"></a><span class="fu">#### Remove the FASTA sequence at the end of prokka gff files</span></span>
<span id="cb19-90"><a href="#cb19-90"></a></span>
<span id="cb19-91"><a href="#cb19-91"></a><span class="in">``` bash</span></span>
<span id="cb19-92"><a href="#cb19-92"></a><span class="cf">for</span> i <span class="kw">in</span> <span class="kw">`</span><span class="fu">cat</span> data/reference_data/pre_qc_assembly_ids.txt<span class="kw">`</span></span>
<span id="cb19-93"><a href="#cb19-93"></a><span class="cf">do</span></span>
<span id="cb19-94"><a href="#cb19-94"></a><span class="fu">sed</span> <span class="at">-n</span> <span class="st">'1,/##FASTA/ {/##FASTA/!p}'</span> data/prokka_annotation/<span class="va">${i}</span>/<span class="va">${i}</span>.gff <span class="op">&gt;</span> <span class="dt">\</span></span>
<span id="cb19-95"><a href="#cb19-95"></a>data/prokka_annotation/<span class="va">${i}</span>/<span class="va">${i}</span>.gff3</span>
<span id="cb19-96"><a href="#cb19-96"></a><span class="cf">done</span></span>
<span id="cb19-97"><a href="#cb19-97"></a></span>
<span id="cb19-98"><a href="#cb19-98"></a><span class="in">```</span></span>
<span id="cb19-99"><a href="#cb19-99"></a></span>
<span id="cb19-100"><a href="#cb19-100"></a><span class="fu">#### FASTA file indexing</span></span>
<span id="cb19-101"><a href="#cb19-101"></a></span>
<span id="cb19-102"><a href="#cb19-102"></a><span class="in">``` bash</span></span>
<span id="cb19-103"><a href="#cb19-103"></a><span class="bu">printf</span> <span class="st">''</span> <span class="op">&gt;</span> analysis/QC/assembly_chr_size.txt</span>
<span id="cb19-104"><a href="#cb19-104"></a><span class="bu">printf</span> <span class="st">''</span> <span class="op">&gt;</span> data/reference_data/pre_qc_genomes_fa.list</span>
<span id="cb19-105"><a href="#cb19-105"></a><span class="ex">conda</span> activate omics_py37</span>
<span id="cb19-106"><a href="#cb19-106"></a></span>
<span id="cb19-107"><a href="#cb19-107"></a><span class="cf">for</span> i <span class="kw">in</span> <span class="kw">`</span><span class="fu">cat</span> data/reference_data/pre_qc_assembly_ids.txt<span class="kw">`</span></span>
<span id="cb19-108"><a href="#cb19-108"></a><span class="cf">do</span></span>
<span id="cb19-109"><a href="#cb19-109"></a>    <span class="ex">samtools</span> faidx data/prokka_annotation/<span class="va">${i}</span>/<span class="va">${i}</span>.fna</span>
<span id="cb19-110"><a href="#cb19-110"></a>    <span class="fu">sort</span> <span class="at">-r</span> <span class="at">-n</span> <span class="at">-k</span> 2,2 data/prokka_annotation/<span class="va">${i}</span>/<span class="va">${i}</span>.fna.fai <span class="kw">|</span> <span class="dt">\</span></span>
<span id="cb19-111"><a href="#cb19-111"></a>    <span class="fu">awk</span>  <span class="at">-v</span> i=<span class="va">${i}</span> <span class="st">'{print i, "\t", $1, "\t", $2}'</span> <span class="op">&gt;&gt;</span> analysis/QC/assembly_chr_size.txt</span>
<span id="cb19-112"><a href="#cb19-112"></a>    <span class="fu">ls</span> data/prokka_annotation/<span class="va">${i}</span>/<span class="va">${i}</span>.fna <span class="op">&gt;&gt;</span> data/reference_data/pre_qc_genomes_fa.list</span>
<span id="cb19-113"><a href="#cb19-113"></a></span>
<span id="cb19-114"><a href="#cb19-114"></a>    <span class="ex">faToTwoBit</span> data/prokka_annotation/<span class="va">${i}</span>/<span class="va">${i}</span>.fna data/prokka_annotation/<span class="va">${i}</span>/<span class="va">${i}</span>.2bit</span>
<span id="cb19-115"><a href="#cb19-115"></a></span>
<span id="cb19-116"><a href="#cb19-116"></a><span class="cf">done</span></span>
<span id="cb19-117"><a href="#cb19-117"></a></span>
<span id="cb19-118"><a href="#cb19-118"></a><span class="in">```</span></span>
<span id="cb19-119"><a href="#cb19-119"></a></span>
<span id="cb19-120"><a href="#cb19-120"></a><span class="fu">### BUSCO protein and genome assembly evaluation</span></span>
<span id="cb19-121"><a href="#cb19-121"></a></span>
<span id="cb19-122"><a href="#cb19-122"></a><span class="in">``` bash</span></span>
<span id="cb19-123"><a href="#cb19-123"></a><span class="co">## !GNU parallel on single server</span></span>
<span id="cb19-124"><a href="#cb19-124"></a><span class="fu">nohup</span> <span class="dt">\</span></span>
<span id="cb19-125"><a href="#cb19-125"></a>cat data/reference_data/temp_assembly_ids.txt <span class="kw">|</span> <span class="dt">\</span></span>
<span id="cb19-126"><a href="#cb19-126"></a><span class="ex">parallel</span> <span class="at">--jobs</span> 1 <span class="at">--workdir</span> <span class="va">$PWD</span> <span class="at">--halt</span> now,fail=1 <span class="at">--keep-order</span> <span class="dt">\</span></span>
<span id="cb19-127"><a href="#cb19-127"></a>--results logs/busco/{} <span class="at">--joblog</span> logs/busco/parallel.log <span class="dt">\</span></span>
<span id="cb19-128"><a href="#cb19-128"></a><span class="va">$PWD</span>/scripts/preprocessing/qc_busco.sh {} <span class="dt">\</span></span>
<span id="cb19-129"><a href="#cb19-129"></a><span class="op">&gt;</span>logs/busco/nohup03.out <span class="dv">2</span><span class="op">&gt;&amp;</span><span class="dv">1</span> <span class="kw">&amp;</span></span>
<span id="cb19-130"><a href="#cb19-130"></a></span>
<span id="cb19-131"><a href="#cb19-131"></a><span class="in">```</span></span>
<span id="cb19-132"><a href="#cb19-132"></a></span>
<span id="cb19-133"><a href="#cb19-133"></a>This is failing with <span class="in">`parallel`</span>, most likely because of the number of open files</span>
<span id="cb19-134"><a href="#cb19-134"></a>exceeding the <span class="in">`ulimit`</span>. Need to debug further. For now, running GNU <span class="in">`parallel`</span></span>
<span id="cb19-135"><a href="#cb19-135"></a>with <span class="in">`--jobs 1`</span> setting is serial mode.</span>
<span id="cb19-136"><a href="#cb19-136"></a></span>
<span id="cb19-137"><a href="#cb19-137"></a><span class="fu">### QUAST assembly evaluation</span></span>
<span id="cb19-138"><a href="#cb19-138"></a></span>
<span id="cb19-139"><a href="#cb19-139"></a><span class="in">``` bash</span></span>
<span id="cb19-140"><a href="#cb19-140"></a><span class="fu">nohup</span> <span class="dt">\</span></span>
<span id="cb19-141"><a href="#cb19-141"></a>cat data/reference_data/temp_assembly_ids.txt <span class="kw">|</span> <span class="dt">\</span></span>
<span id="cb19-142"><a href="#cb19-142"></a><span class="ex">parallel</span> <span class="at">--jobs</span> 6 <span class="at">--workdir</span> <span class="va">$PWD</span> <span class="at">--halt</span> now,fail=1 <span class="dt">\</span></span>
<span id="cb19-143"><a href="#cb19-143"></a>--keep-order <span class="at">--results</span> logs/quast/{} <span class="dt">\</span></span>
<span id="cb19-144"><a href="#cb19-144"></a>--joblog logs/quast/parallel.log <span class="dt">\</span></span>
<span id="cb19-145"><a href="#cb19-145"></a>./scripts/preprocessing/qc_quast.sh {} <span class="dt">\</span></span>
<span id="cb19-146"><a href="#cb19-146"></a><span class="op">&gt;&gt;</span>logs/quast/nohup.out <span class="dv">2</span><span class="op">&gt;&amp;</span><span class="dv">1</span> <span class="kw">&amp;</span></span>
<span id="cb19-147"><a href="#cb19-147"></a><span class="in">```</span></span>
<span id="cb19-148"><a href="#cb19-148"></a></span>
<span id="cb19-149"><a href="#cb19-149"></a><span class="fu">### InterProScan functional annotation</span></span>
<span id="cb19-150"><a href="#cb19-150"></a></span>
<span id="cb19-151"><a href="#cb19-151"></a><span class="in">``` bash</span></span>
<span id="cb19-152"><a href="#cb19-152"></a><span class="co">## GNU parallel on single server</span></span>
<span id="cb19-153"><a href="#cb19-153"></a><span class="fu">nohup</span> <span class="dt">\</span></span>
<span id="cb19-154"><a href="#cb19-154"></a>cat data/reference_data/temp_assembly_ids.txt <span class="kw">|</span> <span class="dt">\</span></span>
<span id="cb19-155"><a href="#cb19-155"></a><span class="ex">parallel</span> <span class="at">--jobs</span> 6 <span class="at">--workdir</span> <span class="va">$PWD</span> <span class="at">--halt</span> now,fail=1 <span class="dt">\</span></span>
<span id="cb19-156"><a href="#cb19-156"></a>--keep-order <span class="at">--results</span> logs/interproscan/{} <span class="dt">\</span></span>
<span id="cb19-157"><a href="#cb19-157"></a>--joblog logs/interproscan/parallel_batch06.log <span class="dt">\</span></span>
<span id="cb19-158"><a href="#cb19-158"></a>./scripts/preprocessing/interproscan_annotation.sh {} <span class="dt">\</span></span>
<span id="cb19-159"><a href="#cb19-159"></a><span class="op">&gt;&gt;</span>logs/interproscan/nohup_batch07.out <span class="dv">2</span><span class="op">&gt;&amp;</span><span class="dv">1</span> <span class="kw">&amp;</span></span>
<span id="cb19-160"><a href="#cb19-160"></a></span>
<span id="cb19-161"><a href="#cb19-161"></a></span>
<span id="cb19-162"><a href="#cb19-162"></a><span class="co">## GNU parallel: submit jobs to another server</span></span>
<span id="cb19-163"><a href="#cb19-163"></a><span class="fu">head</span> <span class="at">-n</span> 10 data/reference_data/sample_subsets/genebank_ass.txt <span class="kw">|</span> <span class="dt">\</span></span>
<span id="cb19-164"><a href="#cb19-164"></a><span class="ex">env_parallel</span> <span class="at">--jobs</span> 4 <span class="at">--workdir</span> <span class="va">$PWD</span> <span class="at">--halt</span> now,fail=1 <span class="dt">\</span></span>
<span id="cb19-165"><a href="#cb19-165"></a>--keep-order <span class="at">--results</span> logs/interproscan/{} <span class="dt">\</span></span>
<span id="cb19-166"><a href="#cb19-166"></a>--joblog logs/interproscan/parallel.log <span class="dt">\</span></span>
<span id="cb19-167"><a href="#cb19-167"></a>--sshlogin 4/waterman.bioinformatics.nl <span class="at">--cleanup</span> <span class="dt">\</span></span>
<span id="cb19-168"><a href="#cb19-168"></a>--env error_exit <span class="at">--env</span> process_start <span class="at">--env</span> TOOLS_PATH <span class="at">--env</span> LUSTRE_HOME <span class="dt">\</span></span>
<span id="cb19-169"><a href="#cb19-169"></a>./scripts/preprocessing/interproscan_annotation.sh {}</span>
<span id="cb19-170"><a href="#cb19-170"></a></span>
<span id="cb19-171"><a href="#cb19-171"></a><span class="in">```</span></span>
<span id="cb19-172"><a href="#cb19-172"></a></span>
<span id="cb19-173"><a href="#cb19-173"></a><span class="fu">### eggNOG annotation</span></span>
<span id="cb19-174"><a href="#cb19-174"></a></span>
<span id="cb19-175"><a href="#cb19-175"></a><span class="in">``` bash</span></span>
<span id="cb19-176"><a href="#cb19-176"></a><span class="ex">conda</span> activate eggnog</span>
<span id="cb19-177"><a href="#cb19-177"></a></span>
<span id="cb19-178"><a href="#cb19-178"></a><span class="co">## download eggNOG data</span></span>
<span id="cb19-179"><a href="#cb19-179"></a><span class="ex">download_eggnog_data.py</span> <span class="at">-M</span> <span class="at">-F</span> <span class="at">-H</span> <span class="at">-d</span> 1 <span class="at">--dbname</span> root <span class="at">-f</span> <span class="at">-y</span></span>
<span id="cb19-180"><a href="#cb19-180"></a><span class="ex">download_eggnog_data.py</span> <span class="at">-D</span> <span class="at">-H</span> <span class="at">-d</span> 2759 <span class="at">--dbname</span> eukaryota <span class="at">-q</span> <span class="at">-y</span></span>
<span id="cb19-181"><a href="#cb19-181"></a><span class="ex">download_eggnog_data.py</span> <span class="at">-D</span> <span class="at">-H</span> <span class="at">-d</span> 2 <span class="at">--dbname</span> bacteria <span class="at">-q</span> <span class="at">-y</span></span>
<span id="cb19-182"><a href="#cb19-182"></a><span class="ex">download_eggnog_data.py</span> <span class="at">-D</span> <span class="at">-H</span> <span class="at">-d</span> 10239 <span class="at">--dbname</span> viruses <span class="at">-q</span> <span class="at">-y</span></span>
<span id="cb19-183"><a href="#cb19-183"></a><span class="ex">download_eggnog_data.py</span> <span class="at">-D</span> <span class="at">-H</span> <span class="at">-d</span> 4751 <span class="at">--dbname</span> fungi <span class="at">-q</span> <span class="at">-y</span></span>
<span id="cb19-184"><a href="#cb19-184"></a><span class="ex">download_eggnog_data.py</span> <span class="at">-D</span> <span class="at">-H</span> <span class="at">-d</span> 2157 <span class="at">--dbname</span> archea <span class="at">-q</span> <span class="at">-y</span></span>
<span id="cb19-185"><a href="#cb19-185"></a></span>
<span id="cb19-186"><a href="#cb19-186"></a><span class="ex">mmseqs</span> createindex <span class="st">"</span><span class="va">$EGGNOG_DATA_DIR</span><span class="st">"</span>/mmseqs/mmseqs.db /local_scratch/parde001/tmp</span>
<span id="cb19-187"><a href="#cb19-187"></a></span>
<span id="cb19-188"><a href="#cb19-188"></a><span class="co">## create mmseq database for bacteria, archea and viruses</span></span>
<span id="cb19-189"><a href="#cb19-189"></a><span class="ex">create_dbs.py</span> <span class="at">-m</span> mmseqs <span class="at">--dbname</span> bact_arch_vir <span class="at">--taxids</span> 2,2157,10239 <span class="at">-y</span></span>
<span id="cb19-190"><a href="#cb19-190"></a></span>
<span id="cb19-191"><a href="#cb19-191"></a><span class="co">## annotate for eggNOG clusters</span></span>
<span id="cb19-192"><a href="#cb19-192"></a><span class="cf">for</span> i <span class="kw">in</span> <span class="kw">`</span><span class="fu">cat</span> data/reference_data/assembly_ids.txt<span class="kw">`</span></span>
<span id="cb19-193"><a href="#cb19-193"></a><span class="cf">do</span></span>
<span id="cb19-194"><a href="#cb19-194"></a>    <span class="va">file_aa</span><span class="op">=</span><span class="st">"data/prokka_annotation/</span><span class="va">${i}</span><span class="st">/</span><span class="va">${i}</span><span class="st">.faa"</span></span>
<span id="cb19-195"><a href="#cb19-195"></a>    </span>
<span id="cb19-196"><a href="#cb19-196"></a>    <span class="ex">emapper.py</span> <span class="at">--cpu</span> 20 <span class="at">-m</span> mmseqs <span class="at">--dbmem</span> <span class="at">--go_evidence</span> all  <span class="dt">\</span></span>
<span id="cb19-197"><a href="#cb19-197"></a>    <span class="at">-i</span> <span class="va">${file_aa}</span> <span class="at">-o</span> <span class="va">${i}</span> <span class="at">--output_dir</span> data/eggnog <span class="dt">\</span></span>
<span id="cb19-198"><a href="#cb19-198"></a>    <span class="at">--scratch_dir</span> /local_scratch/parde001/ <span class="at">--temp_dir</span> /local_scratch/parde001/tmp</span>
<span id="cb19-199"><a href="#cb19-199"></a></span>
<span id="cb19-200"><a href="#cb19-200"></a>    <span class="ex">error_exit</span> <span class="va">$?</span></span>
<span id="cb19-201"><a href="#cb19-201"></a><span class="cf">done</span></span>
<span id="cb19-202"><a href="#cb19-202"></a></span>
<span id="cb19-203"><a href="#cb19-203"></a><span class="in">```</span></span>
<span id="cb19-204"><a href="#cb19-204"></a></span>
<span id="cb19-205"><a href="#cb19-205"></a><span class="fu">### MultiQC</span></span>
<span id="cb19-206"><a href="#cb19-206"></a></span>
<span id="cb19-207"><a href="#cb19-207"></a><span class="in">``` bash</span></span>
<span id="cb19-208"><a href="#cb19-208"></a><span class="co">## QUAST MultiQC</span></span>
<span id="cb19-209"><a href="#cb19-209"></a><span class="fu">nohup</span> <span class="dt">\</span></span>
<span id="cb19-210"><a href="#cb19-210"></a>multiqc <span class="at">-f</span> <span class="at">--filename</span> quast_multiqc <span class="at">--interactive</span> <span class="at">--title</span> <span class="st">"QUAST report"</span> <span class="dt">\</span></span>
<span id="cb19-211"><a href="#cb19-211"></a>--outdir analysis/multiqc/ <span class="at">--module</span> quast data/quast/ <span class="dt">\</span></span>
<span id="cb19-212"><a href="#cb19-212"></a><span class="op">&gt;&gt;</span>nohup.out <span class="dv">2</span><span class="op">&gt;&amp;</span><span class="dv">1</span> <span class="kw">&amp;</span></span>
<span id="cb19-213"><a href="#cb19-213"></a></span>
<span id="cb19-214"><a href="#cb19-214"></a><span class="co">## BUSCO protein MultiQC</span></span>
<span id="cb19-215"><a href="#cb19-215"></a><span class="fu">nohup</span> <span class="dt">\</span></span>
<span id="cb19-216"><a href="#cb19-216"></a>multiqc <span class="at">-f</span> <span class="at">--filename</span> busco_prot_multiqc <span class="at">--interactive</span> <span class="at">--title</span> <span class="st">"BUSCO report"</span> <span class="dt">\</span></span>
<span id="cb19-217"><a href="#cb19-217"></a>--outdir analysis/multiqc/ <span class="at">--module</span> busco data/busco.prot/ <span class="dt">\</span></span>
<span id="cb19-218"><a href="#cb19-218"></a><span class="op">&gt;&gt;</span>nohup.out <span class="dv">2</span><span class="op">&gt;&amp;</span><span class="dv">1</span> <span class="kw">&amp;</span></span>
<span id="cb19-219"><a href="#cb19-219"></a></span>
<span id="cb19-220"><a href="#cb19-220"></a><span class="co">## BUSCO genome MultiQC</span></span>
<span id="cb19-221"><a href="#cb19-221"></a><span class="fu">nohup</span> <span class="dt">\</span></span>
<span id="cb19-222"><a href="#cb19-222"></a>multiqc <span class="at">-f</span> <span class="at">--filename</span> busco_geno_multiqc <span class="at">--interactive</span> <span class="at">--title</span> <span class="st">"BUSCO report"</span> <span class="dt">\</span></span>
<span id="cb19-223"><a href="#cb19-223"></a>--outdir analysis/multiqc/ <span class="at">--module</span> busco data/busco.geno/ <span class="dt">\</span></span>
<span id="cb19-224"><a href="#cb19-224"></a><span class="op">&gt;&gt;</span>nohup.out <span class="dv">2</span><span class="op">&gt;&amp;</span><span class="dv">1</span> <span class="kw">&amp;</span></span>
<span id="cb19-225"><a href="#cb19-225"></a></span>
<span id="cb19-226"><a href="#cb19-226"></a><span class="in">```</span></span>
<span id="cb19-227"><a href="#cb19-227"></a></span>
<span id="cb19-228"><a href="#cb19-228"></a><span class="fu">### fastANI</span></span>
<span id="cb19-229"><a href="#cb19-229"></a></span>
<span id="cb19-230"><a href="#cb19-230"></a><span class="in">``` bash</span></span>
<span id="cb19-231"><a href="#cb19-231"></a><span class="ex">process_start</span> <span class="st">"ANI on all genomes"</span></span>
<span id="cb19-232"><a href="#cb19-232"></a><span class="fu">nohup</span> bash scripts/preprocessing/ANI_processing.sh <span class="dt">\</span></span>
<span id="cb19-233"><a href="#cb19-233"></a>data/reference_data/pre_qc_genomes_fa.list <span class="dt">\</span></span>
<span id="cb19-234"><a href="#cb19-234"></a>analysis/02_fastANI/ANI_results <span class="op">&gt;</span>logs/fastANI.log <span class="dv">2</span><span class="op">&gt;&amp;</span><span class="dv">1</span> <span class="kw">&amp;</span></span>
<span id="cb19-235"><a href="#cb19-235"></a><span class="ex">error_exit</span> <span class="va">$?</span></span>
<span id="cb19-236"><a href="#cb19-236"></a><span class="in">```</span></span>
<span id="cb19-237"><a href="#cb19-237"></a></span>
<span id="cb19-238"><a href="#cb19-238"></a><span class="fu">### Defense-finder</span></span>
<span id="cb19-239"><a href="#cb19-239"></a></span>
<span id="cb19-240"><a href="#cb19-240"></a><span class="in">```bash</span></span>
<span id="cb19-241"><a href="#cb19-241"></a><span class="ex">conda</span> activate defensefinder</span>
<span id="cb19-242"><a href="#cb19-242"></a></span>
<span id="cb19-243"><a href="#cb19-243"></a><span class="cf">for</span> i <span class="kw">in</span> <span class="kw">`</span><span class="fu">cat</span> data/reference_data/assembly_ids.txt<span class="kw">`</span></span>
<span id="cb19-244"><a href="#cb19-244"></a><span class="cf">do</span></span>
<span id="cb19-245"><a href="#cb19-245"></a>    <span class="va">file_aa</span><span class="op">=</span><span class="st">"data/prokka_annotation/</span><span class="va">${i}</span><span class="st">/</span><span class="va">${i}</span><span class="st">.faa"</span></span>
<span id="cb19-246"><a href="#cb19-246"></a>    <span class="va">out_dir</span><span class="op">=</span><span class="st">"data/defense_finder/</span><span class="va">${i}</span><span class="st">"</span></span>
<span id="cb19-247"><a href="#cb19-247"></a>    <span class="fu">mkdir</span> <span class="at">-p</span> <span class="va">${out_dir}</span></span>
<span id="cb19-248"><a href="#cb19-248"></a></span>
<span id="cb19-249"><a href="#cb19-249"></a>    <span class="ex">process_start</span> <span class="st">"defense-finder: </span><span class="va">${i}</span><span class="st">"</span></span>
<span id="cb19-250"><a href="#cb19-250"></a></span>
<span id="cb19-251"><a href="#cb19-251"></a>    <span class="ex">defense-finder</span> run <span class="at">--workers</span> 20 <span class="at">-o</span> <span class="va">${out_dir}</span> <span class="dt">\</span></span>
<span id="cb19-252"><a href="#cb19-252"></a>    <span class="at">--models-dir</span> <span class="va">$TOOLS_PATH</span>/defense-finder-models <span class="va">${file_aa}</span></span>
<span id="cb19-253"><a href="#cb19-253"></a></span>
<span id="cb19-254"><a href="#cb19-254"></a>    <span class="ex">error_exit</span> <span class="va">$?</span></span>
<span id="cb19-255"><a href="#cb19-255"></a><span class="cf">done</span></span>
<span id="cb19-256"><a href="#cb19-256"></a><span class="in">```</span></span>
<span id="cb19-257"><a href="#cb19-257"></a></span>
<span id="cb19-258"><a href="#cb19-258"></a>Combine Defense-finder output into single file</span>
<span id="cb19-259"><a href="#cb19-259"></a></span>
<span id="cb19-260"><a href="#cb19-260"></a><span class="in">```bash</span></span>
<span id="cb19-261"><a href="#cb19-261"></a><span class="co"># combine systems</span></span>
<span id="cb19-262"><a href="#cb19-262"></a><span class="cf">for</span> i <span class="kw">in</span> <span class="kw">`</span><span class="fu">cat</span> data/reference_data/assembly_ids.txt<span class="kw">`</span></span>
<span id="cb19-263"><a href="#cb19-263"></a><span class="cf">do</span></span>
<span id="cb19-264"><a href="#cb19-264"></a>    <span class="fu">awk</span> <span class="at">-v</span> var=<span class="st">"</span><span class="va">${i}</span><span class="st">"</span> <span class="st">'!/^sys_id/ {print var "\t" $0}'</span> data/defense_finder/<span class="va">${i}</span>/<span class="va">${i}</span>_defense_finder_systems.tsv</span>
<span id="cb19-265"><a href="#cb19-265"></a><span class="cf">done</span> <span class="op">&gt;</span> data/defense_finder/defense_systems.tsv</span>
<span id="cb19-266"><a href="#cb19-266"></a></span>
<span id="cb19-267"><a href="#cb19-267"></a><span class="co"># combine genes</span></span>
<span id="cb19-268"><a href="#cb19-268"></a><span class="cf">for</span> i <span class="kw">in</span> <span class="kw">`</span><span class="fu">cat</span> data/reference_data/assembly_ids.txt<span class="kw">`</span></span>
<span id="cb19-269"><a href="#cb19-269"></a><span class="cf">do</span></span>
<span id="cb19-270"><a href="#cb19-270"></a>    <span class="fu">awk</span> <span class="at">-v</span> var=<span class="st">"</span><span class="va">${i}</span><span class="st">"</span> <span class="st">'NR&gt;1 {print var "\t" $0}'</span> data/defense_finder/<span class="va">${i}</span>/<span class="va">${i}</span>_defense_finder_genes.tsv</span>
<span id="cb19-271"><a href="#cb19-271"></a><span class="cf">done</span> <span class="op">&gt;</span> data/defense_finder/defense_systems_genes.tsv</span>
<span id="cb19-272"><a href="#cb19-272"></a><span class="in">```</span></span>
<span id="cb19-273"><a href="#cb19-273"></a></span>
<span id="cb19-274"><a href="#cb19-274"></a><span class="fu">## MMSeq2 database</span></span>
<span id="cb19-275"><a href="#cb19-275"></a></span>
<span id="cb19-276"><a href="#cb19-276"></a><span class="in">```</span></span>
<span id="cb19-277"><a href="#cb19-277"></a><span class="in">conda activate mmseq2</span></span>
<span id="cb19-278"><a href="#cb19-278"></a></span>
<span id="cb19-279"><a href="#cb19-279"></a><span class="in">mmseqs createdb data/pangenomes/pectobacterium.v2/mmseq_panproteome_db</span></span>
<span id="cb19-280"><a href="#cb19-280"></a><span class="in">```</span></span>
<span id="cb19-281"><a href="#cb19-281"></a></span>
<span id="cb19-282"><a href="#cb19-282"></a><span class="fu">## Prepare data for pangenome construction</span></span>
<span id="cb19-283"><a href="#cb19-283"></a></span>
<span id="cb19-284"><a href="#cb19-284"></a><span class="fu">### Extract metadata from NCBI XML files</span></span>
<span id="cb19-285"><a href="#cb19-285"></a></span>
<span id="cb19-286"><a href="#cb19-286"></a><span class="in">```bash</span></span>
<span id="cb19-287"><a href="#cb19-287"></a><span class="ex">Rscript</span> scripts/preprocessing/raw_prebuild_metadata.R</span>
<span id="cb19-288"><a href="#cb19-288"></a><span class="in">```</span></span>
<span id="cb19-289"><a href="#cb19-289"></a></span>
<span id="cb19-290"><a href="#cb19-290"></a><span class="fu">### Correct taxonomy</span></span>
<span id="cb19-291"><a href="#cb19-291"></a></span>
<span id="cb19-292"><a href="#cb19-292"></a><span class="in">```bash</span></span>
<span id="cb19-293"><a href="#cb19-293"></a><span class="ex">Rscript</span> scripts/preprocessing/taxonomy_check_inhouse.R</span>
<span id="cb19-294"><a href="#cb19-294"></a><span class="in">```</span></span>
<span id="cb19-295"><a href="#cb19-295"></a></span>
<span id="cb19-296"><a href="#cb19-296"></a><span class="fu">### Prepare final metadata table with updated taxonomy</span></span>
<span id="cb19-297"><a href="#cb19-297"></a></span>
<span id="cb19-298"><a href="#cb19-298"></a><span class="in">```bash</span></span>
<span id="cb19-299"><a href="#cb19-299"></a><span class="ex">Rscript</span> scripts/preprocessing/build_metadata.R</span>
<span id="cb19-300"><a href="#cb19-300"></a><span class="in">```</span></span>
<span id="cb19-301"><a href="#cb19-301"></a></span>
<span id="cb19-302"><a href="#cb19-302"></a><span class="fu">### Prepare PanTools input files</span></span>
<span id="cb19-303"><a href="#cb19-303"></a></span>
<span id="cb19-304"><a href="#cb19-304"></a><span class="in">```bash</span></span>
<span id="cb19-305"><a href="#cb19-305"></a><span class="ex">Rscript</span> scripts/preprocessing/prepare_PanTools_input.R</span>
<span id="cb19-306"><a href="#cb19-306"></a><span class="in">```</span></span></code><button title="Copy to Clipboard" class="code-copy-button" data-in-quarto-modal=""><i class="bi"></i></button></pre></div>
</div></div></div></div></div>
</div> <!-- /content -->




<footer class="footer"><div class="nav-footer"><div class="nav-footer-center"><div class="toc-actions d-sm-block d-md-none"><ul><li><a href="https://github.com/lakhanp1/Pectobacterium_pangenome/issues/new" class="toc-action"><i class="bi bi-github"></i>Report an issue</a></li></ul></div></div></div></footer></body></html>