<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>

<meta charset="utf-8">
<meta name="generator" content="quarto-1.3.433">

<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">

<meta name="author" content="Lakhansing Pardeshi">

<title>Pectobacterium pangenome data analysis – prophage_analysis</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
  vertical-align: middle;
}
/* CSS for syntax highlighting */
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
  }
pre.numberSource { margin-left: 3em;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
</style>


<script src="../../site_libs/quarto-nav/quarto-nav.js"></script>
<script src="../../site_libs/quarto-nav/headroom.min.js"></script>
<script src="../../site_libs/clipboard/clipboard.min.js"></script>
<script src="../../site_libs/quarto-search/autocomplete.umd.js"></script>
<script src="../../site_libs/quarto-search/fuse.min.js"></script>
<script src="../../site_libs/quarto-search/quarto-search.js"></script>
<meta name="quarto:offset" content="../../">
<link href="../../scripts/analysis/prophage_summary.html" rel="next">
<link href="../../scripts/analysis/ani.clustering.html" rel="prev">
<script src="../../site_libs/quarto-html/quarto.js"></script>
<script src="../../site_libs/quarto-html/popper.min.js"></script>
<script src="../../site_libs/quarto-html/tippy.umd.min.js"></script>
<script src="../../site_libs/quarto-html/anchor.min.js"></script>
<link href="../../site_libs/quarto-html/tippy.css" rel="stylesheet">
<link href="../../site_libs/quarto-html/quarto-syntax-highlighting.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="../../site_libs/bootstrap/bootstrap.min.js"></script>
<link href="../../site_libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="../../site_libs/bootstrap/bootstrap.min.css" rel="stylesheet" id="quarto-bootstrap" data-mode="light">
<script id="quarto-search-options" type="application/json">{
  "location": "sidebar",
  "copy-button": false,
  "collapse-after": 3,
  "panel-placement": "start",
  "type": "textbox",
  "limit": 20,
  "language": {
    "search-no-results-text": "No results",
    "search-matching-documents-text": "matching documents",
    "search-copy-link-title": "Copy link to search",
    "search-hide-matches-text": "Hide additional matches",
    "search-more-match-text": "more match in this document",
    "search-more-matches-text": "more matches in this document",
    "search-clear-button-title": "Clear",
    "search-detached-cancel-button-title": "Cancel",
    "search-submit-button-title": "Submit",
    "search-label": "Search"
  }
}</script>


<link rel="stylesheet" href="../../scripts/utils/quarto_style.css">
</head>

<body class="nav-sidebar docked">


<div id="quarto-search-results"></div>
  <header id="quarto-header" class="headroom fixed-top">
  <nav class="quarto-secondary-nav">
    <div class="container-fluid d-flex">
      <button type="button" class="quarto-btn-toggle btn" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar,#quarto-sidebar-glass" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
        <i class="bi bi-layout-text-sidebar-reverse"></i>
      </button>
      <nav class="quarto-page-breadcrumbs" aria-label="breadcrumb"><ol class="breadcrumb"><li class="breadcrumb-item"><a href="../../scripts/notebooks/prophage_analysis.html">Prophages</a></li><li class="breadcrumb-item"><a href="../../scripts/notebooks/prophage_analysis.html">Analysis</a></li></ol></nav>
      <a class="flex-grow-1" role="button" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar,#quarto-sidebar-glass" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">      
      </a>
      <button type="button" class="btn quarto-search-button" aria-label="" onclick="window.quartoOpenSearch();">
        <i class="bi bi-search"></i>
      </button>
    </div>
  </nav>
</header>
<!-- content -->
<div id="quarto-content" class="quarto-container page-columns page-rows-contents page-layout-article">
<!-- sidebar -->
  <nav id="quarto-sidebar" class="sidebar collapse collapse-horizontal sidebar-navigation docked overflow-auto">
    <div class="pt-lg-2 mt-2 text-left sidebar-header">
    <div class="sidebar-title mb-0 py-0">
      <a href="../../">Pectobacterium pangenome data analysis</a> 
    </div>
      </div>
        <div class="mt-2 flex-shrink-0 align-items-center">
        <div class="sidebar-search">
        <div id="quarto-search" class="" title="Search"></div>
        </div>
        </div>
    <div class="sidebar-menu-container"> 
    <ul class="list-unstyled mt-1">
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../scripts/notebooks/preprocessing.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Preprocessing</span></a>
  </div>
</li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-1" aria-expanded="true">
 <span class="menu-text">Pangenome</span></a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-1" aria-expanded="true" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-1" class="collapse list-unstyled sidebar-section depth1 show">  
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../scripts/notebooks/pangenome_construction.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Build</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../scripts/notebooks/pangenome_analysis.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Analysis</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../scripts/analysis/pangenome_metadata.viz.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Metadata</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../scripts/analysis/ani.clustering.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">ANI</span></a>
  </div>
</li>
      </ul>
  </li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-2" aria-expanded="true">
 <span class="menu-text">Prophages</span></a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-2" aria-expanded="true" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-2" class="collapse list-unstyled sidebar-section depth1 show">  
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../scripts/notebooks/prophage_analysis.html" class="sidebar-item-text sidebar-link active">
 <span class="menu-text">Analysis</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../scripts/analysis/prophage_summary.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Prophage summary</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../scripts/analysis/prophage_clustering.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Clustering</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../scripts/analysis/prophage_cluster_summary.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Clustering summary</span></a>
  </div>
</li>
      </ul>
  </li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../scripts/ms_figures/supplementary_tables.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Supplementary tables</span></a>
  </div>
</li>
    </ul>
    </div>
</nav>
<div id="quarto-sidebar-glass" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar,#quarto-sidebar-glass"></div>
<!-- margin-sidebar -->
    <div id="quarto-margin-sidebar" class="sidebar margin-sidebar">
        <nav id="TOC" role="doc-toc" class="toc-active">
    <h2 id="toc-title">On this page</h2>
   
  <ul>
  <li><a href="#workflow" id="toc-workflow" class="nav-link active" data-scroll-target="#workflow">Workflow</a></li>
  <li><a href="#detect-prophages-in-pangenome" id="toc-detect-prophages-in-pangenome" class="nav-link" data-scroll-target="#detect-prophages-in-pangenome">Detect prophages in pangenome</a>
  <ul class="collapse">
  <li><a href="#genomad-prophage-detection" id="toc-genomad-prophage-detection" class="nav-link" data-scroll-target="#genomad-prophage-detection">genomad prophage detection</a></li>
  <li><a href="#post-processing-of-prophages" id="toc-post-processing-of-prophages" class="nav-link" data-scroll-target="#post-processing-of-prophages">Post processing of prophages</a></li>
  <li><a href="#qc" id="toc-qc" class="nav-link" data-scroll-target="#qc">QC</a></li>
  </ul></li>
  <li><a href="#cluster-prophages" id="toc-cluster-prophages" class="nav-link" data-scroll-target="#cluster-prophages">Cluster prophages</a>
  <ul class="collapse">
  <li><a href="#approach-1" id="toc-approach-1" class="nav-link" data-scroll-target="#approach-1">Approach 1</a></li>
  <li><a href="#approach-2" id="toc-approach-2" class="nav-link" data-scroll-target="#approach-2">Approach 2</a></li>
  <li><a href="#summarize-prophage-clusters" id="toc-summarize-prophage-clusters" class="nav-link" data-scroll-target="#summarize-prophage-clusters">Summarize prophage clusters</a></li>
  </ul></li>
  <li><a href="#individual-prophage-cluster-analysis" id="toc-individual-prophage-cluster-analysis" class="nav-link" data-scroll-target="#individual-prophage-cluster-analysis">Individual prophage cluster analysis</a>
  <ul class="collapse">
  <li><a href="#phage_grp_46-highly-conserved-in-all-but-5-genomes-from-n23-clade" id="toc-phage_grp_46-highly-conserved-in-all-but-5-genomes-from-n23-clade" class="nav-link" data-scroll-target="#phage_grp_46-highly-conserved-in-all-but-5-genomes-from-n23-clade">phage_grp_46: highly conserved in all but 5 genomes from n23 clade</a></li>
  <li><a href="#phage_grp_45-prophage-present-in-all-bl-causing-isolates" id="toc-phage_grp_45-prophage-present-in-all-bl-causing-isolates" class="nav-link" data-scroll-target="#phage_grp_45-prophage-present-in-all-bl-causing-isolates">phage_grp_45: prophage present in all BL-causing isolates</a></li>
  <li><a href="#phage_grp_71" id="toc-phage_grp_71" class="nav-link" data-scroll-target="#phage_grp_71">phage_grp_71</a></li>
  <li><a href="#phage_grp_107" id="toc-phage_grp_107" class="nav-link" data-scroll-target="#phage_grp_107">phage_grp_107</a></li>
  <li><a href="#bl-causing-clade-specific-signatures" id="toc-bl-causing-clade-specific-signatures" class="nav-link" data-scroll-target="#bl-causing-clade-specific-signatures">BL-causing clade specific signatures</a></li>
  <li><a href="#phage_grp_36" id="toc-phage_grp_36" class="nav-link" data-scroll-target="#phage_grp_36">phage_grp_36</a></li>
  <li><a href="#phage_grp_50" id="toc-phage_grp_50" class="nav-link" data-scroll-target="#phage_grp_50">phage_grp_50</a></li>
  <li><a href="#phage_grp_21" id="toc-phage_grp_21" class="nav-link" data-scroll-target="#phage_grp_21">phage_grp_21</a></li>
  <li><a href="#t3ss-missing-from-4-p.-brasiliense-genomes" id="toc-t3ss-missing-from-4-p.-brasiliense-genomes" class="nav-link" data-scroll-target="#t3ss-missing-from-4-p.-brasiliense-genomes">T3SS missing from 4 P. brasiliense genomes</a></li>
  </ul></li>
  <li><a href="#carotovoricin-ctv-cluster-or-phage_grp_1-data-analysis" id="toc-carotovoricin-ctv-cluster-or-phage_grp_1-data-analysis" class="nav-link" data-scroll-target="#carotovoricin-ctv-cluster-or-phage_grp_1-data-analysis">Carotovoricin (CTV) cluster or phage_grp_1 data analysis</a>
  <ul class="collapse">
  <li><a href="#ctv-across-pangenome" id="toc-ctv-across-pangenome" class="nav-link" data-scroll-target="#ctv-across-pangenome">CTV across pangenome</a></li>
  <li><a href="#ctv-in-type-strains" id="toc-ctv-in-type-strains" class="nav-link" data-scroll-target="#ctv-in-type-strains">CTV in type strains</a></li>
  <li><a href="#carotovoricin-cluster-missing-from-genomes" id="toc-carotovoricin-cluster-missing-from-genomes" class="nav-link" data-scroll-target="#carotovoricin-cluster-missing-from-genomes">Carotovoricin cluster missing from genomes</a></li>
  <li><a href="#carotovoricin-cluster-absent-in-some-p.-brasiliense-isolates" id="toc-carotovoricin-cluster-absent-in-some-p.-brasiliense-isolates" class="nav-link" data-scroll-target="#carotovoricin-cluster-absent-in-some-p.-brasiliense-isolates">carotovoricin cluster absent in some <em>P. brasiliense</em> isolates</a></li>
  <li><a href="#ctv-in-p.-versatile" id="toc-ctv-in-p.-versatile" class="nav-link" data-scroll-target="#ctv-in-p.-versatile">CTV in <em>P. versatile</em></a></li>
  <li><a href="#ctv-cluster-in-p.-versatile-collected-from-france" id="toc-ctv-cluster-in-p.-versatile-collected-from-france" class="nav-link" data-scroll-target="#ctv-cluster-in-p.-versatile-collected-from-france">CTV cluster in <em>P. versatile</em> collected from France</a></li>
  </ul></li>
  <li><a href="#carotovoricin-tail-fiber-variation-region-msa" id="toc-carotovoricin-tail-fiber-variation-region-msa" class="nav-link" data-scroll-target="#carotovoricin-tail-fiber-variation-region-msa">Carotovoricin tail fiber variation region MSA</a></li>
  <li><a href="#inter-species-horizontal-gene-transfer-of-ctv" id="toc-inter-species-horizontal-gene-transfer-of-ctv" class="nav-link" data-scroll-target="#inter-species-horizontal-gene-transfer-of-ctv">Inter-species horizontal gene transfer of CTV</a>
  <ul class="collapse">
  <li><a href="#dna-sequence-comparison" id="toc-dna-sequence-comparison" class="nav-link" data-scroll-target="#dna-sequence-comparison">DNA sequence comparison</a></li>
  <li><a href="#tail-fiber-haplotype-examples" id="toc-tail-fiber-haplotype-examples" class="nav-link" data-scroll-target="#tail-fiber-haplotype-examples">Tail fiber haplotype examples</a></li>
  <li><a href="#use-mauve-to-detect-the-structural-variation-and-consevation" id="toc-use-mauve-to-detect-the-structural-variation-and-consevation" class="nav-link" data-scroll-target="#use-mauve-to-detect-the-structural-variation-and-consevation">Use Mauve to detect the structural variation and consevation</a></li>
  <li><a href="#visualize-tail-fiber-locus-hg-sets-on-the-pangenome" id="toc-visualize-tail-fiber-locus-hg-sets-on-the-pangenome" class="nav-link" data-scroll-target="#visualize-tail-fiber-locus-hg-sets-on-the-pangenome">Visualize tail fiber locus HG sets on the pangenome</a></li>
  </ul></li>
  </ul>
<div class="toc-actions"><div><i class="bi bi-github"></i></div><div class="action-links"><p><a href="https://github.com/lakhanp1/Pectobacterium_pangenome/issues/new" class="toc-action">Report an issue</a></p></div></div></nav>
    </div>
<!-- main -->
<main class="content" id="quarto-document-content"><div class="quarto-title-block"><div class="quarto-title-tools-only"><h1></h1><button type="button" class="btn code-tools-button" id="quarto-code-tools-source"><i class="bi"></i> Code</button></div></div>



<p><em># Prophage analysis in </em>Pectobacterium* pangenome</p>
<section id="workflow" class="level2">
<h2 class="anchored" data-anchor-id="workflow">Workflow</h2>
<div class="sourceCode" id="cb1"><pre class="sourceCode numberSource mermaid number-lines code-with-copy"><code class="sourceCode"></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</section>
<section id="detect-prophages-in-pangenome" class="level2">
<h2 class="anchored" data-anchor-id="detect-prophages-in-pangenome">Detect prophages in pangenome</h2>
<section id="genomad-prophage-detection" class="level3">
<h3 class="anchored" data-anchor-id="genomad-prophage-detection">genomad prophage detection</h3>
<div class="sourceCode" id="cb2"><pre class="sourceCode numberSource bash number-lines code-with-copy"><code class="sourceCode bash"><span id="cb2-1"><a href="#cb2-1"></a><span class="co"># run genomad and checkv pipeline</span></span>
<span id="cb2-2"><a href="#cb2-2"></a><span class="fu">nohup</span> <span class="dt">\</span></span>
<span id="cb2-3"><a href="#cb2-3"></a>cat data/reference_data/assembly_ids.txt <span class="kw">|</span> <span class="dt">\</span></span>
<span id="cb2-4"><a href="#cb2-4"></a><span class="ex">parallel</span> <span class="at">--jobs</span> 6 <span class="at">--workdir</span> <span class="va">$PWD</span> <span class="at">--halt</span> now,fail=1 <span class="dt">\</span></span>
<span id="cb2-5"><a href="#cb2-5"></a>--keep-order <span class="at">--results</span> logs/genomad/{} <span class="dt">\</span></span>
<span id="cb2-6"><a href="#cb2-6"></a>--joblog logs/genomad/parallel.log <span class="dt">\</span></span>
<span id="cb2-7"><a href="#cb2-7"></a>scripts/preprocessing/genomad_prophage_annotation.sh {} <span class="dt">\</span></span>
<span id="cb2-8"><a href="#cb2-8"></a><span class="op">&gt;&gt;</span>logs/genomad/nohup.out <span class="dv">2</span><span class="op">&gt;&amp;</span><span class="dv">1</span> <span class="kw">&amp;</span></span>
<span id="cb2-9"><a href="#cb2-9"></a></span>
<span id="cb2-10"><a href="#cb2-10"></a><span class="co"># count of prophage detected in each genome</span></span>
<span id="cb2-11"><a href="#cb2-11"></a><span class="cf">for</span> i <span class="kw">in</span> <span class="kw">`</span><span class="fu">cat</span> data/reference_data/assembly_ids.txt<span class="kw">`</span></span>
<span id="cb2-12"><a href="#cb2-12"></a><span class="cf">do</span></span>
<span id="cb2-13"><a href="#cb2-13"></a>  <span class="bu">printf</span> <span class="st">"</span><span class="va">$i</span><span class="st">\t"</span></span>
<span id="cb2-14"><a href="#cb2-14"></a>  <span class="fu">tail</span> <span class="at">-n</span> +2 data/prophage_genomad/<span class="va">$i</span>/<span class="va">${i}</span>_summary/<span class="va">${i}</span>_virus_summary.tsv <span class="kw">|</span> <span class="fu">wc</span> <span class="at">-l</span></span>
<span id="cb2-15"><a href="#cb2-15"></a><span class="cf">done</span> <span class="kw">|</span> <span class="fu">sort</span> <span class="at">-nr</span> <span class="at">-k2</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<p>Check the checkV contamination.tsv file to see if there is any other contamination than the expected <code>host,viral</code>, <code>viral,host</code> and <code>host,viral,host</code> ones.</p>
<div class="sourceCode" id="cb3"><pre class="sourceCode numberSource bash number-lines code-with-copy"><code class="sourceCode bash"><span id="cb3-1"><a href="#cb3-1"></a><span class="cf">for</span> cn <span class="kw">in</span> <span class="kw">`</span><span class="fu">find</span> <span class="at">-name</span> <span class="st">'*contamination.tsv'</span><span class="kw">`</span></span>
<span id="cb3-2"><a href="#cb3-2"></a><span class="cf">do</span></span>
<span id="cb3-3"><a href="#cb3-3"></a><span class="fu">cat</span> <span class="va">$cn</span> <span class="kw">|</span> <span class="fu">awk</span> <span class="st">'$9 != "NA" &amp;&amp; NR != 1 &amp;&amp; $9 != "host,viral" &amp;&amp; $9 != "viral,host" &amp;&amp; $9 != "host,viral,host"'</span></span>
<span id="cb3-4"><a href="#cb3-4"></a><span class="cf">done</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</section>
<section id="post-processing-of-prophages" class="level3">
<h3 class="anchored" data-anchor-id="post-processing-of-prophages">Post processing of prophages</h3>
<p>Process genomad output and separate prophages into independent FASTA files. Each prophage is assigned a unique identifier with format <code>g_{genome}.vir_{#count}</code>. Prophages are saved in independent FASTA files for downstream analysis, for e.g. ANI clustering of prophages.</p>
<div class="sourceCode" id="cb4"><pre class="sourceCode numberSource bash number-lines code-with-copy"><code class="sourceCode bash"><span id="cb4-1"><a href="#cb4-1"></a><span class="co"># post processing of prophage data</span></span>
<span id="cb4-2"><a href="#cb4-2"></a><span class="ex">Rscript</span> scripts/preprocessing/genomad_checkv_merge.R</span>
<span id="cb4-3"><a href="#cb4-3"></a></span>
<span id="cb4-4"><a href="#cb4-4"></a><span class="co"># save prophages independently</span></span>
<span id="cb4-5"><a href="#cb4-5"></a><span class="fu">bash</span> scripts/preprocessing/prophage_genomes_extract.sh</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<p>Extract homology groups for prophages from the pangenome org.db object.</p>
<div class="sourceCode" id="cb5"><pre class="sourceCode numberSource bash number-lines code-with-copy"><code class="sourceCode bash"><span id="cb5-1"><a href="#cb5-1"></a><span class="ex">Rscript</span> scripts/analysis/prophage_homology_groups.R</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<p>Run ANI on prophage genomes to identify the representative prophages in the pangenome.</p>
<div class="sourceCode" id="cb6"><pre class="sourceCode numberSource bash number-lines code-with-copy"><code class="sourceCode bash"><span id="cb6-1"><a href="#cb6-1"></a><span class="ex">conda</span> activate pantools_v4_3</span>
<span id="cb6-2"><a href="#cb6-2"></a></span>
<span id="cb6-3"><a href="#cb6-3"></a><span class="co"># list prophage FASTA files</span></span>
<span id="cb6-4"><a href="#cb6-4"></a><span class="fu">ls</span> data/prophage_genomad/phage_seqs/<span class="pp">*</span>.fna  <span class="op">&gt;</span> data/prophage_genomad/prophage_fasta.list</span>
<span id="cb6-5"><a href="#cb6-5"></a></span>
<span id="cb6-6"><a href="#cb6-6"></a><span class="co"># run fastANI</span></span>
<span id="cb6-7"><a href="#cb6-7"></a><span class="ex">fastANI</span> <span class="at">--ql</span> data/prophage_genomad/prophage_fasta.list <span class="dt">\</span></span>
<span id="cb6-8"><a href="#cb6-8"></a>--rl data/prophage_genomad/prophage_fasta.list <span class="dt">\</span></span>
<span id="cb6-9"><a href="#cb6-9"></a>--threads 30 <span class="at">--matrix</span> <span class="at">--visualize</span> <span class="at">--fragLen</span> 500 <span class="dt">\</span></span>
<span id="cb6-10"><a href="#cb6-10"></a>-o data/prophage_genomad/prophage_ANI</span>
<span id="cb6-11"><a href="#cb6-11"></a></span>
<span id="cb6-12"><a href="#cb6-12"></a><span class="co"># use mash for distance estimation</span></span>
<span id="cb6-13"><a href="#cb6-13"></a><span class="fu">mkdir</span> data/prophage_genomad/mash</span>
<span id="cb6-14"><a href="#cb6-14"></a></span>
<span id="cb6-15"><a href="#cb6-15"></a><span class="ex">mash</span> sketch <span class="at">-p</span> 30 <span class="at">-s</span> 15000 <span class="at">-k</span> 17 <span class="at">-o</span> data/prophage_genomad/mash/prophages_ref <span class="dt">\</span></span>
<span id="cb6-16"><a href="#cb6-16"></a>-l data/prophage_genomad/prophage_fasta.list <span class="op">&gt;</span>logs/prophage/mash.log <span class="dv">2</span><span class="op">&gt;&amp;</span><span class="dv">1</span></span>
<span id="cb6-17"><a href="#cb6-17"></a></span>
<span id="cb6-18"><a href="#cb6-18"></a><span class="ex">mash</span> info data/prophage_genomad/mash/prophages_ref.msh <span class="op">&gt;</span> <span class="dt">\</span></span>
<span id="cb6-19"><a href="#cb6-19"></a>data/prophage_genomad/mash/prophage_sketch_info.txt</span>
<span id="cb6-20"><a href="#cb6-20"></a></span>
<span id="cb6-21"><a href="#cb6-21"></a><span class="ex">mash</span> dist <span class="at">-p</span> 30 data/prophage_genomad/mash/prophages_ref.msh <span class="dt">\</span></span>
<span id="cb6-22"><a href="#cb6-22"></a>data/prophage_genomad/mash/prophages_ref.msh <span class="op">&gt;</span> <span class="dt">\</span></span>
<span id="cb6-23"><a href="#cb6-23"></a>data/prophage_genomad/mash/prophage_distance.tab</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</section>
<section id="qc" class="level3">
<h3 class="anchored" data-anchor-id="qc">QC</h3>
<p>Identify the fragmented prophages resulting because of the genome assembly that can be merged into a complete prophage. Prepare a consolidated list of prophages which will be used for the subsequent analysis.</p>
<div class="sourceCode" id="cb7"><pre class="sourceCode numberSource bash number-lines code-with-copy"><code class="sourceCode bash"><span id="cb7-1"><a href="#cb7-1"></a><span class="ex">Rscript</span> scripts/analysis/prophage_merge.R</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<p>Summarize the consolidated prophage list to generate various statistics and generate summary plots. Prophages that are fragmented or smaller than the length cutoff are removed and the filtered list of prophages is saved for downstream analysis.</p>
<div class="sourceCode" id="cb8"><pre class="sourceCode numberSource bash number-lines code-with-copy"><code class="sourceCode bash"><span id="cb8-1"><a href="#cb8-1"></a><span class="ex">Rscript</span> scripts/analysis/prophage_summary.R</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<p>Process prophage MASH and ANI results and visualize MASH similarity matrix for the filtered prophages.</p>
<div class="sourceCode" id="cb9"><pre class="sourceCode numberSource bash number-lines code-with-copy"><code class="sourceCode bash"><span id="cb9-1"><a href="#cb9-1"></a><span class="ex">Rscript</span> scripts/analysis/prophage_seq_similarity.qmd</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</section>
</section>
<section id="cluster-prophages" class="level2">
<h2 class="anchored" data-anchor-id="cluster-prophages">Cluster prophages</h2>
<p>Use shared homology groups between two prophages to establish the child-parent relationship between the prophages in pangenome. First, all vs all syntenic Jaccard similarity is calculated between the prophages. This information is later used to cluster the prophages using two different approaches.</p>
<div class="sourceCode" id="cb10"><pre class="sourceCode numberSource bash number-lines code-with-copy"><code class="sourceCode bash"><span id="cb10-1"><a href="#cb10-1"></a><span class="ex">Rscript</span> scripts/analysis/prophage_pairwise_similarity.R</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<section id="approach-1" class="level3">
<h3 class="anchored" data-anchor-id="approach-1">Approach 1</h3>
<p>A heuristic DAG construcion is extracted from the output of the script <code>scripts/analysis/prophage_merge.R</code>. This output includes a best matching parent prophage for each prophage. This is further processed to build a DAG and visualize the clusters.</p>
<p>Post processing of the child-parent relationships to add additional metadata and save a DAG in tabular format:</p>
<div class="sourceCode" id="cb11"><pre class="sourceCode numberSource bash number-lines code-with-copy"><code class="sourceCode bash"><span id="cb11-1"><a href="#cb11-1"></a><span class="ex">Rscript</span> scripts/analysis/prophage_DAG_postprocess.R</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<p>Visualize the clusters of prophages as directed acyclic graphs:</p>
<div class="sourceCode" id="cb12"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb12-1"><a href="#cb12-1"></a>knitr<span class="sc">::</span><span class="fu">knit</span>(<span class="st">"scripts/analysis/prophage_DAG_viz.qmd"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</section>
<section id="approach-2" class="level3">
<h3 class="anchored" data-anchor-id="approach-2">Approach 2</h3>
<p>We first tried to use the Affinity propagation (AP) clustering to group the prophages into clusters and have a set of cluster representatives. However, we encountered some clusters generated by the AP clustering where the cluster members did not even share a homology group. Therefore, we decided not to use AP clustering.</p>
<div class="sourceCode" id="cb13"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb13-1"><a href="#cb13-1"></a>knitr<span class="sc">::</span><span class="fu">knit</span>(<span class="st">"scripts/analysis/prophage_AP_clustering.qmd"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<p>Cluster the all-vs-all syntenic Jaccard distance matrix using “complete” linkage hierarchical clustering and cut the cluster using carotovoricin cluster as a reference.</p>
<div class="sourceCode" id="cb14"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb14-1"><a href="#cb14-1"></a>knitr<span class="sc">::</span><span class="fu">knit</span>(<span class="st">"scripts/analysis/prophage_clustering.qmd"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</section>
<section id="summarize-prophage-clusters" class="level3">
<h3 class="anchored" data-anchor-id="summarize-prophage-clusters">Summarize prophage clusters</h3>
<p>Summarize the representative prophages in the pangenome.</p>
<div class="sourceCode" id="cb15"><pre class="sourceCode numberSource bash number-lines code-with-copy"><code class="sourceCode bash"><span id="cb15-1"><a href="#cb15-1"></a><span class="ex">Rscript</span> scripts/analysis/prophage_representative_summary.qmd</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<p>Visualize the representative prophages across pangenome. Use MASH distance to cluster prophages and ANI distance to cluster the genomes in the pangenome.</p>
<div class="sourceCode" id="cb16"><pre class="sourceCode numberSource bash number-lines code-with-copy"><code class="sourceCode bash"><span id="cb16-1"><a href="#cb16-1"></a><span class="ex">Rscript</span> scripts/analysis/prophage_pangenome_viz.qmd</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<p>Visualize homology group sets for selected prophages as PAV heatmap across the pangenome. Replace the code block in <code>scripts/analysis/homology_group_viz.R</code> script with the following.</p>
<div class="sourceCode" id="cb17"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb17-1"><a href="#cb17-1"></a>analysisName <span class="ot">&lt;-</span> <span class="st">"vir_lineage_prophages"</span></span>
<span id="cb17-2"><a href="#cb17-2"></a></span>
<span id="cb17-3"><a href="#cb17-3"></a>outDir <span class="ot">&lt;-</span> <span class="fu">file.path</span>(confs<span class="sc">$</span>analysis<span class="sc">$</span>prophages<span class="sc">$</span>dir)</span>
<span id="cb17-4"><a href="#cb17-4"></a>outPrefix <span class="ot">&lt;-</span> <span class="fu">file.path</span>(outDir, analysisName)</span>
<span id="cb17-5"><a href="#cb17-5"></a></span>
<span id="cb17-6"><a href="#cb17-6"></a>setIds <span class="ot">&lt;-</span> <span class="fu">c</span>(</span>
<span id="cb17-7"><a href="#cb17-7"></a>  <span class="st">"g_189.vir_2"</span>, <span class="st">"g_400.vir_2"</span>, <span class="st">"g_399.vir_2"</span>, <span class="st">"g_448.vir_2"</span>, <span class="st">"g_189.vir_1"</span>, </span>
<span id="cb17-8"><a href="#cb17-8"></a>  <span class="st">"g_3.vir_4"</span>, <span class="st">"g_438.vir_3"</span>, <span class="st">"g_350.vir_1"</span>,</span>
<span id="cb17-9"><a href="#cb17-9"></a>  <span class="st">"g_172.vir_2"</span>, <span class="st">"g_393.vir_1"</span></span>
<span id="cb17-10"><a href="#cb17-10"></a>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<p>Finally, run the script to generate the plot.</p>
<div class="sourceCode" id="cb18"><pre class="sourceCode numberSource bash number-lines code-with-copy"><code class="sourceCode bash"><span id="cb18-1"><a href="#cb18-1"></a><span class="ex">Rscript</span> scripts/analysis/homology_group_viz.R</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</section>
</section>
<section id="individual-prophage-cluster-analysis" class="level2">
<h2 class="anchored" data-anchor-id="individual-prophage-cluster-analysis">Individual prophage cluster analysis</h2>
<p>Script <code>scripts/analysis/clustersmap_data_prophages.R</code> is used to combine the prophages, homology groups and phylogeny to generate input for <code>clustermap.js</code> visualization tool and generate a report. Below, some configuration is provided for individual prophage clusters.</p>
<section id="phage_grp_46-highly-conserved-in-all-but-5-genomes-from-n23-clade" class="level3">
<h3 class="anchored" data-anchor-id="phage_grp_46-highly-conserved-in-all-but-5-genomes-from-n23-clade">phage_grp_46: highly conserved in all but 5 genomes from n23 clade</h3>
<div class="sourceCode" id="cb19"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb19-1"><a href="#cb19-1"></a>grpToView <span class="ot">&lt;-</span> <span class="st">"phage_grp_46"</span></span>
<span id="cb19-2"><a href="#cb19-2"></a>subSample <span class="ot">&lt;-</span> <span class="cn">TRUE</span></span>
<span id="cb19-3"><a href="#cb19-3"></a>cutHeight <span class="ot">&lt;-</span> <span class="fl">0.5</span></span>
<span id="cb19-4"><a href="#cb19-4"></a>addFlankingRegions <span class="ot">&lt;-</span> <span class="cn">TRUE</span></span>
<span id="cb19-5"><a href="#cb19-5"></a>flankingRegion <span class="ot">&lt;-</span> <span class="dv">5000</span></span>
<span id="cb19-6"><a href="#cb19-6"></a></span>
<span id="cb19-7"><a href="#cb19-7"></a><span class="co"># ordering factor for prophages: host phylogeny, prophage HG PAV, prophage MASH,</span></span>
<span id="cb19-8"><a href="#cb19-8"></a><span class="co"># completeness score</span></span>
<span id="cb19-9"><a href="#cb19-9"></a>clusterOrder <span class="ot">&lt;-</span> <span class="st">"host"</span>  <span class="co"># completeness, host, hg_pav, cluster_mash</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</section>
<section id="phage_grp_45-prophage-present-in-all-bl-causing-isolates" class="level3">
<h3 class="anchored" data-anchor-id="phage_grp_45-prophage-present-in-all-bl-causing-isolates">phage_grp_45: prophage present in all BL-causing isolates</h3>
<div class="sourceCode" id="cb20"><pre class="sourceCode numberSource bash number-lines code-with-copy"><code class="sourceCode bash"><span id="cb20-1"><a href="#cb20-1"></a><span class="ex">Rscript</span> scripts/utils/HG_range_coordinates.R <span class="at">--hgs</span> hg_22429319,hg_22429327 <span class="dt">\</span></span>
<span id="cb20-2"><a href="#cb20-2"></a>--genomes <span class="st">"g_406"</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="sourceCode" id="cb21"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb21-1"><a href="#cb21-1"></a>grpToView <span class="ot">&lt;-</span> <span class="st">"grp_45_mosaicsm"</span></span>
<span id="cb21-2"><a href="#cb21-2"></a>subSample <span class="ot">&lt;-</span> <span class="cn">TRUE</span> </span>
<span id="cb21-3"><a href="#cb21-3"></a>cutHeight <span class="ot">&lt;-</span> <span class="fl">0.5</span></span>
<span id="cb21-4"><a href="#cb21-4"></a>addFlankingRegions <span class="ot">&lt;-</span> <span class="cn">TRUE</span></span>
<span id="cb21-5"><a href="#cb21-5"></a>flankingRegion <span class="ot">&lt;-</span> <span class="dv">7000</span></span>
<span id="cb21-6"><a href="#cb21-6"></a></span>
<span id="cb21-7"><a href="#cb21-7"></a><span class="co"># ordering factor for prophages: host phylogeny, prophage HG PAV, prophage MASH,</span></span>
<span id="cb21-8"><a href="#cb21-8"></a><span class="co"># completeness score</span></span>
<span id="cb21-9"><a href="#cb21-9"></a>clusterOrder <span class="ot">&lt;-</span> <span class="st">"host"</span> <span class="co"># host, hg_pav, cluster_mash</span></span>
<span id="cb21-10"><a href="#cb21-10"></a></span>
<span id="cb21-11"><a href="#cb21-11"></a><span class="co"># a vector of prophage identifiers that will be included in clustermap plot</span></span>
<span id="cb21-12"><a href="#cb21-12"></a>appendPhages <span class="ot">&lt;-</span> <span class="fu">c</span>()</span>
<span id="cb21-13"><a href="#cb21-13"></a></span>
<span id="cb21-14"><a href="#cb21-14"></a><span class="co"># whether to keep custom regions at the bottom or consider during phylogeny</span></span>
<span id="cb21-15"><a href="#cb21-15"></a><span class="co"># based ordering</span></span>
<span id="cb21-16"><a href="#cb21-16"></a>regions_phy_ordered <span class="ot">&lt;-</span> <span class="cn">FALSE</span></span>
<span id="cb21-17"><a href="#cb21-17"></a></span>
<span id="cb21-18"><a href="#cb21-18"></a><span class="co"># regions to append as list of list with following structure</span></span>
<span id="cb21-19"><a href="#cb21-19"></a><span class="co"># list(r1 = list(chr, start, end, genomeId), r2 = list(chr, start, end, genomeId))</span></span>
<span id="cb21-20"><a href="#cb21-20"></a>customRegions <span class="ot">&lt;-</span> <span class="fu">list</span>(</span>
<span id="cb21-21"><a href="#cb21-21"></a>  <span class="at">g_406_reg =</span> <span class="fu">list</span>(</span>
<span id="cb21-22"><a href="#cb21-22"></a>    <span class="at">chr =</span> <span class="st">"NAK641_contig_10_consensus"</span>, <span class="at">start =</span> <span class="dv">671040</span>, <span class="at">end =</span> <span class="dv">674984</span>, <span class="at">genomeId =</span> <span class="st">"g_406"</span></span>
<span id="cb21-23"><a href="#cb21-23"></a>  )</span>
<span id="cb21-24"><a href="#cb21-24"></a>)</span>
<span id="cb21-25"><a href="#cb21-25"></a></span>
<span id="cb21-26"><a href="#cb21-26"></a></span>
<span id="cb21-27"><a href="#cb21-27"></a><span class="co"># optionally, a custom region list can be provided to generate the plot</span></span>
<span id="cb21-28"><a href="#cb21-28"></a>grp <span class="ot">&lt;-</span> <span class="fu">list</span>(</span>
<span id="cb21-29"><a href="#cb21-29"></a>  <span class="at">phage_grp =</span> grpToView,</span>
<span id="cb21-30"><a href="#cb21-30"></a>  <span class="at">members =</span> <span class="fu">c</span>(<span class="st">"g_302.vir_1"</span>, <span class="st">"g_399.vir_1"</span>, <span class="st">"g_400.vir_2"</span>)</span>
<span id="cb21-31"><a href="#cb21-31"></a>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</section>
<section id="phage_grp_71" class="level3">
<h3 class="anchored" data-anchor-id="phage_grp_71">phage_grp_71</h3>
<div class="sourceCode" id="cb22"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb22-1"><a href="#cb22-1"></a>grpToView <span class="ot">&lt;-</span> <span class="st">"phage_grp_71"</span></span>
<span id="cb22-2"><a href="#cb22-2"></a>subSample <span class="ot">&lt;-</span> <span class="cn">FALSE</span></span>
<span id="cb22-3"><a href="#cb22-3"></a>addFlankingRegions <span class="ot">&lt;-</span> <span class="cn">TRUE</span> </span>
<span id="cb22-4"><a href="#cb22-4"></a>flankingRegion <span class="ot">&lt;-</span> <span class="dv">5000</span></span>
<span id="cb22-5"><a href="#cb22-5"></a></span>
<span id="cb22-6"><a href="#cb22-6"></a><span class="co"># ordering factor for prophages: host phylogeny, prophage HG PAV, prophage MASH,</span></span>
<span id="cb22-7"><a href="#cb22-7"></a><span class="co"># completeness score</span></span>
<span id="cb22-8"><a href="#cb22-8"></a>clusterOrder <span class="ot">&lt;-</span> <span class="st">"hg_pav"</span>  <span class="co"># completeness, host, hg_pav, cluster_mash</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</section>
<section id="phage_grp_107" class="level3">
<h3 class="anchored" data-anchor-id="phage_grp_107">phage_grp_107</h3>
<div class="sourceCode" id="cb23"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb23-1"><a href="#cb23-1"></a>grpToView <span class="ot">&lt;-</span> <span class="st">"phage_grp_107"</span></span>
<span id="cb23-2"><a href="#cb23-2"></a>subSample <span class="ot">&lt;-</span> <span class="cn">TRUE</span></span>
<span id="cb23-3"><a href="#cb23-3"></a>cutHeight <span class="ot">&lt;-</span> <span class="fl">0.5</span></span>
<span id="cb23-4"><a href="#cb23-4"></a>addFlankingRegions <span class="ot">&lt;-</span> <span class="cn">TRUE</span></span>
<span id="cb23-5"><a href="#cb23-5"></a>flankingRegion <span class="ot">&lt;-</span> <span class="dv">5000</span></span>
<span id="cb23-6"><a href="#cb23-6"></a></span>
<span id="cb23-7"><a href="#cb23-7"></a><span class="co"># ordering factor for prophages: host phylogeny, prophage HG PAV, prophage MASH,</span></span>
<span id="cb23-8"><a href="#cb23-8"></a><span class="co"># completeness score</span></span>
<span id="cb23-9"><a href="#cb23-9"></a>clusterOrder <span class="ot">&lt;-</span> <span class="st">"host"</span>  <span class="co"># completeness, host, hg_pav, cluster_mash</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</section>
<section id="bl-causing-clade-specific-signatures" class="level3">
<h3 class="anchored" data-anchor-id="bl-causing-clade-specific-signatures">BL-causing clade specific signatures</h3>
<div class="sourceCode" id="cb24"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb24-1"><a href="#cb24-1"></a>grpToView <span class="ot">&lt;-</span> <span class="st">"Pbr_FN_specific"</span></span>
<span id="cb24-2"><a href="#cb24-2"></a>subSample <span class="ot">&lt;-</span> <span class="cn">FALSE</span></span>
<span id="cb24-3"><a href="#cb24-3"></a>cutHeight <span class="ot">&lt;-</span> <span class="fl">0.5</span></span>
<span id="cb24-4"><a href="#cb24-4"></a>addFlankingRegions <span class="ot">&lt;-</span> <span class="cn">FALSE</span></span>
<span id="cb24-5"><a href="#cb24-5"></a>flankingRegion <span class="ot">&lt;-</span> <span class="dv">5000</span></span>
<span id="cb24-6"><a href="#cb24-6"></a></span>
<span id="cb24-7"><a href="#cb24-7"></a><span class="co"># ordering factor for prophages: host phylogeny, prophage HG PAV, prophage MASH,</span></span>
<span id="cb24-8"><a href="#cb24-8"></a><span class="co"># completeness score</span></span>
<span id="cb24-9"><a href="#cb24-9"></a>clusterOrder <span class="ot">&lt;-</span> <span class="st">"default"</span>  <span class="co"># completeness, host, hg_pav, cluster_mash, default</span></span>
<span id="cb24-10"><a href="#cb24-10"></a></span>
<span id="cb24-11"><a href="#cb24-11"></a></span>
<span id="cb24-12"><a href="#cb24-12"></a><span class="co"># optionally, a custom region list can be provided to generate the plot</span></span>
<span id="cb24-13"><a href="#cb24-13"></a>grp <span class="ot">&lt;-</span> <span class="fu">list</span>(</span>
<span id="cb24-14"><a href="#cb24-14"></a>  <span class="at">phage_grp =</span> grpToView,</span>
<span id="cb24-15"><a href="#cb24-15"></a>  <span class="at">members =</span> <span class="fu">c</span>(<span class="st">"g_302.vir_1"</span>, <span class="st">"g_400.vir_2"</span>, <span class="st">"g_409.vir_1"</span>, <span class="st">"g_302.vir_2"</span>)</span>
<span id="cb24-16"><a href="#cb24-16"></a>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</section>
<section id="phage_grp_36" class="level3">
<h3 class="anchored" data-anchor-id="phage_grp_36">phage_grp_36</h3>
<div class="sourceCode" id="cb25"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb25-1"><a href="#cb25-1"></a>grpToView <span class="ot">&lt;-</span> <span class="st">"phage_grp_36"</span></span>
<span id="cb25-2"><a href="#cb25-2"></a>subSample <span class="ot">&lt;-</span> <span class="cn">FALSE</span></span>
<span id="cb25-3"><a href="#cb25-3"></a>cutHeight <span class="ot">&lt;-</span> <span class="fl">1.5</span></span>
<span id="cb25-4"><a href="#cb25-4"></a>addFlankingRegions <span class="ot">&lt;-</span> <span class="cn">TRUE</span></span>
<span id="cb25-5"><a href="#cb25-5"></a>flankingRegion <span class="ot">&lt;-</span> <span class="dv">5000</span></span>
<span id="cb25-6"><a href="#cb25-6"></a></span>
<span id="cb25-7"><a href="#cb25-7"></a><span class="co"># ordering factor for prophages: host phylogeny, prophage HG PAV, prophage MASH,</span></span>
<span id="cb25-8"><a href="#cb25-8"></a><span class="co"># completeness score</span></span>
<span id="cb25-9"><a href="#cb25-9"></a>clusterOrder <span class="ot">&lt;-</span> <span class="st">"host"</span> <span class="co"># completeness, host, hg_pav, cluster_mash</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</section>
<section id="phage_grp_50" class="level3">
<h3 class="anchored" data-anchor-id="phage_grp_50">phage_grp_50</h3>
<div class="sourceCode" id="cb26"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb26-1"><a href="#cb26-1"></a>grpToView <span class="ot">&lt;-</span> <span class="st">"phage_grp_50"</span></span>
<span id="cb26-2"><a href="#cb26-2"></a>subSample <span class="ot">&lt;-</span> <span class="cn">FALSE</span></span>
<span id="cb26-3"><a href="#cb26-3"></a>addFlankingRegions <span class="ot">&lt;-</span> <span class="cn">TRUE</span> </span>
<span id="cb26-4"><a href="#cb26-4"></a>flankingRegion <span class="ot">&lt;-</span> <span class="dv">10000</span></span>
<span id="cb26-5"><a href="#cb26-5"></a></span>
<span id="cb26-6"><a href="#cb26-6"></a><span class="co"># ordering factor for prophages: host phylogeny, prophage HG PAV, prophage MASH,</span></span>
<span id="cb26-7"><a href="#cb26-7"></a><span class="co"># completeness score</span></span>
<span id="cb26-8"><a href="#cb26-8"></a>clusterOrder <span class="ot">&lt;-</span> <span class="st">"cluster_mash"</span>  <span class="co"># completeness, host, hg_pav, cluster_mash</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</section>
<section id="phage_grp_21" class="level3">
<h3 class="anchored" data-anchor-id="phage_grp_21">phage_grp_21</h3>
<div class="sourceCode" id="cb27"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb27-1"><a href="#cb27-1"></a>grpToView <span class="ot">&lt;-</span> <span class="st">"phage_grp_21"</span></span>
<span id="cb27-2"><a href="#cb27-2"></a>subSample <span class="ot">&lt;-</span> <span class="cn">FALSE</span></span>
<span id="cb27-3"><a href="#cb27-3"></a>addFlankingRegions <span class="ot">&lt;-</span> <span class="cn">TRUE</span></span>
<span id="cb27-4"><a href="#cb27-4"></a>flankingRegion <span class="ot">&lt;-</span> <span class="dv">5000</span></span>
<span id="cb27-5"><a href="#cb27-5"></a></span>
<span id="cb27-6"><a href="#cb27-6"></a><span class="co"># ordering factor for prophages: host phylogeny, prophage HG PAV, prophage MASH,</span></span>
<span id="cb27-7"><a href="#cb27-7"></a><span class="co"># completeness score</span></span>
<span id="cb27-8"><a href="#cb27-8"></a>clusterOrder <span class="ot">&lt;-</span> <span class="st">"cluster_mash"</span>  <span class="co"># completeness, host, hg_pav, cluster_mash</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</section>
<section id="t3ss-missing-from-4-p.-brasiliense-genomes" class="level3">
<h3 class="anchored" data-anchor-id="t3ss-missing-from-4-p.-brasiliense-genomes">T3SS missing from 4 P. brasiliense genomes</h3>
<div class="sourceCode" id="cb28"><pre class="sourceCode numberSource bash number-lines code-with-copy"><code class="sourceCode bash"><span id="cb28-1"><a href="#cb28-1"></a><span class="ex">Rscript</span> scripts/utils/HG_range_coordinates.R <span class="at">--hgs</span> hg_22430875,hg_22430876 <span class="at">--genomes</span> g_187,g_368,g_182,g_185,g_236,g_177</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="sourceCode" id="cb29"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb29-1"><a href="#cb29-1"></a>grpToView <span class="ot">&lt;-</span> <span class="st">"t3ss_regions"</span></span>
<span id="cb29-2"><a href="#cb29-2"></a>subSample <span class="ot">&lt;-</span> <span class="cn">FALSE</span></span>
<span id="cb29-3"><a href="#cb29-3"></a>cutHeight <span class="ot">&lt;-</span> <span class="fl">1.5</span></span>
<span id="cb29-4"><a href="#cb29-4"></a>addFlankingRegions <span class="ot">&lt;-</span> <span class="cn">TRUE</span></span>
<span id="cb29-5"><a href="#cb29-5"></a>flankingRegion <span class="ot">&lt;-</span> <span class="dv">6000</span></span>
<span id="cb29-6"><a href="#cb29-6"></a></span>
<span id="cb29-7"><a href="#cb29-7"></a><span class="co"># ordering factor for prophages: host phylogeny, prophage HG PAV, prophage MASH,</span></span>
<span id="cb29-8"><a href="#cb29-8"></a><span class="co"># completeness score</span></span>
<span id="cb29-9"><a href="#cb29-9"></a>clusterOrder <span class="ot">&lt;-</span> <span class="st">"host"</span>  <span class="co"># completeness, host, hg_pav, cluster_mash</span></span>
<span id="cb29-10"><a href="#cb29-10"></a></span>
<span id="cb29-11"><a href="#cb29-11"></a><span class="co"># a vector of prophage identifiers that will be included in clustermap plot</span></span>
<span id="cb29-12"><a href="#cb29-12"></a>appendPhages <span class="ot">&lt;-</span> <span class="fu">c</span>()</span>
<span id="cb29-13"><a href="#cb29-13"></a></span>
<span id="cb29-14"><a href="#cb29-14"></a><span class="co"># regions to append as list of list with following structure</span></span>
<span id="cb29-15"><a href="#cb29-15"></a><span class="co"># list(r1 = list(chr, start, end, genomeId), r2 = list(chr, start, end, genomeId))</span></span>
<span id="cb29-16"><a href="#cb29-16"></a>customRegions <span class="ot">&lt;-</span> <span class="fu">list</span>(</span>
<span id="cb29-17"><a href="#cb29-17"></a>  <span class="at">g_177_reg  =</span> <span class="fu">list</span>(<span class="at">genomeId =</span> <span class="st">"g_177"</span>, <span class="at">chr =</span> <span class="st">"NZ_JACGEP010000027.1"</span>, <span class="at">start =</span> <span class="dv">56850</span>, <span class="at">end =</span> <span class="dv">59337</span>),</span>
<span id="cb29-18"><a href="#cb29-18"></a>  <span class="at">g_182_reg =</span> <span class="fu">list</span>(<span class="at">genomeId =</span> <span class="st">"g_182"</span>, <span class="at">chr =</span> <span class="st">"NZ_JACGZZ010000068.1"</span>, <span class="at">start =</span> <span class="dv">116773</span>, <span class="at">end =</span> <span class="dv">119260</span>),</span>
<span id="cb29-19"><a href="#cb29-19"></a>  <span class="at">g_185_reg =</span> <span class="fu">list</span>(<span class="at">genomeId =</span> <span class="st">"g_185"</span>, <span class="at">chr =</span> <span class="st">"NZ_JACGEN010000027.1"</span>, <span class="at">start =</span> <span class="dv">117747</span>, <span class="at">end =</span> <span class="dv">120234</span>),</span>
<span id="cb29-20"><a href="#cb29-20"></a>  <span class="at">g_236_reg =</span> <span class="fu">list</span>(<span class="at">genomeId =</span> <span class="st">"g_236"</span>, <span class="at">chr =</span> <span class="st">"NZ_JACDSF010000026.1"</span>, <span class="at">start =</span> <span class="dv">2275</span>, <span class="at">end =</span> <span class="dv">4762</span>),</span>
<span id="cb29-21"><a href="#cb29-21"></a>  <span class="at">g_187_reg =</span> <span class="fu">list</span>(<span class="at">genomeId =</span> <span class="st">"g_187"</span>, <span class="at">chr =</span> <span class="st">"NZ_CP059963.1"</span>, <span class="at">start =</span> <span class="dv">2603432</span>, <span class="at">end =</span> <span class="dv">2605919</span>),</span>
<span id="cb29-22"><a href="#cb29-22"></a>  <span class="at">g_368_reg =</span> <span class="fu">list</span>(<span class="at">genomeId =</span> <span class="st">"g_368"</span>, <span class="at">chr =</span> <span class="st">"NZ_JUJT01000001.1"</span>, <span class="at">start =</span> <span class="dv">566617</span>, <span class="at">end =</span> <span class="dv">569104</span>)</span>
<span id="cb29-23"><a href="#cb29-23"></a>)</span>
<span id="cb29-24"><a href="#cb29-24"></a></span>
<span id="cb29-25"><a href="#cb29-25"></a><span class="co"># whether to keep custom regions at the bottom or consider during phylogeny</span></span>
<span id="cb29-26"><a href="#cb29-26"></a><span class="co"># based ordering</span></span>
<span id="cb29-27"><a href="#cb29-27"></a>regions_phy_ordered <span class="ot">&lt;-</span> <span class="cn">TRUE</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</section>
</section>
<section id="carotovoricin-ctv-cluster-or-phage_grp_1-data-analysis" class="level2">
<h2 class="anchored" data-anchor-id="carotovoricin-ctv-cluster-or-phage_grp_1-data-analysis">Carotovoricin (CTV) cluster or phage_grp_1 data analysis</h2>
<section id="ctv-across-pangenome" class="level3">
<h3 class="anchored" data-anchor-id="ctv-across-pangenome">CTV across pangenome</h3>
<div class="sourceCode" id="cb30"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb30-1"><a href="#cb30-1"></a>grpToView <span class="ot">&lt;-</span> <span class="st">"phage_grp_1"</span></span>
<span id="cb30-2"><a href="#cb30-2"></a>subSample <span class="ot">&lt;-</span> <span class="cn">TRUE</span></span>
<span id="cb30-3"><a href="#cb30-3"></a>cutHeight <span class="ot">&lt;-</span> <span class="fl">1.5</span></span>
<span id="cb30-4"><a href="#cb30-4"></a>addFlankingRegions <span class="ot">&lt;-</span> <span class="cn">TRUE</span></span>
<span id="cb30-5"><a href="#cb30-5"></a>flankingRegion <span class="ot">&lt;-</span> <span class="dv">5000</span></span>
<span id="cb30-6"><a href="#cb30-6"></a></span>
<span id="cb30-7"><a href="#cb30-7"></a><span class="co"># ordering factor for prophages: host phylogeny, prophage HG PAV, prophage MASH,</span></span>
<span id="cb30-8"><a href="#cb30-8"></a><span class="co"># completeness score</span></span>
<span id="cb30-9"><a href="#cb30-9"></a>clusterOrder <span class="ot">&lt;-</span> <span class="st">"host"</span>  <span class="co"># completeness, host, hg_pav, cluster_mash</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</section>
<section id="ctv-in-type-strains" class="level3">
<h3 class="anchored" data-anchor-id="ctv-in-type-strains">CTV in type strains</h3>
<p>Include <em>P. atrosepticum</em> (g_385), <em>P. betavasculorum</em> (g_386) and <em>P. cacticida</em> (g_451) genomes that lack CTV cluster.</p>
<div class="sourceCode" id="cb31"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb31-1"><a href="#cb31-1"></a>grpToView <span class="ot">&lt;-</span> <span class="st">"ctv_typeStrains"</span></span>
<span id="cb31-2"><a href="#cb31-2"></a>subSample <span class="ot">&lt;-</span> <span class="cn">FALSE</span></span>
<span id="cb31-3"><a href="#cb31-3"></a>cutHeight <span class="ot">&lt;-</span> <span class="fl">1.5</span></span>
<span id="cb31-4"><a href="#cb31-4"></a>addFlankingRegions <span class="ot">&lt;-</span> <span class="cn">TRUE</span></span>
<span id="cb31-5"><a href="#cb31-5"></a>flankingRegion <span class="ot">&lt;-</span> <span class="dv">6000</span></span>
<span id="cb31-6"><a href="#cb31-6"></a></span>
<span id="cb31-7"><a href="#cb31-7"></a><span class="co"># ordering factor for prophages: host phylogeny, prophage HG PAV, prophage MASH,</span></span>
<span id="cb31-8"><a href="#cb31-8"></a><span class="co"># completeness score</span></span>
<span id="cb31-9"><a href="#cb31-9"></a>clusterOrder <span class="ot">&lt;-</span> <span class="st">"host"</span>  <span class="co"># completeness, host, hg_pav, cluster_mash</span></span>
<span id="cb31-10"><a href="#cb31-10"></a></span>
<span id="cb31-11"><a href="#cb31-11"></a><span class="co"># a vector of prophage identifiers that will be included in clustermap plot</span></span>
<span id="cb31-12"><a href="#cb31-12"></a>appendPhages <span class="ot">&lt;-</span> <span class="fu">c</span>()</span>
<span id="cb31-13"><a href="#cb31-13"></a></span>
<span id="cb31-14"><a href="#cb31-14"></a><span class="co"># regions to append as list of list with following structure</span></span>
<span id="cb31-15"><a href="#cb31-15"></a><span class="co"># list(r1 = list(chr, start, end, genomeId), r2 = list(chr, start, end, genomeId))</span></span>
<span id="cb31-16"><a href="#cb31-16"></a>customRegions <span class="ot">&lt;-</span> <span class="fu">list</span>(</span>
<span id="cb31-17"><a href="#cb31-17"></a>  <span class="at">g_385_reg =</span> <span class="fu">list</span>(</span>
<span id="cb31-18"><a href="#cb31-18"></a>    <span class="at">chr =</span> <span class="st">"NZ_JQHK01000003.1"</span>, <span class="at">start =</span> <span class="dv">203963</span>, <span class="at">end =</span> <span class="dv">207120</span>, <span class="at">genomeId =</span> <span class="st">"g_385"</span></span>
<span id="cb31-19"><a href="#cb31-19"></a>  ),</span>
<span id="cb31-20"><a href="#cb31-20"></a>  <span class="at">g_386_reg =</span> <span class="fu">list</span>(</span>
<span id="cb31-21"><a href="#cb31-21"></a>    <span class="at">chr =</span> <span class="st">"NZ_JQHM01000001.1"</span>, <span class="at">start =</span> <span class="dv">553213</span>, <span class="at">end =</span> <span class="dv">555615</span>, <span class="at">genomeId =</span> <span class="st">"g_386"</span></span>
<span id="cb31-22"><a href="#cb31-22"></a>  ),</span>
<span id="cb31-23"><a href="#cb31-23"></a>  <span class="at">g_451_reg =</span> <span class="fu">list</span>(</span>
<span id="cb31-24"><a href="#cb31-24"></a>    <span class="at">chr =</span> <span class="st">"Contig_2_668.636"</span>, <span class="at">start =</span> <span class="dv">191452</span>, <span class="at">end =</span> <span class="dv">191490</span>, <span class="at">genomeId =</span> <span class="st">"g_451"</span></span>
<span id="cb31-25"><a href="#cb31-25"></a>  )</span>
<span id="cb31-26"><a href="#cb31-26"></a>)</span>
<span id="cb31-27"><a href="#cb31-27"></a></span>
<span id="cb31-28"><a href="#cb31-28"></a><span class="co"># whether to keep custom regions at the bottom or consider during phylogeny</span></span>
<span id="cb31-29"><a href="#cb31-29"></a><span class="co"># based ordering</span></span>
<span id="cb31-30"><a href="#cb31-30"></a>regions_phy_ordered <span class="ot">&lt;-</span> <span class="cn">TRUE</span></span>
<span id="cb31-31"><a href="#cb31-31"></a></span>
<span id="cb31-32"><a href="#cb31-32"></a></span>
<span id="cb31-33"><a href="#cb31-33"></a></span>
<span id="cb31-34"><a href="#cb31-34"></a>grp <span class="ot">&lt;-</span> <span class="fu">list</span>(</span>
<span id="cb31-35"><a href="#cb31-35"></a>  <span class="at">phage_grp =</span> grpToView,</span>
<span id="cb31-36"><a href="#cb31-36"></a>  <span class="at">members =</span> <span class="fu">c</span>(</span>
<span id="cb31-37"><a href="#cb31-37"></a>    <span class="st">"g_345.vir_1"</span>, <span class="st">"g_446.vir_4"</span>, <span class="st">"g_66.vir_3"</span>, <span class="st">"g_222.vir_2"</span>, <span class="st">"g_365.vir_3"</span>,</span>
<span id="cb31-38"><a href="#cb31-38"></a>    <span class="st">"g_442.vir_1"</span>, <span class="st">"g_8.vir_2"</span>, <span class="st">"g_38.vir_2"</span>, <span class="st">"g_273.vir_2"</span>, <span class="st">"g_259.vir_4"</span>,</span>
<span id="cb31-39"><a href="#cb31-39"></a>    <span class="st">"g_305.vir_1"</span>, <span class="st">"g_378.vir_6"</span>, <span class="st">"g_428.vir_1"</span>, <span class="st">"g_248.vir_1"</span>, <span class="st">"g_449.vir_1"</span>,</span>
<span id="cb31-40"><a href="#cb31-40"></a>    <span class="st">"g_54.vir_1"</span>, <span class="st">"g_116.vir_3"</span>, <span class="st">"g_423.vir_3"</span>, <span class="st">"g_375.vir_2"</span>, <span class="st">"g_381.vir_2"</span></span>
<span id="cb31-41"><a href="#cb31-41"></a>  )</span>
<span id="cb31-42"><a href="#cb31-42"></a>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</section>
<section id="carotovoricin-cluster-missing-from-genomes" class="level3">
<h3 class="anchored" data-anchor-id="carotovoricin-cluster-missing-from-genomes">Carotovoricin cluster missing from genomes</h3>
<div class="sourceCode" id="cb32"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb32-1"><a href="#cb32-1"></a>grpToView <span class="ot">&lt;-</span> <span class="st">"ctv_deletion"</span></span>
<span id="cb32-2"><a href="#cb32-2"></a>subSample <span class="ot">&lt;-</span> <span class="cn">TRUE</span></span>
<span id="cb32-3"><a href="#cb32-3"></a>cutHeight <span class="ot">&lt;-</span> <span class="fl">1.5</span></span>
<span id="cb32-4"><a href="#cb32-4"></a>addFlankingRegions <span class="ot">&lt;-</span> <span class="cn">TRUE</span></span>
<span id="cb32-5"><a href="#cb32-5"></a>flankingRegion <span class="ot">&lt;-</span> <span class="dv">5000</span></span>
<span id="cb32-6"><a href="#cb32-6"></a></span>
<span id="cb32-7"><a href="#cb32-7"></a><span class="co"># ordering factor for prophages: host phylogeny, prophage HG PAV, prophage MASH,</span></span>
<span id="cb32-8"><a href="#cb32-8"></a><span class="co"># completeness score</span></span>
<span id="cb32-9"><a href="#cb32-9"></a>clusterOrder <span class="ot">&lt;-</span> <span class="st">"host"</span>  <span class="co"># completeness, host, hg_pav, cluster_mash</span></span>
<span id="cb32-10"><a href="#cb32-10"></a></span>
<span id="cb32-11"><a href="#cb32-11"></a><span class="co"># regions to append as list of list with following structure</span></span>
<span id="cb32-12"><a href="#cb32-12"></a><span class="co"># list(r1 = list(chr, start, end, genomeId), r2 = list(chr, start, end, genomeId))</span></span>
<span id="cb32-13"><a href="#cb32-13"></a>customRegions <span class="ot">&lt;-</span> <span class="fu">list</span>(</span>
<span id="cb32-14"><a href="#cb32-14"></a>  <span class="at">g_177_reg =</span> <span class="fu">list</span>(</span>
<span id="cb32-15"><a href="#cb32-15"></a>    <span class="at">chr =</span> <span class="st">"NZ_JACGEP010000002.1"</span>, <span class="at">start =</span> <span class="dv">102351</span>, <span class="at">end =</span> <span class="dv">102361</span>, <span class="at">genomeId =</span> <span class="st">"g_177"</span></span>
<span id="cb32-16"><a href="#cb32-16"></a>  ),</span>
<span id="cb32-17"><a href="#cb32-17"></a>  <span class="at">g_182_reg =</span> <span class="fu">list</span>(</span>
<span id="cb32-18"><a href="#cb32-18"></a>    <span class="at">chr =</span> <span class="st">"NZ_JACGZZ010000050.1"</span>, <span class="at">start =</span> <span class="dv">81625</span>, <span class="at">end =</span> <span class="dv">81635</span>, <span class="at">genomeId =</span> <span class="st">"g_182"</span></span>
<span id="cb32-19"><a href="#cb32-19"></a>  ),</span>
<span id="cb32-20"><a href="#cb32-20"></a>  <span class="at">g_185_reg =</span> <span class="fu">list</span>(</span>
<span id="cb32-21"><a href="#cb32-21"></a>    <span class="at">chr =</span> <span class="st">"NZ_JACGEN010000006.1"</span>, <span class="at">start =</span> <span class="dv">81657</span>, <span class="at">end =</span> <span class="dv">81667</span>, <span class="at">genomeId =</span> <span class="st">"g_185"</span></span>
<span id="cb32-22"><a href="#cb32-22"></a>  ),</span>
<span id="cb32-23"><a href="#cb32-23"></a>  <span class="at">g_236_reg =</span> <span class="fu">list</span>(</span>
<span id="cb32-24"><a href="#cb32-24"></a>    <span class="at">chr =</span> <span class="st">"NZ_JACDSF010000027.1"</span>, <span class="at">start =</span> <span class="dv">89639</span>, <span class="at">end =</span> <span class="dv">89649</span>, <span class="at">genomeId =</span> <span class="st">"g_236"</span></span>
<span id="cb32-25"><a href="#cb32-25"></a>  ),</span>
<span id="cb32-26"><a href="#cb32-26"></a>  <span class="at">g_385_reg =</span> <span class="fu">list</span>(</span>
<span id="cb32-27"><a href="#cb32-27"></a>    <span class="at">chr =</span> <span class="st">"NZ_JQHK01000003.1"</span>, <span class="at">start =</span> <span class="dv">203963</span>, <span class="at">end =</span> <span class="dv">207120</span>, <span class="at">genomeId =</span> <span class="st">"g_385"</span></span>
<span id="cb32-28"><a href="#cb32-28"></a>  ),</span>
<span id="cb32-29"><a href="#cb32-29"></a>  <span class="at">g_386_reg =</span> <span class="fu">list</span>(</span>
<span id="cb32-30"><a href="#cb32-30"></a>    <span class="at">chr =</span> <span class="st">"NZ_JQHM01000001.1"</span>, <span class="at">start =</span> <span class="dv">553213</span>, <span class="at">end =</span> <span class="dv">555615</span>, <span class="at">genomeId =</span> <span class="st">"g_386"</span></span>
<span id="cb32-31"><a href="#cb32-31"></a>  ),</span>
<span id="cb32-32"><a href="#cb32-32"></a>  <span class="at">g_451_reg =</span> <span class="fu">list</span>(</span>
<span id="cb32-33"><a href="#cb32-33"></a>    <span class="at">chr =</span> <span class="st">"Contig_2_668.636"</span>, <span class="at">start =</span> <span class="dv">191452</span>, <span class="at">end =</span> <span class="dv">191490</span>, <span class="at">genomeId =</span> <span class="st">"g_451"</span></span>
<span id="cb32-34"><a href="#cb32-34"></a>  )</span>
<span id="cb32-35"><a href="#cb32-35"></a>)</span>
<span id="cb32-36"><a href="#cb32-36"></a></span>
<span id="cb32-37"><a href="#cb32-37"></a></span>
<span id="cb32-38"><a href="#cb32-38"></a>grp <span class="ot">&lt;-</span> <span class="fu">list</span>(</span>
<span id="cb32-39"><a href="#cb32-39"></a>  <span class="at">phage_grp =</span> grpToView,</span>
<span id="cb32-40"><a href="#cb32-40"></a>  <span class="at">members =</span> <span class="fu">c</span>(</span>
<span id="cb32-41"><a href="#cb32-41"></a>    <span class="st">"g_345.vir_1"</span>, <span class="st">"g_66.vir_3"</span>, <span class="st">"g_8.vir_2"</span>, <span class="st">"g_381.vir_2"</span></span>
<span id="cb32-42"><a href="#cb32-42"></a>  )</span>
<span id="cb32-43"><a href="#cb32-43"></a>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</section>
<section id="carotovoricin-cluster-absent-in-some-p.-brasiliense-isolates" class="level3">
<h3 class="anchored" data-anchor-id="carotovoricin-cluster-absent-in-some-p.-brasiliense-isolates">carotovoricin cluster absent in some <em>P. brasiliense</em> isolates</h3>
<p>ybiB: hg_22427603</p>
<p>tolC_2: hg_22427641</p>
<p><em>P. brasiliense</em> lacking CTV cluster: g_149, g_177, g_182, g_185, g_236</p>
<div class="sourceCode" id="cb33"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb33-1"><a href="#cb33-1"></a>grpToView <span class="ot">&lt;-</span> <span class="st">"ctv_pbr"</span></span>
<span id="cb33-2"><a href="#cb33-2"></a>subSample <span class="ot">&lt;-</span> <span class="cn">TRUE</span></span>
<span id="cb33-3"><a href="#cb33-3"></a>cutHeight <span class="ot">&lt;-</span> <span class="fl">1.5</span></span>
<span id="cb33-4"><a href="#cb33-4"></a>addFlankingRegions <span class="ot">&lt;-</span> <span class="cn">TRUE</span></span>
<span id="cb33-5"><a href="#cb33-5"></a>flankingRegion <span class="ot">&lt;-</span> <span class="dv">5000</span></span>
<span id="cb33-6"><a href="#cb33-6"></a></span>
<span id="cb33-7"><a href="#cb33-7"></a><span class="co"># ordering factor for prophages: host phylogeny, prophage HG PAV, prophage MASH,</span></span>
<span id="cb33-8"><a href="#cb33-8"></a><span class="co"># completeness score</span></span>
<span id="cb33-9"><a href="#cb33-9"></a>clusterOrder <span class="ot">&lt;-</span> <span class="st">"cluster_mash"</span>  <span class="co"># completeness, host, hg_pav, cluster_mash</span></span>
<span id="cb33-10"><a href="#cb33-10"></a></span>
<span id="cb33-11"><a href="#cb33-11"></a><span class="co"># regions to append as list of list with following structure</span></span>
<span id="cb33-12"><a href="#cb33-12"></a><span class="co"># list(r1 = list(chr, start, end, genomeId), r2 = list(chr, start, end, genomeId))</span></span>
<span id="cb33-13"><a href="#cb33-13"></a>customRegions <span class="ot">&lt;-</span> <span class="fu">list</span>(</span>
<span id="cb33-14"><a href="#cb33-14"></a>  <span class="at">g_177_reg =</span> <span class="fu">list</span>(</span>
<span id="cb33-15"><a href="#cb33-15"></a>    <span class="at">chr =</span> <span class="st">"NZ_JACGEP010000002.1"</span>, <span class="at">start =</span> <span class="dv">102351</span>, <span class="at">end =</span> <span class="dv">102361</span>, <span class="at">genomeId =</span> <span class="st">"g_177"</span></span>
<span id="cb33-16"><a href="#cb33-16"></a>  ),</span>
<span id="cb33-17"><a href="#cb33-17"></a>  <span class="at">g_182_reg =</span> <span class="fu">list</span>(</span>
<span id="cb33-18"><a href="#cb33-18"></a>    <span class="at">chr =</span> <span class="st">"NZ_JACGZZ010000050.1"</span>, <span class="at">start =</span> <span class="dv">81625</span>, <span class="at">end =</span> <span class="dv">81635</span>, <span class="at">genomeId =</span> <span class="st">"g_182"</span></span>
<span id="cb33-19"><a href="#cb33-19"></a>  ),</span>
<span id="cb33-20"><a href="#cb33-20"></a>  <span class="at">g_185_reg =</span> <span class="fu">list</span>(</span>
<span id="cb33-21"><a href="#cb33-21"></a>    <span class="at">chr =</span> <span class="st">"NZ_JACGEN010000006.1"</span>, <span class="at">start =</span> <span class="dv">81657</span>, <span class="at">end =</span> <span class="dv">81667</span>, <span class="at">genomeId =</span> <span class="st">"g_185"</span></span>
<span id="cb33-22"><a href="#cb33-22"></a>  ),</span>
<span id="cb33-23"><a href="#cb33-23"></a>  <span class="at">g_236_reg =</span> <span class="fu">list</span>(</span>
<span id="cb33-24"><a href="#cb33-24"></a>    <span class="at">chr =</span> <span class="st">"NZ_JACDSF010000027.1"</span>, <span class="at">start =</span> <span class="dv">89639</span>, <span class="at">end =</span> <span class="dv">89649</span>, <span class="at">genomeId =</span> <span class="st">"g_236"</span></span>
<span id="cb33-25"><a href="#cb33-25"></a>  )</span>
<span id="cb33-26"><a href="#cb33-26"></a>)</span>
<span id="cb33-27"><a href="#cb33-27"></a></span>
<span id="cb33-28"><a href="#cb33-28"></a></span>
<span id="cb33-29"><a href="#cb33-29"></a>grp <span class="ot">&lt;-</span> <span class="fu">list</span>(</span>
<span id="cb33-30"><a href="#cb33-30"></a>  <span class="at">phage_grp =</span> grpToView,</span>
<span id="cb33-31"><a href="#cb33-31"></a>  <span class="at">members =</span> dplyr<span class="sc">::</span><span class="fu">filter</span>(</span>
<span id="cb33-32"><a href="#cb33-32"></a>    regionClusters,</span>
<span id="cb33-33"><a href="#cb33-33"></a>    SpeciesName <span class="sc">==</span> <span class="st">"P. brasiliense"</span>, nFragments <span class="sc">==</span> <span class="dv">1</span>, phage_grp <span class="sc">==</span> <span class="st">"phage_grp_1"</span>,</span>
<span id="cb33-34"><a href="#cb33-34"></a>    <span class="sc">!</span>prophage_id <span class="sc">%in%</span> <span class="fu">c</span>(<span class="st">"g_408.vir_3"</span>, <span class="st">"g_403.vir_3"</span>, <span class="st">"g_399.vir_3"</span>)</span>
<span id="cb33-35"><a href="#cb33-35"></a>  ) <span class="sc">%&gt;%</span></span>
<span id="cb33-36"><a href="#cb33-36"></a>    dplyr<span class="sc">::</span><span class="fu">pull</span>(prophage_id)</span>
<span id="cb33-37"><a href="#cb33-37"></a>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<p>Prophage clusters found in the ctv-lacking Pbr: phage_grp_30, phage_grp_6, phage_grp_29</p>
</section>
<section id="ctv-in-p.-versatile" class="level3">
<h3 class="anchored" data-anchor-id="ctv-in-p.-versatile">CTV in <em>P. versatile</em></h3>
<div class="sourceCode" id="cb34"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb34-1"><a href="#cb34-1"></a>grpToView <span class="ot">&lt;-</span> <span class="st">"ctv_pvs"</span></span>
<span id="cb34-2"><a href="#cb34-2"></a>subSample <span class="ot">&lt;-</span> <span class="cn">TRUE</span></span>
<span id="cb34-3"><a href="#cb34-3"></a>cutHeight <span class="ot">&lt;-</span> <span class="fl">1.5</span></span>
<span id="cb34-4"><a href="#cb34-4"></a>addFlankingRegions <span class="ot">&lt;-</span> <span class="cn">TRUE</span></span>
<span id="cb34-5"><a href="#cb34-5"></a>flankingRegion <span class="ot">&lt;-</span> <span class="dv">5000</span></span>
<span id="cb34-6"><a href="#cb34-6"></a></span>
<span id="cb34-7"><a href="#cb34-7"></a></span>
<span id="cb34-8"><a href="#cb34-8"></a>grp <span class="ot">&lt;-</span> <span class="fu">list</span>(</span>
<span id="cb34-9"><a href="#cb34-9"></a>  <span class="at">phage_grp =</span> grpToView,</span>
<span id="cb34-10"><a href="#cb34-10"></a>  <span class="at">members =</span> dplyr<span class="sc">::</span><span class="fu">filter</span>(</span>
<span id="cb34-11"><a href="#cb34-11"></a>    regionClusters,</span>
<span id="cb34-12"><a href="#cb34-12"></a>    SpeciesName <span class="sc">==</span> <span class="st">"P. versatile"</span>, nFragments <span class="sc">==</span> <span class="dv">1</span>, phage_grp <span class="sc">==</span> <span class="st">"phage_grp_1"</span></span>
<span id="cb34-13"><a href="#cb34-13"></a>  ) <span class="sc">%&gt;%</span></span>
<span id="cb34-14"><a href="#cb34-14"></a>    dplyr<span class="sc">::</span><span class="fu">pull</span>(prophage_id)</span>
<span id="cb34-15"><a href="#cb34-15"></a>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</section>
<section id="ctv-cluster-in-p.-versatile-collected-from-france" class="level3">
<h3 class="anchored" data-anchor-id="ctv-cluster-in-p.-versatile-collected-from-france">CTV cluster in <em>P. versatile</em> collected from France</h3>
<div class="sourceCode" id="cb35"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb35-1"><a href="#cb35-1"></a>grpToView <span class="ot">&lt;-</span> <span class="st">"ctv_pvs_fr"</span></span>
<span id="cb35-2"><a href="#cb35-2"></a>subSample <span class="ot">&lt;-</span> <span class="cn">TRUE</span></span>
<span id="cb35-3"><a href="#cb35-3"></a>cutHeight <span class="ot">&lt;-</span> <span class="fl">1.5</span></span>
<span id="cb35-4"><a href="#cb35-4"></a>addFlankingRegions <span class="ot">&lt;-</span> <span class="cn">TRUE</span></span>
<span id="cb35-5"><a href="#cb35-5"></a>flankingRegion <span class="ot">&lt;-</span> <span class="dv">5000</span></span>
<span id="cb35-6"><a href="#cb35-6"></a></span>
<span id="cb35-7"><a href="#cb35-7"></a>grp <span class="ot">&lt;-</span> <span class="fu">list</span>(</span>
<span id="cb35-8"><a href="#cb35-8"></a>  <span class="at">phage_grp =</span> grpToView,</span>
<span id="cb35-9"><a href="#cb35-9"></a>  <span class="at">members =</span> dplyr<span class="sc">::</span><span class="fu">filter</span>(</span>
<span id="cb35-10"><a href="#cb35-10"></a>    regionClusters,</span>
<span id="cb35-11"><a href="#cb35-11"></a>    SpeciesName <span class="sc">==</span> <span class="st">"P. versatile"</span>, nFragments <span class="sc">==</span> <span class="dv">1</span>, phage_grp <span class="sc">==</span> <span class="st">"phage_grp_1"</span>,</span>
<span id="cb35-12"><a href="#cb35-12"></a>    geo_loc_country <span class="sc">==</span> <span class="st">"France"</span></span>
<span id="cb35-13"><a href="#cb35-13"></a>  ) <span class="sc">%&gt;%</span></span>
<span id="cb35-14"><a href="#cb35-14"></a>    dplyr<span class="sc">::</span><span class="fu">pull</span>(prophage_id)</span>
<span id="cb35-15"><a href="#cb35-15"></a>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</section>
</section>
<section id="carotovoricin-tail-fiber-variation-region-msa" class="level2">
<h2 class="anchored" data-anchor-id="carotovoricin-tail-fiber-variation-region-msa">Carotovoricin tail fiber variation region MSA</h2>
<p>Extract the region between the two homology groups, <code>hg_22427604</code> and <code>hg_22427603</code> for the carotovoricin cluster. Additionally, generate <code>GFF3</code> files with the homology group, COG, PFAM and other metadata information to visualize.</p>
<div class="sourceCode" id="cb36"><pre class="sourceCode numberSource bash number-lines code-with-copy"><code class="sourceCode bash"><span id="cb36-1"><a href="#cb36-1"></a><span class="co"># Rscript scripts/analysis/clustermap_get_regions.R</span></span>
<span id="cb36-2"><a href="#cb36-2"></a><span class="ex">Rscript</span> scripts/utils/HG_range_coordinates.R <span class="at">--hgs</span> hg_22427604,hg_22427603 <span class="dt">\</span></span>
<span id="cb36-3"><a href="#cb36-3"></a>--genomes <span class="st">"g_302,g_364,g_337,g_439,g_403,g_399,g_408,g_175,g_138,g_368,g_345,</span><span class="dt">\</span></span>
<span id="cb36-4"><a href="#cb36-4"></a><span class="st">g_366,g_308,g_191,g_173,g_155,g_166,g_299,g_438,g_263,g_43,g_391"</span> <span class="dt">\</span></span>
<span id="cb36-5"><a href="#cb36-5"></a>--out ctv_pbr.variable_regions.tab <span class="dt">\</span></span>
<span id="cb36-6"><a href="#cb36-6"></a>--dir analysis/pangenome_v2/prophages/cluster_viz/ctv_pbr</span>
<span id="cb36-7"><a href="#cb36-7"></a></span>
<span id="cb36-8"><a href="#cb36-8"></a><span class="co"># create CTV regions GFF3 files for one species representative each</span></span>
<span id="cb36-9"><a href="#cb36-9"></a><span class="ex">Rscript</span> scripts/utils/HGs_gff_from_regions.R <span class="dt">\</span></span>
<span id="cb36-10"><a href="#cb36-10"></a>--regions analysis/pangenome_v2/carotovoricin/tail_region/hg_regions.tab</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<p>Extract DNA sequence for these 3’ tail fiber regions.</p>
<div class="sourceCode" id="cb37"><pre class="sourceCode numberSource bash number-lines code-with-copy"><code class="sourceCode bash"><span id="cb37-1"><a href="#cb37-1"></a><span class="fu">mkdir</span> analysis/pangenome_v2/prophages/cluster_viz/ctv_pbr/region_fasta</span>
<span id="cb37-2"><a href="#cb37-2"></a></span>
<span id="cb37-3"><a href="#cb37-3"></a><span class="fu">tail</span> <span class="at">-n</span> +2 analysis/pangenome_v2/prophages/cluster_viz/ctv_pbr/ctv_pbr.variable_regions.tab <span class="kw">|</span></span>
<span id="cb37-4"><a href="#cb37-4"></a>  <span class="fu">cut</span> <span class="at">-f</span> 1,3-5 <span class="kw">|</span></span>
<span id="cb37-5"><a href="#cb37-5"></a>  <span class="cf">while</span> <span class="bu">read</span> <span class="va">name</span> <span class="va">sampleId</span> <span class="va">pos</span> <span class="va">strand</span><span class="kw">;</span> <span class="cf">do</span></span>
<span id="cb37-6"><a href="#cb37-6"></a>    <span class="va">rc</span><span class="op">=</span><span class="st">""</span></span>
<span id="cb37-7"><a href="#cb37-7"></a>    <span class="cf">if</span> <span class="bu">[</span> <span class="va">${strand}</span> <span class="ot">==</span> <span class="st">"-"</span> <span class="bu">]</span><span class="kw">;</span> <span class="cf">then</span></span>
<span id="cb37-8"><a href="#cb37-8"></a>        <span class="va">rc</span><span class="op">=</span><span class="st">"--reverse-complement"</span></span>
<span id="cb37-9"><a href="#cb37-9"></a>    <span class="cf">fi</span></span>
<span id="cb37-10"><a href="#cb37-10"></a>    <span class="va">outFa</span><span class="op">=</span><span class="st">"analysis/pangenome_v2/prophages/cluster_viz/ctv_pbr/region_fasta/</span><span class="va">${name}</span><span class="st">.fasta"</span></span>
<span id="cb37-11"><a href="#cb37-11"></a></span>
<span id="cb37-12"><a href="#cb37-12"></a>    <span class="va">seq</span><span class="op">=</span><span class="st">"&gt;</span><span class="va">${name}</span><span class="st"> </span><span class="va">${pos}</span><span class="st">(</span><span class="va">${strand}</span><span class="st">)\n"</span></span>
<span id="cb37-13"><a href="#cb37-13"></a>    <span class="va">seq</span><span class="op">+=</span><span class="kw">`</span><span class="ex">samtools</span> faidx <span class="va">${rc}</span> data/prokka_annotation/<span class="va">${sampleId}</span>/<span class="va">${sampleId}</span>.fna <span class="va">$pos</span> <span class="kw">|</span> <span class="fu">tail</span> <span class="at">-n</span> +2<span class="kw">`</span></span>
<span id="cb37-14"><a href="#cb37-14"></a>    <span class="bu">printf</span> <span class="st">"</span><span class="va">${seq}</span><span class="st">\n"</span> <span class="op">&gt;</span> <span class="va">${outFa}</span></span>
<span id="cb37-15"><a href="#cb37-15"></a>    <span class="bu">printf</span> <span class="st">"</span><span class="va">${seq}</span><span class="st">\n"</span></span>
<span id="cb37-16"><a href="#cb37-16"></a>  <span class="cf">done</span> <span class="op">&gt;</span>analysis/pangenome_v2/prophages/cluster_viz/ctv_pbr/ctv_pbr.variable_regions.fasta</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<p>Mauve aligns sequences provided in <code>fasta</code> and <code>genbank</code> format. However, the sequence annotation is shown only if the alignment input was in <code>genbank</code> format. Therefore, we need to combine the homology group annotation in <code>gff3</code> format with the <code>fasta</code> sequences to generate a <code>genbank</code> formatted files. To do this, Emboss tool <code>seqret</code> was used.</p>
<div class="sourceCode" id="cb38"><pre class="sourceCode numberSource bash number-lines code-with-copy"><code class="sourceCode bash"><span id="cb38-1"><a href="#cb38-1"></a><span class="ex">conda</span> activate omics_py37</span>
<span id="cb38-2"><a href="#cb38-2"></a></span>
<span id="cb38-3"><a href="#cb38-3"></a><span class="bu">cd</span> analysis/pangenome_v2/prophages/cluster_viz/ctv_pbr/region_fasta</span>
<span id="cb38-4"><a href="#cb38-4"></a></span>
<span id="cb38-5"><a href="#cb38-5"></a><span class="cf">for</span> vir <span class="kw">in</span> <span class="kw">`</span><span class="fu">cut</span> <span class="at">-f</span> 1 ../ctv_pbr.variable_regions.tab <span class="kw">|</span> <span class="fu">tail</span> <span class="at">-n</span> +2<span class="kw">`</span></span>
<span id="cb38-6"><a href="#cb38-6"></a><span class="cf">do</span></span>
<span id="cb38-7"><a href="#cb38-7"></a>  <span class="ex">seqret</span> <span class="at">-sequence</span> <span class="va">${vir}</span>.fasta <span class="at">-feature</span> <span class="at">-fformat</span> gff3 <span class="dt">\</span></span>
<span id="cb38-8"><a href="#cb38-8"></a>  <span class="at">-fopenfile</span> <span class="va">${vir}</span>.gff3 <span class="at">-osformat</span> genbank <span class="at">-osextension</span> gbk <span class="dt">\</span></span>
<span id="cb38-9"><a href="#cb38-9"></a>  <span class="at">-osname_outseq</span> <span class="va">${vir}</span> <span class="at">-ofdirectory_outseq</span> gbk_file <span class="at">-auto</span> </span>
<span id="cb38-10"><a href="#cb38-10"></a><span class="cf">done</span></span>
<span id="cb38-11"><a href="#cb38-11"></a></span>
<span id="cb38-12"><a href="#cb38-12"></a><span class="fu">sed</span> <span class="at">-i.bak</span> <span class="at">-r</span> <span class="at">-e</span> <span class="st">'s/(^\s+CDS\s+(complement\()?)[^:]+:([[:digit:]]+\.\.[[:digit:]]+\)?).*/\1\3/'</span> <span class="dt">\</span></span>
<span id="cb38-13"><a href="#cb38-13"></a>  <span class="at">-e</span> <span class="st">'s/\/note="\*([^:]+): /\/\1="/'</span> <span class="pp">*</span>.gbk </span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<p>Align and visualize sequences using Mauve.</p>
<div class="sourceCode" id="cb39"><pre class="sourceCode numberSource bash number-lines code-with-copy"><code class="sourceCode bash"><span id="cb39-1"><a href="#cb39-1"></a><span class="va">mauve_out</span><span class="op">=</span><span class="st">"ctv_pbr"</span></span>
<span id="cb39-2"><a href="#cb39-2"></a></span>
<span id="cb39-3"><a href="#cb39-3"></a><span class="ex">/Applications/Mauve.app/Contents/MacOS/progressiveMauve</span> <span class="dt">\</span></span>
<span id="cb39-4"><a href="#cb39-4"></a>--output=<span class="st">"</span><span class="va">${mauve_out}</span><span class="st">.mauve.xmfa"</span> <span class="dt">\</span></span>
<span id="cb39-5"><a href="#cb39-5"></a>--backbone-output=<span class="st">"</span><span class="va">${mauve_out}</span><span class="st">.mauve.backbone"</span> <span class="dt">\</span></span>
<span id="cb39-6"><a href="#cb39-6"></a>--output-guide-tree=<span class="st">"</span><span class="va">${mauve_out}</span><span class="st">.mauve.guide_tree.newick"</span> <span class="dt">\</span></span>
<span id="cb39-7"><a href="#cb39-7"></a>region_fasta/g_302.vir_3.gbk region_fasta/g_364.vir_2.gbk <span class="dt">\</span></span>
<span id="cb39-8"><a href="#cb39-8"></a>region_fasta/g_337.vir_1.gbk region_fasta/g_439.vir_2.gbk <span class="dt">\</span></span>
<span id="cb39-9"><a href="#cb39-9"></a>region_fasta/g_403.vir_3.gbk region_fasta/g_399.vir_3.gbk <span class="dt">\</span></span>
<span id="cb39-10"><a href="#cb39-10"></a>region_fasta/g_175.vir_3.gbk region_fasta/g_408.vir_3.gbk <span class="dt">\</span></span>
<span id="cb39-11"><a href="#cb39-11"></a>region_fasta/g_138.vir_2.gbk <span class="dt">\</span></span>
<span id="cb39-12"><a href="#cb39-12"></a>region_fasta/g_368.vir_3.gbk region_fasta/g_345.vir_1.gbk <span class="dt">\</span></span>
<span id="cb39-13"><a href="#cb39-13"></a>region_fasta/g_366.vir_3.gbk region_fasta/g_308.vir_2.gbk <span class="dt">\</span></span>
<span id="cb39-14"><a href="#cb39-14"></a>region_fasta/g_191.vir_1.gbk region_fasta/g_173.vir_2.gbk <span class="dt">\</span></span>
<span id="cb39-15"><a href="#cb39-15"></a>region_fasta/g_155.vir_1.gbk region_fasta/g_166.vir_2.gbk <span class="dt">\</span></span>
<span id="cb39-16"><a href="#cb39-16"></a>region_fasta/g_299.vir_1.gbk region_fasta/g_438.vir_2.gbk <span class="dt">\</span></span>
<span id="cb39-17"><a href="#cb39-17"></a>region_fasta/g_263.vir_2.gbk region_fasta/g_43.vir_1.gbk <span class="dt">\</span></span>
<span id="cb39-18"><a href="#cb39-18"></a>region_fasta/g_391.vir_3.gbk  <span class="op">&gt;</span> <span class="st">"</span><span class="va">${mauve_out}</span><span class="st">"</span>.log <span class="dv">2</span><span class="op">&gt;&amp;</span><span class="dv">1</span></span>
<span id="cb39-19"><a href="#cb39-19"></a></span>
<span id="cb39-20"><a href="#cb39-20"></a><span class="va">mauve_out</span><span class="op">=</span><span class="st">"ctv_pbr.ein"</span></span>
<span id="cb39-21"><a href="#cb39-21"></a></span>
<span id="cb39-22"><a href="#cb39-22"></a><span class="ex">/Applications/Mauve.app/Contents/MacOS/progressiveMauve</span> <span class="dt">\</span></span>
<span id="cb39-23"><a href="#cb39-23"></a>--output=<span class="st">"</span><span class="va">${mauve_out}</span><span class="st">.mauve.xmfa"</span> <span class="dt">\</span></span>
<span id="cb39-24"><a href="#cb39-24"></a>--backbone-output=<span class="st">"</span><span class="va">${mauve_out}</span><span class="st">.mauve.backbone"</span> <span class="dt">\</span></span>
<span id="cb39-25"><a href="#cb39-25"></a>--output-guide-tree=<span class="st">"</span><span class="va">${mauve_out}</span><span class="st">.mauve.guide_tree.newick"</span> <span class="dt">\</span></span>
<span id="cb39-26"><a href="#cb39-26"></a>region_fasta/g_345.vir_1.gbk region_fasta/g_366.vir_3.gbk <span class="dt">\</span></span>
<span id="cb39-27"><a href="#cb39-27"></a>region_fasta/g_308.vir_2.gbk region_fasta/g_191.vir_1.gbk <span class="dt">\</span></span>
<span id="cb39-28"><a href="#cb39-28"></a>region_fasta/g_173.vir_2.gbk region_fasta/g_155.vir_1.gbk <span class="dt">\</span></span>
<span id="cb39-29"><a href="#cb39-29"></a>region_fasta/g_166.vir_2.gbk  <span class="op">&gt;</span> <span class="st">"</span><span class="va">${mauve_out}</span><span class="st">"</span>.log <span class="dv">2</span><span class="op">&gt;&amp;</span><span class="dv">1</span></span>
<span id="cb39-30"><a href="#cb39-30"></a></span>
<span id="cb39-31"><a href="#cb39-31"></a></span>
<span id="cb39-32"><a href="#cb39-32"></a><span class="va">mauve_out</span><span class="op">=</span><span class="st">"ctv_pbr.rep"</span></span>
<span id="cb39-33"><a href="#cb39-33"></a></span>
<span id="cb39-34"><a href="#cb39-34"></a><span class="ex">/Applications/Mauve.app/Contents/MacOS/progressiveMauve</span> <span class="dt">\</span></span>
<span id="cb39-35"><a href="#cb39-35"></a>--output=<span class="st">"</span><span class="va">${mauve_out}</span><span class="st">.mauve.xmfa"</span> <span class="dt">\</span></span>
<span id="cb39-36"><a href="#cb39-36"></a>--backbone-output=<span class="st">"</span><span class="va">${mauve_out}</span><span class="st">.mauve.backbone"</span> <span class="dt">\</span></span>
<span id="cb39-37"><a href="#cb39-37"></a>--output-guide-tree=<span class="st">"</span><span class="va">${mauve_out}</span><span class="st">.mauve.guide_tree.newick"</span> <span class="dt">\</span></span>
<span id="cb39-38"><a href="#cb39-38"></a>region_fasta/g_302.vir_3.gbk region_fasta/g_337.vir_1.gbk <span class="dt">\</span></span>
<span id="cb39-39"><a href="#cb39-39"></a>region_fasta/g_175.vir_3.gbk region_fasta/g_173.vir_2.gbk <span class="dt">\</span></span>
<span id="cb39-40"><a href="#cb39-40"></a>region_fasta/g_299.vir_1.gbk region_fasta/g_438.vir_2.gbk <span class="dt">\</span></span>
<span id="cb39-41"><a href="#cb39-41"></a>region_fasta/g_391.vir_3.gbk  <span class="op">&gt;</span> <span class="st">"</span><span class="va">${mauve_out}</span><span class="st">"</span>.log <span class="dv">2</span><span class="op">&gt;&amp;</span><span class="dv">1</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<p>Perform MSA using <code>MAFFT</code></p>
<div class="sourceCode" id="cb40"><pre class="sourceCode numberSource bash number-lines code-with-copy"><code class="sourceCode bash"><span id="cb40-1"><a href="#cb40-1"></a><span class="ex">conda</span> activate pantools_v4_3</span>
<span id="cb40-2"><a href="#cb40-2"></a><span class="bu">cd</span> analysis/pangenome_v2/prophages/cluster_viz/ctv_pbr/</span>
<span id="cb40-3"><a href="#cb40-3"></a></span>
<span id="cb40-4"><a href="#cb40-4"></a><span class="ex">mafft</span> <span class="at">--globalpair</span> <span class="at">--quiet</span> <span class="at">--maxiterate</span> 1000 <span class="at">--treeout</span> ctv_pbr.variable_regions.fasta</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<p>Smash++ pairwise sequence comparison</p>
<div class="sourceCode" id="cb41"><pre class="sourceCode numberSource bash number-lines code-with-copy"><code class="sourceCode bash"><span id="cb41-1"><a href="#cb41-1"></a><span class="ex">conda</span> activate omics_py37</span>
<span id="cb41-2"><a href="#cb41-2"></a></span>
<span id="cb41-3"><a href="#cb41-3"></a><span class="bu">cd</span> analysis/pangenome_v2/prophages/cluster_viz/ctv_pbr/region_fasta</span>
<span id="cb41-4"><a href="#cb41-4"></a></span>
<span id="cb41-5"><a href="#cb41-5"></a><span class="kw">function</span><span class="fu"> smashpp_compare</span></span>
<span id="cb41-6"><a href="#cb41-6"></a><span class="kw">{</span></span>
<span id="cb41-7"><a href="#cb41-7"></a>  <span class="ex">smashpp</span> <span class="at">-r</span> <span class="va">$1</span> <span class="at">-t</span> <span class="va">$2</span></span>
<span id="cb41-8"><a href="#cb41-8"></a>  <span class="ex">smashpp</span> viz <span class="at">-o</span> <span class="va">$(</span><span class="fu">basename</span> <span class="va">$1</span> .fasta<span class="va">)$(</span><span class="fu">basename</span> <span class="va">$2</span> .fasta<span class="va">)</span><span class="st">".svg"</span> <span class="st">"</span><span class="va">${1}</span><span class="st">.</span><span class="va">${2}</span><span class="st">.pos"</span></span>
<span id="cb41-9"><a href="#cb41-9"></a><span class="kw">}</span></span>
<span id="cb41-10"><a href="#cb41-10"></a></span>
<span id="cb41-11"><a href="#cb41-11"></a><span class="bu">export</span> <span class="at">-f</span> <span class="va">smashpp_compare</span></span>
<span id="cb41-12"><a href="#cb41-12"></a></span>
<span id="cb41-13"><a href="#cb41-13"></a><span class="ex">smashpp_compare</span> g_345.vir_1.fasta g_345.vir_1.fasta </span>
<span id="cb41-14"><a href="#cb41-14"></a><span class="ex">smashpp_compare</span> g_345.vir_1.fasta g_366.vir_3.fasta </span>
<span id="cb41-15"><a href="#cb41-15"></a><span class="ex">smashpp_compare</span> g_345.vir_1.fasta g_302.vir_3.fasta</span>
<span id="cb41-16"><a href="#cb41-16"></a><span class="ex">smashpp_compare</span> g_345.vir_1.fasta g_403.vir_3.fasta</span>
<span id="cb41-17"><a href="#cb41-17"></a><span class="ex">smashpp_compare</span> g_345.vir_1.fasta g_173.vir_2.fasta</span>
<span id="cb41-18"><a href="#cb41-18"></a><span class="ex">smashpp_compare</span> g_345.vir_1.fasta g_138.vir_2.fasta</span>
<span id="cb41-19"><a href="#cb41-19"></a><span class="ex">smashpp_compare</span> g_345.vir_1.fasta g_263.vir_2.fasta</span>
<span id="cb41-20"><a href="#cb41-20"></a><span class="ex">smashpp_compare</span> g_345.vir_1.fasta g_155.vir_1.fasta</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<p>Left inverted repeat for Ein: CTCCCGCAAACCTCGGTTTTGGGGAC (CTCCCGCAAACCTCGGTTT)<br>
Left inverted repeat for Ein(rev-com): GTCCCCAAAACCGAGGTTTGCGGGAG (AAACCGAGGTTTGCGGGAG, AAACCGAGGTTTGCG)</p>
<p>Right inverted repeat for Ein: TTCTCGCAAACCTCGGTTTTGGAGAA<br>
Right inverted repeat for Ein(rev): AAGAGGTTTTGGCTCCAAACGCTCTT<br>
Right inverted repeat for Ein(rev-com): TTCTCCAAAACCGAGGTTTGCGAGAA (AAACCGAGGTTTGCGAGAA, AAACCGAGGTTTGCG)<br>
Right inverted repeat for Ein(comp): AAGAGCGTTTGGAGCCAAAACCTCTT</p>
</section>
<section id="inter-species-horizontal-gene-transfer-of-ctv" class="level2">
<h2 class="anchored" data-anchor-id="inter-species-horizontal-gene-transfer-of-ctv">Inter-species horizontal gene transfer of CTV</h2>
<p>Clustermap visualization for some manually selected candidates. Change the code blocks in the script <code>scripts/analysis/clustersmap_data_prophages.R</code> and run.</p>
<div class="sourceCode" id="cb42"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb42-1"><a href="#cb42-1"></a>grpToView <span class="ot">&lt;-</span> <span class="st">"ctv_hgt"</span></span>
<span id="cb42-2"><a href="#cb42-2"></a>subSample <span class="ot">&lt;-</span> <span class="cn">FALSE</span> </span>
<span id="cb42-3"><a href="#cb42-3"></a>cutHeight <span class="ot">&lt;-</span> <span class="fl">1.5</span></span>
<span id="cb42-4"><a href="#cb42-4"></a>addFlankingRegions <span class="ot">&lt;-</span> <span class="cn">TRUE</span></span>
<span id="cb42-5"><a href="#cb42-5"></a>flankingRegion <span class="ot">&lt;-</span> <span class="dv">5000</span></span>
<span id="cb42-6"><a href="#cb42-6"></a></span>
<span id="cb42-7"><a href="#cb42-7"></a><span class="co"># ordering factor for prophages: host phylogeny, prophage HG PAV, prophage MASH,</span></span>
<span id="cb42-8"><a href="#cb42-8"></a><span class="co"># completeness score</span></span>
<span id="cb42-9"><a href="#cb42-9"></a>clusterOrder <span class="ot">&lt;-</span> <span class="st">"hg_pav"</span> <span class="co"># completeness, host, hg_pav, cluster_mash</span></span>
<span id="cb42-10"><a href="#cb42-10"></a></span>
<span id="cb42-11"><a href="#cb42-11"></a></span>
<span id="cb42-12"><a href="#cb42-12"></a><span class="co"># optionally, a custom region list can be provided to generate the plot</span></span>
<span id="cb42-13"><a href="#cb42-13"></a>grp <span class="ot">&lt;-</span> <span class="fu">list</span>(</span>
<span id="cb42-14"><a href="#cb42-14"></a>  <span class="at">phage_grp =</span> grpToView,</span>
<span id="cb42-15"><a href="#cb42-15"></a>  <span class="at">members =</span> <span class="fu">c</span>(</span>
<span id="cb42-16"><a href="#cb42-16"></a>    <span class="st">"g_145.vir_1"</span>, <span class="st">"g_194.vir_1"</span>, <span class="st">"g_429.vir_1"</span>, <span class="st">"g_442.vir_1"</span>, <span class="st">"g_421.vir_1"</span>, </span>
<span id="cb42-17"><a href="#cb42-17"></a>    <span class="st">"g_150.vir_4"</span>, <span class="st">"g_447.vir_1"</span>, <span class="st">"g_434.vir_4"</span>,</span>
<span id="cb42-18"><a href="#cb42-18"></a>    <span class="st">"g_221.vir_3"</span>, <span class="st">"g_53.vir_3"</span>, <span class="st">"g_106.vir_2"</span>, <span class="st">"g_57.vir_1"</span>, <span class="st">"g_125.vir_1"</span></span>
<span id="cb42-19"><a href="#cb42-19"></a>  )</span>
<span id="cb42-20"><a href="#cb42-20"></a>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<p>Run script <code>scripts/utils/HG_range_coordinates.R</code> to extract genomic coordinates for CTV region in selected genomes.</p>
<p>Complete CTV region along with 3 flanking genes:</p>
<div class="sourceCode" id="cb43"><pre class="sourceCode numberSource bash number-lines code-with-copy"><code class="sourceCode bash"><span id="cb43-1"><a href="#cb43-1"></a><span class="ex">Rscript</span> scripts/utils/HG_range_coordinates.R <span class="at">--hgs</span> hg_22427643,hg_22427599 <span class="dt">\</span></span>
<span id="cb43-2"><a href="#cb43-2"></a>--inner_region <span class="at">--dir</span> analysis/pangenome_v2/carotovoricin/ctv_region</span>
<span id="cb43-3"><a href="#cb43-3"></a></span>
<span id="cb43-4"><a href="#cb43-4"></a><span class="co"># create GFF3 files for CTV regions of interest</span></span>
<span id="cb43-5"><a href="#cb43-5"></a><span class="ex">Rscript</span> scripts/utils/HGs_gff_from_regions.R <span class="dt">\</span></span>
<span id="cb43-6"><a href="#cb43-6"></a>--regions analysis/pangenome_v2/carotovoricin/ctv_region/hg_regions.tab <span class="dt">\</span></span>
<span id="cb43-7"><a href="#cb43-7"></a>--genomes g_145,g_194,g_429,g_442,g_421,g_150,g_447,g_434,g_53,g_106,g_57,g_221,g_125</span>
<span id="cb43-8"><a href="#cb43-8"></a></span>
<span id="cb43-9"><a href="#cb43-9"></a><span class="co"># create CTV regions GFF3 files for one species representative each</span></span>
<span id="cb43-10"><a href="#cb43-10"></a><span class="ex">Rscript</span> scripts/utils/HGs_gff_from_regions.R <span class="dt">\</span></span>
<span id="cb43-11"><a href="#cb43-11"></a>--regions analysis/pangenome_v2/carotovoricin/ctv_region/hg_regions.tab <span class="dt">\</span></span>
<span id="cb43-12"><a href="#cb43-12"></a>--genomes g_183,g_423,g_277,g_375,g_116,g_449,g_446,g_378,g_296,g_428,g_259,g_66,g_337,g_442,g_305,g_381,g_8,g_273,g_38</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<p>CTV conserved loci:</p>
<div class="sourceCode" id="cb44"><pre class="sourceCode numberSource bash number-lines code-with-copy"><code class="sourceCode bash"><span id="cb44-1"><a href="#cb44-1"></a><span class="ex">Rscript</span> scripts/utils/HG_range_coordinates.R <span class="at">--hgs</span> hg_22427640,hg_22427604 <span class="dt">\</span></span>
<span id="cb44-2"><a href="#cb44-2"></a>--dir analysis/pangenome_v2/carotovoricin/ctv_conserved</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<p>CTV variable loci:</p>
<div class="sourceCode" id="cb45"><pre class="sourceCode numberSource bash number-lines code-with-copy"><code class="sourceCode bash"><span id="cb45-1"><a href="#cb45-1"></a><span class="ex">Rscript</span> scripts/utils/HG_range_coordinates.R <span class="at">--hgs</span> hg_22427604,hg_22427603 <span class="dt">\</span></span>
<span id="cb45-2"><a href="#cb45-2"></a>--inner_region <span class="at">--dir</span> analysis/pangenome_v2/carotovoricin/ctv_tail <span class="at">--haplotypes</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<p>CTV tape measure protein gene:</p>
<div class="sourceCode" id="cb46"><pre class="sourceCode numberSource bash number-lines code-with-copy"><code class="sourceCode bash"><span id="cb46-1"><a href="#cb46-1"></a><span class="ex">Rscript</span> scripts/utils/HG_range_coordinates.R <span class="at">--hgs</span> hg_22427622,hg_22427616 <span class="dt">\</span></span>
<span id="cb46-2"><a href="#cb46-2"></a>--haplotypes <span class="at">--inner_region</span> <span class="at">--overlapping</span> <span class="dt">\</span></span>
<span id="cb46-3"><a href="#cb46-3"></a>--dir analysis/pangenome_v2/carotovoricin/tape_measure</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<p>Run <code>scripts/analysis/ctv_hgt.qmd</code> script to perform HGT analysis.</p>
<section id="dna-sequence-comparison" class="level3">
<h3 class="anchored" data-anchor-id="dna-sequence-comparison">DNA sequence comparison</h3>
<section id="extract-dna-sequence-for-the-regions" class="level4">
<h4 class="anchored" data-anchor-id="extract-dna-sequence-for-the-regions">Extract DNA sequence for the regions</h4>
<div class="sourceCode" id="cb47"><pre class="sourceCode numberSource bash number-lines code-with-copy"><code class="sourceCode bash"><span id="cb47-1"><a href="#cb47-1"></a><span class="co"># file_regions="analysis/pangenome_v2/carotovoricin/ctv_tail/hg_regions.tab"</span></span>
<span id="cb47-2"><a href="#cb47-2"></a><span class="co"># file_regions="analysis/pangenome_v2/carotovoricin/ctv_conserved/hg_regions.tab"</span></span>
<span id="cb47-3"><a href="#cb47-3"></a><span class="co"># file_regions="analysis/pangenome_v2/carotovoricin/ctv_tail/selected_haplotypes.tab"</span></span>
<span id="cb47-4"><a href="#cb47-4"></a><span class="co"># file_regions="analysis/pangenome_v2/carotovoricin/ctv_tail/selected_haplotypes_flanking.tab"</span></span>
<span id="cb47-5"><a href="#cb47-5"></a><span class="co"># file_regions="analysis/pangenome_v2/carotovoricin/ctv_conserved/selected_conserved.tab"</span></span>
<span id="cb47-6"><a href="#cb47-6"></a><span class="co"># file_regions="analysis/pangenome_v2/carotovoricin/upstream_core/selected_upstream.tab"</span></span>
<span id="cb47-7"><a href="#cb47-7"></a></span>
<span id="cb47-8"><a href="#cb47-8"></a><span class="va">dir_path</span><span class="op">=</span><span class="va">$(</span><span class="fu">dirname</span> <span class="st">"</span><span class="va">${file_regions}</span><span class="st">"</span><span class="va">)</span></span>
<span id="cb47-9"><a href="#cb47-9"></a><span class="va">out_fasta</span><span class="op">=</span><span class="st">"</span><span class="va">${file_regions</span><span class="op">%</span>.<span class="pp">*</span><span class="va">}</span><span class="st">"</span>.fasta</span>
<span id="cb47-10"><a href="#cb47-10"></a></span>
<span id="cb47-11"><a href="#cb47-11"></a><span class="fu">tail</span> <span class="at">-n</span> +2 <span class="va">${file_regions}</span> <span class="kw">|</span></span>
<span id="cb47-12"><a href="#cb47-12"></a>  <span class="fu">cut</span> <span class="at">-f</span> 1,3-5 <span class="kw">|</span></span>
<span id="cb47-13"><a href="#cb47-13"></a>  <span class="cf">while</span> <span class="bu">read</span> <span class="va">name</span> <span class="va">sampleId</span> <span class="va">pos</span> <span class="va">strand</span><span class="kw">;</span> <span class="cf">do</span></span>
<span id="cb47-14"><a href="#cb47-14"></a>    <span class="va">rc</span><span class="op">=</span><span class="st">""</span></span>
<span id="cb47-15"><a href="#cb47-15"></a>    <span class="cf">if</span> <span class="bu">[</span> <span class="va">${strand}</span> <span class="ot">==</span> <span class="st">"-"</span> <span class="bu">]</span><span class="kw">;</span> <span class="cf">then</span></span>
<span id="cb47-16"><a href="#cb47-16"></a>        <span class="va">rc</span><span class="op">=</span><span class="st">"--reverse-complement"</span></span>
<span id="cb47-17"><a href="#cb47-17"></a>    <span class="cf">fi</span></span>
<span id="cb47-18"><a href="#cb47-18"></a>    <span class="va">outFa</span><span class="op">=</span><span class="st">"</span><span class="va">${dir_path}</span><span class="st">/fasta/</span><span class="va">${name}</span><span class="st">.fasta"</span></span>
<span id="cb47-19"><a href="#cb47-19"></a></span>
<span id="cb47-20"><a href="#cb47-20"></a>    <span class="va">seq</span><span class="op">=</span><span class="st">"&gt;</span><span class="va">${name}</span><span class="st">\n"</span></span>
<span id="cb47-21"><a href="#cb47-21"></a>    <span class="va">seq</span><span class="op">+=</span><span class="kw">`</span><span class="ex">samtools</span> faidx <span class="va">${rc}</span> data/prokka_annotation/<span class="va">${sampleId}</span>/<span class="va">${sampleId}</span>.fna <span class="va">$pos</span> <span class="kw">|</span> <span class="fu">tail</span> <span class="at">-n</span> +2<span class="kw">`</span></span>
<span id="cb47-22"><a href="#cb47-22"></a></span>
<span id="cb47-23"><a href="#cb47-23"></a>    <span class="co"># printf "${seq}\n" &gt; ${outFa}</span></span>
<span id="cb47-24"><a href="#cb47-24"></a>    <span class="bu">printf</span> <span class="st">"</span><span class="va">${seq}</span><span class="st">\n"</span></span>
<span id="cb47-25"><a href="#cb47-25"></a>  <span class="cf">done</span> <span class="op">&gt;</span> <span class="va">${out_fasta}</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</section>
<section id="calculate-the-distance-using-mash" class="level4">
<h4 class="anchored" data-anchor-id="calculate-the-distance-using-mash">Calculate the distance using <code>mash</code></h4>
<div class="sourceCode" id="cb48"><pre class="sourceCode numberSource bash number-lines code-with-copy"><code class="sourceCode bash"><span id="cb48-1"><a href="#cb48-1"></a><span class="bu">cd</span> analysis/pangenome_v2/carotovoricin/ctv_tail</span>
<span id="cb48-2"><a href="#cb48-2"></a><span class="ex">mash</span> dist <span class="at">-p</span> 8 <span class="at">-k</span> 12 <span class="at">-s</span> 2000 <span class="at">-i</span> <span class="at">-S</span> 124 hg_regions.fasta hg_regions.fasta <span class="op">&gt;</span> mash_dist.tab</span>
<span id="cb48-3"><a href="#cb48-3"></a></span>
<span id="cb48-4"><a href="#cb48-4"></a><span class="bu">cd</span> analysis/pangenome_v2/carotovoricin/ctv_conserved</span>
<span id="cb48-5"><a href="#cb48-5"></a><span class="ex">mash</span> dist <span class="at">-p</span> 8 <span class="at">-k</span> 12 <span class="at">-s</span> 2000 <span class="at">-i</span> <span class="at">-S</span> 124 hg_regions.fasta hg_regions.fasta <span class="op">&gt;</span> mash_dist.tab</span>
<span id="cb48-6"><a href="#cb48-6"></a></span>
<span id="cb48-7"><a href="#cb48-7"></a><span class="bu">cd</span> analysis/pangenome_v2/carotovoricin/tape_measure</span>
<span id="cb48-8"><a href="#cb48-8"></a><span class="ex">mash</span> dist <span class="at">-p</span> 8 <span class="at">-k</span> 12 <span class="at">-s</span> 2000 <span class="at">-i</span> <span class="at">-S</span> 124 hg_regions.fasta hg_regions.fasta <span class="op">&gt;</span> mash_dist.tab</span>
<span id="cb48-9"><a href="#cb48-9"></a></span>
<span id="cb48-10"><a href="#cb48-10"></a><span class="fu">grep</span> <span class="at">-e</span> <span class="st">'g_(53|106|57|221|125).*g_(53|106|57|221|125)'</span> ctv_<span class="pp">*</span>/ctv_dist.tab</span>
<span id="cb48-11"><a href="#cb48-11"></a><span class="fu">grep</span> <span class="at">-e</span> <span class="st">'g_(150|447|434).*g_(150|447|434)'</span> ctv_<span class="pp">*</span>/ctv_dist.tab</span>
<span id="cb48-12"><a href="#cb48-12"></a><span class="fu">grep</span> <span class="at">-e</span> <span class="st">'g_(53|106|57).*g_(53|106|57)'</span> ctv_<span class="pp">*</span>/ctv_dist.tab</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</section>
<section id="calculate-jaccard-distance-matrix-using-sourmash" class="level4">
<h4 class="anchored" data-anchor-id="calculate-jaccard-distance-matrix-using-sourmash">Calculate Jaccard distance matrix using <code>sourmash</code></h4>
<p>MASH distance is low for haplotype pairs where one is subset of the other, eg, A-B-C-D and A-B. Therefore, another tool was tried to calculate the Jaccard index between sequences.</p>
<div class="sourceCode" id="cb49"><pre class="sourceCode numberSource bash number-lines code-with-copy"><code class="sourceCode bash"><span id="cb49-1"><a href="#cb49-1"></a><span class="co"># file_fa="ctv_conserved/hg_regions.fasta"</span></span>
<span id="cb49-2"><a href="#cb49-2"></a><span class="co"># file_fa="ctv_tail/hg_regions.fasta"</span></span>
<span id="cb49-3"><a href="#cb49-3"></a></span>
<span id="cb49-4"><a href="#cb49-4"></a><span class="va">file_sig</span><span class="op">=</span><span class="st">"</span><span class="va">${file_fa</span><span class="op">%</span>.<span class="pp">*</span><span class="va">}</span><span class="st">"</span>.sig.zip</span>
<span id="cb49-5"><a href="#cb49-5"></a><span class="va">file_dist</span><span class="op">=</span><span class="st">"</span><span class="va">${file_fa</span><span class="op">%</span>.<span class="pp">*</span><span class="va">}</span><span class="st">"</span>.sourmash.csv</span>
<span id="cb49-6"><a href="#cb49-6"></a></span>
<span id="cb49-7"><a href="#cb49-7"></a><span class="co"># # sketch: scaled=5</span></span>
<span id="cb49-8"><a href="#cb49-8"></a><span class="co"># sourmash sketch dna -f -p k=7,k=9,k=11,scaled=5,abund,seed=124 --singleton \</span></span>
<span id="cb49-9"><a href="#cb49-9"></a><span class="co"># -o ${file_sig} ${file_fa}</span></span>
<span id="cb49-10"><a href="#cb49-10"></a></span>
<span id="cb49-11"><a href="#cb49-11"></a><span class="co"># sketch: num=2000</span></span>
<span id="cb49-12"><a href="#cb49-12"></a><span class="ex">sourmash</span> sketch dna <span class="at">-f</span> <span class="at">-p</span> k=7,k=9,k=11,k=13,num=2000,abund,seed=124 <span class="at">--singleton</span> <span class="dt">\</span></span>
<span id="cb49-13"><a href="#cb49-13"></a>-o <span class="va">${file_sig}</span> <span class="va">${file_fa}</span></span>
<span id="cb49-14"><a href="#cb49-14"></a></span>
<span id="cb49-15"><a href="#cb49-15"></a><span class="ex">sourmash</span> compare <span class="at">--ignore-abundance</span> <span class="at">--distance-matrix</span> <span class="at">-p</span> 12 <span class="at">-k</span> 13 <span class="at">--dna</span> <span class="dt">\</span></span>
<span id="cb49-16"><a href="#cb49-16"></a>--csv <span class="va">${file_dist}</span> <span class="va">${file_sig}</span> </span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</section>
<section id="calculate-jaccard-distance-using-dashing" class="level4">
<h4 class="anchored" data-anchor-id="calculate-jaccard-distance-using-dashing">Calculate Jaccard distance using <code>dashing</code></h4>
<div class="sourceCode" id="cb50"><pre class="sourceCode numberSource bash number-lines code-with-copy"><code class="sourceCode bash"><span id="cb50-1"><a href="#cb50-1"></a><span class="co"># dir_fa="ctv_conserved/fasta"</span></span>
<span id="cb50-2"><a href="#cb50-2"></a><span class="co"># dir_fa="ctv_tail/fasta"</span></span>
<span id="cb50-3"><a href="#cb50-3"></a></span>
<span id="cb50-4"><a href="#cb50-4"></a><span class="va">file_dist</span><span class="op">=</span><span class="st">"</span><span class="va">${dir_fa</span><span class="op">%</span>/<span class="pp">*</span><span class="va">}</span><span class="st">/"</span>distance.dashing.txt</span>
<span id="cb50-5"><a href="#cb50-5"></a></span>
<span id="cb50-6"><a href="#cb50-6"></a><span class="ex">dashing</span> cmp <span class="at">-k</span> 11 <span class="at">--seed</span> 124 <span class="at">--nthreads</span> 12 <span class="at">--full-mash-dist</span>  <span class="dt">\</span></span>
<span id="cb50-7"><a href="#cb50-7"></a>-O <span class="va">${file_dist}</span> <span class="va">${dir_fa}</span>/<span class="pp">*</span>.fasta</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</section>
</section>
<section id="tail-fiber-haplotype-examples" class="level3">
<h3 class="anchored" data-anchor-id="tail-fiber-haplotype-examples">Tail fiber haplotype examples</h3>
<p>Extract sequences for some example tail fiber loci haplotypes that are present in multiple <em>Pectobacterium</em> species.</p>
<div class="sourceCode" id="cb51"><pre class="sourceCode numberSource bash number-lines code-with-copy"><code class="sourceCode bash"><span id="cb51-1"><a href="#cb51-1"></a><span class="co"># tail fiber locus</span></span>
<span id="cb51-2"><a href="#cb51-2"></a><span class="ex">Rscript</span> scripts/utils/HG_range_coordinates.R <span class="at">--hgs</span> hg_22427604,hg_22427603 <span class="dt">\</span></span>
<span id="cb51-3"><a href="#cb51-3"></a>--genomes <span class="st">"g_279,g_425,g_149,g_377,g_100,g_106,g_331,g_249,g_125,</span><span class="dt">\</span></span>
<span id="cb51-4"><a href="#cb51-4"></a><span class="st">g_221,g_53,g_395,g_108,g_444,g_160"</span> <span class="dt">\</span></span>
<span id="cb51-5"><a href="#cb51-5"></a>--dir analysis/pangenome_v2/carotovoricin/ctv_tail <span class="dt">\</span></span>
<span id="cb51-6"><a href="#cb51-6"></a>--out <span class="st">"selected_haplotypes_flanking.tab"</span></span>
<span id="cb51-7"><a href="#cb51-7"></a></span>
<span id="cb51-8"><a href="#cb51-8"></a><span class="ex">Rscript</span> scripts/utils/HG_range_coordinates.R <span class="at">--hgs</span> hg_22427604,hg_22427603 <span class="dt">\</span></span>
<span id="cb51-9"><a href="#cb51-9"></a>--genomes <span class="st">"g_279,g_425,g_149,g_377,g_100,g_106,g_331,g_249,g_125,</span><span class="dt">\</span></span>
<span id="cb51-10"><a href="#cb51-10"></a><span class="st">g_221,g_53,g_395,g_108,g_444,g_160"</span> <span class="dt">\</span></span>
<span id="cb51-11"><a href="#cb51-11"></a>--inner_region <span class="at">--dir</span> analysis/pangenome_v2/carotovoricin/ctv_tail <span class="dt">\</span></span>
<span id="cb51-12"><a href="#cb51-12"></a>--out <span class="st">"selected_haplotypes.tab"</span> </span>
<span id="cb51-13"><a href="#cb51-13"></a></span>
<span id="cb51-14"><a href="#cb51-14"></a><span class="co"># conserved CTV locus</span></span>
<span id="cb51-15"><a href="#cb51-15"></a><span class="ex">Rscript</span> scripts/utils/HG_range_coordinates.R <span class="at">--hgs</span> hg_22427640,hg_22427604 <span class="dt">\</span></span>
<span id="cb51-16"><a href="#cb51-16"></a>--genomes <span class="st">"g_279,g_425,g_149,g_377,g_100,g_106,g_331,g_249,g_125,</span><span class="dt">\</span></span>
<span id="cb51-17"><a href="#cb51-17"></a><span class="st">g_221,g_53,g_395,g_108,g_444,g_160"</span> <span class="dt">\</span></span>
<span id="cb51-18"><a href="#cb51-18"></a>--dir analysis/pangenome_v2/carotovoricin/ctv_conserved <span class="dt">\</span></span>
<span id="cb51-19"><a href="#cb51-19"></a>--out <span class="st">"selected_conserved.tab"</span></span>
<span id="cb51-20"><a href="#cb51-20"></a></span>
<span id="cb51-21"><a href="#cb51-21"></a><span class="co"># upstream core</span></span>
<span id="cb51-22"><a href="#cb51-22"></a><span class="ex">Rscript</span> scripts/utils/HG_range_coordinates.R <span class="at">--hgs</span> hg_22427642,hg_22427641 <span class="dt">\</span></span>
<span id="cb51-23"><a href="#cb51-23"></a>--genomes <span class="st">"g_279,g_425,g_149,g_377,g_100,g_106,g_331,g_249,g_125,</span><span class="dt">\</span></span>
<span id="cb51-24"><a href="#cb51-24"></a><span class="st">g_221,g_53,g_395,g_108,g_444,g_160"</span> <span class="dt">\</span></span>
<span id="cb51-25"><a href="#cb51-25"></a>--dir analysis/pangenome_v2/carotovoricin/upstream_core <span class="dt">\</span></span>
<span id="cb51-26"><a href="#cb51-26"></a>--out <span class="st">"selected_upstream.tab"</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<p><a href="#dna-sequence-comparison">DNA sequence comparison of these regions</a></p>
<section id="perform-msa-using-mafft" class="level4">
<h4 class="anchored" data-anchor-id="perform-msa-using-mafft">Perform MSA using <code>MAFFT</code></h4>
<div class="sourceCode" id="cb52"><pre class="sourceCode numberSource bash number-lines code-with-copy"><code class="sourceCode bash"><span id="cb52-1"><a href="#cb52-1"></a><span class="fu">nohup</span> mafft <span class="at">--reorder</span> <span class="at">--allowshift</span> <span class="at">--unalignlevel</span> 0.2 <span class="at">--leavegappyregion</span> <span class="dt">\</span></span>
<span id="cb52-2"><a href="#cb52-2"></a>--maxiterate 0 <span class="at">--globalpair</span> ctv_tail/selected_haplotypes.fasta <span class="dt">\</span></span>
<span id="cb52-3"><a href="#cb52-3"></a><span class="op">&gt;</span> ctv_tail/selected_haplotypes.msa.fasta <span class="dv">2</span><span class="op">&gt;</span>nohup.out <span class="kw">&amp;</span></span>
<span id="cb52-4"><a href="#cb52-4"></a></span>
<span id="cb52-5"><a href="#cb52-5"></a><span class="fu">nohup</span> mafft <span class="at">--reorder</span> <span class="at">--allowshift</span> <span class="at">--unalignlevel</span> 0.2 <span class="at">--leavegappyregion</span> <span class="dt">\</span></span>
<span id="cb52-6"><a href="#cb52-6"></a>--maxiterate 0 <span class="at">--globalpair</span> ctv_conserved/selected_conserved.fasta <span class="dt">\</span></span>
<span id="cb52-7"><a href="#cb52-7"></a><span class="op">&gt;</span> ctv_conserved/selected_conserved.msa.fasta <span class="dv">2</span><span class="op">&gt;</span>nohup.out <span class="kw">&amp;</span></span>
<span id="cb52-8"><a href="#cb52-8"></a></span>
<span id="cb52-9"><a href="#cb52-9"></a><span class="fu">nohup</span> mafft <span class="at">--reorder</span> <span class="at">--allowshift</span> <span class="at">--unalignlevel</span> 0.2 <span class="at">--leavegappyregion</span> <span class="dt">\</span></span>
<span id="cb52-10"><a href="#cb52-10"></a>--maxiterate 0 <span class="at">--globalpair</span> upstream_core/selected_upstream.fasta <span class="dt">\</span></span>
<span id="cb52-11"><a href="#cb52-11"></a><span class="op">&gt;</span> upstream_core/selected_upstream.msa.fasta <span class="dv">2</span><span class="op">&gt;</span>nohup.out <span class="kw">&amp;</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<p>Generate a maximum-likelihood phylogenetic tree for MSAs</p>
<div class="sourceCode" id="cb53"><pre class="sourceCode numberSource bash number-lines code-with-copy"><code class="sourceCode bash"><span id="cb53-1"><a href="#cb53-1"></a><span class="bu">cd</span> cd analysis/pangenome_v2/carotovoricin/</span>
<span id="cb53-2"><a href="#cb53-2"></a></span>
<span id="cb53-3"><a href="#cb53-3"></a><span class="fu">nohup</span> nice iqtree2 <span class="at">-T</span> 40 <span class="at">-s</span> ctv_tail/selected_haplotypes.msa.fasta <span class="at">-B</span> 1000 <span class="dt">\</span></span>
<span id="cb53-4"><a href="#cb53-4"></a>--prefix ctv_tail/iqtree/selected_haplotypes.msa.fasta <span class="op">&gt;&gt;</span> iqtree_tree.log <span class="dv">2</span><span class="op">&gt;&amp;</span><span class="dv">1</span> <span class="kw">&amp;</span></span>
<span id="cb53-5"><a href="#cb53-5"></a></span>
<span id="cb53-6"><a href="#cb53-6"></a><span class="fu">nohup</span> nice iqtree2 <span class="at">-T</span> 40 <span class="at">-s</span> ctv_conserved/selected_conserved.msa.fasta <span class="at">-B</span> 1000 <span class="dt">\</span></span>
<span id="cb53-7"><a href="#cb53-7"></a>--prefix ctv_conserved/iqtree/selected_conserved.msa.fasta <span class="op">&gt;&gt;</span> iqtree_tree.log <span class="dv">2</span><span class="op">&gt;&amp;</span><span class="dv">1</span> <span class="kw">&amp;</span></span>
<span id="cb53-8"><a href="#cb53-8"></a></span>
<span id="cb53-9"><a href="#cb53-9"></a><span class="fu">nohup</span> nice iqtree2 <span class="at">-T</span> 40 <span class="at">-s</span> upstream_core/selected_upstream.msa.fasta <span class="at">-B</span> 1000 <span class="dt">\</span></span>
<span id="cb53-10"><a href="#cb53-10"></a>--prefix upstream_core/iqtree/selected_upstream.msa.fasta <span class="op">&gt;&gt;</span> iqtree_tree.log <span class="dv">2</span><span class="op">&gt;&amp;</span><span class="dv">1</span> <span class="kw">&amp;</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</section>
</section>
<section id="use-mauve-to-detect-the-structural-variation-and-consevation" class="level3">
<h3 class="anchored" data-anchor-id="use-mauve-to-detect-the-structural-variation-and-consevation">Use Mauve to detect the structural variation and consevation</h3>
<p>Mauve aligns sequences provided in <code>fasta</code> and <code>genbank</code> format. However, the sequence annotation is shown only if the alignment input was in <code>genbank</code> format. Therefore, we need to combine the homology group annotation in <code>gff3</code> format with the <code>fasta</code> sequences to generate a <code>genbank</code> formatted files. To do this, Emboss tool <code>seqret</code> was used.</p>
<div class="sourceCode" id="cb54"><pre class="sourceCode numberSource bash number-lines code-with-copy"><code class="sourceCode bash"><span id="cb54-1"><a href="#cb54-1"></a><span class="ex">conda</span> activate omics_py37</span>
<span id="cb54-2"><a href="#cb54-2"></a></span>
<span id="cb54-3"><a href="#cb54-3"></a><span class="cf">for</span> vir <span class="kw">in</span> <span class="kw">`</span><span class="fu">cut</span> <span class="at">-f</span> 1 hg_regions.tab <span class="kw">|</span> <span class="fu">tail</span> <span class="at">-n</span> +2<span class="kw">`</span></span>
<span id="cb54-4"><a href="#cb54-4"></a><span class="cf">do</span></span>
<span id="cb54-5"><a href="#cb54-5"></a>  <span class="ex">seqret</span> <span class="at">-sequence</span> <span class="va">${vir}</span>.fasta <span class="at">-feature</span> <span class="at">-fformat</span> gff3 <span class="dt">\</span></span>
<span id="cb54-6"><a href="#cb54-6"></a>  <span class="at">-fopenfile</span> <span class="va">${vir}</span>.gff3 <span class="at">-osformat</span> genbank <span class="at">-osextension</span> gbk <span class="dt">\</span></span>
<span id="cb54-7"><a href="#cb54-7"></a>  <span class="at">-osname_outseq</span> <span class="va">${vir}</span> <span class="at">-ofdirectory_outseq</span> gbk_file <span class="at">-auto</span> </span>
<span id="cb54-8"><a href="#cb54-8"></a><span class="cf">done</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<p>Correct GenBank file as per Mauve requirement</p>
<div class="sourceCode" id="cb55"><pre class="sourceCode numberSource bash number-lines code-with-copy"><code class="sourceCode bash"><span id="cb55-1"><a href="#cb55-1"></a><span class="fu">sed</span> <span class="at">-i.bak</span> <span class="at">-r</span> <span class="at">-e</span> <span class="st">'s/(^\s+CDS\s+(complement\()?)[^:]+:([[:digit:]]+\.\.[[:digit:]]+\)?).*/\1\3/'</span> <span class="dt">\</span></span>
<span id="cb55-2"><a href="#cb55-2"></a>  <span class="at">-e</span> <span class="st">'s/\/note="\*([^:]+): /\/\1="/'</span> <span class="pp">*</span>.gbk</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<p>Align and visualize sequences using Mauve.</p>
<div class="sourceCode" id="cb56"><pre class="sourceCode numberSource bash number-lines code-with-copy"><code class="sourceCode bash"><span id="cb56-1"><a href="#cb56-1"></a><span class="va">mauve_out</span><span class="op">=</span><span class="st">"ctv_hgt.defaults"</span></span>
<span id="cb56-2"><a href="#cb56-2"></a></span>
<span id="cb56-3"><a href="#cb56-3"></a><span class="ex">progressiveMauve</span> <span class="at">--output</span><span class="op">=</span><span class="st">"</span><span class="va">${mauve_out}</span><span class="st">.mauve.xmfa"</span> <span class="dt">\</span></span>
<span id="cb56-4"><a href="#cb56-4"></a>--backbone-output=<span class="st">"</span><span class="va">${mauve_out}</span><span class="st">.mauve.backbone"</span> <span class="dt">\</span></span>
<span id="cb56-5"><a href="#cb56-5"></a>--output-guide-tree=<span class="st">"</span><span class="va">${mauve_out}</span><span class="st">.mauve.guide_tree.newick"</span> <span class="dt">\</span></span>
<span id="cb56-6"><a href="#cb56-6"></a><span class="pp">*</span>.gbk  <span class="op">&gt;</span> <span class="st">"</span><span class="va">${mauve_out}</span><span class="st">"</span>.mauve.log <span class="dv">2</span><span class="op">&gt;&amp;</span><span class="dv">1</span></span>
<span id="cb56-7"><a href="#cb56-7"></a></span>
<span id="cb56-8"><a href="#cb56-8"></a><span class="co"># mauve with custom settings</span></span>
<span id="cb56-9"><a href="#cb56-9"></a><span class="va">mauve_out</span><span class="op">=</span><span class="st">"ctv_hgt.opt"</span></span>
<span id="cb56-10"><a href="#cb56-10"></a></span>
<span id="cb56-11"><a href="#cb56-11"></a><span class="ex">progressiveMauve</span> <span class="at">--output</span><span class="op">=</span><span class="st">"</span><span class="va">${mauve_out}</span><span class="st">.mauve.xmfa"</span> <span class="dt">\</span></span>
<span id="cb56-12"><a href="#cb56-12"></a>--seed-weight 10 <span class="dt">\</span></span>
<span id="cb56-13"><a href="#cb56-13"></a>--backbone-output=<span class="st">"</span><span class="va">${mauve_out}</span><span class="st">.mauve.backbone"</span> <span class="dt">\</span></span>
<span id="cb56-14"><a href="#cb56-14"></a>--output-guide-tree=<span class="st">"</span><span class="va">${mauve_out}</span><span class="st">.mauve.guide_tree.newick"</span> <span class="dt">\</span></span>
<span id="cb56-15"><a href="#cb56-15"></a><span class="pp">*</span>.gbk  <span class="op">&gt;</span> <span class="st">"</span><span class="va">${mauve_out}</span><span class="st">"</span>.mauve.log <span class="dv">2</span><span class="op">&gt;&amp;</span><span class="dv">1</span></span>
<span id="cb56-16"><a href="#cb56-16"></a></span>
<span id="cb56-17"><a href="#cb56-17"></a><span class="co"># --seed-weight 5 \</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</section>
<section id="visualize-tail-fiber-locus-hg-sets-on-the-pangenome" class="level3">
<h3 class="anchored" data-anchor-id="visualize-tail-fiber-locus-hg-sets-on-the-pangenome">Visualize tail fiber locus HG sets on the pangenome</h3>
<p>R script <code>scripts/analysis/HG_tandem_match.viz.qmd</code> to plot tail fiber locus HGs for a set of genomes from different species.</p>
<div class="sourceCode" id="cb57"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb57-1"><a href="#cb57-1"></a>hgSets <span class="ot">&lt;-</span> <span class="fu">list</span>(</span>
<span id="cb57-2"><a href="#cb57-2"></a>  <span class="at">s1 =</span> <span class="fu">c</span>(<span class="st">"hg_22427604"</span>, <span class="st">"hg_22426828"</span>),</span>
<span id="cb57-3"><a href="#cb57-3"></a>  <span class="at">s2 =</span> <span class="fu">c</span>(<span class="st">"hg_22427604"</span>, <span class="st">"hg_22426818"</span>),</span>
<span id="cb57-4"><a href="#cb57-4"></a>  <span class="at">s3 =</span> <span class="fu">c</span>(<span class="st">"hg_22427604"</span>, <span class="st">"hg_22426820"</span>, <span class="st">"hg_22426802"</span>, <span class="st">"hg_22426815"</span>)</span>
<span id="cb57-5"><a href="#cb57-5"></a>)</span>
<span id="cb57-6"><a href="#cb57-6"></a></span>
<span id="cb57-7"><a href="#cb57-7"></a>outDir <span class="ot">&lt;-</span> <span class="fu">paste</span>(confs<span class="sc">$</span>analysis<span class="sc">$</span>prophages<span class="sc">$</span>dir, <span class="st">"/cluster_viz/ctv_hgt"</span>, <span class="at">sep =</span> <span class="st">""</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<ul>
<li></li>
</ul>


<!-- -->

</section>
</section>

</main> <!-- /main -->
<script id="quarto-html-after-body" type="application/javascript">
window.document.addEventListener("DOMContentLoaded", function (event) {
  const toggleBodyColorMode = (bsSheetEl) => {
    const mode = bsSheetEl.getAttribute("data-mode");
    const bodyEl = window.document.querySelector("body");
    if (mode === "dark") {
      bodyEl.classList.add("quarto-dark");
      bodyEl.classList.remove("quarto-light");
    } else {
      bodyEl.classList.add("quarto-light");
      bodyEl.classList.remove("quarto-dark");
    }
  }
  const toggleBodyColorPrimary = () => {
    const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
    if (bsSheetEl) {
      toggleBodyColorMode(bsSheetEl);
    }
  }
  toggleBodyColorPrimary();  
  const icon = "";
  const anchorJS = new window.AnchorJS();
  anchorJS.options = {
    placement: 'right',
    icon: icon
  };
  anchorJS.add('.anchored');
  const isCodeAnnotation = (el) => {
    for (const clz of el.classList) {
      if (clz.startsWith('code-annotation-')) {                     
        return true;
      }
    }
    return false;
  }
  const clipboard = new window.ClipboardJS('.code-copy-button', {
    text: function(trigger) {
      const codeEl = trigger.previousElementSibling.cloneNode(true);
      for (const childEl of codeEl.children) {
        if (isCodeAnnotation(childEl)) {
          childEl.remove();
        }
      }
      return codeEl.innerText;
    }
  });
  clipboard.on('success', function(e) {
    // button target
    const button = e.trigger;
    // don't keep focus
    button.blur();
    // flash "checked"
    button.classList.add('code-copy-button-checked');
    var currentTitle = button.getAttribute("title");
    button.setAttribute("title", "Copied!");
    let tooltip;
    if (window.bootstrap) {
      button.setAttribute("data-bs-toggle", "tooltip");
      button.setAttribute("data-bs-placement", "left");
      button.setAttribute("data-bs-title", "Copied!");
      tooltip = new bootstrap.Tooltip(button, 
        { trigger: "manual", 
          customClass: "code-copy-button-tooltip",
          offset: [0, -8]});
      tooltip.show();    
    }
    setTimeout(function() {
      if (tooltip) {
        tooltip.hide();
        button.removeAttribute("data-bs-title");
        button.removeAttribute("data-bs-toggle");
        button.removeAttribute("data-bs-placement");
      }
      button.setAttribute("title", currentTitle);
      button.classList.remove('code-copy-button-checked');
    }, 1000);
    // clear code selection
    e.clearSelection();
  });
  const viewSource = window.document.getElementById('quarto-view-source') ||
                     window.document.getElementById('quarto-code-tools-source');
  if (viewSource) {
    const sourceUrl = viewSource.getAttribute("data-quarto-source-url");
    viewSource.addEventListener("click", function(e) {
      if (sourceUrl) {
        // rstudio viewer pane
        if (/\bcapabilities=\b/.test(window.location)) {
          window.open(sourceUrl);
        } else {
          window.location.href = sourceUrl;
        }
      } else {
        const modal = new bootstrap.Modal(document.getElementById('quarto-embedded-source-code-modal'));
        modal.show();
      }
      return false;
    });
  }
  function toggleCodeHandler(show) {
    return function(e) {
      const detailsSrc = window.document.querySelectorAll(".cell > details > .sourceCode");
      for (let i=0; i<detailsSrc.length; i++) {
        const details = detailsSrc[i].parentElement;
        if (show) {
          details.open = true;
        } else {
          details.removeAttribute("open");
        }
      }
      const cellCodeDivs = window.document.querySelectorAll(".cell > .sourceCode");
      const fromCls = show ? "hidden" : "unhidden";
      const toCls = show ? "unhidden" : "hidden";
      for (let i=0; i<cellCodeDivs.length; i++) {
        const codeDiv = cellCodeDivs[i];
        if (codeDiv.classList.contains(fromCls)) {
          codeDiv.classList.remove(fromCls);
          codeDiv.classList.add(toCls);
        } 
      }
      return false;
    }
  }
  const hideAllCode = window.document.getElementById("quarto-hide-all-code");
  if (hideAllCode) {
    hideAllCode.addEventListener("click", toggleCodeHandler(false));
  }
  const showAllCode = window.document.getElementById("quarto-show-all-code");
  if (showAllCode) {
    showAllCode.addEventListener("click", toggleCodeHandler(true));
  }
  function tippyHover(el, contentFn) {
    const config = {
      allowHTML: true,
      content: contentFn,
      maxWidth: 500,
      delay: 100,
      arrow: false,
      appendTo: function(el) {
          return el.parentElement;
      },
      interactive: true,
      interactiveBorder: 10,
      theme: 'quarto',
      placement: 'bottom-start'
    };
    window.tippy(el, config); 
  }
  const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
  for (var i=0; i<noterefs.length; i++) {
    const ref = noterefs[i];
    tippyHover(ref, function() {
      // use id or data attribute instead here
      let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
      try { href = new URL(href).hash; } catch {}
      const id = href.replace(/^#\/?/, "");
      const note = window.document.getElementById(id);
      return note.innerHTML;
    });
  }
      let selectedAnnoteEl;
      const selectorForAnnotation = ( cell, annotation) => {
        let cellAttr = 'data-code-cell="' + cell + '"';
        let lineAttr = 'data-code-annotation="' +  annotation + '"';
        const selector = 'span[' + cellAttr + '][' + lineAttr + ']';
        return selector;
      }
      const selectCodeLines = (annoteEl) => {
        const doc = window.document;
        const targetCell = annoteEl.getAttribute("data-target-cell");
        const targetAnnotation = annoteEl.getAttribute("data-target-annotation");
        const annoteSpan = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
        const lines = annoteSpan.getAttribute("data-code-lines").split(",");
        const lineIds = lines.map((line) => {
          return targetCell + "-" + line;
        })
        let top = null;
        let height = null;
        let parent = null;
        if (lineIds.length > 0) {
            //compute the position of the single el (top and bottom and make a div)
            const el = window.document.getElementById(lineIds[0]);
            top = el.offsetTop;
            height = el.offsetHeight;
            parent = el.parentElement.parentElement;
          if (lineIds.length > 1) {
            const lastEl = window.document.getElementById(lineIds[lineIds.length - 1]);
            const bottom = lastEl.offsetTop + lastEl.offsetHeight;
            height = bottom - top;
          }
          if (top !== null && height !== null && parent !== null) {
            // cook up a div (if necessary) and position it 
            let div = window.document.getElementById("code-annotation-line-highlight");
            if (div === null) {
              div = window.document.createElement("div");
              div.setAttribute("id", "code-annotation-line-highlight");
              div.style.position = 'absolute';
              parent.appendChild(div);
            }
            div.style.top = top - 2 + "px";
            div.style.height = height + 4 + "px";
            let gutterDiv = window.document.getElementById("code-annotation-line-highlight-gutter");
            if (gutterDiv === null) {
              gutterDiv = window.document.createElement("div");
              gutterDiv.setAttribute("id", "code-annotation-line-highlight-gutter");
              gutterDiv.style.position = 'absolute';
              const codeCell = window.document.getElementById(targetCell);
              const gutter = codeCell.querySelector('.code-annotation-gutter');
              gutter.appendChild(gutterDiv);
            }
            gutterDiv.style.top = top - 2 + "px";
            gutterDiv.style.height = height + 4 + "px";
          }
          selectedAnnoteEl = annoteEl;
        }
      };
      const unselectCodeLines = () => {
        const elementsIds = ["code-annotation-line-highlight", "code-annotation-line-highlight-gutter"];
        elementsIds.forEach((elId) => {
          const div = window.document.getElementById(elId);
          if (div) {
            div.remove();
          }
        });
        selectedAnnoteEl = undefined;
      };
      // Attach click handler to the DT
      const annoteDls = window.document.querySelectorAll('dt[data-target-cell]');
      for (const annoteDlNode of annoteDls) {
        annoteDlNode.addEventListener('click', (event) => {
          const clickedEl = event.target;
          if (clickedEl !== selectedAnnoteEl) {
            unselectCodeLines();
            const activeEl = window.document.querySelector('dt[data-target-cell].code-annotation-active');
            if (activeEl) {
              activeEl.classList.remove('code-annotation-active');
            }
            selectCodeLines(clickedEl);
            clickedEl.classList.add('code-annotation-active');
          } else {
            // Unselect the line
            unselectCodeLines();
            clickedEl.classList.remove('code-annotation-active');
          }
        });
      }
  const findCites = (el) => {
    const parentEl = el.parentElement;
    if (parentEl) {
      const cites = parentEl.dataset.cites;
      if (cites) {
        return {
          el,
          cites: cites.split(' ')
        };
      } else {
        return findCites(el.parentElement)
      }
    } else {
      return undefined;
    }
  };
  var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
  for (var i=0; i<bibliorefs.length; i++) {
    const ref = bibliorefs[i];
    const citeInfo = findCites(ref);
    if (citeInfo) {
      tippyHover(citeInfo.el, function() {
        var popup = window.document.createElement('div');
        citeInfo.cites.forEach(function(cite) {
          var citeDiv = window.document.createElement('div');
          citeDiv.classList.add('hanging-indent');
          citeDiv.classList.add('csl-entry');
          var biblioDiv = window.document.getElementById('ref-' + cite);
          if (biblioDiv) {
            citeDiv.innerHTML = biblioDiv.innerHTML;
          }
          popup.appendChild(citeDiv);
        });
        return popup.innerHTML;
      });
    }
  }
});
</script>
<nav class="page-navigation">
  <div class="nav-page nav-page-previous">
      <a href="../../scripts/analysis/ani.clustering.html" class="pagination-link">
        <i class="bi bi-arrow-left-short"></i> <span class="nav-page-text">ANI</span>
      </a>          
  </div>
  <div class="nav-page nav-page-next">
      <a href="../../scripts/analysis/prophage_summary.html" class="pagination-link">
        <span class="nav-page-text">Prophage summary</span> <i class="bi bi-arrow-right-short"></i>
      </a>
  </div>
</nav><div class="modal fade" id="quarto-embedded-source-code-modal" tabindex="-1" aria-labelledby="quarto-embedded-source-code-modal-label" aria-hidden="true"><div class="modal-dialog modal-dialog-scrollable"><div class="modal-content"><div class="modal-header"><h5 class="modal-title" id="quarto-embedded-source-code-modal-label">Source Code</h5><button class="btn-close" data-bs-dismiss="modal"></button></div><div class="modal-body"><div class="">
<div class="sourceCode" id="cb58" data-shortcodes="false"><pre class="sourceCode numberSource markdown number-lines code-with-copy"><code class="sourceCode markdown"><span id="cb58-1"><a href="#cb58-1"></a>*# Prophage analysis in *Pectobacterium* pangenome</span>
<span id="cb58-2"><a href="#cb58-2"></a></span>
<span id="cb58-3"><a href="#cb58-3"></a><span class="fu">## Workflow</span></span>
<span id="cb58-4"><a href="#cb58-4"></a></span>
<span id="cb58-5"><a href="#cb58-5"></a><span class="in">```mermaid</span></span>
<span id="cb58-6"><a href="#cb58-6"></a></span>
<span id="cb58-7"><a href="#cb58-7"></a><span class="in">```</span></span>
<span id="cb58-8"><a href="#cb58-8"></a></span>
<span id="cb58-9"><a href="#cb58-9"></a><span class="fu">## Detect prophages in pangenome</span></span>
<span id="cb58-10"><a href="#cb58-10"></a></span>
<span id="cb58-11"><a href="#cb58-11"></a><span class="fu">### genomad prophage detection</span></span>
<span id="cb58-12"><a href="#cb58-12"></a></span>
<span id="cb58-13"><a href="#cb58-13"></a><span class="in">```bash</span></span>
<span id="cb58-14"><a href="#cb58-14"></a><span class="co"># run genomad and checkv pipeline</span></span>
<span id="cb58-15"><a href="#cb58-15"></a><span class="fu">nohup</span> <span class="dt">\</span></span>
<span id="cb58-16"><a href="#cb58-16"></a>cat data/reference_data/assembly_ids.txt <span class="kw">|</span> <span class="dt">\</span></span>
<span id="cb58-17"><a href="#cb58-17"></a><span class="ex">parallel</span> <span class="at">--jobs</span> 6 <span class="at">--workdir</span> <span class="va">$PWD</span> <span class="at">--halt</span> now,fail=1 <span class="dt">\</span></span>
<span id="cb58-18"><a href="#cb58-18"></a>--keep-order <span class="at">--results</span> logs/genomad/{} <span class="dt">\</span></span>
<span id="cb58-19"><a href="#cb58-19"></a>--joblog logs/genomad/parallel.log <span class="dt">\</span></span>
<span id="cb58-20"><a href="#cb58-20"></a>scripts/preprocessing/genomad_prophage_annotation.sh {} <span class="dt">\</span></span>
<span id="cb58-21"><a href="#cb58-21"></a><span class="op">&gt;&gt;</span>logs/genomad/nohup.out <span class="dv">2</span><span class="op">&gt;&amp;</span><span class="dv">1</span> <span class="kw">&amp;</span></span>
<span id="cb58-22"><a href="#cb58-22"></a></span>
<span id="cb58-23"><a href="#cb58-23"></a><span class="co"># count of prophage detected in each genome</span></span>
<span id="cb58-24"><a href="#cb58-24"></a><span class="cf">for</span> i <span class="kw">in</span> <span class="kw">`</span><span class="fu">cat</span> data/reference_data/assembly_ids.txt<span class="kw">`</span></span>
<span id="cb58-25"><a href="#cb58-25"></a><span class="cf">do</span></span>
<span id="cb58-26"><a href="#cb58-26"></a>  <span class="bu">printf</span> <span class="st">"</span><span class="va">$i</span><span class="st">\t"</span></span>
<span id="cb58-27"><a href="#cb58-27"></a>  <span class="fu">tail</span> <span class="at">-n</span> +2 data/prophage_genomad/<span class="va">$i</span>/<span class="va">${i}</span>_summary/<span class="va">${i}</span>_virus_summary.tsv <span class="kw">|</span> <span class="fu">wc</span> <span class="at">-l</span></span>
<span id="cb58-28"><a href="#cb58-28"></a><span class="cf">done</span> <span class="kw">|</span> <span class="fu">sort</span> <span class="at">-nr</span> <span class="at">-k2</span></span>
<span id="cb58-29"><a href="#cb58-29"></a></span>
<span id="cb58-30"><a href="#cb58-30"></a><span class="in">```</span></span>
<span id="cb58-31"><a href="#cb58-31"></a></span>
<span id="cb58-32"><a href="#cb58-32"></a>Check the checkV contamination.tsv file to see if there is any other contamination</span>
<span id="cb58-33"><a href="#cb58-33"></a>than the expected <span class="in">`host,viral`</span>, <span class="in">`viral,host`</span> and <span class="in">`host,viral,host`</span> ones.</span>
<span id="cb58-34"><a href="#cb58-34"></a></span>
<span id="cb58-35"><a href="#cb58-35"></a><span class="in">```bash</span></span>
<span id="cb58-36"><a href="#cb58-36"></a><span class="cf">for</span> cn <span class="kw">in</span> <span class="kw">`</span><span class="fu">find</span> <span class="at">-name</span> <span class="st">'*contamination.tsv'</span><span class="kw">`</span></span>
<span id="cb58-37"><a href="#cb58-37"></a><span class="cf">do</span></span>
<span id="cb58-38"><a href="#cb58-38"></a><span class="fu">cat</span> <span class="va">$cn</span> <span class="kw">|</span> <span class="fu">awk</span> <span class="st">'$9 != "NA" &amp;&amp; NR != 1 &amp;&amp; $9 != "host,viral" &amp;&amp; $9 != "viral,host" &amp;&amp; $9 != "host,viral,host"'</span></span>
<span id="cb58-39"><a href="#cb58-39"></a><span class="cf">done</span></span>
<span id="cb58-40"><a href="#cb58-40"></a><span class="in">```</span></span>
<span id="cb58-41"><a href="#cb58-41"></a></span>
<span id="cb58-42"><a href="#cb58-42"></a><span class="fu">### Post processing of prophages </span></span>
<span id="cb58-43"><a href="#cb58-43"></a></span>
<span id="cb58-44"><a href="#cb58-44"></a>Process genomad output and separate prophages into independent FASTA files. Each</span>
<span id="cb58-45"><a href="#cb58-45"></a>prophage is assigned a unique identifier with format <span class="in">`g_{genome}.vir_{#count}`</span>.</span>
<span id="cb58-46"><a href="#cb58-46"></a>Prophages are saved in independent FASTA files for downstream analysis, for e.g.</span>
<span id="cb58-47"><a href="#cb58-47"></a>ANI clustering of prophages.</span>
<span id="cb58-48"><a href="#cb58-48"></a></span>
<span id="cb58-49"><a href="#cb58-49"></a><span class="in">```bash</span></span>
<span id="cb58-50"><a href="#cb58-50"></a><span class="co"># post processing of prophage data</span></span>
<span id="cb58-51"><a href="#cb58-51"></a><span class="ex">Rscript</span> scripts/preprocessing/genomad_checkv_merge.R</span>
<span id="cb58-52"><a href="#cb58-52"></a></span>
<span id="cb58-53"><a href="#cb58-53"></a><span class="co"># save prophages independently</span></span>
<span id="cb58-54"><a href="#cb58-54"></a><span class="fu">bash</span> scripts/preprocessing/prophage_genomes_extract.sh</span>
<span id="cb58-55"><a href="#cb58-55"></a><span class="in">```</span></span>
<span id="cb58-56"><a href="#cb58-56"></a></span>
<span id="cb58-57"><a href="#cb58-57"></a>Extract homology groups for prophages from the pangenome org.db object.</span>
<span id="cb58-58"><a href="#cb58-58"></a></span>
<span id="cb58-59"><a href="#cb58-59"></a><span class="in">```bash</span></span>
<span id="cb58-60"><a href="#cb58-60"></a><span class="ex">Rscript</span> scripts/analysis/prophage_homology_groups.R</span>
<span id="cb58-61"><a href="#cb58-61"></a><span class="in">```</span></span>
<span id="cb58-62"><a href="#cb58-62"></a></span>
<span id="cb58-63"><a href="#cb58-63"></a></span>
<span id="cb58-64"><a href="#cb58-64"></a>Run ANI on prophage genomes to identify the representative prophages in the </span>
<span id="cb58-65"><a href="#cb58-65"></a>pangenome.</span>
<span id="cb58-66"><a href="#cb58-66"></a></span>
<span id="cb58-67"><a href="#cb58-67"></a><span class="in">```bash</span></span>
<span id="cb58-68"><a href="#cb58-68"></a><span class="ex">conda</span> activate pantools_v4_3</span>
<span id="cb58-69"><a href="#cb58-69"></a></span>
<span id="cb58-70"><a href="#cb58-70"></a><span class="co"># list prophage FASTA files</span></span>
<span id="cb58-71"><a href="#cb58-71"></a><span class="fu">ls</span> data/prophage_genomad/phage_seqs/<span class="pp">*</span>.fna  <span class="op">&gt;</span> data/prophage_genomad/prophage_fasta.list</span>
<span id="cb58-72"><a href="#cb58-72"></a></span>
<span id="cb58-73"><a href="#cb58-73"></a><span class="co"># run fastANI</span></span>
<span id="cb58-74"><a href="#cb58-74"></a><span class="ex">fastANI</span> <span class="at">--ql</span> data/prophage_genomad/prophage_fasta.list <span class="dt">\</span></span>
<span id="cb58-75"><a href="#cb58-75"></a>--rl data/prophage_genomad/prophage_fasta.list <span class="dt">\</span></span>
<span id="cb58-76"><a href="#cb58-76"></a>--threads 30 <span class="at">--matrix</span> <span class="at">--visualize</span> <span class="at">--fragLen</span> 500 <span class="dt">\</span></span>
<span id="cb58-77"><a href="#cb58-77"></a>-o data/prophage_genomad/prophage_ANI</span>
<span id="cb58-78"><a href="#cb58-78"></a></span>
<span id="cb58-79"><a href="#cb58-79"></a><span class="co"># use mash for distance estimation</span></span>
<span id="cb58-80"><a href="#cb58-80"></a><span class="fu">mkdir</span> data/prophage_genomad/mash</span>
<span id="cb58-81"><a href="#cb58-81"></a></span>
<span id="cb58-82"><a href="#cb58-82"></a><span class="ex">mash</span> sketch <span class="at">-p</span> 30 <span class="at">-s</span> 15000 <span class="at">-k</span> 17 <span class="at">-o</span> data/prophage_genomad/mash/prophages_ref <span class="dt">\</span></span>
<span id="cb58-83"><a href="#cb58-83"></a>-l data/prophage_genomad/prophage_fasta.list <span class="op">&gt;</span>logs/prophage/mash.log <span class="dv">2</span><span class="op">&gt;&amp;</span><span class="dv">1</span></span>
<span id="cb58-84"><a href="#cb58-84"></a></span>
<span id="cb58-85"><a href="#cb58-85"></a><span class="ex">mash</span> info data/prophage_genomad/mash/prophages_ref.msh <span class="op">&gt;</span> <span class="dt">\</span></span>
<span id="cb58-86"><a href="#cb58-86"></a>data/prophage_genomad/mash/prophage_sketch_info.txt</span>
<span id="cb58-87"><a href="#cb58-87"></a></span>
<span id="cb58-88"><a href="#cb58-88"></a><span class="ex">mash</span> dist <span class="at">-p</span> 30 data/prophage_genomad/mash/prophages_ref.msh <span class="dt">\</span></span>
<span id="cb58-89"><a href="#cb58-89"></a>data/prophage_genomad/mash/prophages_ref.msh <span class="op">&gt;</span> <span class="dt">\</span></span>
<span id="cb58-90"><a href="#cb58-90"></a>data/prophage_genomad/mash/prophage_distance.tab</span>
<span id="cb58-91"><a href="#cb58-91"></a></span>
<span id="cb58-92"><a href="#cb58-92"></a><span class="in">```</span></span>
<span id="cb58-93"><a href="#cb58-93"></a></span>
<span id="cb58-94"><a href="#cb58-94"></a><span class="fu">### QC</span></span>
<span id="cb58-95"><a href="#cb58-95"></a></span>
<span id="cb58-96"><a href="#cb58-96"></a>Identify the fragmented prophages resulting because of the genome assembly</span>
<span id="cb58-97"><a href="#cb58-97"></a>that can be merged into a complete prophage. Prepare a consolidated list of </span>
<span id="cb58-98"><a href="#cb58-98"></a>prophages which will be used for the subsequent analysis.</span>
<span id="cb58-99"><a href="#cb58-99"></a></span>
<span id="cb58-100"><a href="#cb58-100"></a><span class="in">```bash</span></span>
<span id="cb58-101"><a href="#cb58-101"></a><span class="ex">Rscript</span> scripts/analysis/prophage_merge.R</span>
<span id="cb58-102"><a href="#cb58-102"></a><span class="in">```</span></span>
<span id="cb58-103"><a href="#cb58-103"></a></span>
<span id="cb58-104"><a href="#cb58-104"></a>Summarize the consolidated prophage list to generate various statistics and </span>
<span id="cb58-105"><a href="#cb58-105"></a>generate summary plots. Prophages that are fragmented or smaller than the length</span>
<span id="cb58-106"><a href="#cb58-106"></a>cutoff are removed and the filtered list of prophages is saved for downstream</span>
<span id="cb58-107"><a href="#cb58-107"></a>analysis.</span>
<span id="cb58-108"><a href="#cb58-108"></a></span>
<span id="cb58-109"><a href="#cb58-109"></a><span class="in">```bash</span></span>
<span id="cb58-110"><a href="#cb58-110"></a><span class="ex">Rscript</span> scripts/analysis/prophage_summary.R</span>
<span id="cb58-111"><a href="#cb58-111"></a><span class="in">```</span></span>
<span id="cb58-112"><a href="#cb58-112"></a></span>
<span id="cb58-113"><a href="#cb58-113"></a>Process prophage MASH and ANI results and visualize MASH similarity matrix for</span>
<span id="cb58-114"><a href="#cb58-114"></a>the filtered prophages.</span>
<span id="cb58-115"><a href="#cb58-115"></a></span>
<span id="cb58-116"><a href="#cb58-116"></a><span class="in">```bash</span></span>
<span id="cb58-117"><a href="#cb58-117"></a><span class="ex">Rscript</span> scripts/analysis/prophage_seq_similarity.qmd</span>
<span id="cb58-118"><a href="#cb58-118"></a><span class="in">```</span></span>
<span id="cb58-119"><a href="#cb58-119"></a></span>
<span id="cb58-120"><a href="#cb58-120"></a></span>
<span id="cb58-121"><a href="#cb58-121"></a><span class="fu">## Cluster prophages</span></span>
<span id="cb58-122"><a href="#cb58-122"></a></span>
<span id="cb58-123"><a href="#cb58-123"></a>Use shared homology groups between two prophages to establish the child-parent</span>
<span id="cb58-124"><a href="#cb58-124"></a>relationship between the prophages in pangenome. First, all vs all syntenic</span>
<span id="cb58-125"><a href="#cb58-125"></a>Jaccard similarity is calculated between the prophages. This information is later</span>
<span id="cb58-126"><a href="#cb58-126"></a>used to cluster the prophages using two different approaches.</span>
<span id="cb58-127"><a href="#cb58-127"></a></span>
<span id="cb58-128"><a href="#cb58-128"></a><span class="in">```bash</span></span>
<span id="cb58-129"><a href="#cb58-129"></a><span class="ex">Rscript</span> scripts/analysis/prophage_pairwise_similarity.R</span>
<span id="cb58-130"><a href="#cb58-130"></a><span class="in">```</span></span>
<span id="cb58-131"><a href="#cb58-131"></a></span>
<span id="cb58-132"><a href="#cb58-132"></a><span class="fu">### Approach 1</span></span>
<span id="cb58-133"><a href="#cb58-133"></a></span>
<span id="cb58-134"><a href="#cb58-134"></a>A heuristic DAG construcion is extracted from the output of the script </span>
<span id="cb58-135"><a href="#cb58-135"></a><span class="in">`scripts/analysis/prophage_merge.R`</span>. This output includes a best matching parent</span>
<span id="cb58-136"><a href="#cb58-136"></a>prophage for each prophage. This is further processed to build a DAG and</span>
<span id="cb58-137"><a href="#cb58-137"></a>visualize the clusters.</span>
<span id="cb58-138"><a href="#cb58-138"></a></span>
<span id="cb58-139"><a href="#cb58-139"></a>Post processing of the child-parent relationships to add additional metadata</span>
<span id="cb58-140"><a href="#cb58-140"></a>and save a DAG in tabular format:</span>
<span id="cb58-141"><a href="#cb58-141"></a></span>
<span id="cb58-142"><a href="#cb58-142"></a><span class="in">```bash</span></span>
<span id="cb58-143"><a href="#cb58-143"></a><span class="ex">Rscript</span> scripts/analysis/prophage_DAG_postprocess.R</span>
<span id="cb58-144"><a href="#cb58-144"></a><span class="in">```</span></span>
<span id="cb58-145"><a href="#cb58-145"></a></span>
<span id="cb58-146"><a href="#cb58-146"></a>Visualize the clusters of prophages as directed acyclic graphs:</span>
<span id="cb58-147"><a href="#cb58-147"></a></span>
<span id="cb58-148"><a href="#cb58-148"></a><span class="in">```r</span></span>
<span id="cb58-149"><a href="#cb58-149"></a>knitr<span class="sc">::</span><span class="fu">knit</span>(<span class="st">"scripts/analysis/prophage_DAG_viz.qmd"</span>)</span>
<span id="cb58-150"><a href="#cb58-150"></a><span class="in">```</span></span>
<span id="cb58-151"><a href="#cb58-151"></a></span>
<span id="cb58-152"><a href="#cb58-152"></a><span class="fu">### Approach 2</span></span>
<span id="cb58-153"><a href="#cb58-153"></a></span>
<span id="cb58-154"><a href="#cb58-154"></a>We first tried to use the Affinity propagation (AP) clustering to group the</span>
<span id="cb58-155"><a href="#cb58-155"></a>prophages into clusters and have a set of cluster representatives. However,</span>
<span id="cb58-156"><a href="#cb58-156"></a>we encountered some clusters generated by the AP clustering where the cluster</span>
<span id="cb58-157"><a href="#cb58-157"></a>members did not even share a homology group. Therefore, we decided not to use</span>
<span id="cb58-158"><a href="#cb58-158"></a>AP clustering.</span>
<span id="cb58-159"><a href="#cb58-159"></a></span>
<span id="cb58-160"><a href="#cb58-160"></a><span class="in">```r</span></span>
<span id="cb58-161"><a href="#cb58-161"></a>knitr<span class="sc">::</span><span class="fu">knit</span>(<span class="st">"scripts/analysis/prophage_AP_clustering.qmd"</span>)</span>
<span id="cb58-162"><a href="#cb58-162"></a><span class="in">```</span></span>
<span id="cb58-163"><a href="#cb58-163"></a></span>
<span id="cb58-164"><a href="#cb58-164"></a>Cluster the all-vs-all syntenic Jaccard distance matrix using "complete" linkage</span>
<span id="cb58-165"><a href="#cb58-165"></a>hierarchical clustering and cut the cluster using carotovoricin cluster as a</span>
<span id="cb58-166"><a href="#cb58-166"></a>reference.</span>
<span id="cb58-167"><a href="#cb58-167"></a></span>
<span id="cb58-168"><a href="#cb58-168"></a><span class="in">```r</span></span>
<span id="cb58-169"><a href="#cb58-169"></a>knitr<span class="sc">::</span><span class="fu">knit</span>(<span class="st">"scripts/analysis/prophage_clustering.qmd"</span>)</span>
<span id="cb58-170"><a href="#cb58-170"></a><span class="in">```</span></span>
<span id="cb58-171"><a href="#cb58-171"></a></span>
<span id="cb58-172"><a href="#cb58-172"></a></span>
<span id="cb58-173"><a href="#cb58-173"></a><span class="fu">### Summarize prophage clusters</span></span>
<span id="cb58-174"><a href="#cb58-174"></a></span>
<span id="cb58-175"><a href="#cb58-175"></a>Summarize the representative prophages in the pangenome.</span>
<span id="cb58-176"><a href="#cb58-176"></a></span>
<span id="cb58-177"><a href="#cb58-177"></a><span class="in">```bash</span></span>
<span id="cb58-178"><a href="#cb58-178"></a><span class="ex">Rscript</span> scripts/analysis/prophage_representative_summary.qmd</span>
<span id="cb58-179"><a href="#cb58-179"></a><span class="in">```</span></span>
<span id="cb58-180"><a href="#cb58-180"></a></span>
<span id="cb58-181"><a href="#cb58-181"></a>Visualize the representative prophages across pangenome. Use MASH distance to</span>
<span id="cb58-182"><a href="#cb58-182"></a>cluster prophages and ANI distance to cluster the genomes in the pangenome.</span>
<span id="cb58-183"><a href="#cb58-183"></a></span>
<span id="cb58-184"><a href="#cb58-184"></a><span class="in">```bash</span></span>
<span id="cb58-185"><a href="#cb58-185"></a><span class="ex">Rscript</span> scripts/analysis/prophage_pangenome_viz.qmd</span>
<span id="cb58-186"><a href="#cb58-186"></a><span class="in">```</span></span>
<span id="cb58-187"><a href="#cb58-187"></a></span>
<span id="cb58-188"><a href="#cb58-188"></a>Visualize homology group sets for selected prophages as PAV heatmap across the</span>
<span id="cb58-189"><a href="#cb58-189"></a>pangenome. Replace the code block in <span class="in">`scripts/analysis/homology_group_viz.R`</span> </span>
<span id="cb58-190"><a href="#cb58-190"></a>script with the following.</span>
<span id="cb58-191"><a href="#cb58-191"></a></span>
<span id="cb58-192"><a href="#cb58-192"></a><span class="in">```r</span></span>
<span id="cb58-193"><a href="#cb58-193"></a>analysisName <span class="ot">&lt;-</span> <span class="st">"vir_lineage_prophages"</span></span>
<span id="cb58-194"><a href="#cb58-194"></a></span>
<span id="cb58-195"><a href="#cb58-195"></a>outDir <span class="ot">&lt;-</span> <span class="fu">file.path</span>(confs<span class="sc">$</span>analysis<span class="sc">$</span>prophages<span class="sc">$</span>dir)</span>
<span id="cb58-196"><a href="#cb58-196"></a>outPrefix <span class="ot">&lt;-</span> <span class="fu">file.path</span>(outDir, analysisName)</span>
<span id="cb58-197"><a href="#cb58-197"></a></span>
<span id="cb58-198"><a href="#cb58-198"></a>setIds <span class="ot">&lt;-</span> <span class="fu">c</span>(</span>
<span id="cb58-199"><a href="#cb58-199"></a>  <span class="st">"g_189.vir_2"</span>, <span class="st">"g_400.vir_2"</span>, <span class="st">"g_399.vir_2"</span>, <span class="st">"g_448.vir_2"</span>, <span class="st">"g_189.vir_1"</span>, </span>
<span id="cb58-200"><a href="#cb58-200"></a>  <span class="st">"g_3.vir_4"</span>, <span class="st">"g_438.vir_3"</span>, <span class="st">"g_350.vir_1"</span>,</span>
<span id="cb58-201"><a href="#cb58-201"></a>  <span class="st">"g_172.vir_2"</span>, <span class="st">"g_393.vir_1"</span></span>
<span id="cb58-202"><a href="#cb58-202"></a>)</span>
<span id="cb58-203"><a href="#cb58-203"></a><span class="in">```</span></span>
<span id="cb58-204"><a href="#cb58-204"></a></span>
<span id="cb58-205"><a href="#cb58-205"></a>Finally, run the script to generate the plot.</span>
<span id="cb58-206"><a href="#cb58-206"></a></span>
<span id="cb58-207"><a href="#cb58-207"></a><span class="in">```bash</span></span>
<span id="cb58-208"><a href="#cb58-208"></a><span class="ex">Rscript</span> scripts/analysis/homology_group_viz.R</span>
<span id="cb58-209"><a href="#cb58-209"></a><span class="in">```</span></span>
<span id="cb58-210"><a href="#cb58-210"></a></span>
<span id="cb58-211"><a href="#cb58-211"></a><span class="fu">## Individual prophage cluster analysis</span></span>
<span id="cb58-212"><a href="#cb58-212"></a></span>
<span id="cb58-213"><a href="#cb58-213"></a>Script <span class="in">`scripts/analysis/clustersmap_data_prophages.R`</span> is used to combine the</span>
<span id="cb58-214"><a href="#cb58-214"></a>prophages, homology groups and phylogeny to generate input for <span class="in">`clustermap.js`</span></span>
<span id="cb58-215"><a href="#cb58-215"></a>visualization tool and generate a report. Below, some configuration is provided</span>
<span id="cb58-216"><a href="#cb58-216"></a>for individual prophage clusters.</span>
<span id="cb58-217"><a href="#cb58-217"></a></span>
<span id="cb58-218"><a href="#cb58-218"></a><span class="fu">### phage_grp_46: highly conserved in all but 5 genomes from n23 clade</span></span>
<span id="cb58-219"><a href="#cb58-219"></a></span>
<span id="cb58-220"><a href="#cb58-220"></a><span class="in">```r</span></span>
<span id="cb58-221"><a href="#cb58-221"></a>grpToView <span class="ot">&lt;-</span> <span class="st">"phage_grp_46"</span></span>
<span id="cb58-222"><a href="#cb58-222"></a>subSample <span class="ot">&lt;-</span> <span class="cn">TRUE</span></span>
<span id="cb58-223"><a href="#cb58-223"></a>cutHeight <span class="ot">&lt;-</span> <span class="fl">0.5</span></span>
<span id="cb58-224"><a href="#cb58-224"></a>addFlankingRegions <span class="ot">&lt;-</span> <span class="cn">TRUE</span></span>
<span id="cb58-225"><a href="#cb58-225"></a>flankingRegion <span class="ot">&lt;-</span> <span class="dv">5000</span></span>
<span id="cb58-226"><a href="#cb58-226"></a></span>
<span id="cb58-227"><a href="#cb58-227"></a><span class="co"># ordering factor for prophages: host phylogeny, prophage HG PAV, prophage MASH,</span></span>
<span id="cb58-228"><a href="#cb58-228"></a><span class="co"># completeness score</span></span>
<span id="cb58-229"><a href="#cb58-229"></a>clusterOrder <span class="ot">&lt;-</span> <span class="st">"host"</span>  <span class="co"># completeness, host, hg_pav, cluster_mash</span></span>
<span id="cb58-230"><a href="#cb58-230"></a><span class="in">```</span></span>
<span id="cb58-231"><a href="#cb58-231"></a></span>
<span id="cb58-232"><a href="#cb58-232"></a><span class="fu">### phage_grp_45: prophage present in all BL-causing isolates</span></span>
<span id="cb58-233"><a href="#cb58-233"></a></span>
<span id="cb58-234"><a href="#cb58-234"></a><span class="in">```bash</span></span>
<span id="cb58-235"><a href="#cb58-235"></a><span class="ex">Rscript</span> scripts/utils/HG_range_coordinates.R <span class="at">--hgs</span> hg_22429319,hg_22429327 <span class="dt">\</span></span>
<span id="cb58-236"><a href="#cb58-236"></a>--genomes <span class="st">"g_406"</span></span>
<span id="cb58-237"><a href="#cb58-237"></a><span class="in">```</span></span>
<span id="cb58-238"><a href="#cb58-238"></a></span>
<span id="cb58-239"><a href="#cb58-239"></a><span class="in">```r</span></span>
<span id="cb58-240"><a href="#cb58-240"></a>grpToView <span class="ot">&lt;-</span> <span class="st">"grp_45_mosaicsm"</span></span>
<span id="cb58-241"><a href="#cb58-241"></a>subSample <span class="ot">&lt;-</span> <span class="cn">TRUE</span> </span>
<span id="cb58-242"><a href="#cb58-242"></a>cutHeight <span class="ot">&lt;-</span> <span class="fl">0.5</span></span>
<span id="cb58-243"><a href="#cb58-243"></a>addFlankingRegions <span class="ot">&lt;-</span> <span class="cn">TRUE</span></span>
<span id="cb58-244"><a href="#cb58-244"></a>flankingRegion <span class="ot">&lt;-</span> <span class="dv">7000</span></span>
<span id="cb58-245"><a href="#cb58-245"></a></span>
<span id="cb58-246"><a href="#cb58-246"></a><span class="co"># ordering factor for prophages: host phylogeny, prophage HG PAV, prophage MASH,</span></span>
<span id="cb58-247"><a href="#cb58-247"></a><span class="co"># completeness score</span></span>
<span id="cb58-248"><a href="#cb58-248"></a>clusterOrder <span class="ot">&lt;-</span> <span class="st">"host"</span> <span class="co"># host, hg_pav, cluster_mash</span></span>
<span id="cb58-249"><a href="#cb58-249"></a></span>
<span id="cb58-250"><a href="#cb58-250"></a><span class="co"># a vector of prophage identifiers that will be included in clustermap plot</span></span>
<span id="cb58-251"><a href="#cb58-251"></a>appendPhages <span class="ot">&lt;-</span> <span class="fu">c</span>()</span>
<span id="cb58-252"><a href="#cb58-252"></a></span>
<span id="cb58-253"><a href="#cb58-253"></a><span class="co"># whether to keep custom regions at the bottom or consider during phylogeny</span></span>
<span id="cb58-254"><a href="#cb58-254"></a><span class="co"># based ordering</span></span>
<span id="cb58-255"><a href="#cb58-255"></a>regions_phy_ordered <span class="ot">&lt;-</span> <span class="cn">FALSE</span></span>
<span id="cb58-256"><a href="#cb58-256"></a></span>
<span id="cb58-257"><a href="#cb58-257"></a><span class="co"># regions to append as list of list with following structure</span></span>
<span id="cb58-258"><a href="#cb58-258"></a><span class="co"># list(r1 = list(chr, start, end, genomeId), r2 = list(chr, start, end, genomeId))</span></span>
<span id="cb58-259"><a href="#cb58-259"></a>customRegions <span class="ot">&lt;-</span> <span class="fu">list</span>(</span>
<span id="cb58-260"><a href="#cb58-260"></a>  <span class="at">g_406_reg =</span> <span class="fu">list</span>(</span>
<span id="cb58-261"><a href="#cb58-261"></a>    <span class="at">chr =</span> <span class="st">"NAK641_contig_10_consensus"</span>, <span class="at">start =</span> <span class="dv">671040</span>, <span class="at">end =</span> <span class="dv">674984</span>, <span class="at">genomeId =</span> <span class="st">"g_406"</span></span>
<span id="cb58-262"><a href="#cb58-262"></a>  )</span>
<span id="cb58-263"><a href="#cb58-263"></a>)</span>
<span id="cb58-264"><a href="#cb58-264"></a></span>
<span id="cb58-265"><a href="#cb58-265"></a></span>
<span id="cb58-266"><a href="#cb58-266"></a><span class="co"># optionally, a custom region list can be provided to generate the plot</span></span>
<span id="cb58-267"><a href="#cb58-267"></a>grp <span class="ot">&lt;-</span> <span class="fu">list</span>(</span>
<span id="cb58-268"><a href="#cb58-268"></a>  <span class="at">phage_grp =</span> grpToView,</span>
<span id="cb58-269"><a href="#cb58-269"></a>  <span class="at">members =</span> <span class="fu">c</span>(<span class="st">"g_302.vir_1"</span>, <span class="st">"g_399.vir_1"</span>, <span class="st">"g_400.vir_2"</span>)</span>
<span id="cb58-270"><a href="#cb58-270"></a>)</span>
<span id="cb58-271"><a href="#cb58-271"></a><span class="in">```</span></span>
<span id="cb58-272"><a href="#cb58-272"></a></span>
<span id="cb58-273"><a href="#cb58-273"></a><span class="fu">### phage_grp_71</span></span>
<span id="cb58-274"><a href="#cb58-274"></a></span>
<span id="cb58-275"><a href="#cb58-275"></a><span class="in">```r</span></span>
<span id="cb58-276"><a href="#cb58-276"></a>grpToView <span class="ot">&lt;-</span> <span class="st">"phage_grp_71"</span></span>
<span id="cb58-277"><a href="#cb58-277"></a>subSample <span class="ot">&lt;-</span> <span class="cn">FALSE</span></span>
<span id="cb58-278"><a href="#cb58-278"></a>addFlankingRegions <span class="ot">&lt;-</span> <span class="cn">TRUE</span> </span>
<span id="cb58-279"><a href="#cb58-279"></a>flankingRegion <span class="ot">&lt;-</span> <span class="dv">5000</span></span>
<span id="cb58-280"><a href="#cb58-280"></a></span>
<span id="cb58-281"><a href="#cb58-281"></a><span class="co"># ordering factor for prophages: host phylogeny, prophage HG PAV, prophage MASH,</span></span>
<span id="cb58-282"><a href="#cb58-282"></a><span class="co"># completeness score</span></span>
<span id="cb58-283"><a href="#cb58-283"></a>clusterOrder <span class="ot">&lt;-</span> <span class="st">"hg_pav"</span>  <span class="co"># completeness, host, hg_pav, cluster_mash</span></span>
<span id="cb58-284"><a href="#cb58-284"></a></span>
<span id="cb58-285"><a href="#cb58-285"></a><span class="in">```</span></span>
<span id="cb58-286"><a href="#cb58-286"></a></span>
<span id="cb58-287"><a href="#cb58-287"></a><span class="fu">### phage_grp_107</span></span>
<span id="cb58-288"><a href="#cb58-288"></a></span>
<span id="cb58-289"><a href="#cb58-289"></a><span class="in">```r</span></span>
<span id="cb58-290"><a href="#cb58-290"></a>grpToView <span class="ot">&lt;-</span> <span class="st">"phage_grp_107"</span></span>
<span id="cb58-291"><a href="#cb58-291"></a>subSample <span class="ot">&lt;-</span> <span class="cn">TRUE</span></span>
<span id="cb58-292"><a href="#cb58-292"></a>cutHeight <span class="ot">&lt;-</span> <span class="fl">0.5</span></span>
<span id="cb58-293"><a href="#cb58-293"></a>addFlankingRegions <span class="ot">&lt;-</span> <span class="cn">TRUE</span></span>
<span id="cb58-294"><a href="#cb58-294"></a>flankingRegion <span class="ot">&lt;-</span> <span class="dv">5000</span></span>
<span id="cb58-295"><a href="#cb58-295"></a></span>
<span id="cb58-296"><a href="#cb58-296"></a><span class="co"># ordering factor for prophages: host phylogeny, prophage HG PAV, prophage MASH,</span></span>
<span id="cb58-297"><a href="#cb58-297"></a><span class="co"># completeness score</span></span>
<span id="cb58-298"><a href="#cb58-298"></a>clusterOrder <span class="ot">&lt;-</span> <span class="st">"host"</span>  <span class="co"># completeness, host, hg_pav, cluster_mash</span></span>
<span id="cb58-299"><a href="#cb58-299"></a></span>
<span id="cb58-300"><a href="#cb58-300"></a><span class="in">```</span></span>
<span id="cb58-301"><a href="#cb58-301"></a></span>
<span id="cb58-302"><a href="#cb58-302"></a><span class="fu">### BL-causing clade specific signatures</span></span>
<span id="cb58-303"><a href="#cb58-303"></a></span>
<span id="cb58-304"><a href="#cb58-304"></a><span class="in">```r</span></span>
<span id="cb58-305"><a href="#cb58-305"></a>grpToView <span class="ot">&lt;-</span> <span class="st">"Pbr_FN_specific"</span></span>
<span id="cb58-306"><a href="#cb58-306"></a>subSample <span class="ot">&lt;-</span> <span class="cn">FALSE</span></span>
<span id="cb58-307"><a href="#cb58-307"></a>cutHeight <span class="ot">&lt;-</span> <span class="fl">0.5</span></span>
<span id="cb58-308"><a href="#cb58-308"></a>addFlankingRegions <span class="ot">&lt;-</span> <span class="cn">FALSE</span></span>
<span id="cb58-309"><a href="#cb58-309"></a>flankingRegion <span class="ot">&lt;-</span> <span class="dv">5000</span></span>
<span id="cb58-310"><a href="#cb58-310"></a></span>
<span id="cb58-311"><a href="#cb58-311"></a><span class="co"># ordering factor for prophages: host phylogeny, prophage HG PAV, prophage MASH,</span></span>
<span id="cb58-312"><a href="#cb58-312"></a><span class="co"># completeness score</span></span>
<span id="cb58-313"><a href="#cb58-313"></a>clusterOrder <span class="ot">&lt;-</span> <span class="st">"default"</span>  <span class="co"># completeness, host, hg_pav, cluster_mash, default</span></span>
<span id="cb58-314"><a href="#cb58-314"></a></span>
<span id="cb58-315"><a href="#cb58-315"></a></span>
<span id="cb58-316"><a href="#cb58-316"></a><span class="co"># optionally, a custom region list can be provided to generate the plot</span></span>
<span id="cb58-317"><a href="#cb58-317"></a>grp <span class="ot">&lt;-</span> <span class="fu">list</span>(</span>
<span id="cb58-318"><a href="#cb58-318"></a>  <span class="at">phage_grp =</span> grpToView,</span>
<span id="cb58-319"><a href="#cb58-319"></a>  <span class="at">members =</span> <span class="fu">c</span>(<span class="st">"g_302.vir_1"</span>, <span class="st">"g_400.vir_2"</span>, <span class="st">"g_409.vir_1"</span>, <span class="st">"g_302.vir_2"</span>)</span>
<span id="cb58-320"><a href="#cb58-320"></a>)</span>
<span id="cb58-321"><a href="#cb58-321"></a><span class="in">```</span></span>
<span id="cb58-322"><a href="#cb58-322"></a></span>
<span id="cb58-323"><a href="#cb58-323"></a><span class="fu">### phage_grp_36</span></span>
<span id="cb58-324"><a href="#cb58-324"></a></span>
<span id="cb58-325"><a href="#cb58-325"></a><span class="in">```r</span></span>
<span id="cb58-326"><a href="#cb58-326"></a>grpToView <span class="ot">&lt;-</span> <span class="st">"phage_grp_36"</span></span>
<span id="cb58-327"><a href="#cb58-327"></a>subSample <span class="ot">&lt;-</span> <span class="cn">FALSE</span></span>
<span id="cb58-328"><a href="#cb58-328"></a>cutHeight <span class="ot">&lt;-</span> <span class="fl">1.5</span></span>
<span id="cb58-329"><a href="#cb58-329"></a>addFlankingRegions <span class="ot">&lt;-</span> <span class="cn">TRUE</span></span>
<span id="cb58-330"><a href="#cb58-330"></a>flankingRegion <span class="ot">&lt;-</span> <span class="dv">5000</span></span>
<span id="cb58-331"><a href="#cb58-331"></a></span>
<span id="cb58-332"><a href="#cb58-332"></a><span class="co"># ordering factor for prophages: host phylogeny, prophage HG PAV, prophage MASH,</span></span>
<span id="cb58-333"><a href="#cb58-333"></a><span class="co"># completeness score</span></span>
<span id="cb58-334"><a href="#cb58-334"></a>clusterOrder <span class="ot">&lt;-</span> <span class="st">"host"</span> <span class="co"># completeness, host, hg_pav, cluster_mash</span></span>
<span id="cb58-335"><a href="#cb58-335"></a><span class="in">```</span></span>
<span id="cb58-336"><a href="#cb58-336"></a></span>
<span id="cb58-337"><a href="#cb58-337"></a><span class="fu">### phage_grp_50</span></span>
<span id="cb58-338"><a href="#cb58-338"></a></span>
<span id="cb58-339"><a href="#cb58-339"></a><span class="in">```r</span></span>
<span id="cb58-340"><a href="#cb58-340"></a>grpToView <span class="ot">&lt;-</span> <span class="st">"phage_grp_50"</span></span>
<span id="cb58-341"><a href="#cb58-341"></a>subSample <span class="ot">&lt;-</span> <span class="cn">FALSE</span></span>
<span id="cb58-342"><a href="#cb58-342"></a>addFlankingRegions <span class="ot">&lt;-</span> <span class="cn">TRUE</span> </span>
<span id="cb58-343"><a href="#cb58-343"></a>flankingRegion <span class="ot">&lt;-</span> <span class="dv">10000</span></span>
<span id="cb58-344"><a href="#cb58-344"></a></span>
<span id="cb58-345"><a href="#cb58-345"></a><span class="co"># ordering factor for prophages: host phylogeny, prophage HG PAV, prophage MASH,</span></span>
<span id="cb58-346"><a href="#cb58-346"></a><span class="co"># completeness score</span></span>
<span id="cb58-347"><a href="#cb58-347"></a>clusterOrder <span class="ot">&lt;-</span> <span class="st">"cluster_mash"</span>  <span class="co"># completeness, host, hg_pav, cluster_mash</span></span>
<span id="cb58-348"><a href="#cb58-348"></a><span class="in">```</span></span>
<span id="cb58-349"><a href="#cb58-349"></a></span>
<span id="cb58-350"><a href="#cb58-350"></a><span class="fu">### phage_grp_21</span></span>
<span id="cb58-351"><a href="#cb58-351"></a></span>
<span id="cb58-352"><a href="#cb58-352"></a><span class="in">```r</span></span>
<span id="cb58-353"><a href="#cb58-353"></a>grpToView <span class="ot">&lt;-</span> <span class="st">"phage_grp_21"</span></span>
<span id="cb58-354"><a href="#cb58-354"></a>subSample <span class="ot">&lt;-</span> <span class="cn">FALSE</span></span>
<span id="cb58-355"><a href="#cb58-355"></a>addFlankingRegions <span class="ot">&lt;-</span> <span class="cn">TRUE</span></span>
<span id="cb58-356"><a href="#cb58-356"></a>flankingRegion <span class="ot">&lt;-</span> <span class="dv">5000</span></span>
<span id="cb58-357"><a href="#cb58-357"></a></span>
<span id="cb58-358"><a href="#cb58-358"></a><span class="co"># ordering factor for prophages: host phylogeny, prophage HG PAV, prophage MASH,</span></span>
<span id="cb58-359"><a href="#cb58-359"></a><span class="co"># completeness score</span></span>
<span id="cb58-360"><a href="#cb58-360"></a>clusterOrder <span class="ot">&lt;-</span> <span class="st">"cluster_mash"</span>  <span class="co"># completeness, host, hg_pav, cluster_mash</span></span>
<span id="cb58-361"><a href="#cb58-361"></a><span class="in">```</span></span>
<span id="cb58-362"><a href="#cb58-362"></a></span>
<span id="cb58-363"><a href="#cb58-363"></a><span class="fu">### T3SS missing from 4 P. brasiliense genomes</span></span>
<span id="cb58-364"><a href="#cb58-364"></a></span>
<span id="cb58-365"><a href="#cb58-365"></a><span class="in">```bash</span></span>
<span id="cb58-366"><a href="#cb58-366"></a><span class="ex">Rscript</span> scripts/utils/HG_range_coordinates.R <span class="at">--hgs</span> hg_22430875,hg_22430876 <span class="at">--genomes</span> g_187,g_368,g_182,g_185,g_236,g_177</span>
<span id="cb58-367"><a href="#cb58-367"></a><span class="in">```</span></span>
<span id="cb58-368"><a href="#cb58-368"></a></span>
<span id="cb58-369"><a href="#cb58-369"></a><span class="in">```r</span></span>
<span id="cb58-370"><a href="#cb58-370"></a>grpToView <span class="ot">&lt;-</span> <span class="st">"t3ss_regions"</span></span>
<span id="cb58-371"><a href="#cb58-371"></a>subSample <span class="ot">&lt;-</span> <span class="cn">FALSE</span></span>
<span id="cb58-372"><a href="#cb58-372"></a>cutHeight <span class="ot">&lt;-</span> <span class="fl">1.5</span></span>
<span id="cb58-373"><a href="#cb58-373"></a>addFlankingRegions <span class="ot">&lt;-</span> <span class="cn">TRUE</span></span>
<span id="cb58-374"><a href="#cb58-374"></a>flankingRegion <span class="ot">&lt;-</span> <span class="dv">6000</span></span>
<span id="cb58-375"><a href="#cb58-375"></a></span>
<span id="cb58-376"><a href="#cb58-376"></a><span class="co"># ordering factor for prophages: host phylogeny, prophage HG PAV, prophage MASH,</span></span>
<span id="cb58-377"><a href="#cb58-377"></a><span class="co"># completeness score</span></span>
<span id="cb58-378"><a href="#cb58-378"></a>clusterOrder <span class="ot">&lt;-</span> <span class="st">"host"</span>  <span class="co"># completeness, host, hg_pav, cluster_mash</span></span>
<span id="cb58-379"><a href="#cb58-379"></a></span>
<span id="cb58-380"><a href="#cb58-380"></a><span class="co"># a vector of prophage identifiers that will be included in clustermap plot</span></span>
<span id="cb58-381"><a href="#cb58-381"></a>appendPhages <span class="ot">&lt;-</span> <span class="fu">c</span>()</span>
<span id="cb58-382"><a href="#cb58-382"></a></span>
<span id="cb58-383"><a href="#cb58-383"></a><span class="co"># regions to append as list of list with following structure</span></span>
<span id="cb58-384"><a href="#cb58-384"></a><span class="co"># list(r1 = list(chr, start, end, genomeId), r2 = list(chr, start, end, genomeId))</span></span>
<span id="cb58-385"><a href="#cb58-385"></a>customRegions <span class="ot">&lt;-</span> <span class="fu">list</span>(</span>
<span id="cb58-386"><a href="#cb58-386"></a>  <span class="at">g_177_reg  =</span> <span class="fu">list</span>(<span class="at">genomeId =</span> <span class="st">"g_177"</span>, <span class="at">chr =</span> <span class="st">"NZ_JACGEP010000027.1"</span>, <span class="at">start =</span> <span class="dv">56850</span>, <span class="at">end =</span> <span class="dv">59337</span>),</span>
<span id="cb58-387"><a href="#cb58-387"></a>  <span class="at">g_182_reg =</span> <span class="fu">list</span>(<span class="at">genomeId =</span> <span class="st">"g_182"</span>, <span class="at">chr =</span> <span class="st">"NZ_JACGZZ010000068.1"</span>, <span class="at">start =</span> <span class="dv">116773</span>, <span class="at">end =</span> <span class="dv">119260</span>),</span>
<span id="cb58-388"><a href="#cb58-388"></a>  <span class="at">g_185_reg =</span> <span class="fu">list</span>(<span class="at">genomeId =</span> <span class="st">"g_185"</span>, <span class="at">chr =</span> <span class="st">"NZ_JACGEN010000027.1"</span>, <span class="at">start =</span> <span class="dv">117747</span>, <span class="at">end =</span> <span class="dv">120234</span>),</span>
<span id="cb58-389"><a href="#cb58-389"></a>  <span class="at">g_236_reg =</span> <span class="fu">list</span>(<span class="at">genomeId =</span> <span class="st">"g_236"</span>, <span class="at">chr =</span> <span class="st">"NZ_JACDSF010000026.1"</span>, <span class="at">start =</span> <span class="dv">2275</span>, <span class="at">end =</span> <span class="dv">4762</span>),</span>
<span id="cb58-390"><a href="#cb58-390"></a>  <span class="at">g_187_reg =</span> <span class="fu">list</span>(<span class="at">genomeId =</span> <span class="st">"g_187"</span>, <span class="at">chr =</span> <span class="st">"NZ_CP059963.1"</span>, <span class="at">start =</span> <span class="dv">2603432</span>, <span class="at">end =</span> <span class="dv">2605919</span>),</span>
<span id="cb58-391"><a href="#cb58-391"></a>  <span class="at">g_368_reg =</span> <span class="fu">list</span>(<span class="at">genomeId =</span> <span class="st">"g_368"</span>, <span class="at">chr =</span> <span class="st">"NZ_JUJT01000001.1"</span>, <span class="at">start =</span> <span class="dv">566617</span>, <span class="at">end =</span> <span class="dv">569104</span>)</span>
<span id="cb58-392"><a href="#cb58-392"></a>)</span>
<span id="cb58-393"><a href="#cb58-393"></a></span>
<span id="cb58-394"><a href="#cb58-394"></a><span class="co"># whether to keep custom regions at the bottom or consider during phylogeny</span></span>
<span id="cb58-395"><a href="#cb58-395"></a><span class="co"># based ordering</span></span>
<span id="cb58-396"><a href="#cb58-396"></a>regions_phy_ordered <span class="ot">&lt;-</span> <span class="cn">TRUE</span></span>
<span id="cb58-397"><a href="#cb58-397"></a><span class="in">```</span></span>
<span id="cb58-398"><a href="#cb58-398"></a></span>
<span id="cb58-399"><a href="#cb58-399"></a><span class="fu">## Carotovoricin (CTV) cluster or phage_grp_1 data analysis</span></span>
<span id="cb58-400"><a href="#cb58-400"></a></span>
<span id="cb58-401"><a href="#cb58-401"></a><span class="fu">### CTV across pangenome</span></span>
<span id="cb58-402"><a href="#cb58-402"></a></span>
<span id="cb58-403"><a href="#cb58-403"></a><span class="in">```r</span></span>
<span id="cb58-404"><a href="#cb58-404"></a>grpToView <span class="ot">&lt;-</span> <span class="st">"phage_grp_1"</span></span>
<span id="cb58-405"><a href="#cb58-405"></a>subSample <span class="ot">&lt;-</span> <span class="cn">TRUE</span></span>
<span id="cb58-406"><a href="#cb58-406"></a>cutHeight <span class="ot">&lt;-</span> <span class="fl">1.5</span></span>
<span id="cb58-407"><a href="#cb58-407"></a>addFlankingRegions <span class="ot">&lt;-</span> <span class="cn">TRUE</span></span>
<span id="cb58-408"><a href="#cb58-408"></a>flankingRegion <span class="ot">&lt;-</span> <span class="dv">5000</span></span>
<span id="cb58-409"><a href="#cb58-409"></a></span>
<span id="cb58-410"><a href="#cb58-410"></a><span class="co"># ordering factor for prophages: host phylogeny, prophage HG PAV, prophage MASH,</span></span>
<span id="cb58-411"><a href="#cb58-411"></a><span class="co"># completeness score</span></span>
<span id="cb58-412"><a href="#cb58-412"></a>clusterOrder <span class="ot">&lt;-</span> <span class="st">"host"</span>  <span class="co"># completeness, host, hg_pav, cluster_mash</span></span>
<span id="cb58-413"><a href="#cb58-413"></a><span class="in">```</span></span>
<span id="cb58-414"><a href="#cb58-414"></a></span>
<span id="cb58-415"><a href="#cb58-415"></a><span class="fu">### CTV in type strains</span></span>
<span id="cb58-416"><a href="#cb58-416"></a></span>
<span id="cb58-417"><a href="#cb58-417"></a>Include *P. atrosepticum* (g_385), *P. betavasculorum* (g_386) and *P. cacticida*</span>
<span id="cb58-418"><a href="#cb58-418"></a>(g_451) genomes that lack CTV cluster.</span>
<span id="cb58-419"><a href="#cb58-419"></a></span>
<span id="cb58-420"><a href="#cb58-420"></a><span class="in">```r</span></span>
<span id="cb58-421"><a href="#cb58-421"></a>grpToView <span class="ot">&lt;-</span> <span class="st">"ctv_typeStrains"</span></span>
<span id="cb58-422"><a href="#cb58-422"></a>subSample <span class="ot">&lt;-</span> <span class="cn">FALSE</span></span>
<span id="cb58-423"><a href="#cb58-423"></a>cutHeight <span class="ot">&lt;-</span> <span class="fl">1.5</span></span>
<span id="cb58-424"><a href="#cb58-424"></a>addFlankingRegions <span class="ot">&lt;-</span> <span class="cn">TRUE</span></span>
<span id="cb58-425"><a href="#cb58-425"></a>flankingRegion <span class="ot">&lt;-</span> <span class="dv">6000</span></span>
<span id="cb58-426"><a href="#cb58-426"></a></span>
<span id="cb58-427"><a href="#cb58-427"></a><span class="co"># ordering factor for prophages: host phylogeny, prophage HG PAV, prophage MASH,</span></span>
<span id="cb58-428"><a href="#cb58-428"></a><span class="co"># completeness score</span></span>
<span id="cb58-429"><a href="#cb58-429"></a>clusterOrder <span class="ot">&lt;-</span> <span class="st">"host"</span>  <span class="co"># completeness, host, hg_pav, cluster_mash</span></span>
<span id="cb58-430"><a href="#cb58-430"></a></span>
<span id="cb58-431"><a href="#cb58-431"></a><span class="co"># a vector of prophage identifiers that will be included in clustermap plot</span></span>
<span id="cb58-432"><a href="#cb58-432"></a>appendPhages <span class="ot">&lt;-</span> <span class="fu">c</span>()</span>
<span id="cb58-433"><a href="#cb58-433"></a></span>
<span id="cb58-434"><a href="#cb58-434"></a><span class="co"># regions to append as list of list with following structure</span></span>
<span id="cb58-435"><a href="#cb58-435"></a><span class="co"># list(r1 = list(chr, start, end, genomeId), r2 = list(chr, start, end, genomeId))</span></span>
<span id="cb58-436"><a href="#cb58-436"></a>customRegions <span class="ot">&lt;-</span> <span class="fu">list</span>(</span>
<span id="cb58-437"><a href="#cb58-437"></a>  <span class="at">g_385_reg =</span> <span class="fu">list</span>(</span>
<span id="cb58-438"><a href="#cb58-438"></a>    <span class="at">chr =</span> <span class="st">"NZ_JQHK01000003.1"</span>, <span class="at">start =</span> <span class="dv">203963</span>, <span class="at">end =</span> <span class="dv">207120</span>, <span class="at">genomeId =</span> <span class="st">"g_385"</span></span>
<span id="cb58-439"><a href="#cb58-439"></a>  ),</span>
<span id="cb58-440"><a href="#cb58-440"></a>  <span class="at">g_386_reg =</span> <span class="fu">list</span>(</span>
<span id="cb58-441"><a href="#cb58-441"></a>    <span class="at">chr =</span> <span class="st">"NZ_JQHM01000001.1"</span>, <span class="at">start =</span> <span class="dv">553213</span>, <span class="at">end =</span> <span class="dv">555615</span>, <span class="at">genomeId =</span> <span class="st">"g_386"</span></span>
<span id="cb58-442"><a href="#cb58-442"></a>  ),</span>
<span id="cb58-443"><a href="#cb58-443"></a>  <span class="at">g_451_reg =</span> <span class="fu">list</span>(</span>
<span id="cb58-444"><a href="#cb58-444"></a>    <span class="at">chr =</span> <span class="st">"Contig_2_668.636"</span>, <span class="at">start =</span> <span class="dv">191452</span>, <span class="at">end =</span> <span class="dv">191490</span>, <span class="at">genomeId =</span> <span class="st">"g_451"</span></span>
<span id="cb58-445"><a href="#cb58-445"></a>  )</span>
<span id="cb58-446"><a href="#cb58-446"></a>)</span>
<span id="cb58-447"><a href="#cb58-447"></a></span>
<span id="cb58-448"><a href="#cb58-448"></a><span class="co"># whether to keep custom regions at the bottom or consider during phylogeny</span></span>
<span id="cb58-449"><a href="#cb58-449"></a><span class="co"># based ordering</span></span>
<span id="cb58-450"><a href="#cb58-450"></a>regions_phy_ordered <span class="ot">&lt;-</span> <span class="cn">TRUE</span></span>
<span id="cb58-451"><a href="#cb58-451"></a></span>
<span id="cb58-452"><a href="#cb58-452"></a></span>
<span id="cb58-453"><a href="#cb58-453"></a></span>
<span id="cb58-454"><a href="#cb58-454"></a>grp <span class="ot">&lt;-</span> <span class="fu">list</span>(</span>
<span id="cb58-455"><a href="#cb58-455"></a>  <span class="at">phage_grp =</span> grpToView,</span>
<span id="cb58-456"><a href="#cb58-456"></a>  <span class="at">members =</span> <span class="fu">c</span>(</span>
<span id="cb58-457"><a href="#cb58-457"></a>    <span class="st">"g_345.vir_1"</span>, <span class="st">"g_446.vir_4"</span>, <span class="st">"g_66.vir_3"</span>, <span class="st">"g_222.vir_2"</span>, <span class="st">"g_365.vir_3"</span>,</span>
<span id="cb58-458"><a href="#cb58-458"></a>    <span class="st">"g_442.vir_1"</span>, <span class="st">"g_8.vir_2"</span>, <span class="st">"g_38.vir_2"</span>, <span class="st">"g_273.vir_2"</span>, <span class="st">"g_259.vir_4"</span>,</span>
<span id="cb58-459"><a href="#cb58-459"></a>    <span class="st">"g_305.vir_1"</span>, <span class="st">"g_378.vir_6"</span>, <span class="st">"g_428.vir_1"</span>, <span class="st">"g_248.vir_1"</span>, <span class="st">"g_449.vir_1"</span>,</span>
<span id="cb58-460"><a href="#cb58-460"></a>    <span class="st">"g_54.vir_1"</span>, <span class="st">"g_116.vir_3"</span>, <span class="st">"g_423.vir_3"</span>, <span class="st">"g_375.vir_2"</span>, <span class="st">"g_381.vir_2"</span></span>
<span id="cb58-461"><a href="#cb58-461"></a>  )</span>
<span id="cb58-462"><a href="#cb58-462"></a>)</span>
<span id="cb58-463"><a href="#cb58-463"></a></span>
<span id="cb58-464"><a href="#cb58-464"></a><span class="in">```</span></span>
<span id="cb58-465"><a href="#cb58-465"></a></span>
<span id="cb58-466"><a href="#cb58-466"></a><span class="fu">### Carotovoricin cluster missing from genomes</span></span>
<span id="cb58-467"><a href="#cb58-467"></a></span>
<span id="cb58-468"><a href="#cb58-468"></a><span class="in">```r</span></span>
<span id="cb58-469"><a href="#cb58-469"></a>grpToView <span class="ot">&lt;-</span> <span class="st">"ctv_deletion"</span></span>
<span id="cb58-470"><a href="#cb58-470"></a>subSample <span class="ot">&lt;-</span> <span class="cn">TRUE</span></span>
<span id="cb58-471"><a href="#cb58-471"></a>cutHeight <span class="ot">&lt;-</span> <span class="fl">1.5</span></span>
<span id="cb58-472"><a href="#cb58-472"></a>addFlankingRegions <span class="ot">&lt;-</span> <span class="cn">TRUE</span></span>
<span id="cb58-473"><a href="#cb58-473"></a>flankingRegion <span class="ot">&lt;-</span> <span class="dv">5000</span></span>
<span id="cb58-474"><a href="#cb58-474"></a></span>
<span id="cb58-475"><a href="#cb58-475"></a><span class="co"># ordering factor for prophages: host phylogeny, prophage HG PAV, prophage MASH,</span></span>
<span id="cb58-476"><a href="#cb58-476"></a><span class="co"># completeness score</span></span>
<span id="cb58-477"><a href="#cb58-477"></a>clusterOrder <span class="ot">&lt;-</span> <span class="st">"host"</span>  <span class="co"># completeness, host, hg_pav, cluster_mash</span></span>
<span id="cb58-478"><a href="#cb58-478"></a></span>
<span id="cb58-479"><a href="#cb58-479"></a><span class="co"># regions to append as list of list with following structure</span></span>
<span id="cb58-480"><a href="#cb58-480"></a><span class="co"># list(r1 = list(chr, start, end, genomeId), r2 = list(chr, start, end, genomeId))</span></span>
<span id="cb58-481"><a href="#cb58-481"></a>customRegions <span class="ot">&lt;-</span> <span class="fu">list</span>(</span>
<span id="cb58-482"><a href="#cb58-482"></a>  <span class="at">g_177_reg =</span> <span class="fu">list</span>(</span>
<span id="cb58-483"><a href="#cb58-483"></a>    <span class="at">chr =</span> <span class="st">"NZ_JACGEP010000002.1"</span>, <span class="at">start =</span> <span class="dv">102351</span>, <span class="at">end =</span> <span class="dv">102361</span>, <span class="at">genomeId =</span> <span class="st">"g_177"</span></span>
<span id="cb58-484"><a href="#cb58-484"></a>  ),</span>
<span id="cb58-485"><a href="#cb58-485"></a>  <span class="at">g_182_reg =</span> <span class="fu">list</span>(</span>
<span id="cb58-486"><a href="#cb58-486"></a>    <span class="at">chr =</span> <span class="st">"NZ_JACGZZ010000050.1"</span>, <span class="at">start =</span> <span class="dv">81625</span>, <span class="at">end =</span> <span class="dv">81635</span>, <span class="at">genomeId =</span> <span class="st">"g_182"</span></span>
<span id="cb58-487"><a href="#cb58-487"></a>  ),</span>
<span id="cb58-488"><a href="#cb58-488"></a>  <span class="at">g_185_reg =</span> <span class="fu">list</span>(</span>
<span id="cb58-489"><a href="#cb58-489"></a>    <span class="at">chr =</span> <span class="st">"NZ_JACGEN010000006.1"</span>, <span class="at">start =</span> <span class="dv">81657</span>, <span class="at">end =</span> <span class="dv">81667</span>, <span class="at">genomeId =</span> <span class="st">"g_185"</span></span>
<span id="cb58-490"><a href="#cb58-490"></a>  ),</span>
<span id="cb58-491"><a href="#cb58-491"></a>  <span class="at">g_236_reg =</span> <span class="fu">list</span>(</span>
<span id="cb58-492"><a href="#cb58-492"></a>    <span class="at">chr =</span> <span class="st">"NZ_JACDSF010000027.1"</span>, <span class="at">start =</span> <span class="dv">89639</span>, <span class="at">end =</span> <span class="dv">89649</span>, <span class="at">genomeId =</span> <span class="st">"g_236"</span></span>
<span id="cb58-493"><a href="#cb58-493"></a>  ),</span>
<span id="cb58-494"><a href="#cb58-494"></a>  <span class="at">g_385_reg =</span> <span class="fu">list</span>(</span>
<span id="cb58-495"><a href="#cb58-495"></a>    <span class="at">chr =</span> <span class="st">"NZ_JQHK01000003.1"</span>, <span class="at">start =</span> <span class="dv">203963</span>, <span class="at">end =</span> <span class="dv">207120</span>, <span class="at">genomeId =</span> <span class="st">"g_385"</span></span>
<span id="cb58-496"><a href="#cb58-496"></a>  ),</span>
<span id="cb58-497"><a href="#cb58-497"></a>  <span class="at">g_386_reg =</span> <span class="fu">list</span>(</span>
<span id="cb58-498"><a href="#cb58-498"></a>    <span class="at">chr =</span> <span class="st">"NZ_JQHM01000001.1"</span>, <span class="at">start =</span> <span class="dv">553213</span>, <span class="at">end =</span> <span class="dv">555615</span>, <span class="at">genomeId =</span> <span class="st">"g_386"</span></span>
<span id="cb58-499"><a href="#cb58-499"></a>  ),</span>
<span id="cb58-500"><a href="#cb58-500"></a>  <span class="at">g_451_reg =</span> <span class="fu">list</span>(</span>
<span id="cb58-501"><a href="#cb58-501"></a>    <span class="at">chr =</span> <span class="st">"Contig_2_668.636"</span>, <span class="at">start =</span> <span class="dv">191452</span>, <span class="at">end =</span> <span class="dv">191490</span>, <span class="at">genomeId =</span> <span class="st">"g_451"</span></span>
<span id="cb58-502"><a href="#cb58-502"></a>  )</span>
<span id="cb58-503"><a href="#cb58-503"></a>)</span>
<span id="cb58-504"><a href="#cb58-504"></a></span>
<span id="cb58-505"><a href="#cb58-505"></a></span>
<span id="cb58-506"><a href="#cb58-506"></a>grp <span class="ot">&lt;-</span> <span class="fu">list</span>(</span>
<span id="cb58-507"><a href="#cb58-507"></a>  <span class="at">phage_grp =</span> grpToView,</span>
<span id="cb58-508"><a href="#cb58-508"></a>  <span class="at">members =</span> <span class="fu">c</span>(</span>
<span id="cb58-509"><a href="#cb58-509"></a>    <span class="st">"g_345.vir_1"</span>, <span class="st">"g_66.vir_3"</span>, <span class="st">"g_8.vir_2"</span>, <span class="st">"g_381.vir_2"</span></span>
<span id="cb58-510"><a href="#cb58-510"></a>  )</span>
<span id="cb58-511"><a href="#cb58-511"></a>)</span>
<span id="cb58-512"><a href="#cb58-512"></a></span>
<span id="cb58-513"><a href="#cb58-513"></a><span class="in">```</span></span>
<span id="cb58-514"><a href="#cb58-514"></a></span>
<span id="cb58-515"><a href="#cb58-515"></a><span class="fu">### carotovoricin cluster absent in some *P. brasiliense* isolates</span></span>
<span id="cb58-516"><a href="#cb58-516"></a></span>
<span id="cb58-517"><a href="#cb58-517"></a>ybiB: hg_22427603</span>
<span id="cb58-518"><a href="#cb58-518"></a></span>
<span id="cb58-519"><a href="#cb58-519"></a>tolC_2: hg_22427641</span>
<span id="cb58-520"><a href="#cb58-520"></a></span>
<span id="cb58-521"><a href="#cb58-521"></a>*P. brasiliense* lacking CTV cluster: g_149, g_177, g_182, g_185, g_236</span>
<span id="cb58-522"><a href="#cb58-522"></a></span>
<span id="cb58-523"><a href="#cb58-523"></a><span class="in">```r</span></span>
<span id="cb58-524"><a href="#cb58-524"></a>grpToView <span class="ot">&lt;-</span> <span class="st">"ctv_pbr"</span></span>
<span id="cb58-525"><a href="#cb58-525"></a>subSample <span class="ot">&lt;-</span> <span class="cn">TRUE</span></span>
<span id="cb58-526"><a href="#cb58-526"></a>cutHeight <span class="ot">&lt;-</span> <span class="fl">1.5</span></span>
<span id="cb58-527"><a href="#cb58-527"></a>addFlankingRegions <span class="ot">&lt;-</span> <span class="cn">TRUE</span></span>
<span id="cb58-528"><a href="#cb58-528"></a>flankingRegion <span class="ot">&lt;-</span> <span class="dv">5000</span></span>
<span id="cb58-529"><a href="#cb58-529"></a></span>
<span id="cb58-530"><a href="#cb58-530"></a><span class="co"># ordering factor for prophages: host phylogeny, prophage HG PAV, prophage MASH,</span></span>
<span id="cb58-531"><a href="#cb58-531"></a><span class="co"># completeness score</span></span>
<span id="cb58-532"><a href="#cb58-532"></a>clusterOrder <span class="ot">&lt;-</span> <span class="st">"cluster_mash"</span>  <span class="co"># completeness, host, hg_pav, cluster_mash</span></span>
<span id="cb58-533"><a href="#cb58-533"></a></span>
<span id="cb58-534"><a href="#cb58-534"></a><span class="co"># regions to append as list of list with following structure</span></span>
<span id="cb58-535"><a href="#cb58-535"></a><span class="co"># list(r1 = list(chr, start, end, genomeId), r2 = list(chr, start, end, genomeId))</span></span>
<span id="cb58-536"><a href="#cb58-536"></a>customRegions <span class="ot">&lt;-</span> <span class="fu">list</span>(</span>
<span id="cb58-537"><a href="#cb58-537"></a>  <span class="at">g_177_reg =</span> <span class="fu">list</span>(</span>
<span id="cb58-538"><a href="#cb58-538"></a>    <span class="at">chr =</span> <span class="st">"NZ_JACGEP010000002.1"</span>, <span class="at">start =</span> <span class="dv">102351</span>, <span class="at">end =</span> <span class="dv">102361</span>, <span class="at">genomeId =</span> <span class="st">"g_177"</span></span>
<span id="cb58-539"><a href="#cb58-539"></a>  ),</span>
<span id="cb58-540"><a href="#cb58-540"></a>  <span class="at">g_182_reg =</span> <span class="fu">list</span>(</span>
<span id="cb58-541"><a href="#cb58-541"></a>    <span class="at">chr =</span> <span class="st">"NZ_JACGZZ010000050.1"</span>, <span class="at">start =</span> <span class="dv">81625</span>, <span class="at">end =</span> <span class="dv">81635</span>, <span class="at">genomeId =</span> <span class="st">"g_182"</span></span>
<span id="cb58-542"><a href="#cb58-542"></a>  ),</span>
<span id="cb58-543"><a href="#cb58-543"></a>  <span class="at">g_185_reg =</span> <span class="fu">list</span>(</span>
<span id="cb58-544"><a href="#cb58-544"></a>    <span class="at">chr =</span> <span class="st">"NZ_JACGEN010000006.1"</span>, <span class="at">start =</span> <span class="dv">81657</span>, <span class="at">end =</span> <span class="dv">81667</span>, <span class="at">genomeId =</span> <span class="st">"g_185"</span></span>
<span id="cb58-545"><a href="#cb58-545"></a>  ),</span>
<span id="cb58-546"><a href="#cb58-546"></a>  <span class="at">g_236_reg =</span> <span class="fu">list</span>(</span>
<span id="cb58-547"><a href="#cb58-547"></a>    <span class="at">chr =</span> <span class="st">"NZ_JACDSF010000027.1"</span>, <span class="at">start =</span> <span class="dv">89639</span>, <span class="at">end =</span> <span class="dv">89649</span>, <span class="at">genomeId =</span> <span class="st">"g_236"</span></span>
<span id="cb58-548"><a href="#cb58-548"></a>  )</span>
<span id="cb58-549"><a href="#cb58-549"></a>)</span>
<span id="cb58-550"><a href="#cb58-550"></a></span>
<span id="cb58-551"><a href="#cb58-551"></a></span>
<span id="cb58-552"><a href="#cb58-552"></a>grp <span class="ot">&lt;-</span> <span class="fu">list</span>(</span>
<span id="cb58-553"><a href="#cb58-553"></a>  <span class="at">phage_grp =</span> grpToView,</span>
<span id="cb58-554"><a href="#cb58-554"></a>  <span class="at">members =</span> dplyr<span class="sc">::</span><span class="fu">filter</span>(</span>
<span id="cb58-555"><a href="#cb58-555"></a>    regionClusters,</span>
<span id="cb58-556"><a href="#cb58-556"></a>    SpeciesName <span class="sc">==</span> <span class="st">"P. brasiliense"</span>, nFragments <span class="sc">==</span> <span class="dv">1</span>, phage_grp <span class="sc">==</span> <span class="st">"phage_grp_1"</span>,</span>
<span id="cb58-557"><a href="#cb58-557"></a>    <span class="sc">!</span>prophage_id <span class="sc">%in%</span> <span class="fu">c</span>(<span class="st">"g_408.vir_3"</span>, <span class="st">"g_403.vir_3"</span>, <span class="st">"g_399.vir_3"</span>)</span>
<span id="cb58-558"><a href="#cb58-558"></a>  ) <span class="sc">%&gt;%</span></span>
<span id="cb58-559"><a href="#cb58-559"></a>    dplyr<span class="sc">::</span><span class="fu">pull</span>(prophage_id)</span>
<span id="cb58-560"><a href="#cb58-560"></a>)</span>
<span id="cb58-561"><a href="#cb58-561"></a></span>
<span id="cb58-562"><a href="#cb58-562"></a><span class="in">```</span></span>
<span id="cb58-563"><a href="#cb58-563"></a></span>
<span id="cb58-564"><a href="#cb58-564"></a>Prophage clusters found in the ctv-lacking Pbr: phage_grp_30, phage_grp_6, phage_grp_29</span>
<span id="cb58-565"><a href="#cb58-565"></a></span>
<span id="cb58-566"><a href="#cb58-566"></a><span class="fu">### CTV in *P. versatile*</span></span>
<span id="cb58-567"><a href="#cb58-567"></a></span>
<span id="cb58-568"><a href="#cb58-568"></a><span class="in">```r</span></span>
<span id="cb58-569"><a href="#cb58-569"></a>grpToView <span class="ot">&lt;-</span> <span class="st">"ctv_pvs"</span></span>
<span id="cb58-570"><a href="#cb58-570"></a>subSample <span class="ot">&lt;-</span> <span class="cn">TRUE</span></span>
<span id="cb58-571"><a href="#cb58-571"></a>cutHeight <span class="ot">&lt;-</span> <span class="fl">1.5</span></span>
<span id="cb58-572"><a href="#cb58-572"></a>addFlankingRegions <span class="ot">&lt;-</span> <span class="cn">TRUE</span></span>
<span id="cb58-573"><a href="#cb58-573"></a>flankingRegion <span class="ot">&lt;-</span> <span class="dv">5000</span></span>
<span id="cb58-574"><a href="#cb58-574"></a></span>
<span id="cb58-575"><a href="#cb58-575"></a></span>
<span id="cb58-576"><a href="#cb58-576"></a>grp <span class="ot">&lt;-</span> <span class="fu">list</span>(</span>
<span id="cb58-577"><a href="#cb58-577"></a>  <span class="at">phage_grp =</span> grpToView,</span>
<span id="cb58-578"><a href="#cb58-578"></a>  <span class="at">members =</span> dplyr<span class="sc">::</span><span class="fu">filter</span>(</span>
<span id="cb58-579"><a href="#cb58-579"></a>    regionClusters,</span>
<span id="cb58-580"><a href="#cb58-580"></a>    SpeciesName <span class="sc">==</span> <span class="st">"P. versatile"</span>, nFragments <span class="sc">==</span> <span class="dv">1</span>, phage_grp <span class="sc">==</span> <span class="st">"phage_grp_1"</span></span>
<span id="cb58-581"><a href="#cb58-581"></a>  ) <span class="sc">%&gt;%</span></span>
<span id="cb58-582"><a href="#cb58-582"></a>    dplyr<span class="sc">::</span><span class="fu">pull</span>(prophage_id)</span>
<span id="cb58-583"><a href="#cb58-583"></a>)</span>
<span id="cb58-584"><a href="#cb58-584"></a><span class="in">```</span></span>
<span id="cb58-585"><a href="#cb58-585"></a></span>
<span id="cb58-586"><a href="#cb58-586"></a><span class="fu">### CTV cluster in *P. versatile* collected from France</span></span>
<span id="cb58-587"><a href="#cb58-587"></a></span>
<span id="cb58-588"><a href="#cb58-588"></a><span class="in">```r</span></span>
<span id="cb58-589"><a href="#cb58-589"></a>grpToView <span class="ot">&lt;-</span> <span class="st">"ctv_pvs_fr"</span></span>
<span id="cb58-590"><a href="#cb58-590"></a>subSample <span class="ot">&lt;-</span> <span class="cn">TRUE</span></span>
<span id="cb58-591"><a href="#cb58-591"></a>cutHeight <span class="ot">&lt;-</span> <span class="fl">1.5</span></span>
<span id="cb58-592"><a href="#cb58-592"></a>addFlankingRegions <span class="ot">&lt;-</span> <span class="cn">TRUE</span></span>
<span id="cb58-593"><a href="#cb58-593"></a>flankingRegion <span class="ot">&lt;-</span> <span class="dv">5000</span></span>
<span id="cb58-594"><a href="#cb58-594"></a></span>
<span id="cb58-595"><a href="#cb58-595"></a>grp <span class="ot">&lt;-</span> <span class="fu">list</span>(</span>
<span id="cb58-596"><a href="#cb58-596"></a>  <span class="at">phage_grp =</span> grpToView,</span>
<span id="cb58-597"><a href="#cb58-597"></a>  <span class="at">members =</span> dplyr<span class="sc">::</span><span class="fu">filter</span>(</span>
<span id="cb58-598"><a href="#cb58-598"></a>    regionClusters,</span>
<span id="cb58-599"><a href="#cb58-599"></a>    SpeciesName <span class="sc">==</span> <span class="st">"P. versatile"</span>, nFragments <span class="sc">==</span> <span class="dv">1</span>, phage_grp <span class="sc">==</span> <span class="st">"phage_grp_1"</span>,</span>
<span id="cb58-600"><a href="#cb58-600"></a>    geo_loc_country <span class="sc">==</span> <span class="st">"France"</span></span>
<span id="cb58-601"><a href="#cb58-601"></a>  ) <span class="sc">%&gt;%</span></span>
<span id="cb58-602"><a href="#cb58-602"></a>    dplyr<span class="sc">::</span><span class="fu">pull</span>(prophage_id)</span>
<span id="cb58-603"><a href="#cb58-603"></a>)</span>
<span id="cb58-604"><a href="#cb58-604"></a><span class="in">```</span></span>
<span id="cb58-605"><a href="#cb58-605"></a></span>
<span id="cb58-606"><a href="#cb58-606"></a><span class="fu">## Carotovoricin tail fiber variation region MSA</span></span>
<span id="cb58-607"><a href="#cb58-607"></a></span>
<span id="cb58-608"><a href="#cb58-608"></a>Extract the region between the two homology groups, <span class="in">`hg_22427604`</span> and <span class="in">`hg_22427603`</span></span>
<span id="cb58-609"><a href="#cb58-609"></a>for the carotovoricin cluster. Additionally, generate <span class="in">`GFF3`</span> files with the</span>
<span id="cb58-610"><a href="#cb58-610"></a>homology group, COG, PFAM and other metadata information to visualize.</span>
<span id="cb58-611"><a href="#cb58-611"></a></span>
<span id="cb58-612"><a href="#cb58-612"></a><span class="in">```bash</span></span>
<span id="cb58-613"><a href="#cb58-613"></a><span class="co"># Rscript scripts/analysis/clustermap_get_regions.R</span></span>
<span id="cb58-614"><a href="#cb58-614"></a><span class="ex">Rscript</span> scripts/utils/HG_range_coordinates.R <span class="at">--hgs</span> hg_22427604,hg_22427603 <span class="dt">\</span></span>
<span id="cb58-615"><a href="#cb58-615"></a>--genomes <span class="st">"g_302,g_364,g_337,g_439,g_403,g_399,g_408,g_175,g_138,g_368,g_345,</span><span class="dt">\</span></span>
<span id="cb58-616"><a href="#cb58-616"></a><span class="st">g_366,g_308,g_191,g_173,g_155,g_166,g_299,g_438,g_263,g_43,g_391"</span> <span class="dt">\</span></span>
<span id="cb58-617"><a href="#cb58-617"></a>--out ctv_pbr.variable_regions.tab <span class="dt">\</span></span>
<span id="cb58-618"><a href="#cb58-618"></a>--dir analysis/pangenome_v2/prophages/cluster_viz/ctv_pbr</span>
<span id="cb58-619"><a href="#cb58-619"></a></span>
<span id="cb58-620"><a href="#cb58-620"></a><span class="co"># create CTV regions GFF3 files for one species representative each</span></span>
<span id="cb58-621"><a href="#cb58-621"></a><span class="ex">Rscript</span> scripts/utils/HGs_gff_from_regions.R <span class="dt">\</span></span>
<span id="cb58-622"><a href="#cb58-622"></a>--regions analysis/pangenome_v2/carotovoricin/tail_region/hg_regions.tab</span>
<span id="cb58-623"><a href="#cb58-623"></a><span class="in">```</span></span>
<span id="cb58-624"><a href="#cb58-624"></a></span>
<span id="cb58-625"><a href="#cb58-625"></a>Extract DNA sequence for these 3' tail fiber regions.</span>
<span id="cb58-626"><a href="#cb58-626"></a></span>
<span id="cb58-627"><a href="#cb58-627"></a><span class="in">```bash</span></span>
<span id="cb58-628"><a href="#cb58-628"></a><span class="fu">mkdir</span> analysis/pangenome_v2/prophages/cluster_viz/ctv_pbr/region_fasta</span>
<span id="cb58-629"><a href="#cb58-629"></a></span>
<span id="cb58-630"><a href="#cb58-630"></a><span class="fu">tail</span> <span class="at">-n</span> +2 analysis/pangenome_v2/prophages/cluster_viz/ctv_pbr/ctv_pbr.variable_regions.tab <span class="kw">|</span></span>
<span id="cb58-631"><a href="#cb58-631"></a>  <span class="fu">cut</span> <span class="at">-f</span> 1,3-5 <span class="kw">|</span></span>
<span id="cb58-632"><a href="#cb58-632"></a>  <span class="cf">while</span> <span class="bu">read</span> <span class="va">name</span> <span class="va">sampleId</span> <span class="va">pos</span> <span class="va">strand</span><span class="kw">;</span> <span class="cf">do</span></span>
<span id="cb58-633"><a href="#cb58-633"></a>    <span class="va">rc</span><span class="op">=</span><span class="st">""</span></span>
<span id="cb58-634"><a href="#cb58-634"></a>    <span class="cf">if</span> <span class="bu">[</span> <span class="va">${strand}</span> <span class="ot">==</span> <span class="st">"-"</span> <span class="bu">]</span><span class="kw">;</span> <span class="cf">then</span></span>
<span id="cb58-635"><a href="#cb58-635"></a>        <span class="va">rc</span><span class="op">=</span><span class="st">"--reverse-complement"</span></span>
<span id="cb58-636"><a href="#cb58-636"></a>    <span class="cf">fi</span></span>
<span id="cb58-637"><a href="#cb58-637"></a>    <span class="va">outFa</span><span class="op">=</span><span class="st">"analysis/pangenome_v2/prophages/cluster_viz/ctv_pbr/region_fasta/</span><span class="va">${name}</span><span class="st">.fasta"</span></span>
<span id="cb58-638"><a href="#cb58-638"></a></span>
<span id="cb58-639"><a href="#cb58-639"></a>    <span class="va">seq</span><span class="op">=</span><span class="st">"&gt;</span><span class="va">${name}</span><span class="st"> </span><span class="va">${pos}</span><span class="st">(</span><span class="va">${strand}</span><span class="st">)\n"</span></span>
<span id="cb58-640"><a href="#cb58-640"></a>    <span class="va">seq</span><span class="op">+=</span><span class="kw">`</span><span class="ex">samtools</span> faidx <span class="va">${rc}</span> data/prokka_annotation/<span class="va">${sampleId}</span>/<span class="va">${sampleId}</span>.fna <span class="va">$pos</span> <span class="kw">|</span> <span class="fu">tail</span> <span class="at">-n</span> +2<span class="kw">`</span></span>
<span id="cb58-641"><a href="#cb58-641"></a>    <span class="bu">printf</span> <span class="st">"</span><span class="va">${seq}</span><span class="st">\n"</span> <span class="op">&gt;</span> <span class="va">${outFa}</span></span>
<span id="cb58-642"><a href="#cb58-642"></a>    <span class="bu">printf</span> <span class="st">"</span><span class="va">${seq}</span><span class="st">\n"</span></span>
<span id="cb58-643"><a href="#cb58-643"></a>  <span class="cf">done</span> <span class="op">&gt;</span>analysis/pangenome_v2/prophages/cluster_viz/ctv_pbr/ctv_pbr.variable_regions.fasta</span>
<span id="cb58-644"><a href="#cb58-644"></a></span>
<span id="cb58-645"><a href="#cb58-645"></a><span class="kw">```</span></span>
<span id="cb58-646"><a href="#cb58-646"></a></span>
<span id="cb58-647"><a href="#cb58-647"></a><span class="ex">Mauve</span> aligns sequences provided in <span class="kw">`</span>fasta<span class="kw">`</span> <span class="ex">and</span> <span class="kw">`</span>genbank<span class="kw">`</span> <span class="ex">format.</span> However, the</span>
<span id="cb58-648"><a href="#cb58-648"></a><span class="ex">sequence</span> annotation is shown only if the alignment input was in <span class="kw">`</span>genbank<span class="kw">`</span></span>
<span id="cb58-649"><a href="#cb58-649"></a><span class="ex">format.</span> Therefore, we need to combine the homology group annotation in <span class="kw">`</span>gff3<span class="kw">`</span></span>
<span id="cb58-650"><a href="#cb58-650"></a><span class="ex">format</span> with the <span class="kw">`</span>fasta<span class="kw">`</span> <span class="ex">sequences</span> to generate a <span class="kw">`</span>genbank<span class="kw">`</span> <span class="ex">formatted</span> files. To</span>
<span id="cb58-651"><a href="#cb58-651"></a><span class="cf">do</span> <span class="ex">this,</span> Emboss tool <span class="kw">`</span>seqret<span class="kw">`</span> <span class="ex">was</span> used.</span>
<span id="cb58-652"><a href="#cb58-652"></a></span>
<span id="cb58-653"><a href="#cb58-653"></a><span class="kw">```</span>bash</span>
<span id="cb58-654"><a href="#cb58-654"></a><span class="ex">conda</span> activate omics_py37</span>
<span id="cb58-655"><a href="#cb58-655"></a></span>
<span id="cb58-656"><a href="#cb58-656"></a><span class="bu">cd</span> analysis/pangenome_v2/prophages/cluster_viz/ctv_pbr/region_fasta</span>
<span id="cb58-657"><a href="#cb58-657"></a></span>
<span id="cb58-658"><a href="#cb58-658"></a><span class="cf">for</span> vir <span class="kw">in</span> <span class="kw">`</span><span class="fu">cut</span> <span class="at">-f</span> 1 ../ctv_pbr.variable_regions.tab <span class="kw">|</span> <span class="fu">tail</span> <span class="at">-n</span> +2<span class="kw">`</span></span>
<span id="cb58-659"><a href="#cb58-659"></a><span class="cf">do</span></span>
<span id="cb58-660"><a href="#cb58-660"></a>  <span class="ex">seqret</span> <span class="at">-sequence</span> <span class="va">${vir}</span>.fasta <span class="at">-feature</span> <span class="at">-fformat</span> gff3 <span class="dt">\</span></span>
<span id="cb58-661"><a href="#cb58-661"></a>  <span class="at">-fopenfile</span> <span class="va">${vir}</span>.gff3 <span class="at">-osformat</span> genbank <span class="at">-osextension</span> gbk <span class="dt">\</span></span>
<span id="cb58-662"><a href="#cb58-662"></a>  <span class="at">-osname_outseq</span> <span class="va">${vir}</span> <span class="at">-ofdirectory_outseq</span> gbk_file <span class="at">-auto</span> </span>
<span id="cb58-663"><a href="#cb58-663"></a><span class="cf">done</span></span>
<span id="cb58-664"><a href="#cb58-664"></a></span>
<span id="cb58-665"><a href="#cb58-665"></a><span class="fu">sed</span> <span class="at">-i.bak</span> <span class="at">-r</span> <span class="at">-e</span> <span class="st">'s/(^\s+CDS\s+(complement\()?)[^:]+:([[:digit:]]+\.\.[[:digit:]]+\)?).*/\1\3/'</span> <span class="dt">\</span></span>
<span id="cb58-666"><a href="#cb58-666"></a>  <span class="at">-e</span> <span class="st">'s/\/note="\*([^:]+): /\/\1="/'</span> <span class="pp">*</span>.gbk </span>
<span id="cb58-667"><a href="#cb58-667"></a></span>
<span id="cb58-668"><a href="#cb58-668"></a><span class="kw">```</span></span>
<span id="cb58-669"><a href="#cb58-669"></a></span>
<span id="cb58-670"><a href="#cb58-670"></a><span class="ex">Align</span> and visualize sequences using Mauve.</span>
<span id="cb58-671"><a href="#cb58-671"></a></span>
<span id="cb58-672"><a href="#cb58-672"></a><span class="kw">```</span>bash</span>
<span id="cb58-673"><a href="#cb58-673"></a><span class="va">mauve_out</span><span class="op">=</span><span class="st">"ctv_pbr"</span></span>
<span id="cb58-674"><a href="#cb58-674"></a></span>
<span id="cb58-675"><a href="#cb58-675"></a><span class="ex">/Applications/Mauve.app/Contents/MacOS/progressiveMauve</span> <span class="dt">\</span></span>
<span id="cb58-676"><a href="#cb58-676"></a>--output=<span class="st">"</span><span class="va">${mauve_out}</span><span class="st">.mauve.xmfa"</span> <span class="dt">\</span></span>
<span id="cb58-677"><a href="#cb58-677"></a>--backbone-output=<span class="st">"</span><span class="va">${mauve_out}</span><span class="st">.mauve.backbone"</span> <span class="dt">\</span></span>
<span id="cb58-678"><a href="#cb58-678"></a>--output-guide-tree=<span class="st">"</span><span class="va">${mauve_out}</span><span class="st">.mauve.guide_tree.newick"</span> <span class="dt">\</span></span>
<span id="cb58-679"><a href="#cb58-679"></a>region_fasta/g_302.vir_3.gbk region_fasta/g_364.vir_2.gbk <span class="dt">\</span></span>
<span id="cb58-680"><a href="#cb58-680"></a>region_fasta/g_337.vir_1.gbk region_fasta/g_439.vir_2.gbk <span class="dt">\</span></span>
<span id="cb58-681"><a href="#cb58-681"></a>region_fasta/g_403.vir_3.gbk region_fasta/g_399.vir_3.gbk <span class="dt">\</span></span>
<span id="cb58-682"><a href="#cb58-682"></a>region_fasta/g_175.vir_3.gbk region_fasta/g_408.vir_3.gbk <span class="dt">\</span></span>
<span id="cb58-683"><a href="#cb58-683"></a>region_fasta/g_138.vir_2.gbk <span class="dt">\</span></span>
<span id="cb58-684"><a href="#cb58-684"></a>region_fasta/g_368.vir_3.gbk region_fasta/g_345.vir_1.gbk <span class="dt">\</span></span>
<span id="cb58-685"><a href="#cb58-685"></a>region_fasta/g_366.vir_3.gbk region_fasta/g_308.vir_2.gbk <span class="dt">\</span></span>
<span id="cb58-686"><a href="#cb58-686"></a>region_fasta/g_191.vir_1.gbk region_fasta/g_173.vir_2.gbk <span class="dt">\</span></span>
<span id="cb58-687"><a href="#cb58-687"></a>region_fasta/g_155.vir_1.gbk region_fasta/g_166.vir_2.gbk <span class="dt">\</span></span>
<span id="cb58-688"><a href="#cb58-688"></a>region_fasta/g_299.vir_1.gbk region_fasta/g_438.vir_2.gbk <span class="dt">\</span></span>
<span id="cb58-689"><a href="#cb58-689"></a>region_fasta/g_263.vir_2.gbk region_fasta/g_43.vir_1.gbk <span class="dt">\</span></span>
<span id="cb58-690"><a href="#cb58-690"></a>region_fasta/g_391.vir_3.gbk  <span class="op">&gt;</span> <span class="st">"</span><span class="va">${mauve_out}</span><span class="st">"</span>.log <span class="dv">2</span><span class="op">&gt;&amp;</span><span class="dv">1</span></span>
<span id="cb58-691"><a href="#cb58-691"></a></span>
<span id="cb58-692"><a href="#cb58-692"></a><span class="va">mauve_out</span><span class="op">=</span><span class="st">"ctv_pbr.ein"</span></span>
<span id="cb58-693"><a href="#cb58-693"></a></span>
<span id="cb58-694"><a href="#cb58-694"></a><span class="ex">/Applications/Mauve.app/Contents/MacOS/progressiveMauve</span> <span class="dt">\</span></span>
<span id="cb58-695"><a href="#cb58-695"></a>--output=<span class="st">"</span><span class="va">${mauve_out}</span><span class="st">.mauve.xmfa"</span> <span class="dt">\</span></span>
<span id="cb58-696"><a href="#cb58-696"></a>--backbone-output=<span class="st">"</span><span class="va">${mauve_out}</span><span class="st">.mauve.backbone"</span> <span class="dt">\</span></span>
<span id="cb58-697"><a href="#cb58-697"></a>--output-guide-tree=<span class="st">"</span><span class="va">${mauve_out}</span><span class="st">.mauve.guide_tree.newick"</span> <span class="dt">\</span></span>
<span id="cb58-698"><a href="#cb58-698"></a>region_fasta/g_345.vir_1.gbk region_fasta/g_366.vir_3.gbk <span class="dt">\</span></span>
<span id="cb58-699"><a href="#cb58-699"></a>region_fasta/g_308.vir_2.gbk region_fasta/g_191.vir_1.gbk <span class="dt">\</span></span>
<span id="cb58-700"><a href="#cb58-700"></a>region_fasta/g_173.vir_2.gbk region_fasta/g_155.vir_1.gbk <span class="dt">\</span></span>
<span id="cb58-701"><a href="#cb58-701"></a>region_fasta/g_166.vir_2.gbk  <span class="op">&gt;</span> <span class="st">"</span><span class="va">${mauve_out}</span><span class="st">"</span>.log <span class="dv">2</span><span class="op">&gt;&amp;</span><span class="dv">1</span></span>
<span id="cb58-702"><a href="#cb58-702"></a></span>
<span id="cb58-703"><a href="#cb58-703"></a></span>
<span id="cb58-704"><a href="#cb58-704"></a><span class="va">mauve_out</span><span class="op">=</span><span class="st">"ctv_pbr.rep"</span></span>
<span id="cb58-705"><a href="#cb58-705"></a></span>
<span id="cb58-706"><a href="#cb58-706"></a><span class="ex">/Applications/Mauve.app/Contents/MacOS/progressiveMauve</span> <span class="dt">\</span></span>
<span id="cb58-707"><a href="#cb58-707"></a>--output=<span class="st">"</span><span class="va">${mauve_out}</span><span class="st">.mauve.xmfa"</span> <span class="dt">\</span></span>
<span id="cb58-708"><a href="#cb58-708"></a>--backbone-output=<span class="st">"</span><span class="va">${mauve_out}</span><span class="st">.mauve.backbone"</span> <span class="dt">\</span></span>
<span id="cb58-709"><a href="#cb58-709"></a>--output-guide-tree=<span class="st">"</span><span class="va">${mauve_out}</span><span class="st">.mauve.guide_tree.newick"</span> <span class="dt">\</span></span>
<span id="cb58-710"><a href="#cb58-710"></a>region_fasta/g_302.vir_3.gbk region_fasta/g_337.vir_1.gbk <span class="dt">\</span></span>
<span id="cb58-711"><a href="#cb58-711"></a>region_fasta/g_175.vir_3.gbk region_fasta/g_173.vir_2.gbk <span class="dt">\</span></span>
<span id="cb58-712"><a href="#cb58-712"></a>region_fasta/g_299.vir_1.gbk region_fasta/g_438.vir_2.gbk <span class="dt">\</span></span>
<span id="cb58-713"><a href="#cb58-713"></a>region_fasta/g_391.vir_3.gbk  <span class="op">&gt;</span> <span class="st">"</span><span class="va">${mauve_out}</span><span class="st">"</span>.log <span class="dv">2</span><span class="op">&gt;&amp;</span><span class="dv">1</span></span>
<span id="cb58-714"><a href="#cb58-714"></a></span>
<span id="cb58-715"><a href="#cb58-715"></a><span class="kw">```</span></span>
<span id="cb58-716"><a href="#cb58-716"></a></span>
<span id="cb58-717"><a href="#cb58-717"></a><span class="ex">Perform</span> MSA using <span class="kw">`</span>MAFFT<span class="kw">`</span></span>
<span id="cb58-718"><a href="#cb58-718"></a></span>
<span id="cb58-719"><a href="#cb58-719"></a><span class="kw">```</span>bash</span>
<span id="cb58-720"><a href="#cb58-720"></a><span class="ex">conda</span> activate pantools_v4_3</span>
<span id="cb58-721"><a href="#cb58-721"></a><span class="bu">cd</span> analysis/pangenome_v2/prophages/cluster_viz/ctv_pbr/</span>
<span id="cb58-722"><a href="#cb58-722"></a></span>
<span id="cb58-723"><a href="#cb58-723"></a><span class="ex">mafft</span> <span class="at">--globalpair</span> <span class="at">--quiet</span> <span class="at">--maxiterate</span> 1000 <span class="at">--treeout</span> ctv_pbr.variable_regions.fasta</span>
<span id="cb58-724"><a href="#cb58-724"></a><span class="kw">```</span></span>
<span id="cb58-725"><a href="#cb58-725"></a></span>
<span id="cb58-726"><a href="#cb58-726"></a><span class="ex">Smash++</span> pairwise sequence comparison</span>
<span id="cb58-727"><a href="#cb58-727"></a></span>
<span id="cb58-728"><a href="#cb58-728"></a><span class="kw">```</span>bash</span>
<span id="cb58-729"><a href="#cb58-729"></a><span class="ex">conda</span> activate omics_py37</span>
<span id="cb58-730"><a href="#cb58-730"></a></span>
<span id="cb58-731"><a href="#cb58-731"></a><span class="bu">cd</span> analysis/pangenome_v2/prophages/cluster_viz/ctv_pbr/region_fasta</span>
<span id="cb58-732"><a href="#cb58-732"></a></span>
<span id="cb58-733"><a href="#cb58-733"></a><span class="kw">function</span><span class="fu"> smashpp_compare</span></span>
<span id="cb58-734"><a href="#cb58-734"></a><span class="kw">{</span></span>
<span id="cb58-735"><a href="#cb58-735"></a>  <span class="ex">smashpp</span> <span class="at">-r</span> <span class="va">$1</span> <span class="at">-t</span> <span class="va">$2</span></span>
<span id="cb58-736"><a href="#cb58-736"></a>  <span class="ex">smashpp</span> viz <span class="at">-o</span> <span class="va">$(</span><span class="fu">basename</span> <span class="va">$1</span> .fasta<span class="va">)$(</span><span class="fu">basename</span> <span class="va">$2</span> .fasta<span class="va">)</span><span class="st">".svg"</span> <span class="st">"</span><span class="va">${1}</span><span class="st">.</span><span class="va">${2}</span><span class="st">.pos"</span></span>
<span id="cb58-737"><a href="#cb58-737"></a><span class="kw">}</span></span>
<span id="cb58-738"><a href="#cb58-738"></a></span>
<span id="cb58-739"><a href="#cb58-739"></a><span class="bu">export</span> <span class="at">-f</span> <span class="va">smashpp_compare</span></span>
<span id="cb58-740"><a href="#cb58-740"></a></span>
<span id="cb58-741"><a href="#cb58-741"></a><span class="ex">smashpp_compare</span> g_345.vir_1.fasta g_345.vir_1.fasta </span>
<span id="cb58-742"><a href="#cb58-742"></a><span class="ex">smashpp_compare</span> g_345.vir_1.fasta g_366.vir_3.fasta </span>
<span id="cb58-743"><a href="#cb58-743"></a><span class="ex">smashpp_compare</span> g_345.vir_1.fasta g_302.vir_3.fasta</span>
<span id="cb58-744"><a href="#cb58-744"></a><span class="ex">smashpp_compare</span> g_345.vir_1.fasta g_403.vir_3.fasta</span>
<span id="cb58-745"><a href="#cb58-745"></a><span class="ex">smashpp_compare</span> g_345.vir_1.fasta g_173.vir_2.fasta</span>
<span id="cb58-746"><a href="#cb58-746"></a><span class="ex">smashpp_compare</span> g_345.vir_1.fasta g_138.vir_2.fasta</span>
<span id="cb58-747"><a href="#cb58-747"></a><span class="ex">smashpp_compare</span> g_345.vir_1.fasta g_263.vir_2.fasta</span>
<span id="cb58-748"><a href="#cb58-748"></a><span class="ex">smashpp_compare</span> g_345.vir_1.fasta g_155.vir_1.fasta</span>
<span id="cb58-749"><a href="#cb58-749"></a></span>
<span id="cb58-750"><a href="#cb58-750"></a><span class="kw">```</span></span>
<span id="cb58-751"><a href="#cb58-751"></a></span>
<span id="cb58-752"><a href="#cb58-752"></a><span class="ex">Left</span> inverted repeat for Ein: CTCCCGCAAACCTCGGTTTTGGGGAC <span class="er">(</span><span class="ex">CTCCCGCAAACCTCGGTTT</span><span class="kw">)</span>  </span>
<span id="cb58-753"><a href="#cb58-753"></a><span class="ex">Left</span> inverted repeat for Ein<span class="er">(</span><span class="ex">rev-com</span><span class="kw">)</span><span class="bu">:</span> GTCCCCAAAACCGAGGTTTGCGGGAG <span class="er">(</span><span class="ex">AAACCGAGGTTTGCGGGAG,</span> AAACCGAGGTTTGCG<span class="kw">)</span>  </span>
<span id="cb58-754"><a href="#cb58-754"></a></span>
<span id="cb58-755"><a href="#cb58-755"></a><span class="ex">Right</span> inverted repeat for Ein: TTCTCGCAAACCTCGGTTTTGGAGAA  </span>
<span id="cb58-756"><a href="#cb58-756"></a><span class="ex">Right</span> inverted repeat for Ein<span class="er">(</span><span class="fu">rev</span><span class="kw">)</span><span class="bu">:</span> AAGAGGTTTTGGCTCCAAACGCTCTT  </span>
<span id="cb58-757"><a href="#cb58-757"></a><span class="ex">Right</span> inverted repeat for Ein<span class="er">(</span><span class="ex">rev-com</span><span class="kw">)</span><span class="bu">:</span> TTCTCCAAAACCGAGGTTTGCGAGAA <span class="er">(</span><span class="ex">AAACCGAGGTTTGCGAGAA,</span> AAACCGAGGTTTGCG<span class="kw">)</span>  </span>
<span id="cb58-758"><a href="#cb58-758"></a><span class="ex">Right</span> inverted repeat for Ein<span class="er">(</span><span class="ex">comp</span><span class="kw">)</span><span class="bu">:</span> AAGAGCGTTTGGAGCCAAAACCTCTT  </span>
<span id="cb58-759"><a href="#cb58-759"></a></span>
<span id="cb58-760"><a href="#cb58-760"></a><span class="co">## Inter-species horizontal gene transfer of CTV</span></span>
<span id="cb58-761"><a href="#cb58-761"></a></span>
<span id="cb58-762"><a href="#cb58-762"></a><span class="ex">Clustermap</span> visualization for some manually selected candidates. Change the code</span>
<span id="cb58-763"><a href="#cb58-763"></a><span class="ex">blocks</span> in the script <span class="kw">`</span>scripts/analysis/clustersmap_data_prophages.R<span class="kw">`</span> <span class="ex">and</span> run.</span>
<span id="cb58-764"><a href="#cb58-764"></a></span>
<span id="cb58-765"><a href="#cb58-765"></a><span class="kw">```</span>r</span>
<span id="cb58-766"><a href="#cb58-766"></a><span class="ex">grpToView</span> <span class="op">&lt;</span>- <span class="st">"ctv_hgt"</span></span>
<span id="cb58-767"><a href="#cb58-767"></a><span class="ex">subSample</span> <span class="op">&lt;</span>- FALSE </span>
<span id="cb58-768"><a href="#cb58-768"></a><span class="ex">cutHeight</span> <span class="op">&lt;</span>- 1.5</span>
<span id="cb58-769"><a href="#cb58-769"></a><span class="ex">addFlankingRegions</span> <span class="op">&lt;</span>- TRUE</span>
<span id="cb58-770"><a href="#cb58-770"></a><span class="ex">flankingRegion</span> <span class="op">&lt;</span>- 5000</span>
<span id="cb58-771"><a href="#cb58-771"></a></span>
<span id="cb58-772"><a href="#cb58-772"></a><span class="co"># ordering factor for prophages: host phylogeny, prophage HG PAV, prophage MASH,</span></span>
<span id="cb58-773"><a href="#cb58-773"></a><span class="co"># completeness score</span></span>
<span id="cb58-774"><a href="#cb58-774"></a><span class="ex">clusterOrder</span> <span class="op">&lt;</span>- <span class="st">"hg_pav"</span> <span class="co"># completeness, host, hg_pav, cluster_mash</span></span>
<span id="cb58-775"><a href="#cb58-775"></a></span>
<span id="cb58-776"><a href="#cb58-776"></a></span>
<span id="cb58-777"><a href="#cb58-777"></a><span class="co"># optionally, a custom region list can be provided to generate the plot</span></span>
<span id="cb58-778"><a href="#cb58-778"></a><span class="ex">grp</span> <span class="op">&lt;</span>- list<span class="er">(</span></span>
<span id="cb58-779"><a href="#cb58-779"></a>  <span class="ex">phage_grp</span> = grpToView,</span>
<span id="cb58-780"><a href="#cb58-780"></a>  <span class="ex">members</span> = c<span class="er">(</span></span>
<span id="cb58-781"><a href="#cb58-781"></a>    <span class="st">"g_145.vir_1"</span><span class="ex">,</span> <span class="st">"g_194.vir_1"</span>, <span class="st">"g_429.vir_1"</span>, <span class="st">"g_442.vir_1"</span>, <span class="st">"g_421.vir_1"</span>, </span>
<span id="cb58-782"><a href="#cb58-782"></a>    <span class="st">"g_150.vir_4"</span><span class="ex">,</span> <span class="st">"g_447.vir_1"</span>, <span class="st">"g_434.vir_4"</span>,</span>
<span id="cb58-783"><a href="#cb58-783"></a>    <span class="st">"g_221.vir_3"</span><span class="ex">,</span> <span class="st">"g_53.vir_3"</span>, <span class="st">"g_106.vir_2"</span>, <span class="st">"g_57.vir_1"</span>, <span class="st">"g_125.vir_1"</span></span>
<span id="cb58-784"><a href="#cb58-784"></a>  <span class="kw">)</span></span>
<span id="cb58-785"><a href="#cb58-785"></a><span class="kw">)</span></span>
<span id="cb58-786"><a href="#cb58-786"></a><span class="kw">```</span></span>
<span id="cb58-787"><a href="#cb58-787"></a></span>
<span id="cb58-788"><a href="#cb58-788"></a><span class="ex">Run</span> script <span class="kw">`</span>scripts/utils/HG_range_coordinates.R<span class="kw">`</span> <span class="ex">to</span> extract genomic coordinates</span>
<span id="cb58-789"><a href="#cb58-789"></a><span class="cf">for</span> CTV <span class="ex">region</span> in selected genomes.</span>
<span id="cb58-790"><a href="#cb58-790"></a></span>
<span id="cb58-791"><a href="#cb58-791"></a><span class="ex">Complete</span> CTV region along with 3 flanking genes:</span>
<span id="cb58-792"><a href="#cb58-792"></a></span>
<span id="cb58-793"><a href="#cb58-793"></a><span class="kw">```</span>bash</span>
<span id="cb58-794"><a href="#cb58-794"></a><span class="ex">Rscript</span> scripts/utils/HG_range_coordinates.R <span class="at">--hgs</span> hg_22427643,hg_22427599 <span class="dt">\</span></span>
<span id="cb58-795"><a href="#cb58-795"></a>--inner_region <span class="at">--dir</span> analysis/pangenome_v2/carotovoricin/ctv_region</span>
<span id="cb58-796"><a href="#cb58-796"></a></span>
<span id="cb58-797"><a href="#cb58-797"></a><span class="co"># create GFF3 files for CTV regions of interest</span></span>
<span id="cb58-798"><a href="#cb58-798"></a><span class="ex">Rscript</span> scripts/utils/HGs_gff_from_regions.R <span class="dt">\</span></span>
<span id="cb58-799"><a href="#cb58-799"></a>--regions analysis/pangenome_v2/carotovoricin/ctv_region/hg_regions.tab <span class="dt">\</span></span>
<span id="cb58-800"><a href="#cb58-800"></a>--genomes g_145,g_194,g_429,g_442,g_421,g_150,g_447,g_434,g_53,g_106,g_57,g_221,g_125</span>
<span id="cb58-801"><a href="#cb58-801"></a></span>
<span id="cb58-802"><a href="#cb58-802"></a><span class="co"># create CTV regions GFF3 files for one species representative each</span></span>
<span id="cb58-803"><a href="#cb58-803"></a><span class="ex">Rscript</span> scripts/utils/HGs_gff_from_regions.R <span class="dt">\</span></span>
<span id="cb58-804"><a href="#cb58-804"></a>--regions analysis/pangenome_v2/carotovoricin/ctv_region/hg_regions.tab <span class="dt">\</span></span>
<span id="cb58-805"><a href="#cb58-805"></a>--genomes g_183,g_423,g_277,g_375,g_116,g_449,g_446,g_378,g_296,g_428,g_259,g_66,g_337,g_442,g_305,g_381,g_8,g_273,g_38</span>
<span id="cb58-806"><a href="#cb58-806"></a><span class="kw">```</span></span>
<span id="cb58-807"><a href="#cb58-807"></a></span>
<span id="cb58-808"><a href="#cb58-808"></a><span class="ex">CTV</span> conserved loci:</span>
<span id="cb58-809"><a href="#cb58-809"></a></span>
<span id="cb58-810"><a href="#cb58-810"></a><span class="kw">```</span>bash</span>
<span id="cb58-811"><a href="#cb58-811"></a><span class="ex">Rscript</span> scripts/utils/HG_range_coordinates.R <span class="at">--hgs</span> hg_22427640,hg_22427604 <span class="dt">\</span></span>
<span id="cb58-812"><a href="#cb58-812"></a>--dir analysis/pangenome_v2/carotovoricin/ctv_conserved</span>
<span id="cb58-813"><a href="#cb58-813"></a><span class="kw">```</span></span>
<span id="cb58-814"><a href="#cb58-814"></a></span>
<span id="cb58-815"><a href="#cb58-815"></a><span class="ex">CTV</span> variable loci:</span>
<span id="cb58-816"><a href="#cb58-816"></a></span>
<span id="cb58-817"><a href="#cb58-817"></a><span class="kw">```</span>bash</span>
<span id="cb58-818"><a href="#cb58-818"></a><span class="ex">Rscript</span> scripts/utils/HG_range_coordinates.R <span class="at">--hgs</span> hg_22427604,hg_22427603 <span class="dt">\</span></span>
<span id="cb58-819"><a href="#cb58-819"></a>--inner_region <span class="at">--dir</span> analysis/pangenome_v2/carotovoricin/ctv_tail <span class="at">--haplotypes</span></span>
<span id="cb58-820"><a href="#cb58-820"></a><span class="kw">```</span></span>
<span id="cb58-821"><a href="#cb58-821"></a></span>
<span id="cb58-822"><a href="#cb58-822"></a><span class="ex">CTV</span> tape measure protein gene:</span>
<span id="cb58-823"><a href="#cb58-823"></a></span>
<span id="cb58-824"><a href="#cb58-824"></a><span class="kw">```</span>bash</span>
<span id="cb58-825"><a href="#cb58-825"></a><span class="ex">Rscript</span> scripts/utils/HG_range_coordinates.R <span class="at">--hgs</span> hg_22427622,hg_22427616 <span class="dt">\</span></span>
<span id="cb58-826"><a href="#cb58-826"></a>--haplotypes <span class="at">--inner_region</span> <span class="at">--overlapping</span> <span class="dt">\</span></span>
<span id="cb58-827"><a href="#cb58-827"></a>--dir analysis/pangenome_v2/carotovoricin/tape_measure</span>
<span id="cb58-828"><a href="#cb58-828"></a><span class="kw">```</span></span>
<span id="cb58-829"><a href="#cb58-829"></a></span>
<span id="cb58-830"><a href="#cb58-830"></a><span class="ex">Run</span> <span class="kw">`</span>scripts/analysis/ctv_hgt.qmd<span class="kw">`</span> <span class="ex">script</span> to perform HGT analysis.</span>
<span id="cb58-831"><a href="#cb58-831"></a></span>
<span id="cb58-832"><a href="#cb58-832"></a><span class="co">### DNA sequence comparison</span></span>
<span id="cb58-833"><a href="#cb58-833"></a></span>
<span id="cb58-834"><a href="#cb58-834"></a><span class="co">#### Extract DNA sequence for the regions</span></span>
<span id="cb58-835"><a href="#cb58-835"></a></span>
<span id="cb58-836"><a href="#cb58-836"></a><span class="kw">```</span>bash</span>
<span id="cb58-837"><a href="#cb58-837"></a><span class="co"># file_regions="analysis/pangenome_v2/carotovoricin/ctv_tail/hg_regions.tab"</span></span>
<span id="cb58-838"><a href="#cb58-838"></a><span class="co"># file_regions="analysis/pangenome_v2/carotovoricin/ctv_conserved/hg_regions.tab"</span></span>
<span id="cb58-839"><a href="#cb58-839"></a><span class="co"># file_regions="analysis/pangenome_v2/carotovoricin/ctv_tail/selected_haplotypes.tab"</span></span>
<span id="cb58-840"><a href="#cb58-840"></a><span class="co"># file_regions="analysis/pangenome_v2/carotovoricin/ctv_tail/selected_haplotypes_flanking.tab"</span></span>
<span id="cb58-841"><a href="#cb58-841"></a><span class="co"># file_regions="analysis/pangenome_v2/carotovoricin/ctv_conserved/selected_conserved.tab"</span></span>
<span id="cb58-842"><a href="#cb58-842"></a><span class="co"># file_regions="analysis/pangenome_v2/carotovoricin/upstream_core/selected_upstream.tab"</span></span>
<span id="cb58-843"><a href="#cb58-843"></a></span>
<span id="cb58-844"><a href="#cb58-844"></a><span class="va">dir_path</span><span class="op">=</span><span class="va">$(</span><span class="fu">dirname</span> <span class="st">"</span><span class="va">${file_regions}</span><span class="st">"</span><span class="va">)</span></span>
<span id="cb58-845"><a href="#cb58-845"></a><span class="va">out_fasta</span><span class="op">=</span><span class="st">"</span><span class="va">${file_regions</span><span class="op">%</span>.<span class="pp">*</span><span class="va">}</span><span class="st">"</span>.fasta</span>
<span id="cb58-846"><a href="#cb58-846"></a></span>
<span id="cb58-847"><a href="#cb58-847"></a><span class="fu">tail</span> <span class="at">-n</span> +2 <span class="va">${file_regions}</span> <span class="kw">|</span></span>
<span id="cb58-848"><a href="#cb58-848"></a>  <span class="fu">cut</span> <span class="at">-f</span> 1,3-5 <span class="kw">|</span></span>
<span id="cb58-849"><a href="#cb58-849"></a>  <span class="cf">while</span> <span class="bu">read</span> <span class="va">name</span> <span class="va">sampleId</span> <span class="va">pos</span> <span class="va">strand</span><span class="kw">;</span> <span class="cf">do</span></span>
<span id="cb58-850"><a href="#cb58-850"></a>    <span class="va">rc</span><span class="op">=</span><span class="st">""</span></span>
<span id="cb58-851"><a href="#cb58-851"></a>    <span class="cf">if</span> <span class="bu">[</span> <span class="va">${strand}</span> <span class="ot">==</span> <span class="st">"-"</span> <span class="bu">]</span><span class="kw">;</span> <span class="cf">then</span></span>
<span id="cb58-852"><a href="#cb58-852"></a>        <span class="va">rc</span><span class="op">=</span><span class="st">"--reverse-complement"</span></span>
<span id="cb58-853"><a href="#cb58-853"></a>    <span class="cf">fi</span></span>
<span id="cb58-854"><a href="#cb58-854"></a>    <span class="va">outFa</span><span class="op">=</span><span class="st">"</span><span class="va">${dir_path}</span><span class="st">/fasta/</span><span class="va">${name}</span><span class="st">.fasta"</span></span>
<span id="cb58-855"><a href="#cb58-855"></a></span>
<span id="cb58-856"><a href="#cb58-856"></a>    <span class="va">seq</span><span class="op">=</span><span class="st">"&gt;</span><span class="va">${name}</span><span class="st">\n"</span></span>
<span id="cb58-857"><a href="#cb58-857"></a>    <span class="va">seq</span><span class="op">+=</span><span class="kw">`</span><span class="ex">samtools</span> faidx <span class="va">${rc}</span> data/prokka_annotation/<span class="va">${sampleId}</span>/<span class="va">${sampleId}</span>.fna <span class="va">$pos</span> <span class="kw">|</span> <span class="fu">tail</span> <span class="at">-n</span> +2<span class="kw">`</span></span>
<span id="cb58-858"><a href="#cb58-858"></a></span>
<span id="cb58-859"><a href="#cb58-859"></a>    <span class="co"># printf "${seq}\n" &gt; ${outFa}</span></span>
<span id="cb58-860"><a href="#cb58-860"></a>    <span class="bu">printf</span> <span class="st">"</span><span class="va">${seq}</span><span class="st">\n"</span></span>
<span id="cb58-861"><a href="#cb58-861"></a>  <span class="cf">done</span> <span class="op">&gt;</span> <span class="va">${out_fasta}</span></span>
<span id="cb58-862"><a href="#cb58-862"></a></span>
<span id="cb58-863"><a href="#cb58-863"></a><span class="kw">```</span></span>
<span id="cb58-864"><a href="#cb58-864"></a></span>
<span id="cb58-865"><a href="#cb58-865"></a><span class="co">#### Calculate the distance using </span><span class="kw">`</span>mash<span class="kw">`</span></span>
<span id="cb58-866"><a href="#cb58-866"></a></span>
<span id="cb58-867"><a href="#cb58-867"></a><span class="kw">```</span>bash</span>
<span id="cb58-868"><a href="#cb58-868"></a><span class="bu">cd</span> analysis/pangenome_v2/carotovoricin/ctv_tail</span>
<span id="cb58-869"><a href="#cb58-869"></a><span class="ex">mash</span> dist <span class="at">-p</span> 8 <span class="at">-k</span> 12 <span class="at">-s</span> 2000 <span class="at">-i</span> <span class="at">-S</span> 124 hg_regions.fasta hg_regions.fasta <span class="op">&gt;</span> mash_dist.tab</span>
<span id="cb58-870"><a href="#cb58-870"></a></span>
<span id="cb58-871"><a href="#cb58-871"></a><span class="bu">cd</span> analysis/pangenome_v2/carotovoricin/ctv_conserved</span>
<span id="cb58-872"><a href="#cb58-872"></a><span class="ex">mash</span> dist <span class="at">-p</span> 8 <span class="at">-k</span> 12 <span class="at">-s</span> 2000 <span class="at">-i</span> <span class="at">-S</span> 124 hg_regions.fasta hg_regions.fasta <span class="op">&gt;</span> mash_dist.tab</span>
<span id="cb58-873"><a href="#cb58-873"></a></span>
<span id="cb58-874"><a href="#cb58-874"></a><span class="bu">cd</span> analysis/pangenome_v2/carotovoricin/tape_measure</span>
<span id="cb58-875"><a href="#cb58-875"></a><span class="ex">mash</span> dist <span class="at">-p</span> 8 <span class="at">-k</span> 12 <span class="at">-s</span> 2000 <span class="at">-i</span> <span class="at">-S</span> 124 hg_regions.fasta hg_regions.fasta <span class="op">&gt;</span> mash_dist.tab</span>
<span id="cb58-876"><a href="#cb58-876"></a></span>
<span id="cb58-877"><a href="#cb58-877"></a><span class="fu">grep</span> <span class="at">-e</span> <span class="st">'g_(53|106|57|221|125).*g_(53|106|57|221|125)'</span> ctv_<span class="pp">*</span>/ctv_dist.tab</span>
<span id="cb58-878"><a href="#cb58-878"></a><span class="fu">grep</span> <span class="at">-e</span> <span class="st">'g_(150|447|434).*g_(150|447|434)'</span> ctv_<span class="pp">*</span>/ctv_dist.tab</span>
<span id="cb58-879"><a href="#cb58-879"></a><span class="fu">grep</span> <span class="at">-e</span> <span class="st">'g_(53|106|57).*g_(53|106|57)'</span> ctv_<span class="pp">*</span>/ctv_dist.tab</span>
<span id="cb58-880"><a href="#cb58-880"></a></span>
<span id="cb58-881"><a href="#cb58-881"></a><span class="kw">```</span></span>
<span id="cb58-882"><a href="#cb58-882"></a></span>
<span id="cb58-883"><a href="#cb58-883"></a><span class="co">#### Calculate Jaccard distance matrix using </span><span class="kw">`</span>sourmash<span class="kw">`</span></span>
<span id="cb58-884"><a href="#cb58-884"></a></span>
<span id="cb58-885"><a href="#cb58-885"></a><span class="ex">MASH</span> distance is low for haplotype pairs where one is subset of the other, eg,</span>
<span id="cb58-886"><a href="#cb58-886"></a><span class="ex">A-B-C-D</span> and A-B. Therefore, another tool was tried to calculate the Jaccard</span>
<span id="cb58-887"><a href="#cb58-887"></a><span class="ex">index</span> between sequences.</span>
<span id="cb58-888"><a href="#cb58-888"></a></span>
<span id="cb58-889"><a href="#cb58-889"></a><span class="kw">```</span>bash</span>
<span id="cb58-890"><a href="#cb58-890"></a><span class="co"># file_fa="ctv_conserved/hg_regions.fasta"</span></span>
<span id="cb58-891"><a href="#cb58-891"></a><span class="co"># file_fa="ctv_tail/hg_regions.fasta"</span></span>
<span id="cb58-892"><a href="#cb58-892"></a></span>
<span id="cb58-893"><a href="#cb58-893"></a><span class="va">file_sig</span><span class="op">=</span><span class="st">"</span><span class="va">${file_fa</span><span class="op">%</span>.<span class="pp">*</span><span class="va">}</span><span class="st">"</span>.sig.zip</span>
<span id="cb58-894"><a href="#cb58-894"></a><span class="va">file_dist</span><span class="op">=</span><span class="st">"</span><span class="va">${file_fa</span><span class="op">%</span>.<span class="pp">*</span><span class="va">}</span><span class="st">"</span>.sourmash.csv</span>
<span id="cb58-895"><a href="#cb58-895"></a></span>
<span id="cb58-896"><a href="#cb58-896"></a><span class="co"># # sketch: scaled=5</span></span>
<span id="cb58-897"><a href="#cb58-897"></a><span class="co"># sourmash sketch dna -f -p k=7,k=9,k=11,scaled=5,abund,seed=124 --singleton \</span></span>
<span id="cb58-898"><a href="#cb58-898"></a><span class="co"># -o ${file_sig} ${file_fa}</span></span>
<span id="cb58-899"><a href="#cb58-899"></a></span>
<span id="cb58-900"><a href="#cb58-900"></a><span class="co"># sketch: num=2000</span></span>
<span id="cb58-901"><a href="#cb58-901"></a><span class="ex">sourmash</span> sketch dna <span class="at">-f</span> <span class="at">-p</span> k=7,k=9,k=11,k=13,num=2000,abund,seed=124 <span class="at">--singleton</span> <span class="dt">\</span></span>
<span id="cb58-902"><a href="#cb58-902"></a>-o <span class="va">${file_sig}</span> <span class="va">${file_fa}</span></span>
<span id="cb58-903"><a href="#cb58-903"></a></span>
<span id="cb58-904"><a href="#cb58-904"></a><span class="ex">sourmash</span> compare <span class="at">--ignore-abundance</span> <span class="at">--distance-matrix</span> <span class="at">-p</span> 12 <span class="at">-k</span> 13 <span class="at">--dna</span> <span class="dt">\</span></span>
<span id="cb58-905"><a href="#cb58-905"></a>--csv <span class="va">${file_dist}</span> <span class="va">${file_sig}</span> </span>
<span id="cb58-906"><a href="#cb58-906"></a></span>
<span id="cb58-907"><a href="#cb58-907"></a><span class="kw">```</span></span>
<span id="cb58-908"><a href="#cb58-908"></a></span>
<span id="cb58-909"><a href="#cb58-909"></a><span class="co">#### Calculate Jaccard distance using </span><span class="kw">`</span>dashing<span class="kw">`</span></span>
<span id="cb58-910"><a href="#cb58-910"></a></span>
<span id="cb58-911"><a href="#cb58-911"></a><span class="kw">```</span>bash</span>
<span id="cb58-912"><a href="#cb58-912"></a><span class="co"># dir_fa="ctv_conserved/fasta"</span></span>
<span id="cb58-913"><a href="#cb58-913"></a><span class="co"># dir_fa="ctv_tail/fasta"</span></span>
<span id="cb58-914"><a href="#cb58-914"></a></span>
<span id="cb58-915"><a href="#cb58-915"></a><span class="va">file_dist</span><span class="op">=</span><span class="st">"</span><span class="va">${dir_fa</span><span class="op">%</span>/<span class="pp">*</span><span class="va">}</span><span class="st">/"</span>distance.dashing.txt</span>
<span id="cb58-916"><a href="#cb58-916"></a></span>
<span id="cb58-917"><a href="#cb58-917"></a><span class="ex">dashing</span> cmp <span class="at">-k</span> 11 <span class="at">--seed</span> 124 <span class="at">--nthreads</span> 12 <span class="at">--full-mash-dist</span>  <span class="dt">\</span></span>
<span id="cb58-918"><a href="#cb58-918"></a>-O <span class="va">${file_dist}</span> <span class="va">${dir_fa}</span>/<span class="pp">*</span>.fasta</span>
<span id="cb58-919"><a href="#cb58-919"></a></span>
<span id="cb58-920"><a href="#cb58-920"></a><span class="kw">```</span></span>
<span id="cb58-921"><a href="#cb58-921"></a></span>
<span id="cb58-922"><a href="#cb58-922"></a><span class="co">### Tail fiber haplotype examples</span></span>
<span id="cb58-923"><a href="#cb58-923"></a></span>
<span id="cb58-924"><a href="#cb58-924"></a><span class="ex">Extract</span> sequences for some example tail fiber loci haplotypes that are present</span>
<span id="cb58-925"><a href="#cb58-925"></a><span class="er">in</span> <span class="ex">multiple</span> <span class="pp">*</span>Pectobacterium<span class="pp">*</span> species.</span>
<span id="cb58-926"><a href="#cb58-926"></a></span>
<span id="cb58-927"><a href="#cb58-927"></a><span class="kw">```</span>bash</span>
<span id="cb58-928"><a href="#cb58-928"></a><span class="co"># tail fiber locus</span></span>
<span id="cb58-929"><a href="#cb58-929"></a><span class="ex">Rscript</span> scripts/utils/HG_range_coordinates.R <span class="at">--hgs</span> hg_22427604,hg_22427603 <span class="dt">\</span></span>
<span id="cb58-930"><a href="#cb58-930"></a>--genomes <span class="st">"g_279,g_425,g_149,g_377,g_100,g_106,g_331,g_249,g_125,</span><span class="dt">\</span></span>
<span id="cb58-931"><a href="#cb58-931"></a><span class="st">g_221,g_53,g_395,g_108,g_444,g_160"</span> <span class="dt">\</span></span>
<span id="cb58-932"><a href="#cb58-932"></a>--dir analysis/pangenome_v2/carotovoricin/ctv_tail <span class="dt">\</span></span>
<span id="cb58-933"><a href="#cb58-933"></a>--out <span class="st">"selected_haplotypes_flanking.tab"</span></span>
<span id="cb58-934"><a href="#cb58-934"></a></span>
<span id="cb58-935"><a href="#cb58-935"></a><span class="ex">Rscript</span> scripts/utils/HG_range_coordinates.R <span class="at">--hgs</span> hg_22427604,hg_22427603 <span class="dt">\</span></span>
<span id="cb58-936"><a href="#cb58-936"></a>--genomes <span class="st">"g_279,g_425,g_149,g_377,g_100,g_106,g_331,g_249,g_125,</span><span class="dt">\</span></span>
<span id="cb58-937"><a href="#cb58-937"></a><span class="st">g_221,g_53,g_395,g_108,g_444,g_160"</span> <span class="dt">\</span></span>
<span id="cb58-938"><a href="#cb58-938"></a>--inner_region <span class="at">--dir</span> analysis/pangenome_v2/carotovoricin/ctv_tail <span class="dt">\</span></span>
<span id="cb58-939"><a href="#cb58-939"></a>--out <span class="st">"selected_haplotypes.tab"</span> </span>
<span id="cb58-940"><a href="#cb58-940"></a></span>
<span id="cb58-941"><a href="#cb58-941"></a><span class="co"># conserved CTV locus</span></span>
<span id="cb58-942"><a href="#cb58-942"></a><span class="ex">Rscript</span> scripts/utils/HG_range_coordinates.R <span class="at">--hgs</span> hg_22427640,hg_22427604 <span class="dt">\</span></span>
<span id="cb58-943"><a href="#cb58-943"></a>--genomes <span class="st">"g_279,g_425,g_149,g_377,g_100,g_106,g_331,g_249,g_125,</span><span class="dt">\</span></span>
<span id="cb58-944"><a href="#cb58-944"></a><span class="st">g_221,g_53,g_395,g_108,g_444,g_160"</span> <span class="dt">\</span></span>
<span id="cb58-945"><a href="#cb58-945"></a>--dir analysis/pangenome_v2/carotovoricin/ctv_conserved <span class="dt">\</span></span>
<span id="cb58-946"><a href="#cb58-946"></a>--out <span class="st">"selected_conserved.tab"</span></span>
<span id="cb58-947"><a href="#cb58-947"></a></span>
<span id="cb58-948"><a href="#cb58-948"></a><span class="co"># upstream core</span></span>
<span id="cb58-949"><a href="#cb58-949"></a><span class="ex">Rscript</span> scripts/utils/HG_range_coordinates.R <span class="at">--hgs</span> hg_22427642,hg_22427641 <span class="dt">\</span></span>
<span id="cb58-950"><a href="#cb58-950"></a>--genomes <span class="st">"g_279,g_425,g_149,g_377,g_100,g_106,g_331,g_249,g_125,</span><span class="dt">\</span></span>
<span id="cb58-951"><a href="#cb58-951"></a><span class="st">g_221,g_53,g_395,g_108,g_444,g_160"</span> <span class="dt">\</span></span>
<span id="cb58-952"><a href="#cb58-952"></a>--dir analysis/pangenome_v2/carotovoricin/upstream_core <span class="dt">\</span></span>
<span id="cb58-953"><a href="#cb58-953"></a>--out <span class="st">"selected_upstream.tab"</span></span>
<span id="cb58-954"><a href="#cb58-954"></a><span class="kw">```</span></span>
<span id="cb58-955"><a href="#cb58-955"></a></span>
<span id="cb58-956"><a href="#cb58-956"></a><span class="ex">[DNA</span> sequence comparison of these regions]<span class="er">(</span><span class="co">#dna-sequence-comparison)</span></span>
<span id="cb58-957"><a href="#cb58-957"></a></span>
<span id="cb58-958"><a href="#cb58-958"></a><span class="co">#### Perform MSA using `MAFFT`</span></span>
<span id="cb58-959"><a href="#cb58-959"></a></span>
<span id="cb58-960"><a href="#cb58-960"></a><span class="kw">```</span><span class="fu">bash</span></span>
<span id="cb58-961"><a href="#cb58-961"></a><span class="fu">nohup</span> mafft <span class="at">--reorder</span> <span class="at">--allowshift</span> <span class="at">--unalignlevel</span> 0.2 <span class="at">--leavegappyregion</span> <span class="dt">\</span></span>
<span id="cb58-962"><a href="#cb58-962"></a>--maxiterate 0 <span class="at">--globalpair</span> ctv_tail/selected_haplotypes.fasta <span class="dt">\</span></span>
<span id="cb58-963"><a href="#cb58-963"></a><span class="op">&gt;</span> ctv_tail/selected_haplotypes.msa.fasta <span class="dv">2</span><span class="op">&gt;</span>nohup.out <span class="kw">&amp;</span></span>
<span id="cb58-964"><a href="#cb58-964"></a></span>
<span id="cb58-965"><a href="#cb58-965"></a><span class="fu">nohup</span> mafft <span class="at">--reorder</span> <span class="at">--allowshift</span> <span class="at">--unalignlevel</span> 0.2 <span class="at">--leavegappyregion</span> <span class="dt">\</span></span>
<span id="cb58-966"><a href="#cb58-966"></a>--maxiterate 0 <span class="at">--globalpair</span> ctv_conserved/selected_conserved.fasta <span class="dt">\</span></span>
<span id="cb58-967"><a href="#cb58-967"></a><span class="op">&gt;</span> ctv_conserved/selected_conserved.msa.fasta <span class="dv">2</span><span class="op">&gt;</span>nohup.out <span class="kw">&amp;</span></span>
<span id="cb58-968"><a href="#cb58-968"></a></span>
<span id="cb58-969"><a href="#cb58-969"></a><span class="fu">nohup</span> mafft <span class="at">--reorder</span> <span class="at">--allowshift</span> <span class="at">--unalignlevel</span> 0.2 <span class="at">--leavegappyregion</span> <span class="dt">\</span></span>
<span id="cb58-970"><a href="#cb58-970"></a>--maxiterate 0 <span class="at">--globalpair</span> upstream_core/selected_upstream.fasta <span class="dt">\</span></span>
<span id="cb58-971"><a href="#cb58-971"></a><span class="op">&gt;</span> upstream_core/selected_upstream.msa.fasta <span class="dv">2</span><span class="op">&gt;</span>nohup.out <span class="kw">&amp;</span></span>
<span id="cb58-972"><a href="#cb58-972"></a><span class="kw">```</span></span>
<span id="cb58-973"><a href="#cb58-973"></a></span>
<span id="cb58-974"><a href="#cb58-974"></a><span class="ex">Generate</span> a maximum-likelihood phylogenetic tree for MSAs</span>
<span id="cb58-975"><a href="#cb58-975"></a></span>
<span id="cb58-976"><a href="#cb58-976"></a><span class="kw">```</span><span class="fu">bash</span></span>
<span id="cb58-977"><a href="#cb58-977"></a><span class="bu">cd</span> cd analysis/pangenome_v2/carotovoricin/</span>
<span id="cb58-978"><a href="#cb58-978"></a></span>
<span id="cb58-979"><a href="#cb58-979"></a><span class="fu">nohup</span> nice iqtree2 <span class="at">-T</span> 40 <span class="at">-s</span> ctv_tail/selected_haplotypes.msa.fasta <span class="at">-B</span> 1000 <span class="dt">\</span></span>
<span id="cb58-980"><a href="#cb58-980"></a>--prefix ctv_tail/iqtree/selected_haplotypes.msa.fasta <span class="op">&gt;&gt;</span> iqtree_tree.log <span class="dv">2</span><span class="op">&gt;&amp;</span><span class="dv">1</span> <span class="kw">&amp;</span></span>
<span id="cb58-981"><a href="#cb58-981"></a></span>
<span id="cb58-982"><a href="#cb58-982"></a><span class="fu">nohup</span> nice iqtree2 <span class="at">-T</span> 40 <span class="at">-s</span> ctv_conserved/selected_conserved.msa.fasta <span class="at">-B</span> 1000 <span class="dt">\</span></span>
<span id="cb58-983"><a href="#cb58-983"></a>--prefix ctv_conserved/iqtree/selected_conserved.msa.fasta <span class="op">&gt;&gt;</span> iqtree_tree.log <span class="dv">2</span><span class="op">&gt;&amp;</span><span class="dv">1</span> <span class="kw">&amp;</span></span>
<span id="cb58-984"><a href="#cb58-984"></a></span>
<span id="cb58-985"><a href="#cb58-985"></a><span class="fu">nohup</span> nice iqtree2 <span class="at">-T</span> 40 <span class="at">-s</span> upstream_core/selected_upstream.msa.fasta <span class="at">-B</span> 1000 <span class="dt">\</span></span>
<span id="cb58-986"><a href="#cb58-986"></a>--prefix upstream_core/iqtree/selected_upstream.msa.fasta <span class="op">&gt;&gt;</span> iqtree_tree.log <span class="dv">2</span><span class="op">&gt;&amp;</span><span class="dv">1</span> <span class="kw">&amp;</span></span>
<span id="cb58-987"><a href="#cb58-987"></a></span>
<span id="cb58-988"><a href="#cb58-988"></a><span class="kw">```</span></span>
<span id="cb58-989"><a href="#cb58-989"></a></span>
<span id="cb58-990"><a href="#cb58-990"></a></span>
<span id="cb58-991"><a href="#cb58-991"></a></span>
<span id="cb58-992"><a href="#cb58-992"></a><span class="co">### Use Mauve to detect the structural variation and consevation</span></span>
<span id="cb58-993"><a href="#cb58-993"></a></span>
<span id="cb58-994"><a href="#cb58-994"></a><span class="ex">Mauve</span> aligns sequences provided in <span class="kw">`</span><span class="ex">fasta</span><span class="kw">`</span> and <span class="kw">`</span><span class="ex">genbank</span><span class="kw">`</span> format. However, the</span>
<span id="cb58-995"><a href="#cb58-995"></a><span class="ex">sequence</span> annotation is shown only if the alignment input was in <span class="kw">`</span><span class="ex">genbank</span><span class="kw">`</span></span>
<span id="cb58-996"><a href="#cb58-996"></a><span class="ex">format.</span> Therefore, we need to combine the homology group annotation in <span class="kw">`</span><span class="ex">gff3</span><span class="kw">`</span></span>
<span id="cb58-997"><a href="#cb58-997"></a><span class="ex">format</span> with the <span class="kw">`</span><span class="ex">fasta</span><span class="kw">`</span> sequences to generate a <span class="kw">`</span><span class="ex">genbank</span><span class="kw">`</span> formatted files. To</span>
<span id="cb58-998"><a href="#cb58-998"></a><span class="cf">do</span> <span class="ex">this,</span> Emboss tool <span class="kw">`</span><span class="ex">seqret</span><span class="kw">`</span> was used.</span>
<span id="cb58-999"><a href="#cb58-999"></a></span>
<span id="cb58-1000"><a href="#cb58-1000"></a><span class="kw">```</span><span class="fu">bash</span></span>
<span id="cb58-1001"><a href="#cb58-1001"></a><span class="ex">conda</span> activate omics_py37</span>
<span id="cb58-1002"><a href="#cb58-1002"></a></span>
<span id="cb58-1003"><a href="#cb58-1003"></a><span class="cf">for</span> vir <span class="kw">in</span> <span class="kw">`</span>cut <span class="at">-f</span> 1 hg_regions.tab <span class="kw">|</span> <span class="fu">tail</span> <span class="at">-n</span> +2<span class="kw">`</span></span>
<span id="cb58-1004"><a href="#cb58-1004"></a><span class="cf">do</span></span>
<span id="cb58-1005"><a href="#cb58-1005"></a>  <span class="ex">seqret</span> <span class="at">-sequence</span> <span class="va">${vir}</span>.fasta <span class="at">-feature</span> <span class="at">-fformat</span> gff3 <span class="dt">\</span></span>
<span id="cb58-1006"><a href="#cb58-1006"></a>  <span class="at">-fopenfile</span> <span class="va">${vir}</span>.gff3 <span class="at">-osformat</span> genbank <span class="at">-osextension</span> gbk <span class="dt">\</span></span>
<span id="cb58-1007"><a href="#cb58-1007"></a>  <span class="at">-osname_outseq</span> <span class="va">${vir}</span> <span class="at">-ofdirectory_outseq</span> gbk_file <span class="at">-auto</span> </span>
<span id="cb58-1008"><a href="#cb58-1008"></a><span class="cf">done</span></span>
<span id="cb58-1009"><a href="#cb58-1009"></a><span class="kw">```</span></span>
<span id="cb58-1010"><a href="#cb58-1010"></a></span>
<span id="cb58-1011"><a href="#cb58-1011"></a><span class="ex">Correct</span> GenBank file as per Mauve requirement</span>
<span id="cb58-1012"><a href="#cb58-1012"></a></span>
<span id="cb58-1013"><a href="#cb58-1013"></a><span class="kw">```</span><span class="fu">bash</span></span>
<span id="cb58-1014"><a href="#cb58-1014"></a><span class="fu">sed</span> <span class="at">-i.bak</span> <span class="at">-r</span> <span class="at">-e</span> <span class="st">'s/(^\s+CDS\s+(complement\()?)[^:]+:([[:digit:]]+\.\.[[:digit:]]+\)?).*/\1\3/'</span> <span class="dt">\</span></span>
<span id="cb58-1015"><a href="#cb58-1015"></a>  <span class="at">-e</span> <span class="st">'s/\/note="\*([^:]+): /\/\1="/'</span> <span class="pp">*</span>.gbk</span>
<span id="cb58-1016"><a href="#cb58-1016"></a><span class="kw">```</span></span>
<span id="cb58-1017"><a href="#cb58-1017"></a></span>
<span id="cb58-1018"><a href="#cb58-1018"></a><span class="ex">Align</span> and visualize sequences using Mauve.</span>
<span id="cb58-1019"><a href="#cb58-1019"></a></span>
<span id="cb58-1020"><a href="#cb58-1020"></a><span class="kw">```</span><span class="fu">bash</span></span>
<span id="cb58-1021"><a href="#cb58-1021"></a><span class="va">mauve_out</span><span class="op">=</span><span class="st">"ctv_hgt.defaults"</span></span>
<span id="cb58-1022"><a href="#cb58-1022"></a></span>
<span id="cb58-1023"><a href="#cb58-1023"></a><span class="ex">progressiveMauve</span> <span class="at">--output</span><span class="op">=</span><span class="st">"</span><span class="va">${mauve_out}</span><span class="st">.mauve.xmfa"</span> <span class="dt">\</span></span>
<span id="cb58-1024"><a href="#cb58-1024"></a>--backbone-output=<span class="st">"</span><span class="va">${mauve_out}</span><span class="st">.mauve.backbone"</span> <span class="dt">\</span></span>
<span id="cb58-1025"><a href="#cb58-1025"></a>--output-guide-tree=<span class="st">"</span><span class="va">${mauve_out}</span><span class="st">.mauve.guide_tree.newick"</span> <span class="dt">\</span></span>
<span id="cb58-1026"><a href="#cb58-1026"></a><span class="pp">*</span>.gbk  <span class="op">&gt;</span> <span class="st">"</span><span class="va">${mauve_out}</span><span class="st">"</span>.mauve.log <span class="dv">2</span><span class="op">&gt;&amp;</span><span class="dv">1</span></span>
<span id="cb58-1027"><a href="#cb58-1027"></a></span>
<span id="cb58-1028"><a href="#cb58-1028"></a><span class="co"># mauve with custom settings</span></span>
<span id="cb58-1029"><a href="#cb58-1029"></a><span class="va">mauve_out</span><span class="op">=</span><span class="st">"ctv_hgt.opt"</span></span>
<span id="cb58-1030"><a href="#cb58-1030"></a></span>
<span id="cb58-1031"><a href="#cb58-1031"></a><span class="ex">progressiveMauve</span> <span class="at">--output</span><span class="op">=</span><span class="st">"</span><span class="va">${mauve_out}</span><span class="st">.mauve.xmfa"</span> <span class="dt">\</span></span>
<span id="cb58-1032"><a href="#cb58-1032"></a>--seed-weight 10 <span class="dt">\</span></span>
<span id="cb58-1033"><a href="#cb58-1033"></a>--backbone-output=<span class="st">"</span><span class="va">${mauve_out}</span><span class="st">.mauve.backbone"</span> <span class="dt">\</span></span>
<span id="cb58-1034"><a href="#cb58-1034"></a>--output-guide-tree=<span class="st">"</span><span class="va">${mauve_out}</span><span class="st">.mauve.guide_tree.newick"</span> <span class="dt">\</span></span>
<span id="cb58-1035"><a href="#cb58-1035"></a><span class="pp">*</span>.gbk  <span class="op">&gt;</span> <span class="st">"</span><span class="va">${mauve_out}</span><span class="st">"</span>.mauve.log <span class="dv">2</span><span class="op">&gt;&amp;</span><span class="dv">1</span></span>
<span id="cb58-1036"><a href="#cb58-1036"></a></span>
<span id="cb58-1037"><a href="#cb58-1037"></a><span class="co"># --seed-weight 5 \</span></span>
<span id="cb58-1038"><a href="#cb58-1038"></a></span>
<span id="cb58-1039"><a href="#cb58-1039"></a><span class="kw">```</span></span>
<span id="cb58-1040"><a href="#cb58-1040"></a></span>
<span id="cb58-1041"><a href="#cb58-1041"></a><span class="co">### Visualize tail fiber locus HG sets on the pangenome</span></span>
<span id="cb58-1042"><a href="#cb58-1042"></a></span>
<span id="cb58-1043"><a href="#cb58-1043"></a><span class="ex">R</span> script <span class="kw">`</span><span class="ex">scripts/analysis/HG_tandem_match.viz.qmd</span><span class="kw">`</span> to plot tail fiber locus HGs</span>
<span id="cb58-1044"><a href="#cb58-1044"></a><span class="cf">for</span> a <span class="bu">set</span> of genomes from different species.</span>
<span id="cb58-1045"><a href="#cb58-1045"></a></span>
<span id="cb58-1046"><a href="#cb58-1046"></a><span class="kw">```</span><span class="ex">r</span></span>
<span id="cb58-1047"><a href="#cb58-1047"></a><span class="ex">hgSets</span> <span class="op">&lt;</span>- list<span class="er">(</span></span>
<span id="cb58-1048"><a href="#cb58-1048"></a>  <span class="ex">s1</span> = c<span class="er">(</span><span class="st">"hg_22427604"</span><span class="ex">,</span> <span class="st">"hg_22426828"</span><span class="kw">)</span><span class="ex">,</span></span>
<span id="cb58-1049"><a href="#cb58-1049"></a>  <span class="ex">s2</span> = c<span class="er">(</span><span class="st">"hg_22427604"</span><span class="ex">,</span> <span class="st">"hg_22426818"</span><span class="kw">)</span><span class="ex">,</span></span>
<span id="cb58-1050"><a href="#cb58-1050"></a>  <span class="ex">s3</span> = c<span class="er">(</span><span class="st">"hg_22427604"</span><span class="ex">,</span> <span class="st">"hg_22426820"</span>, <span class="st">"hg_22426802"</span>, <span class="st">"hg_22426815"</span><span class="kw">)</span></span>
<span id="cb58-1051"><a href="#cb58-1051"></a><span class="kw">)</span></span>
<span id="cb58-1052"><a href="#cb58-1052"></a></span>
<span id="cb58-1053"><a href="#cb58-1053"></a><span class="ex">outDir</span> <span class="op">&lt;</span>- paste<span class="er">(</span><span class="ex">confs</span><span class="va">$analysis$prophages$dir</span><span class="ex">,</span> <span class="st">"/cluster_viz/ctv_hgt"</span>, sep = <span class="st">""</span><span class="kw">)</span></span>
<span id="cb58-1054"><a href="#cb58-1054"></a><span class="kw">```</span></span>
<span id="cb58-1055"><a href="#cb58-1055"></a></span>
<span id="cb58-1056"><a href="#cb58-1056"></a><span class="ex">*</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div></div></div></div></div>
</div> <!-- /content -->



</body></html>