<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>

<meta charset="utf-8">
<meta name="generator" content="quarto-1.5.57">

<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">

<meta name="author" content="Lakhansing Pardeshi">

<title>prophage_analysis – Pectobacterium pangenome data analysis</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
  vertical-align: middle;
}
/* CSS for syntax highlighting */
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { display: inline-block; text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
  }
pre.numberSource { margin-left: 3em;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
</style>


<script src="../../site_libs/quarto-nav/quarto-nav.js"></script>
<script src="../../site_libs/quarto-nav/headroom.min.js"></script>
<script src="../../site_libs/clipboard/clipboard.min.js"></script>
<script src="../../site_libs/quarto-search/autocomplete.umd.js"></script>
<script src="../../site_libs/quarto-search/fuse.min.js"></script>
<script src="../../site_libs/quarto-search/quarto-search.js"></script>
<meta name="quarto:offset" content="../../">
<link href="../../scripts/analysis/prophage_summary.html" rel="next">
<link href="../../scripts/analysis/phylo_viz_tree.html" rel="prev">
<script src="../../site_libs/quarto-html/quarto.js"></script>
<script src="../../site_libs/quarto-html/popper.min.js"></script>
<script src="../../site_libs/quarto-html/tippy.umd.min.js"></script>
<script src="../../site_libs/quarto-html/anchor.min.js"></script>
<link href="../../site_libs/quarto-html/tippy.css" rel="stylesheet">
<link href="../../site_libs/quarto-html/quarto-syntax-highlighting.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="../../site_libs/bootstrap/bootstrap.min.js"></script>
<link href="../../site_libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="../../site_libs/bootstrap/bootstrap.min.css" rel="stylesheet" id="quarto-bootstrap" data-mode="light">
<script id="quarto-search-options" type="application/json">{
  "location": "sidebar",
  "copy-button": false,
  "collapse-after": 3,
  "panel-placement": "start",
  "type": "textbox",
  "limit": 50,
  "keyboard-shortcut": [
    "f",
    "/",
    "s"
  ],
  "show-item-context": false,
  "language": {
    "search-no-results-text": "No results",
    "search-matching-documents-text": "matching documents",
    "search-copy-link-title": "Copy link to search",
    "search-hide-matches-text": "Hide additional matches",
    "search-more-match-text": "more match in this document",
    "search-more-matches-text": "more matches in this document",
    "search-clear-button-title": "Clear",
    "search-text-placeholder": "",
    "search-detached-cancel-button-title": "Cancel",
    "search-submit-button-title": "Submit",
    "search-label": "Search"
  }
}</script>


<link rel="stylesheet" href="../../scripts/utils/quarto_style.css">
<meta property="og:title" content="Pectobacterium pangenome data analysis">
<meta property="og:site_name" content="Pectobacterium pangenome data analysis">
</head>

<body class="nav-sidebar docked">

<div id="quarto-search-results"></div>
  <header id="quarto-header" class="headroom fixed-top">
  <nav class="quarto-secondary-nav">
    <div class="container-fluid d-flex">
      <button type="button" class="quarto-btn-toggle btn" data-bs-toggle="collapse" role="button" data-bs-target=".quarto-sidebar-collapse-item" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
        <i class="bi bi-layout-text-sidebar-reverse"></i>
      </button>
        <nav class="quarto-page-breadcrumbs" aria-label="breadcrumb"><ol class="breadcrumb"><li class="breadcrumb-item"><a href="../../scripts/notebooks/prophage_analysis.html">Prophages</a></li><li class="breadcrumb-item"><a href="../../scripts/notebooks/prophage_analysis.html">Analysis</a></li></ol></nav>
        <a class="flex-grow-1" role="navigation" data-bs-toggle="collapse" data-bs-target=".quarto-sidebar-collapse-item" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">      
        </a>
      <button type="button" class="btn quarto-search-button" aria-label="Search" onclick="window.quartoOpenSearch();">
        <i class="bi bi-search"></i>
      </button>
    </div>
  </nav>
</header>
<!-- content -->
<div id="quarto-content" class="quarto-container page-columns page-rows-contents page-layout-article">
<!-- sidebar -->
  <nav id="quarto-sidebar" class="sidebar collapse collapse-horizontal quarto-sidebar-collapse-item sidebar-navigation docked overflow-auto">
    <div class="pt-lg-2 mt-2 text-left sidebar-header">
    <div class="sidebar-title mb-0 py-0">
      <a href="../../">Pectobacterium pangenome data analysis</a> 
        <div class="sidebar-tools-main">
    <div class="dropdown">
      <a href="" title="" id="quarto-navigation-tool-dropdown-0" class="quarto-navigation-tool dropdown-toggle px-1" data-bs-toggle="dropdown" aria-expanded="false" role="link" aria-label=""><i class="bi bi-github"></i></a>
      <ul class="dropdown-menu" aria-labelledby="quarto-navigation-tool-dropdown-0">
          <li>
            <a class="dropdown-item sidebar-tools-main-item" href="https://github.com/lakhanp1/Pectobacterium_pangenome">
            Source Code
            </a>
          </li>
          <li>
            <a class="dropdown-item sidebar-tools-main-item" href="https://github.com/lakhanp1/Pectobacterium_pangenome/issues/new">
            Report a bug
            </a>
          </li>
      </ul>
    </div>
    <a href="https://x.com/lakhanp01" title="" class="quarto-navigation-tool px-1" aria-label=""><i class="bi bi-twitter"></i></a>
</div>
    </div>
      </div>
        <div class="mt-2 flex-shrink-0 align-items-center">
        <div class="sidebar-search">
        <div id="quarto-search" class="" title="Search"></div>
        </div>
        </div>
    <div class="sidebar-menu-container"> 
    <ul class="list-unstyled mt-1">
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../scripts/notebooks/index.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Pectobacterium genus pangenome</span></a>
  </div>
</li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-1" role="navigation" aria-expanded="true">
 <span class="menu-text">Preprocessing</span></a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-1" role="navigation" aria-expanded="true" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-1" class="collapse list-unstyled sidebar-section depth1 show">  
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../scripts/notebooks/preprocessing.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Preprocessing</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../scripts/analysis/ani_clustering.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">ANI</span></a>
  </div>
</li>
      </ul>
  </li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-2" role="navigation" aria-expanded="true">
 <span class="menu-text">Pangenome</span></a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-2" role="navigation" aria-expanded="true" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-2" class="collapse list-unstyled sidebar-section depth1 show">  
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../scripts/analysis/pangenome_metadata_viz.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Pangenome metadata</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../scripts/notebooks/pangenome_construction.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Build</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../scripts/notebooks/pangenome_analysis.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Analysis</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../scripts/analysis/HG_summary.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Pangenome summary</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../scripts/analysis/phylo_viz_tree.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Phylogeny</span></a>
  </div>
</li>
      </ul>
  </li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-3" role="navigation" aria-expanded="true">
 <span class="menu-text">Prophages</span></a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-3" role="navigation" aria-expanded="true" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-3" class="collapse list-unstyled sidebar-section depth1 show">  
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../scripts/notebooks/prophage_analysis.html" class="sidebar-item-text sidebar-link active">
 <span class="menu-text">Analysis</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../scripts/analysis/prophage_summary.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Prophage summary</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../scripts/analysis/synteny_demo.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Syntenic Jaccard index</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../scripts/analysis/prophage_clustering.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Clustering</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../scripts/analysis/prophage_cluster_summary.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Clustering summary</span></a>
  </div>
</li>
          <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start collapsed" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-4" role="navigation" aria-expanded="false">
 <span class="menu-text">Prophage clusters</span></a>
          <a class="sidebar-item-toggle text-start collapsed" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-4" role="navigation" aria-expanded="false" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-4" class="collapse list-unstyled sidebar-section depth2 ">  
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../scripts/ms_figures/phage_grp_45.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Phage group 45</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../scripts/ms_figures/phage_grp_46.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Phage group 46</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../scripts/ms_figures/phage_grp_107.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Phage group 107</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../scripts/ms_figures/phage_grp_71.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Phage group 71</span></a>
  </div>
</li>
      </ul>
  </li>
      </ul>
  </li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../scripts/ms_figures/supplementary_tables.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Supplementary tables</span></a>
  </div>
</li>
    </ul>
    </div>
</nav>
<div id="quarto-sidebar-glass" class="quarto-sidebar-collapse-item" data-bs-toggle="collapse" data-bs-target=".quarto-sidebar-collapse-item"></div>
<!-- margin-sidebar -->
    <div id="quarto-margin-sidebar" class="sidebar margin-sidebar">
        <nav id="TOC" role="doc-toc" class="toc-active">
    <h2 id="toc-title">On this page</h2>
   
  <ul>
  <li><a href="#prophage-analysis-in-pectobacterium-pangenome" id="toc-prophage-analysis-in-pectobacterium-pangenome" class="nav-link active" data-scroll-target="#prophage-analysis-in-pectobacterium-pangenome">Prophage analysis in <em>Pectobacterium</em> pangenome</a>
  <ul class="collapse">
  <li><a href="#detect-prophages-in-genome" id="toc-detect-prophages-in-genome" class="nav-link" data-scroll-target="#detect-prophages-in-genome">Detect prophages in genome</a>
  <ul class="collapse">
  <li><a href="#genomad-prophage-detection" id="toc-genomad-prophage-detection" class="nav-link" data-scroll-target="#genomad-prophage-detection">genomad prophage detection</a></li>
  <li><a href="#post-processing-of-prophages" id="toc-post-processing-of-prophages" class="nav-link" data-scroll-target="#post-processing-of-prophages">Post processing of prophages</a></li>
  <li><a href="#qc" id="toc-qc" class="nav-link" data-scroll-target="#qc">QC</a></li>
  </ul></li>
  <li><a href="#cluster-prophages" id="toc-cluster-prophages" class="nav-link" data-scroll-target="#cluster-prophages">Cluster prophages</a>
  <ul class="collapse">
  <li><a href="#approach-1" id="toc-approach-1" class="nav-link" data-scroll-target="#approach-1">Approach 1</a></li>
  <li><a href="#approach-2" id="toc-approach-2" class="nav-link" data-scroll-target="#approach-2">Approach 2</a></li>
  <li><a href="#summarize-prophage-clusters" id="toc-summarize-prophage-clusters" class="nav-link" data-scroll-target="#summarize-prophage-clusters">Summarize prophage clusters</a></li>
  </ul></li>
  <li><a href="#individual-prophage-cluster-analysis" id="toc-individual-prophage-cluster-analysis" class="nav-link" data-scroll-target="#individual-prophage-cluster-analysis">Individual prophage cluster analysis</a>
  <ul class="collapse">
  <li><a href="#phage_grp_46-highly-conserved-in-all-but-5-genomes-from-n23-clade" id="toc-phage_grp_46-highly-conserved-in-all-but-5-genomes-from-n23-clade" class="nav-link" data-scroll-target="#phage_grp_46-highly-conserved-in-all-but-5-genomes-from-n23-clade">phage_grp_46: highly conserved in all but 5 genomes from n23 clade</a></li>
  <li><a href="#phage_grp_45-prophage-present-in-all-bl-causing-isolates" id="toc-phage_grp_45-prophage-present-in-all-bl-causing-isolates" class="nav-link" data-scroll-target="#phage_grp_45-prophage-present-in-all-bl-causing-isolates">phage_grp_45: prophage present in all BL-causing isolates</a></li>
  <li><a href="#phage_grp_45-and-phage_grp_343-mosaicsm" id="toc-phage_grp_45-and-phage_grp_343-mosaicsm" class="nav-link" data-scroll-target="#phage_grp_45-and-phage_grp_343-mosaicsm">phage_grp_45 and phage_grp_343 mosaicsm</a></li>
  <li><a href="#phage_grp_71" id="toc-phage_grp_71" class="nav-link" data-scroll-target="#phage_grp_71">phage_grp_71</a></li>
  <li><a href="#phage_grp_107" id="toc-phage_grp_107" class="nav-link" data-scroll-target="#phage_grp_107">phage_grp_107</a></li>
  <li><a href="#bl-causing-clade-specific-signatures" id="toc-bl-causing-clade-specific-signatures" class="nav-link" data-scroll-target="#bl-causing-clade-specific-signatures">BL-causing clade specific signatures</a></li>
  <li><a href="#phage_grp_36" id="toc-phage_grp_36" class="nav-link" data-scroll-target="#phage_grp_36">phage_grp_36</a></li>
  <li><a href="#phage_grp_50" id="toc-phage_grp_50" class="nav-link" data-scroll-target="#phage_grp_50">phage_grp_50</a></li>
  <li><a href="#phage_grp_21" id="toc-phage_grp_21" class="nav-link" data-scroll-target="#phage_grp_21">phage_grp_21</a></li>
  <li><a href="#t3ss-missing-from-4-p.-brasiliense-genomes" id="toc-t3ss-missing-from-4-p.-brasiliense-genomes" class="nav-link" data-scroll-target="#t3ss-missing-from-4-p.-brasiliense-genomes">T3SS missing from 4 P. brasiliense genomes</a></li>
  </ul></li>
  </ul></li>
  </ul>
<div class="toc-actions"><ul><li><a href="https://github.com/lakhanp1/Pectobacterium_pangenome/issues/new" class="toc-action"><i class="bi bi-github"></i>Report an issue</a></li></ul></div></nav>
    </div>
<!-- main -->
<main class="content" id="quarto-document-content"><div class="quarto-title-block"><nav class="quarto-page-breadcrumbs quarto-title-breadcrumbs d-none d-lg-block" aria-label="breadcrumb"><ol class="breadcrumb"><li class="breadcrumb-item"><a href="../../scripts/notebooks/prophage_analysis.html">Prophages</a></li><li class="breadcrumb-item"><a href="../../scripts/notebooks/prophage_analysis.html">Analysis</a></li></ol></nav><div class="quarto-title-tools-only"><h1></h1><button type="button" class="btn code-tools-button" id="quarto-code-tools-source"><i class="bi"></i> Code</button></div></div>

<header id="title-block-header" class="quarto-title-block default">
<div class="quarto-title">
</div>



<div class="quarto-title-meta">

    <div>
    <div class="quarto-title-meta-heading">Author</div>
    <div class="quarto-title-meta-contents">
             <p>Lakhansing Pardeshi </p>
          </div>
  </div>
    
  
    
  </div>
  


</header>


<section id="prophage-analysis-in-pectobacterium-pangenome" class="level1">
<h1>Prophage analysis in <em>Pectobacterium</em> pangenome</h1>
<section id="detect-prophages-in-genome" class="level2">
<h2 class="anchored" data-anchor-id="detect-prophages-in-genome">Detect prophages in genome</h2>
<section id="genomad-prophage-detection" class="level3">
<h3 class="anchored" data-anchor-id="genomad-prophage-detection">genomad prophage detection</h3>
<div class="sourceCode" id="cb1"><pre class="sourceCode numberSource bash number-lines code-with-copy"><code class="sourceCode bash"><span id="cb1-1"><a href="#cb1-1"></a><span class="co"># run genomad and checkv pipeline</span></span>
<span id="cb1-2"><a href="#cb1-2"></a><span class="fu">nohup</span> <span class="dt">\</span></span>
<span id="cb1-3"><a href="#cb1-3"></a>cat data/reference_data/assembly_ids.txt <span class="kw">|</span> <span class="dt">\</span></span>
<span id="cb1-4"><a href="#cb1-4"></a><span class="ex">parallel</span> <span class="at">--jobs</span> 6 <span class="at">--workdir</span> <span class="va">$PWD</span> <span class="at">--halt</span> now,fail=1 <span class="dt">\</span></span>
<span id="cb1-5"><a href="#cb1-5"></a>--keep-order <span class="at">--results</span> logs/genomad/{} <span class="dt">\</span></span>
<span id="cb1-6"><a href="#cb1-6"></a>--joblog logs/genomad/parallel.log <span class="dt">\</span></span>
<span id="cb1-7"><a href="#cb1-7"></a>scripts/preprocessing/genomad_prophage_annotation.sh {} <span class="dt">\</span></span>
<span id="cb1-8"><a href="#cb1-8"></a><span class="op">&gt;&gt;</span>logs/genomad/nohup.out <span class="dv">2</span><span class="op">&gt;&amp;</span><span class="dv">1</span> <span class="kw">&amp;</span></span>
<span id="cb1-9"><a href="#cb1-9"></a></span>
<span id="cb1-10"><a href="#cb1-10"></a><span class="co"># count of prophage detected in each genome</span></span>
<span id="cb1-11"><a href="#cb1-11"></a><span class="cf">for</span> i <span class="kw">in</span> <span class="kw">`</span><span class="fu">cat</span> data/reference_data/assembly_ids.txt<span class="kw">`</span></span>
<span id="cb1-12"><a href="#cb1-12"></a><span class="cf">do</span></span>
<span id="cb1-13"><a href="#cb1-13"></a>  <span class="bu">printf</span> <span class="st">"</span><span class="va">$i</span><span class="st">\t"</span></span>
<span id="cb1-14"><a href="#cb1-14"></a>  <span class="fu">tail</span> <span class="at">-n</span> +2 data/prophage_genomad/<span class="va">$i</span>/<span class="va">${i}</span>_summary/<span class="va">${i}</span>_virus_summary.tsv <span class="kw">|</span> <span class="fu">wc</span> <span class="at">-l</span></span>
<span id="cb1-15"><a href="#cb1-15"></a><span class="cf">done</span> <span class="kw">|</span> <span class="fu">sort</span> <span class="at">-nr</span> <span class="at">-k2</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<p>Check the checkV contamination.tsv file to see if there is any other contamination than the expected <code>host,viral</code>, <code>viral,host</code> and <code>host,viral,host</code> ones.</p>
<div class="sourceCode" id="cb2"><pre class="sourceCode numberSource bash number-lines code-with-copy"><code class="sourceCode bash"><span id="cb2-1"><a href="#cb2-1"></a><span class="cf">for</span> cn <span class="kw">in</span> <span class="kw">`</span><span class="fu">find</span> <span class="at">-name</span> <span class="st">'*contamination.tsv'</span><span class="kw">`</span></span>
<span id="cb2-2"><a href="#cb2-2"></a><span class="cf">do</span></span>
<span id="cb2-3"><a href="#cb2-3"></a><span class="fu">cat</span> <span class="va">$cn</span> <span class="kw">|</span> <span class="fu">awk</span> <span class="st">'$9 != "NA" &amp;&amp; NR != 1 &amp;&amp; $9 != "host,viral" &amp;&amp; $9 != "viral,host" &amp;&amp; $9 != "host,viral,host"'</span></span>
<span id="cb2-4"><a href="#cb2-4"></a><span class="cf">done</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</section>
<section id="post-processing-of-prophages" class="level3">
<h3 class="anchored" data-anchor-id="post-processing-of-prophages">Post processing of prophages</h3>
<p>Process genomad output and separate prophages into independent FASTA files. Each prophage is assigned a unique identifier with format <code>g_{genome}.vir_{#count}</code>. Prophages are saved in independent FASTA files for downstream analysis, for e.g. ANI clustering of prophages.</p>
<div class="sourceCode" id="cb3"><pre class="sourceCode numberSource bash number-lines code-with-copy"><code class="sourceCode bash"><span id="cb3-1"><a href="#cb3-1"></a><span class="co"># post processing of prophage data</span></span>
<span id="cb3-2"><a href="#cb3-2"></a><span class="ex">Rscript</span> scripts/preprocessing/genomad_checkv_merge.R</span>
<span id="cb3-3"><a href="#cb3-3"></a></span>
<span id="cb3-4"><a href="#cb3-4"></a><span class="co"># save prophages independently</span></span>
<span id="cb3-5"><a href="#cb3-5"></a><span class="fu">bash</span> scripts/preprocessing/prophage_genomes_extract.sh</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<p>Output: - yaml: <code>confs$data$prophages$files$data</code> - file: <code>data/prophage_genomad/prophage_summary.tab</code></p>
<p>Extract homology groups for prophages from the pangenome org.db object.</p>
<div class="sourceCode" id="cb4"><pre class="sourceCode numberSource bash number-lines code-with-copy"><code class="sourceCode bash"><span id="cb4-1"><a href="#cb4-1"></a><span class="ex">Rscript</span> scripts/analysis/prophage_homology_groups.R</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<p>Output: - yaml: <code>confs$analysis$prophages$preprocessing$files$raw_prophage_hg</code> - file: <code>analysis/pangenome_v2/prophages/preprocessing/prophage_hgs.tab</code></p>
<p>Run ANI and MASH on prophage genomes.</p>
<div class="sourceCode" id="cb5"><pre class="sourceCode numberSource bash number-lines code-with-copy"><code class="sourceCode bash"><span id="cb5-1"><a href="#cb5-1"></a><span class="ex">conda</span> activate pantools_v4_3</span>
<span id="cb5-2"><a href="#cb5-2"></a></span>
<span id="cb5-3"><a href="#cb5-3"></a><span class="co"># list prophage FASTA files</span></span>
<span id="cb5-4"><a href="#cb5-4"></a><span class="fu">ls</span> data/prophage_genomad/phage_seqs/<span class="pp">*</span>.fna  <span class="op">&gt;</span> data/prophage_genomad/prophage_fasta.list</span>
<span id="cb5-5"><a href="#cb5-5"></a></span>
<span id="cb5-6"><a href="#cb5-6"></a><span class="co"># run fastANI</span></span>
<span id="cb5-7"><a href="#cb5-7"></a><span class="ex">fastANI</span> <span class="at">--ql</span> data/prophage_genomad/prophage_fasta.list <span class="dt">\</span></span>
<span id="cb5-8"><a href="#cb5-8"></a>--rl data/prophage_genomad/prophage_fasta.list <span class="dt">\</span></span>
<span id="cb5-9"><a href="#cb5-9"></a>--threads 30 <span class="at">--matrix</span> <span class="at">--visualize</span> <span class="at">--fragLen</span> 500 <span class="dt">\</span></span>
<span id="cb5-10"><a href="#cb5-10"></a>-o data/prophage_genomad/prophage_ANI</span>
<span id="cb5-11"><a href="#cb5-11"></a></span>
<span id="cb5-12"><a href="#cb5-12"></a><span class="co"># use mash for distance estimation</span></span>
<span id="cb5-13"><a href="#cb5-13"></a><span class="fu">mkdir</span> data/prophage_genomad/mash</span>
<span id="cb5-14"><a href="#cb5-14"></a></span>
<span id="cb5-15"><a href="#cb5-15"></a><span class="ex">mash</span> sketch <span class="at">-p</span> 30 <span class="at">-s</span> 15000 <span class="at">-k</span> 17 <span class="at">-o</span> data/prophage_genomad/mash/prophages_ref <span class="dt">\</span></span>
<span id="cb5-16"><a href="#cb5-16"></a>-l data/prophage_genomad/prophage_fasta.list <span class="op">&gt;</span>logs/prophage/mash.log <span class="dv">2</span><span class="op">&gt;&amp;</span><span class="dv">1</span></span>
<span id="cb5-17"><a href="#cb5-17"></a></span>
<span id="cb5-18"><a href="#cb5-18"></a><span class="ex">mash</span> info data/prophage_genomad/mash/prophages_ref.msh <span class="op">&gt;</span> <span class="dt">\</span></span>
<span id="cb5-19"><a href="#cb5-19"></a>data/prophage_genomad/mash/prophage_sketch_info.txt</span>
<span id="cb5-20"><a href="#cb5-20"></a></span>
<span id="cb5-21"><a href="#cb5-21"></a><span class="ex">mash</span> dist <span class="at">-p</span> 30 data/prophage_genomad/mash/prophages_ref.msh <span class="dt">\</span></span>
<span id="cb5-22"><a href="#cb5-22"></a>data/prophage_genomad/mash/prophages_ref.msh <span class="op">&gt;</span> <span class="dt">\</span></span>
<span id="cb5-23"><a href="#cb5-23"></a>data/prophage_genomad/mash/prophage_distance.tab</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</section>
<section id="qc" class="level3">
<h3 class="anchored" data-anchor-id="qc">QC</h3>
<p>Identify the fragmented prophages resulting because of the genome assembly that can be merged into a complete prophage. Prepare a consolidated list of prophages which will be used for the subsequent analysis.</p>
<div class="sourceCode" id="cb6"><pre class="sourceCode numberSource bash number-lines code-with-copy"><code class="sourceCode bash"><span id="cb6-1"><a href="#cb6-1"></a><span class="ex">Rscript</span> scripts/analysis/prophage_merge.R</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<p>Output: - yaml: <code>confs$analysis$prophages$preprocessing$files$consolidated</code> - file: <code>analysis/prophages/preprocessing/consolidated_phages.tab</code></p>
<p>Summarize the consolidated prophage list to generate various statistics and generate summary plots. Prophages that are fragmented or smaller than the length cutoff are removed and the filtered list of prophages is saved for downstream analysis.</p>
<div class="sourceCode" id="cb7"><pre class="sourceCode numberSource bash number-lines code-with-copy"><code class="sourceCode bash"><span id="cb7-1"><a href="#cb7-1"></a><span class="ex">Rscript</span> scripts/analysis/prophage_summary.R</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<p>Output: - yaml: <code>confs$analysis$prophages$preprocessing$files$prophage_pool</code> - File: <code>analysis/pangenome_v2/prophages/preprocessing/prophages_pool.tab</code></p>
<p>Process prophage MASH and ANI results and visualize MASH similarity matrix for the filtered prophages.</p>
<div class="sourceCode" id="cb8"><pre class="sourceCode numberSource bash number-lines code-with-copy"><code class="sourceCode bash"><span id="cb8-1"><a href="#cb8-1"></a><span class="ex">quarto</span> render scripts/analysis/prophage_seq_similarity.qmd <span class="at">--execute-dir</span> ./</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</section>
</section>
<section id="cluster-prophages" class="level2">
<h2 class="anchored" data-anchor-id="cluster-prophages">Cluster prophages</h2>
<p>Use shared homology groups between two prophages to establish the child-parent relationship between the prophages in pangenome. First, all vs all syntenic Jaccard similarity is calculated between the prophages. This information is later used to cluster the prophages using two different approaches.</p>
<div class="sourceCode" id="cb9"><pre class="sourceCode numberSource bash number-lines code-with-copy"><code class="sourceCode bash"><span id="cb9-1"><a href="#cb9-1"></a><span class="ex">Rscript</span> scripts/analysis/prophage_pairwise_similarity.R</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<section id="approach-1" class="level3">
<h3 class="anchored" data-anchor-id="approach-1">Approach 1</h3>
<p>A heuristic DAG construcion is extracted from the output of the script <code>scripts/analysis/prophage_merge.R</code>. This output includes a best matching parent prophage for each prophage. This is further processed to build a DAG and visualize the clusters.</p>
<p>Post processing of the child-parent relationships to add additional metadata and save a DAG in tabular format:</p>
<div class="sourceCode" id="cb10"><pre class="sourceCode numberSource bash number-lines code-with-copy"><code class="sourceCode bash"><span id="cb10-1"><a href="#cb10-1"></a><span class="ex">Rscript</span> scripts/analysis/prophage_DAG_postprocess.R</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<p>Visualize the clusters of prophages as directed acyclic graphs:</p>
<div class="sourceCode" id="cb11"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb11-1"><a href="#cb11-1"></a>knitr<span class="sc">::</span><span class="fu">knit</span>(<span class="st">"scripts/analysis/prophage_DAG_viz.qmd"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</section>
<section id="approach-2" class="level3">
<h3 class="anchored" data-anchor-id="approach-2">Approach 2</h3>
<p>We first tried to use the Affinity propagation (AP) clustering to group the prophages into clusters and have a set of cluster representatives. However, we encountered some clusters generated by the AP clustering where the cluster members did not even share a homology group. Therefore, we decided not to use AP clustering.</p>
<div class="sourceCode" id="cb12"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb12-1"><a href="#cb12-1"></a>knitr<span class="sc">::</span><span class="fu">knit</span>(<span class="st">"scripts/analysis/prophage_AP_clustering.qmd"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<p>Cluster the all-vs-all syntenic Jaccard distance matrix using “complete” linkage hierarchical clustering and cut the cluster using carotovoricin cluster as a reference.</p>
<div class="sourceCode" id="cb13"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb13-1"><a href="#cb13-1"></a>knitr<span class="sc">::</span><span class="fu">knit</span>(<span class="st">"scripts/analysis/prophage_clustering.qmd"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<p>Output: - yaml: <code>confs$analysis$prophages$files$clusters</code> - file: <code>analysis/pangenome_v2/prophages/prophages_clusters.tab</code></p>
</section>
<section id="summarize-prophage-clusters" class="level3">
<h3 class="anchored" data-anchor-id="summarize-prophage-clusters">Summarize prophage clusters</h3>
<p>Summarize the representative prophages in the pangenome.</p>
<div class="sourceCode" id="cb14"><pre class="sourceCode numberSource bash number-lines code-with-copy"><code class="sourceCode bash"><span id="cb14-1"><a href="#cb14-1"></a><span class="ex">Rscript</span> scripts/analysis/prophage_cluster_summary.qmd</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<p>Visualize the representative prophages across pangenome. Use MASH distance to cluster prophages and ANI distance to cluster the genomes in the pangenome.</p>
<div class="callout callout-style-default callout-caution callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Caution
</div>
</div>
<div class="callout-body-container callout-body">
<p>Need a fix. Do not run!</p>
<div class="sourceCode" id="cb15" data-eval="FALSE"><pre class="sourceCode numberSource bash number-lines code-with-copy"><code class="sourceCode bash"><span id="cb15-1"><a href="#cb15-1"></a><span class="ex">quarto</span> render scripts/analysis/prophage_pangenome_viz.qmd <span class="at">--execute-dir</span> ./</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</div>
<p>Visualize homology group sets for selected prophages as PAV heatmap across the pangenome. Replace the code block in <code>scripts/analysis/homology_group_viz.R</code> script with the following.</p>
<div class="sourceCode" id="cb16"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb16-1"><a href="#cb16-1"></a>analysisName <span class="ot">&lt;-</span> <span class="st">"vir_lineage_prophages"</span></span>
<span id="cb16-2"><a href="#cb16-2"></a></span>
<span id="cb16-3"><a href="#cb16-3"></a>outDir <span class="ot">&lt;-</span> <span class="fu">file.path</span>(confs<span class="sc">$</span>analysis<span class="sc">$</span>prophages<span class="sc">$</span>dir)</span>
<span id="cb16-4"><a href="#cb16-4"></a>outPrefix <span class="ot">&lt;-</span> <span class="fu">file.path</span>(outDir, analysisName)</span>
<span id="cb16-5"><a href="#cb16-5"></a></span>
<span id="cb16-6"><a href="#cb16-6"></a>setIds <span class="ot">&lt;-</span> <span class="fu">c</span>(</span>
<span id="cb16-7"><a href="#cb16-7"></a>  <span class="st">"g_189.vir_2"</span>, <span class="st">"g_400.vir_2"</span>, <span class="st">"g_399.vir_2"</span>, <span class="st">"g_448.vir_2"</span>, <span class="st">"g_189.vir_1"</span>, </span>
<span id="cb16-8"><a href="#cb16-8"></a>  <span class="st">"g_3.vir_4"</span>, <span class="st">"g_438.vir_3"</span>, <span class="st">"g_350.vir_1"</span>,</span>
<span id="cb16-9"><a href="#cb16-9"></a>  <span class="st">"g_172.vir_2"</span>, <span class="st">"g_393.vir_1"</span></span>
<span id="cb16-10"><a href="#cb16-10"></a>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<p>Finally, run the script to generate the plot.</p>
<div class="sourceCode" id="cb17"><pre class="sourceCode numberSource bash number-lines code-with-copy"><code class="sourceCode bash"><span id="cb17-1"><a href="#cb17-1"></a><span class="ex">Rscript</span> scripts/analysis/homology_group_viz.R</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</section>
</section>
<section id="individual-prophage-cluster-analysis" class="level2">
<h2 class="anchored" data-anchor-id="individual-prophage-cluster-analysis">Individual prophage cluster analysis</h2>
<p>Script <code>scripts/analysis/clustersmap_data_prophages.R</code> is used to combine the prophages, homology groups and phylogeny to generate input for <code>clustermap.js</code> visualization tool and generate a report. Below, some configuration is provided for individual prophage clusters.</p>
<section id="phage_grp_46-highly-conserved-in-all-but-5-genomes-from-n23-clade" class="level3">
<h3 class="anchored" data-anchor-id="phage_grp_46-highly-conserved-in-all-but-5-genomes-from-n23-clade">phage_grp_46: highly conserved in all but 5 genomes from n23 clade</h3>
<div class="sourceCode" id="cb18"><pre class="sourceCode numberSource bash number-lines code-with-copy"><code class="sourceCode bash"><span id="cb18-1"><a href="#cb18-1"></a><span class="ex">Rscript</span> scripts/utils/HG_range_coordinates.R <span class="at">--hgs</span> hg_22429500,hg_22429499 <span class="dt">\</span></span>
<span id="cb18-2"><a href="#cb18-2"></a>--inner_region <span class="at">--genomes</span> <span class="st">"g_63,g_190,g_193,g_411,g_415,g_406,g_194,g_263"</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="sourceCode" id="cb19"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb19-1"><a href="#cb19-1"></a>cluster_title <span class="ot">&lt;-</span> <span class="st">"phage_grp_46"</span></span>
<span id="cb19-2"><a href="#cb19-2"></a>outDir <span class="ot">&lt;-</span> <span class="fu">paste</span>(confs<span class="sc">$</span>analysis<span class="sc">$</span>prophages<span class="sc">$</span>dir, <span class="st">"/cluster_viz/"</span>, cluster_title, <span class="at">sep =</span> <span class="st">""</span>)</span>
<span id="cb19-3"><a href="#cb19-3"></a>hg_color_categories <span class="ot">&lt;-</span> confs<span class="sc">$</span>analysis<span class="sc">$</span>prophages<span class="sc">$</span>files<span class="sc">$</span>hg_broad_functions</span>
<span id="cb19-4"><a href="#cb19-4"></a></span>
<span id="cb19-5"><a href="#cb19-5"></a><span class="co"># a vector of prophage identifiers that will be included in clustermap plot</span></span>
<span id="cb19-6"><a href="#cb19-6"></a>region_cluster <span class="ot">&lt;-</span> <span class="st">"phage_grp_46"</span></span>
<span id="cb19-7"><a href="#cb19-7"></a>other_regions <span class="ot">&lt;-</span> <span class="fu">character</span>(<span class="dv">0</span>)</span>
<span id="cb19-8"><a href="#cb19-8"></a></span>
<span id="cb19-9"><a href="#cb19-9"></a>subSample <span class="ot">&lt;-</span> <span class="cn">TRUE</span></span>
<span id="cb19-10"><a href="#cb19-10"></a>cutHeight <span class="ot">&lt;-</span> <span class="fl">1.5</span></span>
<span id="cb19-11"><a href="#cb19-11"></a>addFlankingRegions <span class="ot">&lt;-</span> <span class="cn">TRUE</span></span>
<span id="cb19-12"><a href="#cb19-12"></a>flankingRegion <span class="ot">&lt;-</span> <span class="dv">5000</span></span>
<span id="cb19-13"><a href="#cb19-13"></a></span>
<span id="cb19-14"><a href="#cb19-14"></a><span class="co"># ordering factor for prophages: "host" phylogeny, "hg_pav" for prophage HG PAV,</span></span>
<span id="cb19-15"><a href="#cb19-15"></a><span class="co"># "cluster_mash" for prophage MASH and "default" to use the provided</span></span>
<span id="cb19-16"><a href="#cb19-16"></a>clusterOrder <span class="ot">&lt;-</span> <span class="st">"host"</span> <span class="co"># host, hg_pav, cluster_mash, default</span></span>
<span id="cb19-17"><a href="#cb19-17"></a></span>
<span id="cb19-18"><a href="#cb19-18"></a><span class="co"># whether to keep custom regions at the bottom or consider during phylogeny</span></span>
<span id="cb19-19"><a href="#cb19-19"></a><span class="co"># based ordering</span></span>
<span id="cb19-20"><a href="#cb19-20"></a>regions_phy_ordered <span class="ot">&lt;-</span> <span class="cn">TRUE</span> </span>
<span id="cb19-21"><a href="#cb19-21"></a></span>
<span id="cb19-22"><a href="#cb19-22"></a><span class="co"># regions to append as list of list with following structure</span></span>
<span id="cb19-23"><a href="#cb19-23"></a><span class="co"># list(r1 = list(chr, start, end, genomeId), r2 = list(chr, start, end, genomeId))</span></span>
<span id="cb19-24"><a href="#cb19-24"></a>customRegions <span class="ot">&lt;-</span> <span class="fu">list</span>(</span>
<span id="cb19-25"><a href="#cb19-25"></a>  <span class="at">g_190_reg =</span> <span class="fu">list</span>(<span class="at">chr =</span> <span class="st">"NZ_CP059960.1"</span>, <span class="at">start =</span> <span class="dv">1997124</span>, <span class="at">end =</span> <span class="dv">2000186</span>, <span class="at">genomeId =</span> <span class="st">"g_190"</span>),</span>
<span id="cb19-26"><a href="#cb19-26"></a>  <span class="at">g_193_reg =</span> <span class="fu">list</span>(<span class="at">chr =</span> <span class="st">"NZ_CP059957.1"</span>, <span class="at">start =</span> <span class="dv">5005351</span>, <span class="at">end =</span> <span class="dv">5008413</span>, <span class="at">genomeId =</span> <span class="st">"g_193"</span>),</span>
<span id="cb19-27"><a href="#cb19-27"></a>  <span class="at">g_411_reg =</span> <span class="fu">list</span>(<span class="at">chr =</span> <span class="st">"NAK682_contig_2_consensus"</span>, <span class="at">start =</span> <span class="dv">461100</span>, <span class="at">end =</span> <span class="dv">464162</span>, <span class="at">genomeId =</span> <span class="st">"g_411"</span>),</span>
<span id="cb19-28"><a href="#cb19-28"></a>  <span class="at">g_415_reg =</span> <span class="fu">list</span>(<span class="at">chr =</span> <span class="st">"NAK701_contig_5_consensus"</span>, <span class="at">start =</span> <span class="dv">415799</span>, <span class="at">end =</span> <span class="dv">418861</span>, <span class="at">genomeId =</span> <span class="st">"g_415"</span>),</span>
<span id="cb19-29"><a href="#cb19-29"></a>  <span class="at">g_63_reg =</span> <span class="fu">list</span>(<span class="at">chr =</span> <span class="st">"NZ_CP092039.1"</span>, <span class="at">start =</span> <span class="dv">4288654</span>, <span class="at">end =</span> <span class="dv">4291716</span>, <span class="at">genomeId =</span> <span class="st">"g_63"</span>),</span>
<span id="cb19-30"><a href="#cb19-30"></a>  <span class="at">g_406_reg =</span> <span class="fu">list</span>(<span class="at">chr =</span> <span class="st">"NAK641_contig_10_consensus"</span>, <span class="at">start =</span> <span class="dv">469664</span>, <span class="at">end =</span> <span class="dv">474521</span>, <span class="at">genomeId =</span> <span class="st">"g_406"</span>),</span>
<span id="cb19-31"><a href="#cb19-31"></a>  <span class="at">g_194_reg =</span> <span class="fu">list</span>(<span class="at">chr =</span> <span class="st">"NZ_CP059956.1"</span>, <span class="at">start =</span> <span class="dv">4163454</span>, <span class="at">end =</span> <span class="dv">4167352</span>, <span class="at">genomeId =</span> <span class="st">"g_194"</span>)</span>
<span id="cb19-32"><a href="#cb19-32"></a>  <span class="co"># g_263_reg = list(chr = "NZ_CP047495.1", start = 956237, end = 956907, genomeId = "g_263")</span></span>
<span id="cb19-33"><a href="#cb19-33"></a>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</section>
<section id="phage_grp_45-prophage-present-in-all-bl-causing-isolates" class="level3">
<h3 class="anchored" data-anchor-id="phage_grp_45-prophage-present-in-all-bl-causing-isolates">phage_grp_45: prophage present in all BL-causing isolates</h3>
<div class="sourceCode" id="cb20"><pre class="sourceCode numberSource bash number-lines code-with-copy"><code class="sourceCode bash"><span id="cb20-1"><a href="#cb20-1"></a><span class="ex">Rscript</span> scripts/utils/HG_range_coordinates.R <span class="at">--hgs</span> hg_22429319,hg_22429327 <span class="dt">\</span></span>
<span id="cb20-2"><a href="#cb20-2"></a>--genomes <span class="st">"g_406,g_194"</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="sourceCode" id="cb21"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb21-1"><a href="#cb21-1"></a>cluster_title <span class="ot">&lt;-</span> <span class="st">"phage_grp_45"</span></span>
<span id="cb21-2"><a href="#cb21-2"></a>outDir <span class="ot">&lt;-</span> <span class="fu">paste</span>(confs<span class="sc">$</span>analysis<span class="sc">$</span>prophages<span class="sc">$</span>dir, <span class="st">"/cluster_viz/"</span>, cluster_title, <span class="at">sep =</span> <span class="st">""</span>)</span>
<span id="cb21-3"><a href="#cb21-3"></a>hg_color_categories <span class="ot">&lt;-</span> confs<span class="sc">$</span>analysis<span class="sc">$</span>prophages<span class="sc">$</span>files<span class="sc">$</span>hg_broad_functions</span>
<span id="cb21-4"><a href="#cb21-4"></a></span>
<span id="cb21-5"><a href="#cb21-5"></a><span class="co"># a vector of prophage identifiers that will be included in clustermap plot</span></span>
<span id="cb21-6"><a href="#cb21-6"></a>region_cluster <span class="ot">&lt;-</span> <span class="st">"phage_grp_45"</span></span>
<span id="cb21-7"><a href="#cb21-7"></a>other_regions <span class="ot">&lt;-</span> <span class="fu">character</span>(<span class="dv">0</span>)</span>
<span id="cb21-8"><a href="#cb21-8"></a></span>
<span id="cb21-9"><a href="#cb21-9"></a>subSample <span class="ot">&lt;-</span> <span class="cn">FALSE</span> </span>
<span id="cb21-10"><a href="#cb21-10"></a>cutHeight <span class="ot">&lt;-</span> <span class="fl">1.5</span></span>
<span id="cb21-11"><a href="#cb21-11"></a>addFlankingRegions <span class="ot">&lt;-</span> <span class="cn">TRUE</span></span>
<span id="cb21-12"><a href="#cb21-12"></a>flankingRegion <span class="ot">&lt;-</span> <span class="dv">5000</span></span>
<span id="cb21-13"><a href="#cb21-13"></a></span>
<span id="cb21-14"><a href="#cb21-14"></a><span class="co"># ordering factor for prophages: "host" phylogeny, "hg_pav" for prophage HG PAV,</span></span>
<span id="cb21-15"><a href="#cb21-15"></a><span class="co"># "cluster_mash" for prophage MASH and "default" to use the provided</span></span>
<span id="cb21-16"><a href="#cb21-16"></a>clusterOrder <span class="ot">&lt;-</span> <span class="st">"host"</span> <span class="co"># host, hg_pav, cluster_mash, default</span></span>
<span id="cb21-17"><a href="#cb21-17"></a></span>
<span id="cb21-18"><a href="#cb21-18"></a><span class="co"># whether to keep custom regions at the bottom or consider during phylogeny</span></span>
<span id="cb21-19"><a href="#cb21-19"></a><span class="co"># based ordering</span></span>
<span id="cb21-20"><a href="#cb21-20"></a>regions_phy_ordered <span class="ot">&lt;-</span> <span class="cn">TRUE</span> </span>
<span id="cb21-21"><a href="#cb21-21"></a></span>
<span id="cb21-22"><a href="#cb21-22"></a><span class="co"># regions to append as list of list with following structure</span></span>
<span id="cb21-23"><a href="#cb21-23"></a><span class="co"># list(r1 = list(chr, start, end, genomeId), r2 = list(chr, start, end, genomeId))</span></span>
<span id="cb21-24"><a href="#cb21-24"></a>customRegions <span class="ot">&lt;-</span> <span class="fu">list</span>(</span>
<span id="cb21-25"><a href="#cb21-25"></a>  <span class="at">g_406_reg =</span> <span class="fu">list</span>(<span class="at">chr =</span> <span class="st">"NAK641_contig_10_consensus"</span>, <span class="at">start =</span> <span class="dv">671040</span>, <span class="at">end =</span> <span class="dv">674984</span>, <span class="at">genomeId =</span> <span class="st">"g_406"</span>),</span>
<span id="cb21-26"><a href="#cb21-26"></a>  <span class="at">g_194_reg =</span> <span class="fu">list</span>(<span class="at">chr =</span> <span class="st">"NZ_CP059956.1"</span>, <span class="at">start =</span> <span class="dv">3940785</span>, <span class="at">end =</span> <span class="dv">3958060</span>, <span class="at">genomeId =</span> <span class="st">"g_194"</span>)</span>
<span id="cb21-27"><a href="#cb21-27"></a>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</section>
<section id="phage_grp_45-and-phage_grp_343-mosaicsm" class="level3">
<h3 class="anchored" data-anchor-id="phage_grp_45-and-phage_grp_343-mosaicsm">phage_grp_45 and phage_grp_343 mosaicsm</h3>
<div class="sourceCode" id="cb22"><pre class="sourceCode numberSource bash number-lines code-with-copy"><code class="sourceCode bash"><span id="cb22-1"><a href="#cb22-1"></a><span class="ex">Rscript</span> scripts/utils/HG_range_coordinates.R <span class="at">--hgs</span> hg_22429319,hg_22429327 <span class="dt">\</span></span>
<span id="cb22-2"><a href="#cb22-2"></a>--genomes <span class="st">"g_406,g_194"</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="sourceCode" id="cb23"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb23-1"><a href="#cb23-1"></a>cluster_title <span class="ot">&lt;-</span> <span class="st">"grp_45_mosaicsm"</span></span>
<span id="cb23-2"><a href="#cb23-2"></a>outDir <span class="ot">&lt;-</span> <span class="fu">paste</span>(confs<span class="sc">$</span>analysis<span class="sc">$</span>prophages<span class="sc">$</span>dir, <span class="st">"/cluster_viz/"</span>, cluster_title, <span class="at">sep =</span> <span class="st">""</span>)</span>
<span id="cb23-3"><a href="#cb23-3"></a>hg_color_categories <span class="ot">&lt;-</span> confs<span class="sc">$</span>analysis<span class="sc">$</span>prophages<span class="sc">$</span>files<span class="sc">$</span>hg_broad_functions</span>
<span id="cb23-4"><a href="#cb23-4"></a></span>
<span id="cb23-5"><a href="#cb23-5"></a><span class="co"># a vector of prophage identifiers that will be included in clustermap plot</span></span>
<span id="cb23-6"><a href="#cb23-6"></a>region_cluster <span class="ot">&lt;-</span> <span class="cn">NA</span></span>
<span id="cb23-7"><a href="#cb23-7"></a>other_regions <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">"g_302.vir_1"</span>, <span class="st">"g_399.vir_1"</span>, <span class="st">"g_400.vir_2"</span>)</span>
<span id="cb23-8"><a href="#cb23-8"></a></span>
<span id="cb23-9"><a href="#cb23-9"></a>subSample <span class="ot">&lt;-</span> <span class="cn">TRUE</span></span>
<span id="cb23-10"><a href="#cb23-10"></a>cutHeight <span class="ot">&lt;-</span> <span class="fl">0.5</span></span>
<span id="cb23-11"><a href="#cb23-11"></a>addFlankingRegions <span class="ot">&lt;-</span> <span class="cn">TRUE</span></span>
<span id="cb23-12"><a href="#cb23-12"></a>flankingRegion <span class="ot">&lt;-</span> <span class="dv">7000</span></span>
<span id="cb23-13"><a href="#cb23-13"></a></span>
<span id="cb23-14"><a href="#cb23-14"></a><span class="co"># ordering factor for prophages: "host" phylogeny, "hg_pav" for prophage HG PAV,</span></span>
<span id="cb23-15"><a href="#cb23-15"></a><span class="co"># "cluster_mash" for prophage MASH and "default" to use the provided</span></span>
<span id="cb23-16"><a href="#cb23-16"></a>clusterOrder <span class="ot">&lt;-</span> <span class="st">"host"</span> <span class="co"># host, hg_pav, cluster_mash, default</span></span>
<span id="cb23-17"><a href="#cb23-17"></a></span>
<span id="cb23-18"><a href="#cb23-18"></a><span class="co"># whether to keep custom regions at the bottom or consider during phylogeny</span></span>
<span id="cb23-19"><a href="#cb23-19"></a><span class="co"># based ordering</span></span>
<span id="cb23-20"><a href="#cb23-20"></a>regions_phy_ordered <span class="ot">&lt;-</span> <span class="cn">FALSE</span></span>
<span id="cb23-21"><a href="#cb23-21"></a></span>
<span id="cb23-22"><a href="#cb23-22"></a><span class="co"># regions to append as list of list with following structure</span></span>
<span id="cb23-23"><a href="#cb23-23"></a><span class="co"># list(r1 = list(chr, start, end, genomeId), r2 = list(chr, start, end, genomeId))</span></span>
<span id="cb23-24"><a href="#cb23-24"></a>customRegions <span class="ot">&lt;-</span> <span class="fu">list</span>(</span>
<span id="cb23-25"><a href="#cb23-25"></a>  <span class="at">g_406_reg =</span> <span class="fu">list</span>(<span class="at">chr =</span> <span class="st">"NAK641_contig_10_consensus"</span>, <span class="at">start =</span> <span class="dv">671040</span>, <span class="at">end =</span> <span class="dv">674984</span>, <span class="at">genomeId =</span> <span class="st">"g_406"</span>),</span>
<span id="cb23-26"><a href="#cb23-26"></a>  <span class="at">g_194_reg =</span> <span class="fu">list</span>(<span class="at">chr =</span> <span class="st">"NZ_CP059956.1"</span>, <span class="at">start =</span> <span class="dv">3940785</span>, <span class="at">end =</span> <span class="dv">3958060</span>, <span class="at">genomeId =</span> <span class="st">"g_194"</span>)</span>
<span id="cb23-27"><a href="#cb23-27"></a>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</section>
<section id="phage_grp_71" class="level3">
<h3 class="anchored" data-anchor-id="phage_grp_71">phage_grp_71</h3>
<div class="sourceCode" id="cb24"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb24-1"><a href="#cb24-1"></a>cluster_title <span class="ot">&lt;-</span> <span class="st">"phage_grp_71"</span></span>
<span id="cb24-2"><a href="#cb24-2"></a>outDir <span class="ot">&lt;-</span> <span class="fu">paste</span>(confs<span class="sc">$</span>analysis<span class="sc">$</span>prophages<span class="sc">$</span>dir, <span class="st">"/cluster_viz/"</span>, cluster_title, <span class="at">sep =</span> <span class="st">""</span>)</span>
<span id="cb24-3"><a href="#cb24-3"></a>hg_color_categories <span class="ot">&lt;-</span> confs<span class="sc">$</span>analysis<span class="sc">$</span>prophages<span class="sc">$</span>files<span class="sc">$</span>hg_broad_functions</span>
<span id="cb24-4"><a href="#cb24-4"></a></span>
<span id="cb24-5"><a href="#cb24-5"></a><span class="co"># a vector of prophage identifiers that will be included in clustermap plot</span></span>
<span id="cb24-6"><a href="#cb24-6"></a>region_cluster <span class="ot">&lt;-</span> <span class="st">"phage_grp_71"</span></span>
<span id="cb24-7"><a href="#cb24-7"></a>other_regions <span class="ot">&lt;-</span> <span class="fu">c</span>()</span>
<span id="cb24-8"><a href="#cb24-8"></a></span>
<span id="cb24-9"><a href="#cb24-9"></a>subSample <span class="ot">&lt;-</span> <span class="cn">FALSE</span></span>
<span id="cb24-10"><a href="#cb24-10"></a>cutHeight <span class="ot">&lt;-</span> <span class="fl">0.5</span></span>
<span id="cb24-11"><a href="#cb24-11"></a>addFlankingRegions <span class="ot">&lt;-</span> <span class="cn">TRUE</span></span>
<span id="cb24-12"><a href="#cb24-12"></a>flankingRegion <span class="ot">&lt;-</span> <span class="dv">5000</span></span>
<span id="cb24-13"><a href="#cb24-13"></a></span>
<span id="cb24-14"><a href="#cb24-14"></a><span class="co"># ordering factor for prophages: "host" phylogeny, "hg_pav" for prophage HG PAV,</span></span>
<span id="cb24-15"><a href="#cb24-15"></a><span class="co"># "cluster_mash" for prophage MASH and "default" to use the provided</span></span>
<span id="cb24-16"><a href="#cb24-16"></a>clusterOrder <span class="ot">&lt;-</span> <span class="st">"hg_pav"</span> <span class="co"># host, hg_pav, cluster_mash, default</span></span>
<span id="cb24-17"><a href="#cb24-17"></a></span>
<span id="cb24-18"><a href="#cb24-18"></a><span class="co"># whether to keep custom regions at the bottom or consider during phylogeny</span></span>
<span id="cb24-19"><a href="#cb24-19"></a><span class="co"># based ordering</span></span>
<span id="cb24-20"><a href="#cb24-20"></a>regions_phy_ordered <span class="ot">&lt;-</span> <span class="cn">FALSE</span></span>
<span id="cb24-21"><a href="#cb24-21"></a></span>
<span id="cb24-22"><a href="#cb24-22"></a><span class="co"># regions to append as list of list with following structure</span></span>
<span id="cb24-23"><a href="#cb24-23"></a><span class="co"># list(r1 = list(chr, start, end, genomeId), r2 = list(chr, start, end, genomeId))</span></span>
<span id="cb24-24"><a href="#cb24-24"></a>customRegions <span class="ot">&lt;-</span> <span class="fu">list</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</section>
<section id="phage_grp_107" class="level3">
<h3 class="anchored" data-anchor-id="phage_grp_107">phage_grp_107</h3>
<div class="sourceCode" id="cb25"><pre class="sourceCode numberSource bash number-lines code-with-copy"><code class="sourceCode bash"><span id="cb25-1"><a href="#cb25-1"></a><span class="ex">Rscript</span> scripts/utils/HG_range_coordinates.R <span class="at">--hgs</span> hg_22427004,hg_22426993 <span class="dt">\</span></span>
<span id="cb25-2"><a href="#cb25-2"></a>--inner_region <span class="at">--genomes</span> <span class="st">"g_63,g_194"</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="sourceCode" id="cb26"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb26-1"><a href="#cb26-1"></a>cluster_title <span class="ot">&lt;-</span> <span class="st">"phage_grp_107"</span></span>
<span id="cb26-2"><a href="#cb26-2"></a>outDir <span class="ot">&lt;-</span> <span class="fu">paste</span>(confs<span class="sc">$</span>analysis<span class="sc">$</span>prophages<span class="sc">$</span>dir, <span class="st">"/cluster_viz"</span>, <span class="at">sep =</span> <span class="st">""</span>)</span>
<span id="cb26-3"><a href="#cb26-3"></a>hg_color_categories <span class="ot">&lt;-</span> confs<span class="sc">$</span>analysis<span class="sc">$</span>prophages<span class="sc">$</span>files<span class="sc">$</span>hg_broad_functions</span>
<span id="cb26-4"><a href="#cb26-4"></a></span>
<span id="cb26-5"><a href="#cb26-5"></a><span class="co"># a vector of prophage identifiers that will be included in clustermap plot</span></span>
<span id="cb26-6"><a href="#cb26-6"></a>region_cluster <span class="ot">&lt;-</span> <span class="st">"phage_grp_107"</span></span>
<span id="cb26-7"><a href="#cb26-7"></a>other_regions <span class="ot">&lt;-</span> <span class="fu">character</span>(<span class="dv">0</span>)</span>
<span id="cb26-8"><a href="#cb26-8"></a></span>
<span id="cb26-9"><a href="#cb26-9"></a>subSample <span class="ot">&lt;-</span> <span class="cn">FALSE</span></span>
<span id="cb26-10"><a href="#cb26-10"></a>cutHeight <span class="ot">&lt;-</span> <span class="fl">1.5</span></span>
<span id="cb26-11"><a href="#cb26-11"></a>addFlankingRegions <span class="ot">&lt;-</span> <span class="cn">TRUE</span></span>
<span id="cb26-12"><a href="#cb26-12"></a>flankingRegion <span class="ot">&lt;-</span> <span class="dv">5000</span></span>
<span id="cb26-13"><a href="#cb26-13"></a></span>
<span id="cb26-14"><a href="#cb26-14"></a><span class="co"># ordering factor for prophages: "host" phylogeny, "hg_pav" for prophage HG PAV,</span></span>
<span id="cb26-15"><a href="#cb26-15"></a><span class="co"># "cluster_mash" for prophage MASH and "default" to use the provided</span></span>
<span id="cb26-16"><a href="#cb26-16"></a>clusterOrder <span class="ot">&lt;-</span> <span class="st">"host"</span> <span class="co"># host, hg_pav, cluster_mash, default</span></span>
<span id="cb26-17"><a href="#cb26-17"></a></span>
<span id="cb26-18"><a href="#cb26-18"></a><span class="co"># whether to keep custom regions at the bottom or consider during phylogeny</span></span>
<span id="cb26-19"><a href="#cb26-19"></a><span class="co"># based ordering</span></span>
<span id="cb26-20"><a href="#cb26-20"></a>regions_phy_ordered <span class="ot">&lt;-</span> <span class="cn">TRUE</span> </span>
<span id="cb26-21"><a href="#cb26-21"></a></span>
<span id="cb26-22"><a href="#cb26-22"></a><span class="co"># regions to append as list of list with following structure</span></span>
<span id="cb26-23"><a href="#cb26-23"></a><span class="co"># list(r1 = list(chr, start, end, genomeId), r2 = list(chr, start, end, genomeId))</span></span>
<span id="cb26-24"><a href="#cb26-24"></a>customRegions <span class="ot">&lt;-</span> <span class="fu">list</span>(</span>
<span id="cb26-25"><a href="#cb26-25"></a>  <span class="at">g_63_reg =</span> <span class="fu">list</span>(<span class="at">chr =</span> <span class="st">"NZ_CP092039.1"</span>, <span class="at">start =</span> <span class="dv">2572138</span>, <span class="at">end =</span> <span class="dv">2573569</span>, <span class="at">genomeId =</span> <span class="st">"g_63"</span>),</span>
<span id="cb26-26"><a href="#cb26-26"></a>  <span class="at">g_194_reg =</span> <span class="fu">list</span>(<span class="at">chr =</span> <span class="st">"NZ_CP059956.1"</span>, <span class="at">start =</span> <span class="dv">2620331</span>, <span class="at">end =</span> <span class="dv">2621761</span>, <span class="at">genomeId =</span> <span class="st">"g_194"</span>)</span>
<span id="cb26-27"><a href="#cb26-27"></a>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</section>
<section id="bl-causing-clade-specific-signatures" class="level3">
<h3 class="anchored" data-anchor-id="bl-causing-clade-specific-signatures">BL-causing clade specific signatures</h3>
<div class="sourceCode" id="cb27"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb27-1"><a href="#cb27-1"></a>grpToView <span class="ot">&lt;-</span> <span class="st">"Pbr_FN_specific"</span></span>
<span id="cb27-2"><a href="#cb27-2"></a>subSample <span class="ot">&lt;-</span> <span class="cn">FALSE</span></span>
<span id="cb27-3"><a href="#cb27-3"></a>cutHeight <span class="ot">&lt;-</span> <span class="fl">0.5</span></span>
<span id="cb27-4"><a href="#cb27-4"></a>addFlankingRegions <span class="ot">&lt;-</span> <span class="cn">FALSE</span></span>
<span id="cb27-5"><a href="#cb27-5"></a>flankingRegion <span class="ot">&lt;-</span> <span class="dv">5000</span></span>
<span id="cb27-6"><a href="#cb27-6"></a></span>
<span id="cb27-7"><a href="#cb27-7"></a><span class="co"># ordering factor for prophages: host phylogeny, prophage HG PAV, prophage MASH,</span></span>
<span id="cb27-8"><a href="#cb27-8"></a><span class="co"># completeness score</span></span>
<span id="cb27-9"><a href="#cb27-9"></a>clusterOrder <span class="ot">&lt;-</span> <span class="st">"default"</span>  <span class="co"># completeness, host, hg_pav, cluster_mash, default</span></span>
<span id="cb27-10"><a href="#cb27-10"></a></span>
<span id="cb27-11"><a href="#cb27-11"></a></span>
<span id="cb27-12"><a href="#cb27-12"></a><span class="co"># optionally, a custom region list can be provided to generate the plot</span></span>
<span id="cb27-13"><a href="#cb27-13"></a>grp <span class="ot">&lt;-</span> <span class="fu">list</span>(</span>
<span id="cb27-14"><a href="#cb27-14"></a>  <span class="at">phage_grp =</span> grpToView,</span>
<span id="cb27-15"><a href="#cb27-15"></a>  <span class="at">members =</span> <span class="fu">c</span>(<span class="st">"g_302.vir_1"</span>, <span class="st">"g_400.vir_2"</span>, <span class="st">"g_409.vir_1"</span>, <span class="st">"g_302.vir_2"</span>)</span>
<span id="cb27-16"><a href="#cb27-16"></a>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</section>
<section id="phage_grp_36" class="level3">
<h3 class="anchored" data-anchor-id="phage_grp_36">phage_grp_36</h3>
<div class="sourceCode" id="cb28"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb28-1"><a href="#cb28-1"></a>grpToView <span class="ot">&lt;-</span> <span class="st">"phage_grp_36"</span></span>
<span id="cb28-2"><a href="#cb28-2"></a>subSample <span class="ot">&lt;-</span> <span class="cn">FALSE</span></span>
<span id="cb28-3"><a href="#cb28-3"></a>cutHeight <span class="ot">&lt;-</span> <span class="fl">1.5</span></span>
<span id="cb28-4"><a href="#cb28-4"></a>addFlankingRegions <span class="ot">&lt;-</span> <span class="cn">TRUE</span></span>
<span id="cb28-5"><a href="#cb28-5"></a>flankingRegion <span class="ot">&lt;-</span> <span class="dv">5000</span></span>
<span id="cb28-6"><a href="#cb28-6"></a></span>
<span id="cb28-7"><a href="#cb28-7"></a><span class="co"># ordering factor for prophages: host phylogeny, prophage HG PAV, prophage MASH,</span></span>
<span id="cb28-8"><a href="#cb28-8"></a><span class="co"># completeness score</span></span>
<span id="cb28-9"><a href="#cb28-9"></a>clusterOrder <span class="ot">&lt;-</span> <span class="st">"host"</span> <span class="co"># completeness, host, hg_pav, cluster_mash</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</section>
<section id="phage_grp_50" class="level3">
<h3 class="anchored" data-anchor-id="phage_grp_50">phage_grp_50</h3>
<div class="sourceCode" id="cb29"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb29-1"><a href="#cb29-1"></a>grpToView <span class="ot">&lt;-</span> <span class="st">"phage_grp_50"</span></span>
<span id="cb29-2"><a href="#cb29-2"></a>subSample <span class="ot">&lt;-</span> <span class="cn">FALSE</span></span>
<span id="cb29-3"><a href="#cb29-3"></a>addFlankingRegions <span class="ot">&lt;-</span> <span class="cn">TRUE</span> </span>
<span id="cb29-4"><a href="#cb29-4"></a>flankingRegion <span class="ot">&lt;-</span> <span class="dv">10000</span></span>
<span id="cb29-5"><a href="#cb29-5"></a></span>
<span id="cb29-6"><a href="#cb29-6"></a><span class="co"># ordering factor for prophages: host phylogeny, prophage HG PAV, prophage MASH,</span></span>
<span id="cb29-7"><a href="#cb29-7"></a><span class="co"># completeness score</span></span>
<span id="cb29-8"><a href="#cb29-8"></a>clusterOrder <span class="ot">&lt;-</span> <span class="st">"cluster_mash"</span>  <span class="co"># completeness, host, hg_pav, cluster_mash</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</section>
<section id="phage_grp_21" class="level3">
<h3 class="anchored" data-anchor-id="phage_grp_21">phage_grp_21</h3>
<div class="sourceCode" id="cb30"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb30-1"><a href="#cb30-1"></a>grpToView <span class="ot">&lt;-</span> <span class="st">"phage_grp_21"</span></span>
<span id="cb30-2"><a href="#cb30-2"></a>subSample <span class="ot">&lt;-</span> <span class="cn">FALSE</span></span>
<span id="cb30-3"><a href="#cb30-3"></a>addFlankingRegions <span class="ot">&lt;-</span> <span class="cn">TRUE</span></span>
<span id="cb30-4"><a href="#cb30-4"></a>flankingRegion <span class="ot">&lt;-</span> <span class="dv">5000</span></span>
<span id="cb30-5"><a href="#cb30-5"></a></span>
<span id="cb30-6"><a href="#cb30-6"></a><span class="co"># ordering factor for prophages: host phylogeny, prophage HG PAV, prophage MASH,</span></span>
<span id="cb30-7"><a href="#cb30-7"></a><span class="co"># completeness score</span></span>
<span id="cb30-8"><a href="#cb30-8"></a>clusterOrder <span class="ot">&lt;-</span> <span class="st">"cluster_mash"</span>  <span class="co"># completeness, host, hg_pav, cluster_mash</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</section>
<section id="t3ss-missing-from-4-p.-brasiliense-genomes" class="level3">
<h3 class="anchored" data-anchor-id="t3ss-missing-from-4-p.-brasiliense-genomes">T3SS missing from 4 P. brasiliense genomes</h3>
<div class="sourceCode" id="cb31"><pre class="sourceCode numberSource bash number-lines code-with-copy"><code class="sourceCode bash"><span id="cb31-1"><a href="#cb31-1"></a><span class="ex">Rscript</span> scripts/utils/HG_range_coordinates.R <span class="at">--hgs</span> hg_22430875,hg_22430876 <span class="at">--genomes</span> g_187,g_368,g_182,g_185,g_236,g_177</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="sourceCode" id="cb32"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb32-1"><a href="#cb32-1"></a>grpToView <span class="ot">&lt;-</span> <span class="st">"t3ss_regions"</span></span>
<span id="cb32-2"><a href="#cb32-2"></a>subSample <span class="ot">&lt;-</span> <span class="cn">FALSE</span></span>
<span id="cb32-3"><a href="#cb32-3"></a>cutHeight <span class="ot">&lt;-</span> <span class="fl">1.5</span></span>
<span id="cb32-4"><a href="#cb32-4"></a>addFlankingRegions <span class="ot">&lt;-</span> <span class="cn">TRUE</span></span>
<span id="cb32-5"><a href="#cb32-5"></a>flankingRegion <span class="ot">&lt;-</span> <span class="dv">6000</span></span>
<span id="cb32-6"><a href="#cb32-6"></a></span>
<span id="cb32-7"><a href="#cb32-7"></a><span class="co"># ordering factor for prophages: host phylogeny, prophage HG PAV, prophage MASH,</span></span>
<span id="cb32-8"><a href="#cb32-8"></a><span class="co"># completeness score</span></span>
<span id="cb32-9"><a href="#cb32-9"></a>clusterOrder <span class="ot">&lt;-</span> <span class="st">"host"</span>  <span class="co"># completeness, host, hg_pav, cluster_mash</span></span>
<span id="cb32-10"><a href="#cb32-10"></a></span>
<span id="cb32-11"><a href="#cb32-11"></a><span class="co"># a vector of prophage identifiers that will be included in clustermap plot</span></span>
<span id="cb32-12"><a href="#cb32-12"></a>appendPhages <span class="ot">&lt;-</span> <span class="fu">c</span>()</span>
<span id="cb32-13"><a href="#cb32-13"></a></span>
<span id="cb32-14"><a href="#cb32-14"></a><span class="co"># regions to append as list of list with following structure</span></span>
<span id="cb32-15"><a href="#cb32-15"></a><span class="co"># list(r1 = list(chr, start, end, genomeId), r2 = list(chr, start, end, genomeId))</span></span>
<span id="cb32-16"><a href="#cb32-16"></a>customRegions <span class="ot">&lt;-</span> <span class="fu">list</span>(</span>
<span id="cb32-17"><a href="#cb32-17"></a>  <span class="at">g_177_reg  =</span> <span class="fu">list</span>(<span class="at">genomeId =</span> <span class="st">"g_177"</span>, <span class="at">chr =</span> <span class="st">"NZ_JACGEP010000027.1"</span>, <span class="at">start =</span> <span class="dv">56850</span>, <span class="at">end =</span> <span class="dv">59337</span>),</span>
<span id="cb32-18"><a href="#cb32-18"></a>  <span class="at">g_182_reg =</span> <span class="fu">list</span>(<span class="at">genomeId =</span> <span class="st">"g_182"</span>, <span class="at">chr =</span> <span class="st">"NZ_JACGZZ010000068.1"</span>, <span class="at">start =</span> <span class="dv">116773</span>, <span class="at">end =</span> <span class="dv">119260</span>),</span>
<span id="cb32-19"><a href="#cb32-19"></a>  <span class="at">g_185_reg =</span> <span class="fu">list</span>(<span class="at">genomeId =</span> <span class="st">"g_185"</span>, <span class="at">chr =</span> <span class="st">"NZ_JACGEN010000027.1"</span>, <span class="at">start =</span> <span class="dv">117747</span>, <span class="at">end =</span> <span class="dv">120234</span>),</span>
<span id="cb32-20"><a href="#cb32-20"></a>  <span class="at">g_236_reg =</span> <span class="fu">list</span>(<span class="at">genomeId =</span> <span class="st">"g_236"</span>, <span class="at">chr =</span> <span class="st">"NZ_JACDSF010000026.1"</span>, <span class="at">start =</span> <span class="dv">2275</span>, <span class="at">end =</span> <span class="dv">4762</span>),</span>
<span id="cb32-21"><a href="#cb32-21"></a>  <span class="at">g_187_reg =</span> <span class="fu">list</span>(<span class="at">genomeId =</span> <span class="st">"g_187"</span>, <span class="at">chr =</span> <span class="st">"NZ_CP059963.1"</span>, <span class="at">start =</span> <span class="dv">2603432</span>, <span class="at">end =</span> <span class="dv">2605919</span>),</span>
<span id="cb32-22"><a href="#cb32-22"></a>  <span class="at">g_368_reg =</span> <span class="fu">list</span>(<span class="at">genomeId =</span> <span class="st">"g_368"</span>, <span class="at">chr =</span> <span class="st">"NZ_JUJT01000001.1"</span>, <span class="at">start =</span> <span class="dv">566617</span>, <span class="at">end =</span> <span class="dv">569104</span>)</span>
<span id="cb32-23"><a href="#cb32-23"></a>)</span>
<span id="cb32-24"><a href="#cb32-24"></a></span>
<span id="cb32-25"><a href="#cb32-25"></a><span class="co"># whether to keep custom regions at the bottom or consider during phylogeny</span></span>
<span id="cb32-26"><a href="#cb32-26"></a><span class="co"># based ordering</span></span>
<span id="cb32-27"><a href="#cb32-27"></a>regions_phy_ordered <span class="ot">&lt;-</span> <span class="cn">TRUE</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>


<!-- -->

</section>
</section>
</section>

<a onclick="window.scrollTo(0, 0); return false;" role="button" id="quarto-back-to-top"><i class="bi bi-arrow-up"></i> Back to top</a></main> <!-- /main -->
<script id="quarto-html-after-body" type="application/javascript">
window.document.addEventListener("DOMContentLoaded", function (event) {
  const toggleBodyColorMode = (bsSheetEl) => {
    const mode = bsSheetEl.getAttribute("data-mode");
    const bodyEl = window.document.querySelector("body");
    if (mode === "dark") {
      bodyEl.classList.add("quarto-dark");
      bodyEl.classList.remove("quarto-light");
    } else {
      bodyEl.classList.add("quarto-light");
      bodyEl.classList.remove("quarto-dark");
    }
  }
  const toggleBodyColorPrimary = () => {
    const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
    if (bsSheetEl) {
      toggleBodyColorMode(bsSheetEl);
    }
  }
  toggleBodyColorPrimary();  
  const icon = "";
  const anchorJS = new window.AnchorJS();
  anchorJS.options = {
    placement: 'right',
    icon: icon
  };
  anchorJS.add('.anchored');
  const isCodeAnnotation = (el) => {
    for (const clz of el.classList) {
      if (clz.startsWith('code-annotation-')) {                     
        return true;
      }
    }
    return false;
  }
  const onCopySuccess = function(e) {
    // button target
    const button = e.trigger;
    // don't keep focus
    button.blur();
    // flash "checked"
    button.classList.add('code-copy-button-checked');
    var currentTitle = button.getAttribute("title");
    button.setAttribute("title", "Copied!");
    let tooltip;
    if (window.bootstrap) {
      button.setAttribute("data-bs-toggle", "tooltip");
      button.setAttribute("data-bs-placement", "left");
      button.setAttribute("data-bs-title", "Copied!");
      tooltip = new bootstrap.Tooltip(button, 
        { trigger: "manual", 
          customClass: "code-copy-button-tooltip",
          offset: [0, -8]});
      tooltip.show();    
    }
    setTimeout(function() {
      if (tooltip) {
        tooltip.hide();
        button.removeAttribute("data-bs-title");
        button.removeAttribute("data-bs-toggle");
        button.removeAttribute("data-bs-placement");
      }
      button.setAttribute("title", currentTitle);
      button.classList.remove('code-copy-button-checked');
    }, 1000);
    // clear code selection
    e.clearSelection();
  }
  const getTextToCopy = function(trigger) {
      const codeEl = trigger.previousElementSibling.cloneNode(true);
      for (const childEl of codeEl.children) {
        if (isCodeAnnotation(childEl)) {
          childEl.remove();
        }
      }
      return codeEl.innerText;
  }
  const clipboard = new window.ClipboardJS('.code-copy-button:not([data-in-quarto-modal])', {
    text: getTextToCopy
  });
  clipboard.on('success', onCopySuccess);
  if (window.document.getElementById('quarto-embedded-source-code-modal')) {
    // For code content inside modals, clipBoardJS needs to be initialized with a container option
    // TODO: Check when it could be a function (https://github.com/zenorocha/clipboard.js/issues/860)
    const clipboardModal = new window.ClipboardJS('.code-copy-button[data-in-quarto-modal]', {
      text: getTextToCopy,
      container: window.document.getElementById('quarto-embedded-source-code-modal')
    });
    clipboardModal.on('success', onCopySuccess);
  }
  const viewSource = window.document.getElementById('quarto-view-source') ||
                     window.document.getElementById('quarto-code-tools-source');
  if (viewSource) {
    const sourceUrl = viewSource.getAttribute("data-quarto-source-url");
    viewSource.addEventListener("click", function(e) {
      if (sourceUrl) {
        // rstudio viewer pane
        if (/\bcapabilities=\b/.test(window.location)) {
          window.open(sourceUrl);
        } else {
          window.location.href = sourceUrl;
        }
      } else {
        const modal = new bootstrap.Modal(document.getElementById('quarto-embedded-source-code-modal'));
        modal.show();
      }
      return false;
    });
  }
  function toggleCodeHandler(show) {
    return function(e) {
      const detailsSrc = window.document.querySelectorAll(".cell > details > .sourceCode");
      for (let i=0; i<detailsSrc.length; i++) {
        const details = detailsSrc[i].parentElement;
        if (show) {
          details.open = true;
        } else {
          details.removeAttribute("open");
        }
      }
      const cellCodeDivs = window.document.querySelectorAll(".cell > .sourceCode");
      const fromCls = show ? "hidden" : "unhidden";
      const toCls = show ? "unhidden" : "hidden";
      for (let i=0; i<cellCodeDivs.length; i++) {
        const codeDiv = cellCodeDivs[i];
        if (codeDiv.classList.contains(fromCls)) {
          codeDiv.classList.remove(fromCls);
          codeDiv.classList.add(toCls);
        } 
      }
      return false;
    }
  }
  const hideAllCode = window.document.getElementById("quarto-hide-all-code");
  if (hideAllCode) {
    hideAllCode.addEventListener("click", toggleCodeHandler(false));
  }
  const showAllCode = window.document.getElementById("quarto-show-all-code");
  if (showAllCode) {
    showAllCode.addEventListener("click", toggleCodeHandler(true));
  }
    var localhostRegex = new RegExp(/^(?:http|https):\/\/localhost\:?[0-9]*\//);
    var mailtoRegex = new RegExp(/^mailto:/);
      var filterRegex = new RegExp('/' + window.location.host + '/');
    var isInternal = (href) => {
        return filterRegex.test(href) || localhostRegex.test(href) || mailtoRegex.test(href);
    }
    // Inspect non-navigation links and adorn them if external
 	var links = window.document.querySelectorAll('a[href]:not(.nav-link):not(.navbar-brand):not(.toc-action):not(.sidebar-link):not(.sidebar-item-toggle):not(.pagination-link):not(.no-external):not([aria-hidden]):not(.dropdown-item):not(.quarto-navigation-tool):not(.about-link)');
    for (var i=0; i<links.length; i++) {
      const link = links[i];
      if (!isInternal(link.href)) {
        // undo the damage that might have been done by quarto-nav.js in the case of
        // links that we want to consider external
        if (link.dataset.originalHref !== undefined) {
          link.href = link.dataset.originalHref;
        }
      }
    }
  function tippyHover(el, contentFn, onTriggerFn, onUntriggerFn) {
    const config = {
      allowHTML: true,
      maxWidth: 500,
      delay: 100,
      arrow: false,
      appendTo: function(el) {
          return el.parentElement;
      },
      interactive: true,
      interactiveBorder: 10,
      theme: 'quarto',
      placement: 'bottom-start',
    };
    if (contentFn) {
      config.content = contentFn;
    }
    if (onTriggerFn) {
      config.onTrigger = onTriggerFn;
    }
    if (onUntriggerFn) {
      config.onUntrigger = onUntriggerFn;
    }
    window.tippy(el, config); 
  }
  const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
  for (var i=0; i<noterefs.length; i++) {
    const ref = noterefs[i];
    tippyHover(ref, function() {
      // use id or data attribute instead here
      let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
      try { href = new URL(href).hash; } catch {}
      const id = href.replace(/^#\/?/, "");
      const note = window.document.getElementById(id);
      if (note) {
        return note.innerHTML;
      } else {
        return "";
      }
    });
  }
  const xrefs = window.document.querySelectorAll('a.quarto-xref');
  const processXRef = (id, note) => {
    // Strip column container classes
    const stripColumnClz = (el) => {
      el.classList.remove("page-full", "page-columns");
      if (el.children) {
        for (const child of el.children) {
          stripColumnClz(child);
        }
      }
    }
    stripColumnClz(note)
    if (id === null || id.startsWith('sec-')) {
      // Special case sections, only their first couple elements
      const container = document.createElement("div");
      if (note.children && note.children.length > 2) {
        container.appendChild(note.children[0].cloneNode(true));
        for (let i = 1; i < note.children.length; i++) {
          const child = note.children[i];
          if (child.tagName === "P" && child.innerText === "") {
            continue;
          } else {
            container.appendChild(child.cloneNode(true));
            break;
          }
        }
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(container);
        }
        return container.innerHTML
      } else {
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(note);
        }
        return note.innerHTML;
      }
    } else {
      // Remove any anchor links if they are present
      const anchorLink = note.querySelector('a.anchorjs-link');
      if (anchorLink) {
        anchorLink.remove();
      }
      if (window.Quarto?.typesetMath) {
        window.Quarto.typesetMath(note);
      }
      // TODO in 1.5, we should make sure this works without a callout special case
      if (note.classList.contains("callout")) {
        return note.outerHTML;
      } else {
        return note.innerHTML;
      }
    }
  }
  for (var i=0; i<xrefs.length; i++) {
    const xref = xrefs[i];
    tippyHover(xref, undefined, function(instance) {
      instance.disable();
      let url = xref.getAttribute('href');
      let hash = undefined; 
      if (url.startsWith('#')) {
        hash = url;
      } else {
        try { hash = new URL(url).hash; } catch {}
      }
      if (hash) {
        const id = hash.replace(/^#\/?/, "");
        const note = window.document.getElementById(id);
        if (note !== null) {
          try {
            const html = processXRef(id, note.cloneNode(true));
            instance.setContent(html);
          } finally {
            instance.enable();
            instance.show();
          }
        } else {
          // See if we can fetch this
          fetch(url.split('#')[0])
          .then(res => res.text())
          .then(html => {
            const parser = new DOMParser();
            const htmlDoc = parser.parseFromString(html, "text/html");
            const note = htmlDoc.getElementById(id);
            if (note !== null) {
              const html = processXRef(id, note);
              instance.setContent(html);
            } 
          }).finally(() => {
            instance.enable();
            instance.show();
          });
        }
      } else {
        // See if we can fetch a full url (with no hash to target)
        // This is a special case and we should probably do some content thinning / targeting
        fetch(url)
        .then(res => res.text())
        .then(html => {
          const parser = new DOMParser();
          const htmlDoc = parser.parseFromString(html, "text/html");
          const note = htmlDoc.querySelector('main.content');
          if (note !== null) {
            // This should only happen for chapter cross references
            // (since there is no id in the URL)
            // remove the first header
            if (note.children.length > 0 && note.children[0].tagName === "HEADER") {
              note.children[0].remove();
            }
            const html = processXRef(null, note);
            instance.setContent(html);
          } 
        }).finally(() => {
          instance.enable();
          instance.show();
        });
      }
    }, function(instance) {
    });
  }
      let selectedAnnoteEl;
      const selectorForAnnotation = ( cell, annotation) => {
        let cellAttr = 'data-code-cell="' + cell + '"';
        let lineAttr = 'data-code-annotation="' +  annotation + '"';
        const selector = 'span[' + cellAttr + '][' + lineAttr + ']';
        return selector;
      }
      const selectCodeLines = (annoteEl) => {
        const doc = window.document;
        const targetCell = annoteEl.getAttribute("data-target-cell");
        const targetAnnotation = annoteEl.getAttribute("data-target-annotation");
        const annoteSpan = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
        const lines = annoteSpan.getAttribute("data-code-lines").split(",");
        const lineIds = lines.map((line) => {
          return targetCell + "-" + line;
        })
        let top = null;
        let height = null;
        let parent = null;
        if (lineIds.length > 0) {
            //compute the position of the single el (top and bottom and make a div)
            const el = window.document.getElementById(lineIds[0]);
            top = el.offsetTop;
            height = el.offsetHeight;
            parent = el.parentElement.parentElement;
          if (lineIds.length > 1) {
            const lastEl = window.document.getElementById(lineIds[lineIds.length - 1]);
            const bottom = lastEl.offsetTop + lastEl.offsetHeight;
            height = bottom - top;
          }
          if (top !== null && height !== null && parent !== null) {
            // cook up a div (if necessary) and position it 
            let div = window.document.getElementById("code-annotation-line-highlight");
            if (div === null) {
              div = window.document.createElement("div");
              div.setAttribute("id", "code-annotation-line-highlight");
              div.style.position = 'absolute';
              parent.appendChild(div);
            }
            div.style.top = top - 2 + "px";
            div.style.height = height + 4 + "px";
            div.style.left = 0;
            let gutterDiv = window.document.getElementById("code-annotation-line-highlight-gutter");
            if (gutterDiv === null) {
              gutterDiv = window.document.createElement("div");
              gutterDiv.setAttribute("id", "code-annotation-line-highlight-gutter");
              gutterDiv.style.position = 'absolute';
              const codeCell = window.document.getElementById(targetCell);
              const gutter = codeCell.querySelector('.code-annotation-gutter');
              gutter.appendChild(gutterDiv);
            }
            gutterDiv.style.top = top - 2 + "px";
            gutterDiv.style.height = height + 4 + "px";
          }
          selectedAnnoteEl = annoteEl;
        }
      };
      const unselectCodeLines = () => {
        const elementsIds = ["code-annotation-line-highlight", "code-annotation-line-highlight-gutter"];
        elementsIds.forEach((elId) => {
          const div = window.document.getElementById(elId);
          if (div) {
            div.remove();
          }
        });
        selectedAnnoteEl = undefined;
      };
        // Handle positioning of the toggle
    window.addEventListener(
      "resize",
      throttle(() => {
        elRect = undefined;
        if (selectedAnnoteEl) {
          selectCodeLines(selectedAnnoteEl);
        }
      }, 10)
    );
    function throttle(fn, ms) {
    let throttle = false;
    let timer;
      return (...args) => {
        if(!throttle) { // first call gets through
            fn.apply(this, args);
            throttle = true;
        } else { // all the others get throttled
            if(timer) clearTimeout(timer); // cancel #2
            timer = setTimeout(() => {
              fn.apply(this, args);
              timer = throttle = false;
            }, ms);
        }
      };
    }
      // Attach click handler to the DT
      const annoteDls = window.document.querySelectorAll('dt[data-target-cell]');
      for (const annoteDlNode of annoteDls) {
        annoteDlNode.addEventListener('click', (event) => {
          const clickedEl = event.target;
          if (clickedEl !== selectedAnnoteEl) {
            unselectCodeLines();
            const activeEl = window.document.querySelector('dt[data-target-cell].code-annotation-active');
            if (activeEl) {
              activeEl.classList.remove('code-annotation-active');
            }
            selectCodeLines(clickedEl);
            clickedEl.classList.add('code-annotation-active');
          } else {
            // Unselect the line
            unselectCodeLines();
            clickedEl.classList.remove('code-annotation-active');
          }
        });
      }
  const findCites = (el) => {
    const parentEl = el.parentElement;
    if (parentEl) {
      const cites = parentEl.dataset.cites;
      if (cites) {
        return {
          el,
          cites: cites.split(' ')
        };
      } else {
        return findCites(el.parentElement)
      }
    } else {
      return undefined;
    }
  };
  var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
  for (var i=0; i<bibliorefs.length; i++) {
    const ref = bibliorefs[i];
    const citeInfo = findCites(ref);
    if (citeInfo) {
      tippyHover(citeInfo.el, function() {
        var popup = window.document.createElement('div');
        citeInfo.cites.forEach(function(cite) {
          var citeDiv = window.document.createElement('div');
          citeDiv.classList.add('hanging-indent');
          citeDiv.classList.add('csl-entry');
          var biblioDiv = window.document.getElementById('ref-' + cite);
          if (biblioDiv) {
            citeDiv.innerHTML = biblioDiv.innerHTML;
          }
          popup.appendChild(citeDiv);
        });
        return popup.innerHTML;
      });
    }
  }
});
</script>
<nav class="page-navigation">
  <div class="nav-page nav-page-previous">
      <a href="../../scripts/analysis/phylo_viz_tree.html" class="pagination-link" aria-label="Phylogeny">
        <i class="bi bi-arrow-left-short"></i> <span class="nav-page-text">Phylogeny</span>
      </a>          
  </div>
  <div class="nav-page nav-page-next">
      <a href="../../scripts/analysis/prophage_summary.html" class="pagination-link" aria-label="Prophage summary">
        <span class="nav-page-text">Prophage summary</span> <i class="bi bi-arrow-right-short"></i>
      </a>
  </div>
</nav><div class="modal fade" id="quarto-embedded-source-code-modal" tabindex="-1" aria-labelledby="quarto-embedded-source-code-modal-label" aria-hidden="true"><div class="modal-dialog modal-dialog-scrollable"><div class="modal-content"><div class="modal-header"><h5 class="modal-title" id="quarto-embedded-source-code-modal-label">Source Code</h5><button class="btn-close" data-bs-dismiss="modal"></button></div><div class="modal-body"><div class="">
<div class="sourceCode" id="cb33" data-shortcodes="false"><pre class="sourceCode numberSource markdown number-lines code-with-copy"><code class="sourceCode markdown"><span id="cb33-1"><a href="#cb33-1"></a><span class="fu"># Prophage analysis in *Pectobacterium* pangenome</span></span>
<span id="cb33-2"><a href="#cb33-2"></a></span>
<span id="cb33-3"><a href="#cb33-3"></a><span class="fu">## Detect prophages in genome </span></span>
<span id="cb33-4"><a href="#cb33-4"></a></span>
<span id="cb33-5"><a href="#cb33-5"></a><span class="fu">### genomad prophage detection</span></span>
<span id="cb33-6"><a href="#cb33-6"></a></span>
<span id="cb33-7"><a href="#cb33-7"></a><span class="in">```bash</span></span>
<span id="cb33-8"><a href="#cb33-8"></a><span class="co"># run genomad and checkv pipeline</span></span>
<span id="cb33-9"><a href="#cb33-9"></a><span class="fu">nohup</span> <span class="dt">\</span></span>
<span id="cb33-10"><a href="#cb33-10"></a>cat data/reference_data/assembly_ids.txt <span class="kw">|</span> <span class="dt">\</span></span>
<span id="cb33-11"><a href="#cb33-11"></a><span class="ex">parallel</span> <span class="at">--jobs</span> 6 <span class="at">--workdir</span> <span class="va">$PWD</span> <span class="at">--halt</span> now,fail=1 <span class="dt">\</span></span>
<span id="cb33-12"><a href="#cb33-12"></a>--keep-order <span class="at">--results</span> logs/genomad/{} <span class="dt">\</span></span>
<span id="cb33-13"><a href="#cb33-13"></a>--joblog logs/genomad/parallel.log <span class="dt">\</span></span>
<span id="cb33-14"><a href="#cb33-14"></a>scripts/preprocessing/genomad_prophage_annotation.sh {} <span class="dt">\</span></span>
<span id="cb33-15"><a href="#cb33-15"></a><span class="op">&gt;&gt;</span>logs/genomad/nohup.out <span class="dv">2</span><span class="op">&gt;&amp;</span><span class="dv">1</span> <span class="kw">&amp;</span></span>
<span id="cb33-16"><a href="#cb33-16"></a></span>
<span id="cb33-17"><a href="#cb33-17"></a><span class="co"># count of prophage detected in each genome</span></span>
<span id="cb33-18"><a href="#cb33-18"></a><span class="cf">for</span> i <span class="kw">in</span> <span class="kw">`</span><span class="fu">cat</span> data/reference_data/assembly_ids.txt<span class="kw">`</span></span>
<span id="cb33-19"><a href="#cb33-19"></a><span class="cf">do</span></span>
<span id="cb33-20"><a href="#cb33-20"></a>  <span class="bu">printf</span> <span class="st">"</span><span class="va">$i</span><span class="st">\t"</span></span>
<span id="cb33-21"><a href="#cb33-21"></a>  <span class="fu">tail</span> <span class="at">-n</span> +2 data/prophage_genomad/<span class="va">$i</span>/<span class="va">${i}</span>_summary/<span class="va">${i}</span>_virus_summary.tsv <span class="kw">|</span> <span class="fu">wc</span> <span class="at">-l</span></span>
<span id="cb33-22"><a href="#cb33-22"></a><span class="cf">done</span> <span class="kw">|</span> <span class="fu">sort</span> <span class="at">-nr</span> <span class="at">-k2</span></span>
<span id="cb33-23"><a href="#cb33-23"></a></span>
<span id="cb33-24"><a href="#cb33-24"></a><span class="in">```</span></span>
<span id="cb33-25"><a href="#cb33-25"></a></span>
<span id="cb33-26"><a href="#cb33-26"></a>Check the checkV contamination.tsv file to see if there is any other contamination</span>
<span id="cb33-27"><a href="#cb33-27"></a>than the expected <span class="in">`host,viral`</span>, <span class="in">`viral,host`</span> and <span class="in">`host,viral,host`</span> ones.</span>
<span id="cb33-28"><a href="#cb33-28"></a></span>
<span id="cb33-29"><a href="#cb33-29"></a><span class="in">```bash</span></span>
<span id="cb33-30"><a href="#cb33-30"></a><span class="cf">for</span> cn <span class="kw">in</span> <span class="kw">`</span><span class="fu">find</span> <span class="at">-name</span> <span class="st">'*contamination.tsv'</span><span class="kw">`</span></span>
<span id="cb33-31"><a href="#cb33-31"></a><span class="cf">do</span></span>
<span id="cb33-32"><a href="#cb33-32"></a><span class="fu">cat</span> <span class="va">$cn</span> <span class="kw">|</span> <span class="fu">awk</span> <span class="st">'$9 != "NA" &amp;&amp; NR != 1 &amp;&amp; $9 != "host,viral" &amp;&amp; $9 != "viral,host" &amp;&amp; $9 != "host,viral,host"'</span></span>
<span id="cb33-33"><a href="#cb33-33"></a><span class="cf">done</span></span>
<span id="cb33-34"><a href="#cb33-34"></a><span class="in">```</span></span>
<span id="cb33-35"><a href="#cb33-35"></a></span>
<span id="cb33-36"><a href="#cb33-36"></a><span class="fu">### Post processing of prophages </span></span>
<span id="cb33-37"><a href="#cb33-37"></a></span>
<span id="cb33-38"><a href="#cb33-38"></a>Process genomad output and separate prophages into independent FASTA files. Each</span>
<span id="cb33-39"><a href="#cb33-39"></a>prophage is assigned a unique identifier with format <span class="in">`g_{genome}.vir_{#count}`</span>.</span>
<span id="cb33-40"><a href="#cb33-40"></a>Prophages are saved in independent FASTA files for downstream analysis, for e.g.</span>
<span id="cb33-41"><a href="#cb33-41"></a>ANI clustering of prophages.</span>
<span id="cb33-42"><a href="#cb33-42"></a></span>
<span id="cb33-43"><a href="#cb33-43"></a><span class="in">```bash</span></span>
<span id="cb33-44"><a href="#cb33-44"></a><span class="co"># post processing of prophage data</span></span>
<span id="cb33-45"><a href="#cb33-45"></a><span class="ex">Rscript</span> scripts/preprocessing/genomad_checkv_merge.R</span>
<span id="cb33-46"><a href="#cb33-46"></a></span>
<span id="cb33-47"><a href="#cb33-47"></a><span class="co"># save prophages independently</span></span>
<span id="cb33-48"><a href="#cb33-48"></a><span class="fu">bash</span> scripts/preprocessing/prophage_genomes_extract.sh</span>
<span id="cb33-49"><a href="#cb33-49"></a><span class="in">```</span></span>
<span id="cb33-50"><a href="#cb33-50"></a></span>
<span id="cb33-51"><a href="#cb33-51"></a>Output:</span>
<span id="cb33-52"><a href="#cb33-52"></a><span class="ss">- </span>yaml: <span class="in">`confs$data$prophages$files$data`</span></span>
<span id="cb33-53"><a href="#cb33-53"></a><span class="ss">- </span>file: <span class="in">`data/prophage_genomad/prophage_summary.tab`</span></span>
<span id="cb33-54"><a href="#cb33-54"></a></span>
<span id="cb33-55"><a href="#cb33-55"></a>Extract homology groups for prophages from the pangenome org.db object.</span>
<span id="cb33-56"><a href="#cb33-56"></a></span>
<span id="cb33-57"><a href="#cb33-57"></a><span class="in">```bash</span></span>
<span id="cb33-58"><a href="#cb33-58"></a><span class="ex">Rscript</span> scripts/analysis/prophage_homology_groups.R</span>
<span id="cb33-59"><a href="#cb33-59"></a><span class="in">```</span></span>
<span id="cb33-60"><a href="#cb33-60"></a></span>
<span id="cb33-61"><a href="#cb33-61"></a>Output:</span>
<span id="cb33-62"><a href="#cb33-62"></a><span class="ss">- </span>yaml: <span class="in">`confs$analysis$prophages$preprocessing$files$raw_prophage_hg`</span></span>
<span id="cb33-63"><a href="#cb33-63"></a><span class="ss">- </span>file: <span class="in">`analysis/pangenome_v2/prophages/preprocessing/prophage_hgs.tab`</span></span>
<span id="cb33-64"><a href="#cb33-64"></a></span>
<span id="cb33-65"><a href="#cb33-65"></a>Run ANI and MASH on prophage genomes.</span>
<span id="cb33-66"><a href="#cb33-66"></a></span>
<span id="cb33-67"><a href="#cb33-67"></a><span class="in">```bash</span></span>
<span id="cb33-68"><a href="#cb33-68"></a><span class="ex">conda</span> activate pantools_v4_3</span>
<span id="cb33-69"><a href="#cb33-69"></a></span>
<span id="cb33-70"><a href="#cb33-70"></a><span class="co"># list prophage FASTA files</span></span>
<span id="cb33-71"><a href="#cb33-71"></a><span class="fu">ls</span> data/prophage_genomad/phage_seqs/<span class="pp">*</span>.fna  <span class="op">&gt;</span> data/prophage_genomad/prophage_fasta.list</span>
<span id="cb33-72"><a href="#cb33-72"></a></span>
<span id="cb33-73"><a href="#cb33-73"></a><span class="co"># run fastANI</span></span>
<span id="cb33-74"><a href="#cb33-74"></a><span class="ex">fastANI</span> <span class="at">--ql</span> data/prophage_genomad/prophage_fasta.list <span class="dt">\</span></span>
<span id="cb33-75"><a href="#cb33-75"></a>--rl data/prophage_genomad/prophage_fasta.list <span class="dt">\</span></span>
<span id="cb33-76"><a href="#cb33-76"></a>--threads 30 <span class="at">--matrix</span> <span class="at">--visualize</span> <span class="at">--fragLen</span> 500 <span class="dt">\</span></span>
<span id="cb33-77"><a href="#cb33-77"></a>-o data/prophage_genomad/prophage_ANI</span>
<span id="cb33-78"><a href="#cb33-78"></a></span>
<span id="cb33-79"><a href="#cb33-79"></a><span class="co"># use mash for distance estimation</span></span>
<span id="cb33-80"><a href="#cb33-80"></a><span class="fu">mkdir</span> data/prophage_genomad/mash</span>
<span id="cb33-81"><a href="#cb33-81"></a></span>
<span id="cb33-82"><a href="#cb33-82"></a><span class="ex">mash</span> sketch <span class="at">-p</span> 30 <span class="at">-s</span> 15000 <span class="at">-k</span> 17 <span class="at">-o</span> data/prophage_genomad/mash/prophages_ref <span class="dt">\</span></span>
<span id="cb33-83"><a href="#cb33-83"></a>-l data/prophage_genomad/prophage_fasta.list <span class="op">&gt;</span>logs/prophage/mash.log <span class="dv">2</span><span class="op">&gt;&amp;</span><span class="dv">1</span></span>
<span id="cb33-84"><a href="#cb33-84"></a></span>
<span id="cb33-85"><a href="#cb33-85"></a><span class="ex">mash</span> info data/prophage_genomad/mash/prophages_ref.msh <span class="op">&gt;</span> <span class="dt">\</span></span>
<span id="cb33-86"><a href="#cb33-86"></a>data/prophage_genomad/mash/prophage_sketch_info.txt</span>
<span id="cb33-87"><a href="#cb33-87"></a></span>
<span id="cb33-88"><a href="#cb33-88"></a><span class="ex">mash</span> dist <span class="at">-p</span> 30 data/prophage_genomad/mash/prophages_ref.msh <span class="dt">\</span></span>
<span id="cb33-89"><a href="#cb33-89"></a>data/prophage_genomad/mash/prophages_ref.msh <span class="op">&gt;</span> <span class="dt">\</span></span>
<span id="cb33-90"><a href="#cb33-90"></a>data/prophage_genomad/mash/prophage_distance.tab</span>
<span id="cb33-91"><a href="#cb33-91"></a></span>
<span id="cb33-92"><a href="#cb33-92"></a><span class="in">```</span></span>
<span id="cb33-93"><a href="#cb33-93"></a></span>
<span id="cb33-94"><a href="#cb33-94"></a><span class="fu">### QC</span></span>
<span id="cb33-95"><a href="#cb33-95"></a></span>
<span id="cb33-96"><a href="#cb33-96"></a>Identify the fragmented prophages resulting because of the genome assembly</span>
<span id="cb33-97"><a href="#cb33-97"></a>that can be merged into a complete prophage. Prepare a consolidated list of</span>
<span id="cb33-98"><a href="#cb33-98"></a>prophages which will be used for the subsequent analysis.</span>
<span id="cb33-99"><a href="#cb33-99"></a></span>
<span id="cb33-100"><a href="#cb33-100"></a><span class="in">```bash</span></span>
<span id="cb33-101"><a href="#cb33-101"></a><span class="ex">Rscript</span> scripts/analysis/prophage_merge.R</span>
<span id="cb33-102"><a href="#cb33-102"></a><span class="in">```</span></span>
<span id="cb33-103"><a href="#cb33-103"></a></span>
<span id="cb33-104"><a href="#cb33-104"></a>Output:</span>
<span id="cb33-105"><a href="#cb33-105"></a><span class="ss">- </span>yaml: <span class="in">`confs$analysis$prophages$preprocessing$files$consolidated`</span></span>
<span id="cb33-106"><a href="#cb33-106"></a><span class="ss">- </span>file: <span class="in">`analysis/prophages/preprocessing/consolidated_phages.tab`</span></span>
<span id="cb33-107"><a href="#cb33-107"></a></span>
<span id="cb33-108"><a href="#cb33-108"></a>Summarize the consolidated prophage list to generate various statistics and </span>
<span id="cb33-109"><a href="#cb33-109"></a>generate summary plots. Prophages that are fragmented or smaller than the length</span>
<span id="cb33-110"><a href="#cb33-110"></a>cutoff are removed and the filtered list of prophages is saved for downstream</span>
<span id="cb33-111"><a href="#cb33-111"></a>analysis.</span>
<span id="cb33-112"><a href="#cb33-112"></a></span>
<span id="cb33-113"><a href="#cb33-113"></a><span class="in">```bash</span></span>
<span id="cb33-114"><a href="#cb33-114"></a><span class="ex">Rscript</span> scripts/analysis/prophage_summary.R</span>
<span id="cb33-115"><a href="#cb33-115"></a><span class="in">```</span></span>
<span id="cb33-116"><a href="#cb33-116"></a></span>
<span id="cb33-117"><a href="#cb33-117"></a>Output:</span>
<span id="cb33-118"><a href="#cb33-118"></a><span class="ss">- </span>yaml: <span class="in">`confs$analysis$prophages$preprocessing$files$prophage_pool`</span></span>
<span id="cb33-119"><a href="#cb33-119"></a><span class="ss">- </span>File: <span class="in">`analysis/pangenome_v2/prophages/preprocessing/prophages_pool.tab`</span></span>
<span id="cb33-120"><a href="#cb33-120"></a></span>
<span id="cb33-121"><a href="#cb33-121"></a>Process prophage MASH and ANI results and visualize MASH similarity matrix for</span>
<span id="cb33-122"><a href="#cb33-122"></a>the filtered prophages.</span>
<span id="cb33-123"><a href="#cb33-123"></a></span>
<span id="cb33-124"><a href="#cb33-124"></a><span class="in">```bash</span></span>
<span id="cb33-125"><a href="#cb33-125"></a><span class="ex">quarto</span> render scripts/analysis/prophage_seq_similarity.qmd <span class="at">--execute-dir</span> ./</span>
<span id="cb33-126"><a href="#cb33-126"></a><span class="in">```</span></span>
<span id="cb33-127"><a href="#cb33-127"></a></span>
<span id="cb33-128"><a href="#cb33-128"></a><span class="fu">## Cluster prophages</span></span>
<span id="cb33-129"><a href="#cb33-129"></a></span>
<span id="cb33-130"><a href="#cb33-130"></a>Use shared homology groups between two prophages to establish the child-parent</span>
<span id="cb33-131"><a href="#cb33-131"></a>relationship between the prophages in pangenome. First, all vs all syntenic</span>
<span id="cb33-132"><a href="#cb33-132"></a>Jaccard similarity is calculated between the prophages. This information is later</span>
<span id="cb33-133"><a href="#cb33-133"></a>used to cluster the prophages using two different approaches.</span>
<span id="cb33-134"><a href="#cb33-134"></a></span>
<span id="cb33-135"><a href="#cb33-135"></a><span class="in">```bash</span></span>
<span id="cb33-136"><a href="#cb33-136"></a><span class="ex">Rscript</span> scripts/analysis/prophage_pairwise_similarity.R</span>
<span id="cb33-137"><a href="#cb33-137"></a><span class="in">```</span></span>
<span id="cb33-138"><a href="#cb33-138"></a></span>
<span id="cb33-139"><a href="#cb33-139"></a><span class="fu">### Approach 1</span></span>
<span id="cb33-140"><a href="#cb33-140"></a></span>
<span id="cb33-141"><a href="#cb33-141"></a>A heuristic DAG construcion is extracted from the output of the script </span>
<span id="cb33-142"><a href="#cb33-142"></a><span class="in">`scripts/analysis/prophage_merge.R`</span>. This output includes a best matching parent</span>
<span id="cb33-143"><a href="#cb33-143"></a>prophage for each prophage. This is further processed to build a DAG and</span>
<span id="cb33-144"><a href="#cb33-144"></a>visualize the clusters.</span>
<span id="cb33-145"><a href="#cb33-145"></a></span>
<span id="cb33-146"><a href="#cb33-146"></a>Post processing of the child-parent relationships to add additional metadata</span>
<span id="cb33-147"><a href="#cb33-147"></a>and save a DAG in tabular format:</span>
<span id="cb33-148"><a href="#cb33-148"></a></span>
<span id="cb33-149"><a href="#cb33-149"></a><span class="in">```bash</span></span>
<span id="cb33-150"><a href="#cb33-150"></a><span class="ex">Rscript</span> scripts/analysis/prophage_DAG_postprocess.R</span>
<span id="cb33-151"><a href="#cb33-151"></a><span class="in">```</span></span>
<span id="cb33-152"><a href="#cb33-152"></a></span>
<span id="cb33-153"><a href="#cb33-153"></a>Visualize the clusters of prophages as directed acyclic graphs:</span>
<span id="cb33-154"><a href="#cb33-154"></a></span>
<span id="cb33-155"><a href="#cb33-155"></a><span class="in">```r</span></span>
<span id="cb33-156"><a href="#cb33-156"></a>knitr<span class="sc">::</span><span class="fu">knit</span>(<span class="st">"scripts/analysis/prophage_DAG_viz.qmd"</span>)</span>
<span id="cb33-157"><a href="#cb33-157"></a><span class="in">```</span></span>
<span id="cb33-158"><a href="#cb33-158"></a></span>
<span id="cb33-159"><a href="#cb33-159"></a><span class="fu">### Approach 2</span></span>
<span id="cb33-160"><a href="#cb33-160"></a></span>
<span id="cb33-161"><a href="#cb33-161"></a>We first tried to use the Affinity propagation (AP) clustering to group the</span>
<span id="cb33-162"><a href="#cb33-162"></a>prophages into clusters and have a set of cluster representatives. However,</span>
<span id="cb33-163"><a href="#cb33-163"></a>we encountered some clusters generated by the AP clustering where the cluster</span>
<span id="cb33-164"><a href="#cb33-164"></a>members did not even share a homology group. Therefore, we decided not to use</span>
<span id="cb33-165"><a href="#cb33-165"></a>AP clustering.</span>
<span id="cb33-166"><a href="#cb33-166"></a></span>
<span id="cb33-167"><a href="#cb33-167"></a><span class="in">```r</span></span>
<span id="cb33-168"><a href="#cb33-168"></a>knitr<span class="sc">::</span><span class="fu">knit</span>(<span class="st">"scripts/analysis/prophage_AP_clustering.qmd"</span>)</span>
<span id="cb33-169"><a href="#cb33-169"></a><span class="in">```</span></span>
<span id="cb33-170"><a href="#cb33-170"></a></span>
<span id="cb33-171"><a href="#cb33-171"></a>Cluster the all-vs-all syntenic Jaccard distance matrix using "complete" linkage</span>
<span id="cb33-172"><a href="#cb33-172"></a>hierarchical clustering and cut the cluster using carotovoricin cluster as a</span>
<span id="cb33-173"><a href="#cb33-173"></a>reference.</span>
<span id="cb33-174"><a href="#cb33-174"></a></span>
<span id="cb33-175"><a href="#cb33-175"></a><span class="in">```r</span></span>
<span id="cb33-176"><a href="#cb33-176"></a>knitr<span class="sc">::</span><span class="fu">knit</span>(<span class="st">"scripts/analysis/prophage_clustering.qmd"</span>)</span>
<span id="cb33-177"><a href="#cb33-177"></a><span class="in">```</span></span>
<span id="cb33-178"><a href="#cb33-178"></a></span>
<span id="cb33-179"><a href="#cb33-179"></a>Output:</span>
<span id="cb33-180"><a href="#cb33-180"></a><span class="ss">- </span>yaml: <span class="in">`confs$analysis$prophages$files$clusters`</span></span>
<span id="cb33-181"><a href="#cb33-181"></a><span class="ss">- </span>file: <span class="in">`analysis/pangenome_v2/prophages/prophages_clusters.tab`</span></span>
<span id="cb33-182"><a href="#cb33-182"></a></span>
<span id="cb33-183"><a href="#cb33-183"></a><span class="fu">### Summarize prophage clusters</span></span>
<span id="cb33-184"><a href="#cb33-184"></a></span>
<span id="cb33-185"><a href="#cb33-185"></a>Summarize the representative prophages in the pangenome.</span>
<span id="cb33-186"><a href="#cb33-186"></a></span>
<span id="cb33-187"><a href="#cb33-187"></a><span class="in">```bash</span></span>
<span id="cb33-188"><a href="#cb33-188"></a><span class="ex">Rscript</span> scripts/analysis/prophage_cluster_summary.qmd</span>
<span id="cb33-189"><a href="#cb33-189"></a><span class="in">```</span></span>
<span id="cb33-190"><a href="#cb33-190"></a></span>
<span id="cb33-191"><a href="#cb33-191"></a>Visualize the representative prophages across pangenome. Use MASH distance to</span>
<span id="cb33-192"><a href="#cb33-192"></a>cluster prophages and ANI distance to cluster the genomes in the pangenome.</span>
<span id="cb33-193"><a href="#cb33-193"></a></span>
<span id="cb33-194"><a href="#cb33-194"></a>:::{.callout-caution}</span>
<span id="cb33-195"><a href="#cb33-195"></a>Need a fix. Do not run!</span>
<span id="cb33-196"><a href="#cb33-196"></a></span>
<span id="cb33-197"><a href="#cb33-197"></a><span class="in">```bash {eval=FALSE}</span></span>
<span id="cb33-198"><a href="#cb33-198"></a><span class="in">quarto render scripts/analysis/prophage_pangenome_viz.qmd --execute-dir ./</span></span>
<span id="cb33-199"><a href="#cb33-199"></a><span class="in">```</span></span>
<span id="cb33-200"><a href="#cb33-200"></a></span>
<span id="cb33-201"><a href="#cb33-201"></a>:::</span>
<span id="cb33-202"><a href="#cb33-202"></a></span>
<span id="cb33-203"><a href="#cb33-203"></a>Visualize homology group sets for selected prophages as PAV heatmap across the</span>
<span id="cb33-204"><a href="#cb33-204"></a>pangenome. Replace the code block in <span class="in">`scripts/analysis/homology_group_viz.R`</span> </span>
<span id="cb33-205"><a href="#cb33-205"></a>script with the following.</span>
<span id="cb33-206"><a href="#cb33-206"></a></span>
<span id="cb33-207"><a href="#cb33-207"></a><span class="in">```r</span></span>
<span id="cb33-208"><a href="#cb33-208"></a>analysisName <span class="ot">&lt;-</span> <span class="st">"vir_lineage_prophages"</span></span>
<span id="cb33-209"><a href="#cb33-209"></a></span>
<span id="cb33-210"><a href="#cb33-210"></a>outDir <span class="ot">&lt;-</span> <span class="fu">file.path</span>(confs<span class="sc">$</span>analysis<span class="sc">$</span>prophages<span class="sc">$</span>dir)</span>
<span id="cb33-211"><a href="#cb33-211"></a>outPrefix <span class="ot">&lt;-</span> <span class="fu">file.path</span>(outDir, analysisName)</span>
<span id="cb33-212"><a href="#cb33-212"></a></span>
<span id="cb33-213"><a href="#cb33-213"></a>setIds <span class="ot">&lt;-</span> <span class="fu">c</span>(</span>
<span id="cb33-214"><a href="#cb33-214"></a>  <span class="st">"g_189.vir_2"</span>, <span class="st">"g_400.vir_2"</span>, <span class="st">"g_399.vir_2"</span>, <span class="st">"g_448.vir_2"</span>, <span class="st">"g_189.vir_1"</span>, </span>
<span id="cb33-215"><a href="#cb33-215"></a>  <span class="st">"g_3.vir_4"</span>, <span class="st">"g_438.vir_3"</span>, <span class="st">"g_350.vir_1"</span>,</span>
<span id="cb33-216"><a href="#cb33-216"></a>  <span class="st">"g_172.vir_2"</span>, <span class="st">"g_393.vir_1"</span></span>
<span id="cb33-217"><a href="#cb33-217"></a>)</span>
<span id="cb33-218"><a href="#cb33-218"></a><span class="in">```</span></span>
<span id="cb33-219"><a href="#cb33-219"></a></span>
<span id="cb33-220"><a href="#cb33-220"></a>Finally, run the script to generate the plot.</span>
<span id="cb33-221"><a href="#cb33-221"></a></span>
<span id="cb33-222"><a href="#cb33-222"></a><span class="in">```bash</span></span>
<span id="cb33-223"><a href="#cb33-223"></a><span class="ex">Rscript</span> scripts/analysis/homology_group_viz.R</span>
<span id="cb33-224"><a href="#cb33-224"></a><span class="in">```</span></span>
<span id="cb33-225"><a href="#cb33-225"></a></span>
<span id="cb33-226"><a href="#cb33-226"></a><span class="fu">## Individual prophage cluster analysis</span></span>
<span id="cb33-227"><a href="#cb33-227"></a></span>
<span id="cb33-228"><a href="#cb33-228"></a>Script <span class="in">`scripts/analysis/clustersmap_data_prophages.R`</span> is used to combine the</span>
<span id="cb33-229"><a href="#cb33-229"></a>prophages, homology groups and phylogeny to generate input for <span class="in">`clustermap.js`</span></span>
<span id="cb33-230"><a href="#cb33-230"></a>visualization tool and generate a report. Below, some configuration is provided</span>
<span id="cb33-231"><a href="#cb33-231"></a>for individual prophage clusters.</span>
<span id="cb33-232"><a href="#cb33-232"></a></span>
<span id="cb33-233"><a href="#cb33-233"></a><span class="fu">### phage_grp_46: highly conserved in all but 5 genomes from n23 clade</span></span>
<span id="cb33-234"><a href="#cb33-234"></a></span>
<span id="cb33-235"><a href="#cb33-235"></a><span class="in">```bash</span></span>
<span id="cb33-236"><a href="#cb33-236"></a><span class="ex">Rscript</span> scripts/utils/HG_range_coordinates.R <span class="at">--hgs</span> hg_22429500,hg_22429499 <span class="dt">\</span></span>
<span id="cb33-237"><a href="#cb33-237"></a>--inner_region <span class="at">--genomes</span> <span class="st">"g_63,g_190,g_193,g_411,g_415,g_406,g_194,g_263"</span></span>
<span id="cb33-238"><a href="#cb33-238"></a><span class="in">```</span></span>
<span id="cb33-239"><a href="#cb33-239"></a></span>
<span id="cb33-240"><a href="#cb33-240"></a><span class="in">```r</span></span>
<span id="cb33-241"><a href="#cb33-241"></a>cluster_title <span class="ot">&lt;-</span> <span class="st">"phage_grp_46"</span></span>
<span id="cb33-242"><a href="#cb33-242"></a>outDir <span class="ot">&lt;-</span> <span class="fu">paste</span>(confs<span class="sc">$</span>analysis<span class="sc">$</span>prophages<span class="sc">$</span>dir, <span class="st">"/cluster_viz/"</span>, cluster_title, <span class="at">sep =</span> <span class="st">""</span>)</span>
<span id="cb33-243"><a href="#cb33-243"></a>hg_color_categories <span class="ot">&lt;-</span> confs<span class="sc">$</span>analysis<span class="sc">$</span>prophages<span class="sc">$</span>files<span class="sc">$</span>hg_broad_functions</span>
<span id="cb33-244"><a href="#cb33-244"></a></span>
<span id="cb33-245"><a href="#cb33-245"></a><span class="co"># a vector of prophage identifiers that will be included in clustermap plot</span></span>
<span id="cb33-246"><a href="#cb33-246"></a>region_cluster <span class="ot">&lt;-</span> <span class="st">"phage_grp_46"</span></span>
<span id="cb33-247"><a href="#cb33-247"></a>other_regions <span class="ot">&lt;-</span> <span class="fu">character</span>(<span class="dv">0</span>)</span>
<span id="cb33-248"><a href="#cb33-248"></a></span>
<span id="cb33-249"><a href="#cb33-249"></a>subSample <span class="ot">&lt;-</span> <span class="cn">TRUE</span></span>
<span id="cb33-250"><a href="#cb33-250"></a>cutHeight <span class="ot">&lt;-</span> <span class="fl">1.5</span></span>
<span id="cb33-251"><a href="#cb33-251"></a>addFlankingRegions <span class="ot">&lt;-</span> <span class="cn">TRUE</span></span>
<span id="cb33-252"><a href="#cb33-252"></a>flankingRegion <span class="ot">&lt;-</span> <span class="dv">5000</span></span>
<span id="cb33-253"><a href="#cb33-253"></a></span>
<span id="cb33-254"><a href="#cb33-254"></a><span class="co"># ordering factor for prophages: "host" phylogeny, "hg_pav" for prophage HG PAV,</span></span>
<span id="cb33-255"><a href="#cb33-255"></a><span class="co"># "cluster_mash" for prophage MASH and "default" to use the provided</span></span>
<span id="cb33-256"><a href="#cb33-256"></a>clusterOrder <span class="ot">&lt;-</span> <span class="st">"host"</span> <span class="co"># host, hg_pav, cluster_mash, default</span></span>
<span id="cb33-257"><a href="#cb33-257"></a></span>
<span id="cb33-258"><a href="#cb33-258"></a><span class="co"># whether to keep custom regions at the bottom or consider during phylogeny</span></span>
<span id="cb33-259"><a href="#cb33-259"></a><span class="co"># based ordering</span></span>
<span id="cb33-260"><a href="#cb33-260"></a>regions_phy_ordered <span class="ot">&lt;-</span> <span class="cn">TRUE</span> </span>
<span id="cb33-261"><a href="#cb33-261"></a></span>
<span id="cb33-262"><a href="#cb33-262"></a><span class="co"># regions to append as list of list with following structure</span></span>
<span id="cb33-263"><a href="#cb33-263"></a><span class="co"># list(r1 = list(chr, start, end, genomeId), r2 = list(chr, start, end, genomeId))</span></span>
<span id="cb33-264"><a href="#cb33-264"></a>customRegions <span class="ot">&lt;-</span> <span class="fu">list</span>(</span>
<span id="cb33-265"><a href="#cb33-265"></a>  <span class="at">g_190_reg =</span> <span class="fu">list</span>(<span class="at">chr =</span> <span class="st">"NZ_CP059960.1"</span>, <span class="at">start =</span> <span class="dv">1997124</span>, <span class="at">end =</span> <span class="dv">2000186</span>, <span class="at">genomeId =</span> <span class="st">"g_190"</span>),</span>
<span id="cb33-266"><a href="#cb33-266"></a>  <span class="at">g_193_reg =</span> <span class="fu">list</span>(<span class="at">chr =</span> <span class="st">"NZ_CP059957.1"</span>, <span class="at">start =</span> <span class="dv">5005351</span>, <span class="at">end =</span> <span class="dv">5008413</span>, <span class="at">genomeId =</span> <span class="st">"g_193"</span>),</span>
<span id="cb33-267"><a href="#cb33-267"></a>  <span class="at">g_411_reg =</span> <span class="fu">list</span>(<span class="at">chr =</span> <span class="st">"NAK682_contig_2_consensus"</span>, <span class="at">start =</span> <span class="dv">461100</span>, <span class="at">end =</span> <span class="dv">464162</span>, <span class="at">genomeId =</span> <span class="st">"g_411"</span>),</span>
<span id="cb33-268"><a href="#cb33-268"></a>  <span class="at">g_415_reg =</span> <span class="fu">list</span>(<span class="at">chr =</span> <span class="st">"NAK701_contig_5_consensus"</span>, <span class="at">start =</span> <span class="dv">415799</span>, <span class="at">end =</span> <span class="dv">418861</span>, <span class="at">genomeId =</span> <span class="st">"g_415"</span>),</span>
<span id="cb33-269"><a href="#cb33-269"></a>  <span class="at">g_63_reg =</span> <span class="fu">list</span>(<span class="at">chr =</span> <span class="st">"NZ_CP092039.1"</span>, <span class="at">start =</span> <span class="dv">4288654</span>, <span class="at">end =</span> <span class="dv">4291716</span>, <span class="at">genomeId =</span> <span class="st">"g_63"</span>),</span>
<span id="cb33-270"><a href="#cb33-270"></a>  <span class="at">g_406_reg =</span> <span class="fu">list</span>(<span class="at">chr =</span> <span class="st">"NAK641_contig_10_consensus"</span>, <span class="at">start =</span> <span class="dv">469664</span>, <span class="at">end =</span> <span class="dv">474521</span>, <span class="at">genomeId =</span> <span class="st">"g_406"</span>),</span>
<span id="cb33-271"><a href="#cb33-271"></a>  <span class="at">g_194_reg =</span> <span class="fu">list</span>(<span class="at">chr =</span> <span class="st">"NZ_CP059956.1"</span>, <span class="at">start =</span> <span class="dv">4163454</span>, <span class="at">end =</span> <span class="dv">4167352</span>, <span class="at">genomeId =</span> <span class="st">"g_194"</span>)</span>
<span id="cb33-272"><a href="#cb33-272"></a>  <span class="co"># g_263_reg = list(chr = "NZ_CP047495.1", start = 956237, end = 956907, genomeId = "g_263")</span></span>
<span id="cb33-273"><a href="#cb33-273"></a>)</span>
<span id="cb33-274"><a href="#cb33-274"></a><span class="in">```</span></span>
<span id="cb33-275"><a href="#cb33-275"></a></span>
<span id="cb33-276"><a href="#cb33-276"></a><span class="fu">### phage_grp_45: prophage present in all BL-causing isolates</span></span>
<span id="cb33-277"><a href="#cb33-277"></a></span>
<span id="cb33-278"><a href="#cb33-278"></a><span class="in">```bash</span></span>
<span id="cb33-279"><a href="#cb33-279"></a><span class="ex">Rscript</span> scripts/utils/HG_range_coordinates.R <span class="at">--hgs</span> hg_22429319,hg_22429327 <span class="dt">\</span></span>
<span id="cb33-280"><a href="#cb33-280"></a>--genomes <span class="st">"g_406,g_194"</span></span>
<span id="cb33-281"><a href="#cb33-281"></a><span class="in">```</span></span>
<span id="cb33-282"><a href="#cb33-282"></a></span>
<span id="cb33-283"><a href="#cb33-283"></a><span class="in">```r</span></span>
<span id="cb33-284"><a href="#cb33-284"></a>cluster_title <span class="ot">&lt;-</span> <span class="st">"phage_grp_45"</span></span>
<span id="cb33-285"><a href="#cb33-285"></a>outDir <span class="ot">&lt;-</span> <span class="fu">paste</span>(confs<span class="sc">$</span>analysis<span class="sc">$</span>prophages<span class="sc">$</span>dir, <span class="st">"/cluster_viz/"</span>, cluster_title, <span class="at">sep =</span> <span class="st">""</span>)</span>
<span id="cb33-286"><a href="#cb33-286"></a>hg_color_categories <span class="ot">&lt;-</span> confs<span class="sc">$</span>analysis<span class="sc">$</span>prophages<span class="sc">$</span>files<span class="sc">$</span>hg_broad_functions</span>
<span id="cb33-287"><a href="#cb33-287"></a></span>
<span id="cb33-288"><a href="#cb33-288"></a><span class="co"># a vector of prophage identifiers that will be included in clustermap plot</span></span>
<span id="cb33-289"><a href="#cb33-289"></a>region_cluster <span class="ot">&lt;-</span> <span class="st">"phage_grp_45"</span></span>
<span id="cb33-290"><a href="#cb33-290"></a>other_regions <span class="ot">&lt;-</span> <span class="fu">character</span>(<span class="dv">0</span>)</span>
<span id="cb33-291"><a href="#cb33-291"></a></span>
<span id="cb33-292"><a href="#cb33-292"></a>subSample <span class="ot">&lt;-</span> <span class="cn">FALSE</span> </span>
<span id="cb33-293"><a href="#cb33-293"></a>cutHeight <span class="ot">&lt;-</span> <span class="fl">1.5</span></span>
<span id="cb33-294"><a href="#cb33-294"></a>addFlankingRegions <span class="ot">&lt;-</span> <span class="cn">TRUE</span></span>
<span id="cb33-295"><a href="#cb33-295"></a>flankingRegion <span class="ot">&lt;-</span> <span class="dv">5000</span></span>
<span id="cb33-296"><a href="#cb33-296"></a></span>
<span id="cb33-297"><a href="#cb33-297"></a><span class="co"># ordering factor for prophages: "host" phylogeny, "hg_pav" for prophage HG PAV,</span></span>
<span id="cb33-298"><a href="#cb33-298"></a><span class="co"># "cluster_mash" for prophage MASH and "default" to use the provided</span></span>
<span id="cb33-299"><a href="#cb33-299"></a>clusterOrder <span class="ot">&lt;-</span> <span class="st">"host"</span> <span class="co"># host, hg_pav, cluster_mash, default</span></span>
<span id="cb33-300"><a href="#cb33-300"></a></span>
<span id="cb33-301"><a href="#cb33-301"></a><span class="co"># whether to keep custom regions at the bottom or consider during phylogeny</span></span>
<span id="cb33-302"><a href="#cb33-302"></a><span class="co"># based ordering</span></span>
<span id="cb33-303"><a href="#cb33-303"></a>regions_phy_ordered <span class="ot">&lt;-</span> <span class="cn">TRUE</span> </span>
<span id="cb33-304"><a href="#cb33-304"></a></span>
<span id="cb33-305"><a href="#cb33-305"></a><span class="co"># regions to append as list of list with following structure</span></span>
<span id="cb33-306"><a href="#cb33-306"></a><span class="co"># list(r1 = list(chr, start, end, genomeId), r2 = list(chr, start, end, genomeId))</span></span>
<span id="cb33-307"><a href="#cb33-307"></a>customRegions <span class="ot">&lt;-</span> <span class="fu">list</span>(</span>
<span id="cb33-308"><a href="#cb33-308"></a>  <span class="at">g_406_reg =</span> <span class="fu">list</span>(<span class="at">chr =</span> <span class="st">"NAK641_contig_10_consensus"</span>, <span class="at">start =</span> <span class="dv">671040</span>, <span class="at">end =</span> <span class="dv">674984</span>, <span class="at">genomeId =</span> <span class="st">"g_406"</span>),</span>
<span id="cb33-309"><a href="#cb33-309"></a>  <span class="at">g_194_reg =</span> <span class="fu">list</span>(<span class="at">chr =</span> <span class="st">"NZ_CP059956.1"</span>, <span class="at">start =</span> <span class="dv">3940785</span>, <span class="at">end =</span> <span class="dv">3958060</span>, <span class="at">genomeId =</span> <span class="st">"g_194"</span>)</span>
<span id="cb33-310"><a href="#cb33-310"></a>)</span>
<span id="cb33-311"><a href="#cb33-311"></a><span class="in">```</span></span>
<span id="cb33-312"><a href="#cb33-312"></a></span>
<span id="cb33-313"><a href="#cb33-313"></a><span class="fu">### phage_grp_45 and phage_grp_343 mosaicsm</span></span>
<span id="cb33-314"><a href="#cb33-314"></a></span>
<span id="cb33-315"><a href="#cb33-315"></a><span class="in">```bash</span></span>
<span id="cb33-316"><a href="#cb33-316"></a><span class="ex">Rscript</span> scripts/utils/HG_range_coordinates.R <span class="at">--hgs</span> hg_22429319,hg_22429327 <span class="dt">\</span></span>
<span id="cb33-317"><a href="#cb33-317"></a>--genomes <span class="st">"g_406,g_194"</span></span>
<span id="cb33-318"><a href="#cb33-318"></a><span class="in">```</span></span>
<span id="cb33-319"><a href="#cb33-319"></a></span>
<span id="cb33-320"><a href="#cb33-320"></a><span class="in">```r</span></span>
<span id="cb33-321"><a href="#cb33-321"></a>cluster_title <span class="ot">&lt;-</span> <span class="st">"grp_45_mosaicsm"</span></span>
<span id="cb33-322"><a href="#cb33-322"></a>outDir <span class="ot">&lt;-</span> <span class="fu">paste</span>(confs<span class="sc">$</span>analysis<span class="sc">$</span>prophages<span class="sc">$</span>dir, <span class="st">"/cluster_viz/"</span>, cluster_title, <span class="at">sep =</span> <span class="st">""</span>)</span>
<span id="cb33-323"><a href="#cb33-323"></a>hg_color_categories <span class="ot">&lt;-</span> confs<span class="sc">$</span>analysis<span class="sc">$</span>prophages<span class="sc">$</span>files<span class="sc">$</span>hg_broad_functions</span>
<span id="cb33-324"><a href="#cb33-324"></a></span>
<span id="cb33-325"><a href="#cb33-325"></a><span class="co"># a vector of prophage identifiers that will be included in clustermap plot</span></span>
<span id="cb33-326"><a href="#cb33-326"></a>region_cluster <span class="ot">&lt;-</span> <span class="cn">NA</span></span>
<span id="cb33-327"><a href="#cb33-327"></a>other_regions <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">"g_302.vir_1"</span>, <span class="st">"g_399.vir_1"</span>, <span class="st">"g_400.vir_2"</span>)</span>
<span id="cb33-328"><a href="#cb33-328"></a></span>
<span id="cb33-329"><a href="#cb33-329"></a>subSample <span class="ot">&lt;-</span> <span class="cn">TRUE</span></span>
<span id="cb33-330"><a href="#cb33-330"></a>cutHeight <span class="ot">&lt;-</span> <span class="fl">0.5</span></span>
<span id="cb33-331"><a href="#cb33-331"></a>addFlankingRegions <span class="ot">&lt;-</span> <span class="cn">TRUE</span></span>
<span id="cb33-332"><a href="#cb33-332"></a>flankingRegion <span class="ot">&lt;-</span> <span class="dv">7000</span></span>
<span id="cb33-333"><a href="#cb33-333"></a></span>
<span id="cb33-334"><a href="#cb33-334"></a><span class="co"># ordering factor for prophages: "host" phylogeny, "hg_pav" for prophage HG PAV,</span></span>
<span id="cb33-335"><a href="#cb33-335"></a><span class="co"># "cluster_mash" for prophage MASH and "default" to use the provided</span></span>
<span id="cb33-336"><a href="#cb33-336"></a>clusterOrder <span class="ot">&lt;-</span> <span class="st">"host"</span> <span class="co"># host, hg_pav, cluster_mash, default</span></span>
<span id="cb33-337"><a href="#cb33-337"></a></span>
<span id="cb33-338"><a href="#cb33-338"></a><span class="co"># whether to keep custom regions at the bottom or consider during phylogeny</span></span>
<span id="cb33-339"><a href="#cb33-339"></a><span class="co"># based ordering</span></span>
<span id="cb33-340"><a href="#cb33-340"></a>regions_phy_ordered <span class="ot">&lt;-</span> <span class="cn">FALSE</span></span>
<span id="cb33-341"><a href="#cb33-341"></a></span>
<span id="cb33-342"><a href="#cb33-342"></a><span class="co"># regions to append as list of list with following structure</span></span>
<span id="cb33-343"><a href="#cb33-343"></a><span class="co"># list(r1 = list(chr, start, end, genomeId), r2 = list(chr, start, end, genomeId))</span></span>
<span id="cb33-344"><a href="#cb33-344"></a>customRegions <span class="ot">&lt;-</span> <span class="fu">list</span>(</span>
<span id="cb33-345"><a href="#cb33-345"></a>  <span class="at">g_406_reg =</span> <span class="fu">list</span>(<span class="at">chr =</span> <span class="st">"NAK641_contig_10_consensus"</span>, <span class="at">start =</span> <span class="dv">671040</span>, <span class="at">end =</span> <span class="dv">674984</span>, <span class="at">genomeId =</span> <span class="st">"g_406"</span>),</span>
<span id="cb33-346"><a href="#cb33-346"></a>  <span class="at">g_194_reg =</span> <span class="fu">list</span>(<span class="at">chr =</span> <span class="st">"NZ_CP059956.1"</span>, <span class="at">start =</span> <span class="dv">3940785</span>, <span class="at">end =</span> <span class="dv">3958060</span>, <span class="at">genomeId =</span> <span class="st">"g_194"</span>)</span>
<span id="cb33-347"><a href="#cb33-347"></a>)</span>
<span id="cb33-348"><a href="#cb33-348"></a><span class="in">```</span></span>
<span id="cb33-349"><a href="#cb33-349"></a></span>
<span id="cb33-350"><a href="#cb33-350"></a><span class="fu">### phage_grp_71</span></span>
<span id="cb33-351"><a href="#cb33-351"></a></span>
<span id="cb33-352"><a href="#cb33-352"></a><span class="in">```r</span></span>
<span id="cb33-353"><a href="#cb33-353"></a>cluster_title <span class="ot">&lt;-</span> <span class="st">"phage_grp_71"</span></span>
<span id="cb33-354"><a href="#cb33-354"></a>outDir <span class="ot">&lt;-</span> <span class="fu">paste</span>(confs<span class="sc">$</span>analysis<span class="sc">$</span>prophages<span class="sc">$</span>dir, <span class="st">"/cluster_viz/"</span>, cluster_title, <span class="at">sep =</span> <span class="st">""</span>)</span>
<span id="cb33-355"><a href="#cb33-355"></a>hg_color_categories <span class="ot">&lt;-</span> confs<span class="sc">$</span>analysis<span class="sc">$</span>prophages<span class="sc">$</span>files<span class="sc">$</span>hg_broad_functions</span>
<span id="cb33-356"><a href="#cb33-356"></a></span>
<span id="cb33-357"><a href="#cb33-357"></a><span class="co"># a vector of prophage identifiers that will be included in clustermap plot</span></span>
<span id="cb33-358"><a href="#cb33-358"></a>region_cluster <span class="ot">&lt;-</span> <span class="st">"phage_grp_71"</span></span>
<span id="cb33-359"><a href="#cb33-359"></a>other_regions <span class="ot">&lt;-</span> <span class="fu">c</span>()</span>
<span id="cb33-360"><a href="#cb33-360"></a></span>
<span id="cb33-361"><a href="#cb33-361"></a>subSample <span class="ot">&lt;-</span> <span class="cn">FALSE</span></span>
<span id="cb33-362"><a href="#cb33-362"></a>cutHeight <span class="ot">&lt;-</span> <span class="fl">0.5</span></span>
<span id="cb33-363"><a href="#cb33-363"></a>addFlankingRegions <span class="ot">&lt;-</span> <span class="cn">TRUE</span></span>
<span id="cb33-364"><a href="#cb33-364"></a>flankingRegion <span class="ot">&lt;-</span> <span class="dv">5000</span></span>
<span id="cb33-365"><a href="#cb33-365"></a></span>
<span id="cb33-366"><a href="#cb33-366"></a><span class="co"># ordering factor for prophages: "host" phylogeny, "hg_pav" for prophage HG PAV,</span></span>
<span id="cb33-367"><a href="#cb33-367"></a><span class="co"># "cluster_mash" for prophage MASH and "default" to use the provided</span></span>
<span id="cb33-368"><a href="#cb33-368"></a>clusterOrder <span class="ot">&lt;-</span> <span class="st">"hg_pav"</span> <span class="co"># host, hg_pav, cluster_mash, default</span></span>
<span id="cb33-369"><a href="#cb33-369"></a></span>
<span id="cb33-370"><a href="#cb33-370"></a><span class="co"># whether to keep custom regions at the bottom or consider during phylogeny</span></span>
<span id="cb33-371"><a href="#cb33-371"></a><span class="co"># based ordering</span></span>
<span id="cb33-372"><a href="#cb33-372"></a>regions_phy_ordered <span class="ot">&lt;-</span> <span class="cn">FALSE</span></span>
<span id="cb33-373"><a href="#cb33-373"></a></span>
<span id="cb33-374"><a href="#cb33-374"></a><span class="co"># regions to append as list of list with following structure</span></span>
<span id="cb33-375"><a href="#cb33-375"></a><span class="co"># list(r1 = list(chr, start, end, genomeId), r2 = list(chr, start, end, genomeId))</span></span>
<span id="cb33-376"><a href="#cb33-376"></a>customRegions <span class="ot">&lt;-</span> <span class="fu">list</span>()</span>
<span id="cb33-377"><a href="#cb33-377"></a></span>
<span id="cb33-378"><a href="#cb33-378"></a><span class="in">```</span></span>
<span id="cb33-379"><a href="#cb33-379"></a></span>
<span id="cb33-380"><a href="#cb33-380"></a><span class="fu">### phage_grp_107</span></span>
<span id="cb33-381"><a href="#cb33-381"></a></span>
<span id="cb33-382"><a href="#cb33-382"></a><span class="in">```bash</span></span>
<span id="cb33-383"><a href="#cb33-383"></a><span class="ex">Rscript</span> scripts/utils/HG_range_coordinates.R <span class="at">--hgs</span> hg_22427004,hg_22426993 <span class="dt">\</span></span>
<span id="cb33-384"><a href="#cb33-384"></a>--inner_region <span class="at">--genomes</span> <span class="st">"g_63,g_194"</span></span>
<span id="cb33-385"><a href="#cb33-385"></a><span class="in">```</span></span>
<span id="cb33-386"><a href="#cb33-386"></a></span>
<span id="cb33-387"><a href="#cb33-387"></a><span class="in">```r</span></span>
<span id="cb33-388"><a href="#cb33-388"></a>cluster_title <span class="ot">&lt;-</span> <span class="st">"phage_grp_107"</span></span>
<span id="cb33-389"><a href="#cb33-389"></a>outDir <span class="ot">&lt;-</span> <span class="fu">paste</span>(confs<span class="sc">$</span>analysis<span class="sc">$</span>prophages<span class="sc">$</span>dir, <span class="st">"/cluster_viz"</span>, <span class="at">sep =</span> <span class="st">""</span>)</span>
<span id="cb33-390"><a href="#cb33-390"></a>hg_color_categories <span class="ot">&lt;-</span> confs<span class="sc">$</span>analysis<span class="sc">$</span>prophages<span class="sc">$</span>files<span class="sc">$</span>hg_broad_functions</span>
<span id="cb33-391"><a href="#cb33-391"></a></span>
<span id="cb33-392"><a href="#cb33-392"></a><span class="co"># a vector of prophage identifiers that will be included in clustermap plot</span></span>
<span id="cb33-393"><a href="#cb33-393"></a>region_cluster <span class="ot">&lt;-</span> <span class="st">"phage_grp_107"</span></span>
<span id="cb33-394"><a href="#cb33-394"></a>other_regions <span class="ot">&lt;-</span> <span class="fu">character</span>(<span class="dv">0</span>)</span>
<span id="cb33-395"><a href="#cb33-395"></a></span>
<span id="cb33-396"><a href="#cb33-396"></a>subSample <span class="ot">&lt;-</span> <span class="cn">FALSE</span></span>
<span id="cb33-397"><a href="#cb33-397"></a>cutHeight <span class="ot">&lt;-</span> <span class="fl">1.5</span></span>
<span id="cb33-398"><a href="#cb33-398"></a>addFlankingRegions <span class="ot">&lt;-</span> <span class="cn">TRUE</span></span>
<span id="cb33-399"><a href="#cb33-399"></a>flankingRegion <span class="ot">&lt;-</span> <span class="dv">5000</span></span>
<span id="cb33-400"><a href="#cb33-400"></a></span>
<span id="cb33-401"><a href="#cb33-401"></a><span class="co"># ordering factor for prophages: "host" phylogeny, "hg_pav" for prophage HG PAV,</span></span>
<span id="cb33-402"><a href="#cb33-402"></a><span class="co"># "cluster_mash" for prophage MASH and "default" to use the provided</span></span>
<span id="cb33-403"><a href="#cb33-403"></a>clusterOrder <span class="ot">&lt;-</span> <span class="st">"host"</span> <span class="co"># host, hg_pav, cluster_mash, default</span></span>
<span id="cb33-404"><a href="#cb33-404"></a></span>
<span id="cb33-405"><a href="#cb33-405"></a><span class="co"># whether to keep custom regions at the bottom or consider during phylogeny</span></span>
<span id="cb33-406"><a href="#cb33-406"></a><span class="co"># based ordering</span></span>
<span id="cb33-407"><a href="#cb33-407"></a>regions_phy_ordered <span class="ot">&lt;-</span> <span class="cn">TRUE</span> </span>
<span id="cb33-408"><a href="#cb33-408"></a></span>
<span id="cb33-409"><a href="#cb33-409"></a><span class="co"># regions to append as list of list with following structure</span></span>
<span id="cb33-410"><a href="#cb33-410"></a><span class="co"># list(r1 = list(chr, start, end, genomeId), r2 = list(chr, start, end, genomeId))</span></span>
<span id="cb33-411"><a href="#cb33-411"></a>customRegions <span class="ot">&lt;-</span> <span class="fu">list</span>(</span>
<span id="cb33-412"><a href="#cb33-412"></a>  <span class="at">g_63_reg =</span> <span class="fu">list</span>(<span class="at">chr =</span> <span class="st">"NZ_CP092039.1"</span>, <span class="at">start =</span> <span class="dv">2572138</span>, <span class="at">end =</span> <span class="dv">2573569</span>, <span class="at">genomeId =</span> <span class="st">"g_63"</span>),</span>
<span id="cb33-413"><a href="#cb33-413"></a>  <span class="at">g_194_reg =</span> <span class="fu">list</span>(<span class="at">chr =</span> <span class="st">"NZ_CP059956.1"</span>, <span class="at">start =</span> <span class="dv">2620331</span>, <span class="at">end =</span> <span class="dv">2621761</span>, <span class="at">genomeId =</span> <span class="st">"g_194"</span>)</span>
<span id="cb33-414"><a href="#cb33-414"></a>)</span>
<span id="cb33-415"><a href="#cb33-415"></a></span>
<span id="cb33-416"><a href="#cb33-416"></a><span class="in">```</span></span>
<span id="cb33-417"><a href="#cb33-417"></a></span>
<span id="cb33-418"><a href="#cb33-418"></a><span class="fu">### BL-causing clade specific signatures</span></span>
<span id="cb33-419"><a href="#cb33-419"></a></span>
<span id="cb33-420"><a href="#cb33-420"></a><span class="in">```r</span></span>
<span id="cb33-421"><a href="#cb33-421"></a>grpToView <span class="ot">&lt;-</span> <span class="st">"Pbr_FN_specific"</span></span>
<span id="cb33-422"><a href="#cb33-422"></a>subSample <span class="ot">&lt;-</span> <span class="cn">FALSE</span></span>
<span id="cb33-423"><a href="#cb33-423"></a>cutHeight <span class="ot">&lt;-</span> <span class="fl">0.5</span></span>
<span id="cb33-424"><a href="#cb33-424"></a>addFlankingRegions <span class="ot">&lt;-</span> <span class="cn">FALSE</span></span>
<span id="cb33-425"><a href="#cb33-425"></a>flankingRegion <span class="ot">&lt;-</span> <span class="dv">5000</span></span>
<span id="cb33-426"><a href="#cb33-426"></a></span>
<span id="cb33-427"><a href="#cb33-427"></a><span class="co"># ordering factor for prophages: host phylogeny, prophage HG PAV, prophage MASH,</span></span>
<span id="cb33-428"><a href="#cb33-428"></a><span class="co"># completeness score</span></span>
<span id="cb33-429"><a href="#cb33-429"></a>clusterOrder <span class="ot">&lt;-</span> <span class="st">"default"</span>  <span class="co"># completeness, host, hg_pav, cluster_mash, default</span></span>
<span id="cb33-430"><a href="#cb33-430"></a></span>
<span id="cb33-431"><a href="#cb33-431"></a></span>
<span id="cb33-432"><a href="#cb33-432"></a><span class="co"># optionally, a custom region list can be provided to generate the plot</span></span>
<span id="cb33-433"><a href="#cb33-433"></a>grp <span class="ot">&lt;-</span> <span class="fu">list</span>(</span>
<span id="cb33-434"><a href="#cb33-434"></a>  <span class="at">phage_grp =</span> grpToView,</span>
<span id="cb33-435"><a href="#cb33-435"></a>  <span class="at">members =</span> <span class="fu">c</span>(<span class="st">"g_302.vir_1"</span>, <span class="st">"g_400.vir_2"</span>, <span class="st">"g_409.vir_1"</span>, <span class="st">"g_302.vir_2"</span>)</span>
<span id="cb33-436"><a href="#cb33-436"></a>)</span>
<span id="cb33-437"><a href="#cb33-437"></a><span class="in">```</span></span>
<span id="cb33-438"><a href="#cb33-438"></a></span>
<span id="cb33-439"><a href="#cb33-439"></a><span class="fu">### phage_grp_36</span></span>
<span id="cb33-440"><a href="#cb33-440"></a></span>
<span id="cb33-441"><a href="#cb33-441"></a><span class="in">```r</span></span>
<span id="cb33-442"><a href="#cb33-442"></a>grpToView <span class="ot">&lt;-</span> <span class="st">"phage_grp_36"</span></span>
<span id="cb33-443"><a href="#cb33-443"></a>subSample <span class="ot">&lt;-</span> <span class="cn">FALSE</span></span>
<span id="cb33-444"><a href="#cb33-444"></a>cutHeight <span class="ot">&lt;-</span> <span class="fl">1.5</span></span>
<span id="cb33-445"><a href="#cb33-445"></a>addFlankingRegions <span class="ot">&lt;-</span> <span class="cn">TRUE</span></span>
<span id="cb33-446"><a href="#cb33-446"></a>flankingRegion <span class="ot">&lt;-</span> <span class="dv">5000</span></span>
<span id="cb33-447"><a href="#cb33-447"></a></span>
<span id="cb33-448"><a href="#cb33-448"></a><span class="co"># ordering factor for prophages: host phylogeny, prophage HG PAV, prophage MASH,</span></span>
<span id="cb33-449"><a href="#cb33-449"></a><span class="co"># completeness score</span></span>
<span id="cb33-450"><a href="#cb33-450"></a>clusterOrder <span class="ot">&lt;-</span> <span class="st">"host"</span> <span class="co"># completeness, host, hg_pav, cluster_mash</span></span>
<span id="cb33-451"><a href="#cb33-451"></a><span class="in">```</span></span>
<span id="cb33-452"><a href="#cb33-452"></a></span>
<span id="cb33-453"><a href="#cb33-453"></a><span class="fu">### phage_grp_50</span></span>
<span id="cb33-454"><a href="#cb33-454"></a></span>
<span id="cb33-455"><a href="#cb33-455"></a><span class="in">```r</span></span>
<span id="cb33-456"><a href="#cb33-456"></a>grpToView <span class="ot">&lt;-</span> <span class="st">"phage_grp_50"</span></span>
<span id="cb33-457"><a href="#cb33-457"></a>subSample <span class="ot">&lt;-</span> <span class="cn">FALSE</span></span>
<span id="cb33-458"><a href="#cb33-458"></a>addFlankingRegions <span class="ot">&lt;-</span> <span class="cn">TRUE</span> </span>
<span id="cb33-459"><a href="#cb33-459"></a>flankingRegion <span class="ot">&lt;-</span> <span class="dv">10000</span></span>
<span id="cb33-460"><a href="#cb33-460"></a></span>
<span id="cb33-461"><a href="#cb33-461"></a><span class="co"># ordering factor for prophages: host phylogeny, prophage HG PAV, prophage MASH,</span></span>
<span id="cb33-462"><a href="#cb33-462"></a><span class="co"># completeness score</span></span>
<span id="cb33-463"><a href="#cb33-463"></a>clusterOrder <span class="ot">&lt;-</span> <span class="st">"cluster_mash"</span>  <span class="co"># completeness, host, hg_pav, cluster_mash</span></span>
<span id="cb33-464"><a href="#cb33-464"></a><span class="in">```</span></span>
<span id="cb33-465"><a href="#cb33-465"></a></span>
<span id="cb33-466"><a href="#cb33-466"></a><span class="fu">### phage_grp_21</span></span>
<span id="cb33-467"><a href="#cb33-467"></a></span>
<span id="cb33-468"><a href="#cb33-468"></a><span class="in">```r</span></span>
<span id="cb33-469"><a href="#cb33-469"></a>grpToView <span class="ot">&lt;-</span> <span class="st">"phage_grp_21"</span></span>
<span id="cb33-470"><a href="#cb33-470"></a>subSample <span class="ot">&lt;-</span> <span class="cn">FALSE</span></span>
<span id="cb33-471"><a href="#cb33-471"></a>addFlankingRegions <span class="ot">&lt;-</span> <span class="cn">TRUE</span></span>
<span id="cb33-472"><a href="#cb33-472"></a>flankingRegion <span class="ot">&lt;-</span> <span class="dv">5000</span></span>
<span id="cb33-473"><a href="#cb33-473"></a></span>
<span id="cb33-474"><a href="#cb33-474"></a><span class="co"># ordering factor for prophages: host phylogeny, prophage HG PAV, prophage MASH,</span></span>
<span id="cb33-475"><a href="#cb33-475"></a><span class="co"># completeness score</span></span>
<span id="cb33-476"><a href="#cb33-476"></a>clusterOrder <span class="ot">&lt;-</span> <span class="st">"cluster_mash"</span>  <span class="co"># completeness, host, hg_pav, cluster_mash</span></span>
<span id="cb33-477"><a href="#cb33-477"></a><span class="in">```</span></span>
<span id="cb33-478"><a href="#cb33-478"></a></span>
<span id="cb33-479"><a href="#cb33-479"></a><span class="fu">### T3SS missing from 4 P. brasiliense genomes</span></span>
<span id="cb33-480"><a href="#cb33-480"></a></span>
<span id="cb33-481"><a href="#cb33-481"></a><span class="in">```bash</span></span>
<span id="cb33-482"><a href="#cb33-482"></a><span class="ex">Rscript</span> scripts/utils/HG_range_coordinates.R <span class="at">--hgs</span> hg_22430875,hg_22430876 <span class="at">--genomes</span> g_187,g_368,g_182,g_185,g_236,g_177</span>
<span id="cb33-483"><a href="#cb33-483"></a><span class="in">```</span></span>
<span id="cb33-484"><a href="#cb33-484"></a></span>
<span id="cb33-485"><a href="#cb33-485"></a><span class="in">```r</span></span>
<span id="cb33-486"><a href="#cb33-486"></a>grpToView <span class="ot">&lt;-</span> <span class="st">"t3ss_regions"</span></span>
<span id="cb33-487"><a href="#cb33-487"></a>subSample <span class="ot">&lt;-</span> <span class="cn">FALSE</span></span>
<span id="cb33-488"><a href="#cb33-488"></a>cutHeight <span class="ot">&lt;-</span> <span class="fl">1.5</span></span>
<span id="cb33-489"><a href="#cb33-489"></a>addFlankingRegions <span class="ot">&lt;-</span> <span class="cn">TRUE</span></span>
<span id="cb33-490"><a href="#cb33-490"></a>flankingRegion <span class="ot">&lt;-</span> <span class="dv">6000</span></span>
<span id="cb33-491"><a href="#cb33-491"></a></span>
<span id="cb33-492"><a href="#cb33-492"></a><span class="co"># ordering factor for prophages: host phylogeny, prophage HG PAV, prophage MASH,</span></span>
<span id="cb33-493"><a href="#cb33-493"></a><span class="co"># completeness score</span></span>
<span id="cb33-494"><a href="#cb33-494"></a>clusterOrder <span class="ot">&lt;-</span> <span class="st">"host"</span>  <span class="co"># completeness, host, hg_pav, cluster_mash</span></span>
<span id="cb33-495"><a href="#cb33-495"></a></span>
<span id="cb33-496"><a href="#cb33-496"></a><span class="co"># a vector of prophage identifiers that will be included in clustermap plot</span></span>
<span id="cb33-497"><a href="#cb33-497"></a>appendPhages <span class="ot">&lt;-</span> <span class="fu">c</span>()</span>
<span id="cb33-498"><a href="#cb33-498"></a></span>
<span id="cb33-499"><a href="#cb33-499"></a><span class="co"># regions to append as list of list with following structure</span></span>
<span id="cb33-500"><a href="#cb33-500"></a><span class="co"># list(r1 = list(chr, start, end, genomeId), r2 = list(chr, start, end, genomeId))</span></span>
<span id="cb33-501"><a href="#cb33-501"></a>customRegions <span class="ot">&lt;-</span> <span class="fu">list</span>(</span>
<span id="cb33-502"><a href="#cb33-502"></a>  <span class="at">g_177_reg  =</span> <span class="fu">list</span>(<span class="at">genomeId =</span> <span class="st">"g_177"</span>, <span class="at">chr =</span> <span class="st">"NZ_JACGEP010000027.1"</span>, <span class="at">start =</span> <span class="dv">56850</span>, <span class="at">end =</span> <span class="dv">59337</span>),</span>
<span id="cb33-503"><a href="#cb33-503"></a>  <span class="at">g_182_reg =</span> <span class="fu">list</span>(<span class="at">genomeId =</span> <span class="st">"g_182"</span>, <span class="at">chr =</span> <span class="st">"NZ_JACGZZ010000068.1"</span>, <span class="at">start =</span> <span class="dv">116773</span>, <span class="at">end =</span> <span class="dv">119260</span>),</span>
<span id="cb33-504"><a href="#cb33-504"></a>  <span class="at">g_185_reg =</span> <span class="fu">list</span>(<span class="at">genomeId =</span> <span class="st">"g_185"</span>, <span class="at">chr =</span> <span class="st">"NZ_JACGEN010000027.1"</span>, <span class="at">start =</span> <span class="dv">117747</span>, <span class="at">end =</span> <span class="dv">120234</span>),</span>
<span id="cb33-505"><a href="#cb33-505"></a>  <span class="at">g_236_reg =</span> <span class="fu">list</span>(<span class="at">genomeId =</span> <span class="st">"g_236"</span>, <span class="at">chr =</span> <span class="st">"NZ_JACDSF010000026.1"</span>, <span class="at">start =</span> <span class="dv">2275</span>, <span class="at">end =</span> <span class="dv">4762</span>),</span>
<span id="cb33-506"><a href="#cb33-506"></a>  <span class="at">g_187_reg =</span> <span class="fu">list</span>(<span class="at">genomeId =</span> <span class="st">"g_187"</span>, <span class="at">chr =</span> <span class="st">"NZ_CP059963.1"</span>, <span class="at">start =</span> <span class="dv">2603432</span>, <span class="at">end =</span> <span class="dv">2605919</span>),</span>
<span id="cb33-507"><a href="#cb33-507"></a>  <span class="at">g_368_reg =</span> <span class="fu">list</span>(<span class="at">genomeId =</span> <span class="st">"g_368"</span>, <span class="at">chr =</span> <span class="st">"NZ_JUJT01000001.1"</span>, <span class="at">start =</span> <span class="dv">566617</span>, <span class="at">end =</span> <span class="dv">569104</span>)</span>
<span id="cb33-508"><a href="#cb33-508"></a>)</span>
<span id="cb33-509"><a href="#cb33-509"></a></span>
<span id="cb33-510"><a href="#cb33-510"></a><span class="co"># whether to keep custom regions at the bottom or consider during phylogeny</span></span>
<span id="cb33-511"><a href="#cb33-511"></a><span class="co"># based ordering</span></span>
<span id="cb33-512"><a href="#cb33-512"></a>regions_phy_ordered <span class="ot">&lt;-</span> <span class="cn">TRUE</span></span>
<span id="cb33-513"><a href="#cb33-513"></a><span class="in">```</span></span></code><button title="Copy to Clipboard" class="code-copy-button" data-in-quarto-modal=""><i class="bi"></i></button></pre></div>
</div></div></div></div></div>
</div> <!-- /content -->




<footer class="footer"><div class="nav-footer"><div class="nav-footer-center"><div class="toc-actions d-sm-block d-md-none"><ul><li><a href="https://github.com/lakhanp1/Pectobacterium_pangenome/issues/new" class="toc-action"><i class="bi bi-github"></i>Report an issue</a></li></ul></div></div></div></footer></body></html>